// @c8y/ngx-components/esm5/repository/configuration/configuration-new-modal.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { AlertService, DroppedFile, gettext } from '@c8y/ngx-components';
import { uniqBy } from 'lodash';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { pipe } from 'rxjs';
import { map } from 'rxjs/operators';
import { RepositoryType } from '../repository.model';
import { RepositoryService } from '../repository.service';
var ConfigurationNewModalComponent = /** @class */ (function () {
    function ConfigurationNewModalComponent(repositoryService, bsModalRef, alert) {
        var _this = this;
        this.repositoryService = repositoryService;
        this.bsModalRef = bsModalRef;
        this.alert = alert;
        this.binary = {
            file: undefined,
            url: undefined
        };
        this.pattern = '';
        this.mo = {};
        this.isLoading = false;
        this.result = new Promise(function (resolve, reject) {
            _this._save = resolve;
            _this._cancel = reject;
        });
    }
    ConfigurationNewModalComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.repositoryService.listConfigurations()];
                    case 1:
                        _a.configs = _b.sent();
                        this.setPipe('');
                        return [2 /*return*/];
                }
            });
        });
    };
    ConfigurationNewModalComponent.prototype.cancel = function () {
        this.bsModalRef.hide();
        this._cancel();
    };
    ConfigurationNewModalComponent.prototype.setPipe = function (filterStr) {
        this.pattern = filterStr;
        this.filterPipe = pipe(map(function (data) { return uniqBy(data, 'configurationType'); }), map(function (data) {
            return data.filter(function (mo) {
                return mo.configurationType &&
                    mo.configurationType.toLowerCase().indexOf(filterStr.toLowerCase()) > -1;
            });
        }));
    };
    ConfigurationNewModalComponent.prototype.uploadFile = function (dropped, ngForm) {
        this.binary.file = dropped[0].file;
        this.binary.url = undefined;
        ngForm.form.markAsDirty();
    };
    ConfigurationNewModalComponent.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, selected, versionOrName, description, binary, deviceType, ex_1;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, 3, 4]);
                        this.isLoading = true;
                        _a = this, selected = _a.selected, versionOrName = _a.versionOrName, description = _a.description, binary = _a.binary, deviceType = _a.deviceType;
                        return [4 /*yield*/, this.repositoryService.save({ selected: selected, versionOrName: versionOrName, description: description, binary: binary, deviceType: deviceType }, RepositoryType.CONFIGURATION, this.mo)];
                    case 1:
                        _b.sent();
                        this.alert.success(this.mo.id ? gettext('Configuration updated.') : gettext('Configuration created.'));
                        this.bsModalRef.hide();
                        this._save();
                        return [3 /*break*/, 4];
                    case 2:
                        ex_1 = _b.sent();
                        this.alert.addServerFailure(ex_1);
                        this._cancel();
                        return [3 /*break*/, 4];
                    case 3:
                        this.isLoading = false;
                        return [7 /*endfinally*/];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ConfigurationNewModalComponent.ctorParameters = function () { return [
        { type: RepositoryService },
        { type: BsModalRef },
        { type: AlertService }
    ]; };
    ConfigurationNewModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-configuration-new-modal',
            template: "<div class=\"modal-header bg-primary text-white text-center\">\n  <h1 c8yIcon=\"cogs\" style=\"font-size: 55px;\"></h1>\n  <h4 class=\"p-t-16 text-uppercase\" translate *ngIf=\"mo.id\">Update configuration</h4>\n  <h4 class=\"p-t-16 text-uppercase\" translate *ngIf=\"!mo.id\">Add configuration</h4>\n</div>\n\n<form #configurationForm=\"ngForm\" (ngSubmit)=\"configurationForm.form.valid && save()\">\n  <div class=\"modal-body\">\n    <c8y-form-group>\n      <label translate>Name</label>\n      <input\n        type=\"text\"\n        class=\"form-control\"\n        placeholder=\"e.g. hosts\"\n        autocomplete=\"off\"\n        required\n        maxlength=\"254\"\n        [(ngModel)]=\"versionOrName\"\n        name=\"versionOrName\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate>Device type</label>\n      <input\n        type=\"text\"\n        class=\"form-control\"\n        placeholder=\"e.g. Posix\"\n        maxlength=\"254\"\n        autocomplete=\"off\"\n        [(ngModel)]=\"deviceType\"\n        name=\"deviceType\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate>Description</label>\n      <input\n        type=\"text\"\n        class=\"form-control\"\n        placeholder=\"e.g. Host configuration Posix\"\n        maxlength=\"254\"\n        autocomplete=\"off\"\n        [(ngModel)]=\"description\"\n        name=\"description\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate>Configuration type</label>\n      <c8y-typeahead\n        [(selected)]=\"selected\"\n        placeholder=\"e.g. ssh\"\n        (onSearch)=\"setPipe($event)\"\n        displayProperty=\"configurationType\"\n      >\n        <c8y-li\n          *c8yFor=\"let config of configs; pipe: filterPipe; notFound: notFoundTemplate\"\n          class=\"p-l-8 p-r-8 c8y-list__item--link\"\n          (click)=\"selected = config; setPipe('')\"\n          [active]=\"selected === config\"\n        >\n          <c8y-highlight\n            [text]=\"config.configurationType || '--'\"\n            [pattern]=\"pattern\"\n          ></c8y-highlight>\n        </c8y-li>\n        <ng-template #notFoundTemplate>\n          <c8y-li class=\"bg-gray-lighter p-8\" *ngIf=\"pattern.length > 0\">\n            <span translate>No match found, add new`configuration`?</span>\n            <button type=\"button\" class=\"btn btn-primary btn-xs m-l-16\" translate>\n              Create new`configuration`\n            </button>\n          </c8y-li>\n        </ng-template>\n      </c8y-typeahead>\n    </c8y-form-group>\n\n    <c8y-drop-area\n      (dropped)=\"uploadFile($event, configurationForm)\"\n      [loadingMessage]=\"'Importing, please wait.' | translate\"\n      [title]=\"'Add a configuration file' | translate\"\n      [loading]=\"isLoading\"\n    >\n    </c8y-drop-area>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button (click)=\"cancel()\" type=\"button\" class=\"btn btn-default\" [disabled]=\"isLoading\">\n      <span translate>Cancel</span>\n    </button>\n    <button\n      type=\"button\"\n      class=\"btn btn-primary\"\n      type=\"submit\"\n      [disabled]=\"!configurationForm.valid || configurationForm.pristine || isLoading\"\n    >\n      <span translate *ngIf=\"!mo.id\">Add configuration</span>\n      <span translate *ngIf=\"mo.id\">Update configuration</span>\n    </button>\n  </div>\n</form>\n"
        })
    ], ConfigurationNewModalComponent);
    return ConfigurationNewModalComponent;
}());
export { ConfigurationNewModalComponent };
