// @c8y/client/lib/esm2015/src/inventory/InventoryService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Service, QueriesUtil } from '../core';
import { InventoryBinaryService } from './InventoryBinaryService';
/**
 * Possible types of a child.
 */
export var ChildType;
(function (ChildType) {
    ChildType["ASSETS"] = "childAssets";
    ChildType["DEVICES"] = "childDevices";
    ChildType["ADDITIONS"] = "childAdditions";
})(ChildType || (ChildType = {}));
/**
 * This class allows for managing managed objects and different child types, see [[ChildType]].
 */
export class InventoryService extends Service {
    constructor(client, realtime) {
        super(client, realtime);
        this.baseUrl = 'inventory';
        this.listUrl = 'managedObjects';
        this.propertyName = 'managedObjects';
        this.channel = '/managedobjects/*';
        this.inventoriesQueryParamName = 'query';
        this.devicesQueryParamName = 'q';
        this.queriesUtil = new QueriesUtil();
        this.binary = new InventoryBinaryService(client);
    }
    /**
     * Gets the details of managed object
     *
     * @param {IdReference} managedObjectOrId ManagedObject or Id of the ManagedObject.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *
     *    const managedObjId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.detail(managedObjId);
     *   })();
     * ```
     */
    detail(managedObjectOrId) {
        const _super = Object.create(null, {
            detail: { get: () => super.detail }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.detail.call(this, managedObjectOrId);
        });
    }
    /**
     * Gets the details of managed object as observable
     *
     * @param {IdReference} managedObjectOrId ManagedObject or Id of the ManagedObject.
     * @param {IObservableOptions} options The observable options.
     *
     * @returns Data wrapped as Observable
     *
     * @example
     * ```typescript
     *
     *   const managedObjId: number = 1;
     *
     *   const detail$ = inventoryService.detail$(managedObjId);
     *   detail$.subscribe((data) => console.log(data));
     * ```
     *
     * @deprecated This function will be removed soon in favor of decoupling
     * this library from rxjs. Use [[Realtime]] for realtime connection and
     * rxjs `defer()` or `from()` to composite an observable on your own.
     */
    detail$(entityOrId, options) {
        return super.detail$(entityOrId, options);
    }
    /**
     * Gets a list as observable.
     *
     * @param {object} filter Object containing filters for querying
     * @param {object} options To configure the observable
     *
     * @returns Data wrapped as Observable.
     *
     * @example
     * ```typescript
     *
     *   const list$ = inventoryService.list$();
     *   list$.subscribe((data) => console.log(data));
     * ```
     *
     * @deprecated This function will be removed soon in favor of decoupling
     * this library from rxjs. Use [[Realtime]] for realtime connection and
     * rxjs `defer()` or `from()` to composite an observable on your own.
     */
    list$(filter = {}, options = {}) {
        return super.list$(filter, options);
    }
    /**
     * Creates a new managed object.
     *
     * @param {Partial<IManagedObject>} managedObject
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *
     *  const partialManagedObj: Partial<IManagedObject> = {
     *    customFragment: 'yourData'
     *  };
     *
     *  (async () => {
     *    const {data, res} = await inventoryService.create(partialManagedObj);
     *  })();
     * ```
     */
    create(managedObject) {
        const _super = Object.create(null, {
            create: { get: () => super.create }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.create.call(this, managedObject);
        });
    }
    /**
     * Updates managed object data.
     *
     * @param {Partial<IManagedObject>} managedObject Managed object is partially updatable.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *
     *  const partialUpdateObject: Partial<IManagedObject> = {
     *    customFragment: 'Changed data',
     *    name: 'Name'
     *  };
     *
     *  (async () => {
     *    const {data, res} = await inventoryService.update(partialUpdateObject);
     *  })();
     * ```
     */
    update(managedObject) {
        const _super = Object.create(null, {
            update: { get: () => super.update }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.update.call(this, managedObject);
        });
    }
    /**
     * Gets the list of managed objects filtered by parameters.
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {object} filter Object containing filters for querying managed objects.
     *
     * @example
     * ```typescript
     *
     *  const filter: object = {
     *     pageSize: 100,
     *     withTotalPages: true
     *   };
     *
     *   (async () => {
     *     const {data, res, paging} = await inventoryService.list(filter);
     *   })();
     * ```
     */
    list(filter = {}) {
        const _super = Object.create(null, {
            list: { get: () => super.list }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.list.call(this, filter);
        });
    }
    /**
     * Gets the list of all managed objects filtered and sorted by given query.
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {object} filter Object containing filters for querying managed objects.
     *
     * @example
     * ```typescript
     *
     *  const filter: object = {
     *     pageSize: 100,
     *     withTotalPages: true
     *   };
     *
     *  const query = {
     *      name: 'MY-NAM*'
     *  }
     *
     *   (async () => {
     *     const {data, res, paging} = await inventoryService.listQuery(query, filter);
     *   })();
     * ```
     */
    listQuery(query, filter = {}) {
        const _super = Object.create(null, {
            list: { get: () => super.list }
        });
        return __awaiter(this, void 0, void 0, function* () {
            filter[this.inventoriesQueryParamName] = this.queriesUtil.buildQuery(query);
            return _super.list.call(this, filter);
        });
    }
    /**
     * Gets the list of all devices filtered and sorted by given query.
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {object} filter Object containing filters for querying devices.
     *
     * @example
     * ```typescript
     *
     *  const filter: object = {
     *     pageSize: 100,
     *     withTotalPages: true
     *   };
     *
     *  const query = {
     *      name: 'MY-NAM*'
     *  }
     *
     *   (async () => {
     *     const {data, res, paging} = await inventoryService.listQueryDevices(query, filter);
     *   })();
     * ```
     */
    listQueryDevices(query, filter = {}) {
        const _super = Object.create(null, {
            list: { get: () => super.list }
        });
        return __awaiter(this, void 0, void 0, function* () {
            filter[this.devicesQueryParamName] = this.queriesUtil.buildQuery(query);
            return _super.list.call(this, filter);
        });
    }
    /**
     * Removes managed object with given id.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IdReference} managedObjectOrId ManagedObject or Id of the ManagedObject.
     *
     * @example
     * ```typescript
     *
     *   const managedObjectId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.delete(managedObjectId);
     *   })();
     * ```
     */
    delete(managedObjectOrId) {
        const _super = Object.create(null, {
            delete: { get: () => super.delete }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.delete.call(this, managedObjectOrId);
        });
    }
    /**
     * Gets a list of child additions from a given managed object (parent)
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {IdReference} parentReference
     * @param {object} filter
     *
     * @example
     * ```typescript
     *    const parentReferenceId: IdReference = 1;
     *
     *    const filter: object = {
     *      pageSize: 100,
     *      withTotalPages: true
     *    };
     *
     *    (async () => {
     *      const {data, res, paging} = await inventoryService.childAdditionsList(parentReferenceId, filter);
     *    })();
     * ```
     */
    childAdditionsList(parentReference, filter = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.listChildren(ChildType.ADDITIONS, parentReference, filter);
        });
    }
    /**
     * Creates a new managed object as child addition to another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {Partial<IManagedObject>} managedObject
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const mOAsChildAddition: Partial<IManagedObject> = {
     *      name: 'Child addition MO',
     *      type: 'new type',
     *      ...
     *    };
     *
     *    // This is the identifier of the managed object which should be the parent of
     *    // mOAsChildAddition, see above.
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childAdditionsCreate(mOAsChildAddition, parentReferenceId);
     *    })();
     * ```
     */
    childAdditionsCreate(managedObject, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.createChild(ChildType.ADDITIONS, managedObject, parentReference);
        });
    }
    /**
     * Adds an existing managed object as child addition to another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IdReference} childReference
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const childRef: number = 2;
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childAdditionsCreate(childRef, parentReferenceId);
     *    })();
     * ```
     */
    childAdditionsAdd(childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.addChild(ChildType.ADDITIONS, childReference, parentReference);
        });
    }
    /**
     * Removes an existing managed object as child addition from another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IdReference} childReference
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const childRef: number = 2;
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childAdditionsRemove(childRef, parentReferenceId);
     *    })();
     * ```
     */
    childAdditionsRemove(childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.removeChild(ChildType.ADDITIONS, childReference, parentReference);
        });
    }
    /**
     * Gets a list of child assets from a given managed object (parent)
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {IdReference} parentReference
     * @param {object} filter
     *
     * @example
     * ```typescript
     *
     *    const parentReferenceId: IdReference = 1;
     *
     *    const filter: object = {
     *      pageSize: 100,
     *      withTotalPages: true
     *    };
     *
     *    (async () => {
     *      const {data, res, paging} = await inventoryService.childAssetsList(parentReferenceId, filter);
     *    })();
     * ```
     */
    childAssetsList(parentReference, filter = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.listChildren(ChildType.ASSETS, parentReference, filter);
        });
    }
    /**
     * Creates a new managed object as child asset to another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {Partial<IManagedObject>} managedObject
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const mOAsChildAsset: Partial<IManagedObject> = {
     *      name: 'Child asset MO',
     *      type: 'new type',
     *      ...
     *    };
     *
     *    // This is the identifier of the managed object which should be the parent of
     *    // mOAsChildAsset, see above.
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childAdditionsCreate(mOAsChildAddition, parentReferenceId);
     *    })();
     * ```
     */
    childAssetsCreate(managedObject, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.createChild(ChildType.ASSETS, managedObject, parentReference);
        });
    }
    /**
     * Adds an existing managed object as child asset to another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IdReference} childReference
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const childRef: number = 2;
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childAssetsAdd(childRef, parentReferenceId);
     *    })();
     * ```
     */
    childAssetsAdd(childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.addChild(ChildType.ASSETS, childReference, parentReference);
        });
    }
    /**
     * Removes an existing managed object as child asset from another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IdReference} childReference
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const childRef: number = 2;
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childAssetsRemove(childRef, parentReferenceId);
     *    })();
     * ```
     */
    childAssetsRemove(childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.removeChild(ChildType.ASSETS, childReference, parentReference);
        });
    }
    /**
     * Gets a list of child devices from a given managed object (parent)
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {IdReference} parentReference
     * @param {object} filter
     *
     * @example
     * ```typescript
     *
     *    const parentReferenceId: IdReference = 1;
     *
     *    const filter: object = {
     *      pageSize: 100,
     *      withTotalPages: true
     *    };
     *
     *    (async () => {
     *      const {data, res, paging} = await inventoryService.childDevicesList(parentReferenceId, filter);
     *    })();
     * ```
     */
    childDevicesList(parentReference, filter = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.listChildren(ChildType.DEVICES, parentReference, filter);
        });
    }
    /**
     * Creates a new managed object as child device to another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {Partial<IManagedObject>} managedObject
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const mOAsChildDevice: Partial<IManagedObject> = {
     *      name: 'Child device MO',
     *      type: 'new type',
     *      ...
     *    };
     *
     *    // This is the identifier of the managed object which should be the parent of
     *    // mOAsChildDevice, see above.
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childDevicesCreate(mOAsChildDevice, parentReferenceId);
     *    })();
     * ```
     */
    childDevicesCreate(managedObject, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.createChild(ChildType.DEVICES, managedObject, parentReference);
        });
    }
    /**
     * Adds an existing managed object as child device to another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IdReference} childReference
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const childRef: number = 2;
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childDevicesAdd(childRef, parentReferenceId);
     *    })();
     * ```
     */
    childDevicesAdd(childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.addChild(ChildType.DEVICES, childReference, parentReference);
        });
    }
    /**
     * Removes an existing managed object as child device from another managed object (parent)
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IdReference} childReference
     * @param {IdReference} parentReference
     *
     * @example
     * ```typescript
     *
     *    const childRef: number = 2;
     *    const parentReferenceId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await inventoryService.childDevicesRemove(childRef, parentReferenceId);
     *    })();
     * ```
     */
    childDevicesRemove(childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.removeChild(ChildType.DEVICES, childReference, parentReference);
        });
    }
    onBeforeUpdate(objWithId) {
        delete objWithId.lastUpdated;
        return objWithId;
    }
    onBeforeCreate(managedObject) {
        delete managedObject.id;
        delete managedObject.lastUpdated;
        return managedObject;
    }
    getChildrenUrl(type, parentReference) {
        return `${this.getDetailUrl(parentReference)}/${type}`;
    }
    getChildUrl(type, childReference, parentReference) {
        const typeofChildReference = typeof childReference;
        const childId = this.getIdString(childReference);
        return `${this.getChildrenUrl(type, parentReference)}/${childId}`;
    }
    listChildren(type, parentReference, filter = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const headers = { 'content-type': 'application/json' };
            const url = this.getChildrenUrl(type, parentReference);
            const res = yield this.fetch(url, { headers, params: filter });
            const json = yield res.json();
            const data = json.references.map(ref => ref.managedObject);
            const paging = this.getPaging(json, filter);
            return { res, data, paging };
        });
    }
    createChild(type, managedObject, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = this.getChildrenUrl(type, parentReference);
            const method = 'POST';
            const body = JSON.stringify(this.onBeforeCreate(managedObject));
            const headers = { 'content-type': this.mimeType('managedObject'), accept: 'application/json' };
            const res = yield this.fetch(url, { method, body, headers });
            const data = yield res.json();
            return { res, data };
        });
    }
    addChild(type, childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = this.getChildrenUrl(type, parentReference);
            const method = 'POST';
            const childId = this.getIdString(childReference);
            const body = JSON.stringify({ managedObject: { id: String(childId) } });
            const headers = {
                accept: 'application/json',
                'content-type': this.mimeType('managedObjectReference')
            };
            const res = yield this.fetch(url, { method, body, headers });
            let data = yield res.json();
            data = data.managedObject;
            return { res, data };
        });
    }
    removeChild(type, childReference, parentReference) {
        return __awaiter(this, void 0, void 0, function* () {
            const childId = this.getIdString(childReference);
            const url = `${this.getChildrenUrl(type, parentReference)}/${childId}`;
            const method = 'DELETE';
            const headers = { accept: 'application/json' };
            const res = yield this.fetch(url, { method, headers });
            const data = null;
            return { res, data };
        });
    }
}
