// @c8y/ngx-components/fesm2015/c8y-ngx-components-protocol-impact.js
import { __decorate, __awaiter } from 'tslib';
import { Injectable, ViewContainerRef, ViewChild, TemplateRef, Component, NgModule } from '@angular/core';
import { gettext, AlertService, CoreModule, HOOK_ACTION_BAR } from '@c8y/ngx-components';
import { FetchClient, ApplicationService } from '@c8y/client';
import { get } from 'lodash-es';
import { Router } from '@angular/router';

let ImpactSubscriptionService = class ImpactSubscriptionService {
    constructor(client) {
        this.client = client;
        this.microserviceUrl = '/service/impact';
        this.header = { 'Content-Type': 'application/json' };
    }
    refreshDeviceResources(deviceId) {
        const options = {
            method: 'PUT',
            headers: this.header
        };
        return this.client.fetch(`${this.microserviceUrl}/refresh/${deviceId}`, options);
    }
};
ImpactSubscriptionService.ctorParameters = () => [
    { type: FetchClient }
];
ImpactSubscriptionService = __decorate([
    Injectable()
], ImpactSubscriptionService);

let RefreshActionComponent = class RefreshActionComponent {
    constructor(vcRef, router, impactService, alert) {
        this.vcRef = vcRef;
        this.router = router;
        this.impactService = impactService;
        this.alert = alert;
    }
    ngOnInit() {
        this.vcRef.createEmbeddedView(this.templateCopy);
    }
    refresh() {
        return __awaiter(this, void 0, void 0, function* () {
            // TODO This is only a dirty hack to retrieve deviceId from URL;
            // In fact contextData should be provided for this component by a resolver?
            const url = this.router &&
                this.router.routerState &&
                this.router.routerState.snapshot &&
                this.router.routerState.snapshot.url;
            const deviceId = url && (/^\/device\/(\d+)\/.*$/gi.exec(url) || [])[1];
            if (deviceId) {
                this.requestInProgress = true;
                try {
                    const res = yield this.impactService.refreshDeviceResources(deviceId);
                    if (res && res.status !== 200) {
                        const data = res.json ? yield res.json() : undefined;
                        this.alert.addServerFailure({ data, res });
                    }
                    else {
                        this.alert.success(gettext('Device resource refresh scheduled.'));
                    }
                }
                catch (ex) {
                    this.alert.addServerFailure(ex);
                }
                this.requestInProgress = false;
            }
            else {
                this.alert.danger(gettext('Could not find device ID in URL.'));
            }
        });
    }
};
RefreshActionComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Router },
    { type: ImpactSubscriptionService },
    { type: AlertService }
];
__decorate([
    ViewChild('templateCopy', { read: TemplateRef, static: true })
], RefreshActionComponent.prototype, "templateCopy", void 0);
RefreshActionComponent = __decorate([
    Component({
        selector: 'c8y-impact-refresh-action',
        template: "<ng-template #templateCopy>\n  <li>\n    <button\n      class=\"btn btn-link\"\n      title=\"{{ 'Refresh device resources' | translate }}\"\n      (click)=\"refresh()\"\n    >\n      <i class=\"fa fa-refresh\" [ngClass]=\"{ 'fa-spin': requestInProgress }\"></i>&nbsp;<span\n        translate\n        >Refresh device resources</span\n      >\n    </button>\n  </li>\n</ng-template>\n"
    })
], RefreshActionComponent);

var ImpactActionFactory_1;
let ImpactActionFactory = ImpactActionFactory_1 = class ImpactActionFactory {
    constructor(applicationService) {
        this.applicationService = applicationService;
    }
    get(activeRoute) {
        return __awaiter(this, void 0, void 0, function* () {
            const actions = [];
            const data = !activeRoute.parent || activeRoute.snapshot.data.context
                ? activeRoute.snapshot.data
                : activeRoute.parent.snapshot.data;
            const { contextData } = data;
            const isDeviceInfoTab = get(activeRoute, 'snapshot.url[0].path') === 'device-info';
            const showRefreshActionButton = isDeviceInfoTab &&
                contextData &&
                contextData.c8y_ImpactResourceInfo &&
                /* call application service only for relevant devices to reduce number of service calls! */
                (yield this.applicationService.isAvailable(ImpactActionFactory_1.applicationName)).data;
            if (showRefreshActionButton) {
                actions.push({
                    priority: 500,
                    placement: 'right',
                    template: RefreshActionComponent
                });
            }
            return actions;
        });
    }
};
ImpactActionFactory.applicationName = 'impact';
ImpactActionFactory.ctorParameters = () => [
    { type: ApplicationService }
];
ImpactActionFactory = ImpactActionFactory_1 = __decorate([
    Injectable()
], ImpactActionFactory);

let ImpactProtocolModule = class ImpactProtocolModule {
};
ImpactProtocolModule = __decorate([
    NgModule({
        declarations: [RefreshActionComponent],
        imports: [CoreModule],
        providers: [
            ImpactSubscriptionService,
            { provide: HOOK_ACTION_BAR, useClass: ImpactActionFactory, multi: true }
        ],
        entryComponents: [RefreshActionComponent]
    })
], ImpactProtocolModule);

/**
 * Generated bundle index. Do not edit.
 */

export { ImpactProtocolModule, RefreshActionComponent as ɵa, ImpactSubscriptionService as ɵb, ImpactActionFactory as ɵc };
