// @c8y/ngx-components/esm5/sms-gateway/sms-gateway-navigation.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { ApplicationService } from '@c8y/client';
var SmsGatewayNavigationFactory = /** @class */ (function () {
    function SmsGatewayNavigationFactory(applicationService) {
        this.applicationService = applicationService;
    }
    SmsGatewayNavigationFactory_1 = SmsGatewayNavigationFactory;
    SmsGatewayNavigationFactory.prototype.get = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var showSMSProvider;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.nodeItem) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.applicationService.isAvailable(SmsGatewayNavigationFactory_1.applicationName)];
                    case 1:
                        showSMSProvider = (_a.sent()).data;
                        if (showSMSProvider) {
                            this.nodeItem = new NavigatorNode({
                                label: gettext('SMS provider'),
                                path: 'smsgateway',
                                icon: 'envelope-o',
                                parent: gettext('Settings'),
                                priority: 1000
                            });
                        }
                        _a.label = 2;
                    case 2: return [2 /*return*/, this.nodeItem];
                }
            });
        });
    };
    var SmsGatewayNavigationFactory_1;
    SmsGatewayNavigationFactory.applicationName = 'sms-gateway';
    SmsGatewayNavigationFactory.ctorParameters = function () { return [
        { type: ApplicationService }
    ]; };
    SmsGatewayNavigationFactory = SmsGatewayNavigationFactory_1 = tslib_1.__decorate([
        Injectable()
    ], SmsGatewayNavigationFactory);
    return SmsGatewayNavigationFactory;
}());
export { SmsGatewayNavigationFactory };
