// @c8y/ngx-components/esm5/core/router/context-route.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { ContextRouteService } from './context-route.service';
var ContextRouteComponent = /** @class */ (function () {
    function ContextRouteComponent(route, contextRouteService) {
        this.route = route;
        this.contextRouteService = contextRouteService;
    }
    ContextRouteComponent.prototype.ngOnInit = function () {
        this.contextRouteService.init(this.route);
    };
    ContextRouteComponent.prototype.ngOnDestroy = function () {
        this.contextRouteService.destroy();
    };
    ContextRouteComponent.prototype.refreshTabs = function () {
        this.contextRouteService.refreshContext();
    };
    ContextRouteComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: ContextRouteService }
    ]; };
    ContextRouteComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-context-route',
            template: "<router-outlet></router-outlet>\n"
        })
    ], ContextRouteComponent);
    return ContextRouteComponent;
}());
export { ContextRouteComponent };
