// @c8y/ngx-components/esm5/core/router/context-route.guard.js
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { from, of } from 'rxjs';
import { catchError, map, shareReplay, tap } from 'rxjs/operators';
import { ViewContextServices } from './view-context.service';
import { InventoryService } from '../../api';
var ContextRouteGuard = /** @class */ (function () {
    function ContextRouteGuard(injector) {
        this.injector = injector;
        this.caches = new Map();
    }
    ContextRouteGuard.prototype.canActivate = function (route) {
        var id = route.paramMap.get('id');
        var service = this.service(route);
        this.clearDetailCached(service, id);
        return this.detailCached(service, id).pipe(tap(function (contextData) { return (route.data = tslib_1.__assign({}, route.data, { contextData: contextData })); }), map(function () { return true; }), catchError(function (ex) { return of(false); }));
    };
    ContextRouteGuard.prototype.service = function (route) {
        var serviceInstance = ViewContextServices.contextToService(route.data.context);
        return this.injector.get(serviceInstance);
    };
    ContextRouteGuard.prototype.detailCached = function (service, id) {
        var cache = this.getCache(service);
        var cached = cache[id];
        if (!cached) {
            var params = service instanceof InventoryService ? { withChildren: false } : undefined;
            cached = from(service.detail(id, params)).pipe(map(function (_a) {
                var data = _a.data;
                return data;
            }), shareReplay(1));
            cache[id] = cached;
        }
        return cached;
    };
    ContextRouteGuard.prototype.clearDetailCached = function (service, id) {
        var cache = this.getCache(service);
        delete cache[id];
    };
    ContextRouteGuard.prototype.getCache = function (service) {
        var cache = this.caches.get(service);
        if (!cache) {
            cache = {};
            this.caches.set(service, cache);
        }
        return cache;
    };
    ContextRouteGuard.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    ContextRouteGuard = tslib_1.__decorate([
        Injectable()
    ], ContextRouteGuard);
    return ContextRouteGuard;
}());
export { ContextRouteGuard };
