// @c8y/ngx-components/esm2015/core/action/action.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { ActionService } from './action.service';
let ActionComponent = class ActionComponent {
    constructor(actionService) {
        this.actionService = actionService;
        this.disabled = false;
        this.priority = 0;
    }
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.actionService.add(this));
    }
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.actionService.remove(this);
    }
};
ActionComponent.ctorParameters = () => [
    { type: ActionService }
];
tslib_1.__decorate([
    Input()
], ActionComponent.prototype, "disabled", void 0);
tslib_1.__decorate([
    Input()
], ActionComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Input()
], ActionComponent.prototype, "priority", void 0);
tslib_1.__decorate([
    Input()
], ActionComponent.prototype, "icon", void 0);
tslib_1.__decorate([
    ViewChild('template', { static: false })
], ActionComponent.prototype, "template", void 0);
ActionComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-action',
        template: "<ng-template #template>\n  <li>\n    <ng-content></ng-content>\n  </li>\n</ng-template>"
    })
], ActionComponent);
export { ActionComponent };
