// @c8y/ngx-components/esm5/upgrade/ng1/downgraded.services.js
import { downgradeInjectable } from '@angular/upgrade/static';
import { AlertService, AppStateService, DocsService, HeaderService, PasswordService, UserMenuService } from '@c8y/ngx-components';
import { ApiService } from '@c8y/ngx-components/api';
import { BulkOperationDetailsService } from '@c8y/ngx-components/operations';
import { TranslateService } from '@ngx-translate/core';
import { BridgeService } from '../bridge.service';
import { ServerMessagesService } from './serverMessages.service';
export var bridgeService = downgradeInjectable(BridgeService);
export var appStateService = downgradeInjectable(AppStateService);
export var headerService = downgradeInjectable(HeaderService);
export var alertsService = downgradeInjectable(AlertService);
export var userMenuService = downgradeInjectable(UserMenuService);
export var apiService = downgradeInjectable(ApiService);
export var docsService = downgradeInjectable(DocsService);
export var passwordService = downgradeInjectable(PasswordService);
export var translateService = downgradeInjectable(TranslateService);
export var serverMessagesService = downgradeInjectable(ServerMessagesService);
export var bulkOperationDetailsService = downgradeInjectable(BulkOperationDetailsService);
