// @c8y/ngx-components/esm5/upgrade/ng1/http.interceptor.js
function c8yNg1HttpInterceptor($q, c8yLoadingIndicator, c8yApiService) {
    'ngInject';
    function request(config) {
        var url = config.url, method = config.method;
        c8yApiService.onStart({ url: url, method: method, options: config });
        return config;
    }
    function requestError(rejection) {
        finishRequest(rejection);
        return $q.reject(rejection);
    }
    function response(res) {
        finishRequest(res);
        return res;
    }
    function responseError(rejection) {
        finishRequest(rejection);
        c8yLoadingIndicator.responseError(rejection);
        return $q.reject(rejection);
    }
    function finishRequest(res) {
        var _a = res.config, url = _a.url, method = _a.method;
        c8yApiService.onFinish({
            url: url,
            method: method,
            response: res,
            options: res.config
        });
    }
    return {
        request: request,
        requestError: requestError,
        response: response,
        responseError: responseError
    };
}
export { c8yNg1HttpInterceptor };
