// @c8y/ngx-components/esm5/operations/factories/tabs.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
var TabsFactory = /** @class */ (function () {
    function TabsFactory(router) {
        this.router = router;
    }
    TabsFactory.prototype.get = function () {
        var tabs = [];
        if (this.router.url.match(/devicecontrol/g)) {
            tabs.push({
                path: 'devicecontrol/single',
                label: gettext('Single operations'),
                icon: 'c8y-icon c8y-icon-device-control',
                priority: 1100,
                orientation: 'horizontal'
            });
            tabs.push({
                path: 'devicecontrol/bulk',
                label: gettext('Bulk operations'),
                icon: 'c8y-icon c8y-icon-energy',
                priority: 1050,
                orientation: 'horizontal'
            });
        }
        return tabs;
    };
    TabsFactory.ctorParameters = function () { return [
        { type: Router }
    ]; };
    TabsFactory = tslib_1.__decorate([
        Injectable()
    ], TabsFactory);
    return TabsFactory;
}());
export { TabsFactory };
