// @c8y/ngx-components/esm5/device-lists/device-lists.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, HOOK_NAVIGATOR_NODES, CoreModule } from '@c8y/ngx-components';
import { DeviceListsNavigationFactory } from './device-lists-navigation.factory';
import { DeviceGridExampleGuard } from './device-grid-example-guard';
import { DeviceGridExampleComponent } from './device-grid-example.component';
var routes = [
    {
        path: 'grids/device-grid',
        component: DeviceGridExampleComponent,
        canActivate: [DeviceGridExampleGuard]
    }
];
var DeviceListsModule = /** @class */ (function () {
    function DeviceListsModule() {
    }
    DeviceListsModule = tslib_1.__decorate([
        NgModule({
            imports: [CoreModule, CommonModule, RouterModule.forRoot(routes, { useHash: true })],
            declarations: [DeviceGridExampleComponent],
            entryComponents: [DeviceGridExampleComponent],
            exports: [],
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: DeviceListsNavigationFactory,
                    multi: true
                },
                DeviceGridExampleGuard
            ]
        })
    ], DeviceListsModule);
    return DeviceListsModule;
}());
export { DeviceListsModule };
