// @c8y/ngx-components/esm2015/upgrade/dashboard/copy-action.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
let CopyActionComponent = class CopyActionComponent {
    constructor(vcRef, dashboardBridgeService) {
        this.vcRef = vcRef;
        this.dashboardBridgeService = dashboardBridgeService;
    }
    ngOnInit() {
        this.vcRef.createEmbeddedView(this.templateCopy);
    }
    copyDashboard() {
        this.dashboardBridgeService.copyDashboard();
    }
};
CopyActionComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: DashboardBridgeService }
];
tslib_1.__decorate([
    ViewChild('templateCopy', { read: TemplateRef, static: true })
], CopyActionComponent.prototype, "templateCopy", void 0);
CopyActionComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-copy-action-legacy',
        template: `
    <ng-template #templateCopy>
      <li>
        <button title="{{'Copy dashboard' | translate}}" (click)="copyDashboard()">
          <i class="fa fw fa-clone"></i>&nbsp;<span translate>Copy dashboard</span>
        </button>
      </li>
    </ng-template>
  `
    })
], CopyActionComponent);
export { CopyActionComponent };
