// @c8y/client/lib/esm2015/src/operation/OperationBulkService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Service } from '../core';
/**
 * This class allows for managing bulk operations.
 */
export class OperationBulkService extends Service {
    constructor() {
        super(...arguments);
        this.baseUrl = 'devicecontrol';
        this.listUrl = 'bulkoperations';
        this.propertyName = 'bulkOperations';
    }
    /**
     * Gets the details of selected bulk operation.
     *
     * @param {string|number|IIdentified} entityOrId Entity or Id of the entity.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *
     *    const bulkOperationId: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await operationBulkService.detail(bulkOperationId);
     *   })();
     * ```
     */
    detail(operationOrId) {
        const _super = Object.create(null, {
            detail: { get: () => super.detail }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.detail.call(this, operationOrId);
        });
    }
    /**
     * Creates a new operation.
     *
     * @param {Partial<IOperationBulk>} operation Operation object with mandantory fragments.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *
     *  const mandantoryObject: Partial<IOperationBulk> = {
     *    creationRamp: 15,
     *    groupId: '149044',
     *    operationPrototype: {
     *      c8y_Restart: {},
     *      description: 'Restart device',
     *      deviceId: device.id,
     *      status: 'PENDING'
     *    },
     *     startDate: '2018-02-15T16:01:00.000Z'
     *  };
     *
     *  (async () => {
     *    const {data, res} = await operationBulkService.create(mandantoryObject);
     *  })();
     * ```
     */
    create(operation) {
        const _super = Object.create(null, {
            create: { get: () => super.create }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.create.call(this, operation);
        });
    }
    /**
     * Updates a new operation.
     *
     * @param {Partial<IOperationBulk>} operation Operation object.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *
     *  const updateObject: Partial<IOperationBulk> = {
     *    creationRamp: 15,
     *    groupId: '149044',
     *    operationPrototype: {
     *      c8y_Restart: {},
     *      description: 'Restart device',
     *      deviceId: device.id,
     *      status: 'PENDING'
     *    },
     *     startDate: '2018-02-15T16:01:00.000Z'
     *  };
     *
     *  (async () => {
     *    const {data, res} = await operationBulkService.update(updateObject);
     *  })();
     * ```
     */
    update(entity) {
        const _super = Object.create(null, {
            update: { get: () => super.update }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.update.call(this, entity);
        });
    }
    /**
     * Gets the list of bulk operations filtered by parameters.
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {object} filter Object containing filters for querying alarms.
     *
     * @example
     * ```typescript
     *
     *  const filter: object = {
     *     pageSize: 100,
     *     withTotalPages: true
     *   };
     *
     *   (async () => {
     *     const {data, res, paging} = await operationBulkService.list(filter);
     *   })();
     * ```
     */
    list(filter = {}) {
        const _super = Object.create(null, {
            list: { get: () => super.list }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.list.call(this, filter);
        });
    }
    /**
     * Gets a list as observable.
     *
     * @param {object} filter Object containing filters for querying
     * @param {object} options To configure the observable
     *
     * @returns Data wrapped as Observable.
     *
     * @example
     * ```typescript
     *
     *   const list$ = operationBulkService.list$();
     *   list$.subscribe((data) => console.log(data));
     * ```
     *
     * @deprecated This function will be removed soon in favor of decoupling
     * this library from rxjs. Use [[Realtime]] for realtime connection and
     * rxjs `defer()` or `from()` to composite an observable on your own.
     */
    list$(filter = {}, options = {}) {
        return super.list$(filter, options);
    }
    /**
     * Removes an bulk operation with given id.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {string | number | IIdentified):} operationOrId Operation object or id.
     *
     * @example
     * ```typescript
     *
     *   const id: number = 1;
     *
     *    (async () => {
     *      const {data, res} = await operationBulkService.delete(id);
     *   })();
     * ```
     */
    delete(operationOrId) {
        const _super = Object.create(null, {
            delete: { get: () => super.delete }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.delete.call(this, operationOrId);
        });
    }
}
