// @c8y/ngx-components/esm5/protocol-impact/refresh-action.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { AlertService, gettext } from '@c8y/ngx-components';
import { ImpactSubscriptionService } from './impact-subscription.service';
var RefreshActionComponent = /** @class */ (function () {
    function RefreshActionComponent(vcRef, router, impactService, alert) {
        this.vcRef = vcRef;
        this.router = router;
        this.impactService = impactService;
        this.alert = alert;
    }
    RefreshActionComponent.prototype.ngOnInit = function () {
        this.vcRef.createEmbeddedView(this.templateCopy);
    };
    RefreshActionComponent.prototype.refresh = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var url, deviceId, res, data, _a, ex_1;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        url = this.router &&
                            this.router.routerState &&
                            this.router.routerState.snapshot &&
                            this.router.routerState.snapshot.url;
                        deviceId = url && (/^\/device\/(\d+)\/.*$/gi.exec(url) || [])[1];
                        if (!deviceId) return [3 /*break*/, 10];
                        this.requestInProgress = true;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 8, , 9]);
                        return [4 /*yield*/, this.impactService.refreshDeviceResources(deviceId)];
                    case 2:
                        res = _b.sent();
                        if (!(res && res.status !== 200)) return [3 /*break*/, 6];
                        if (!res.json) return [3 /*break*/, 4];
                        return [4 /*yield*/, res.json()];
                    case 3:
                        _a = _b.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        _a = undefined;
                        _b.label = 5;
                    case 5:
                        data = _a;
                        this.alert.addServerFailure({ data: data, res: res });
                        return [3 /*break*/, 7];
                    case 6:
                        this.alert.success(gettext('Device resource refresh scheduled.'));
                        _b.label = 7;
                    case 7: return [3 /*break*/, 9];
                    case 8:
                        ex_1 = _b.sent();
                        this.alert.addServerFailure(ex_1);
                        return [3 /*break*/, 9];
                    case 9:
                        this.requestInProgress = false;
                        return [3 /*break*/, 11];
                    case 10:
                        this.alert.danger(gettext('Could not find device ID in URL.'));
                        _b.label = 11;
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    RefreshActionComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: Router },
        { type: ImpactSubscriptionService },
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        ViewChild('templateCopy', { read: TemplateRef, static: true })
    ], RefreshActionComponent.prototype, "templateCopy", void 0);
    RefreshActionComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-impact-refresh-action',
            template: "<ng-template #templateCopy>\n  <li>\n    <button\n      class=\"btn btn-link\"\n      title=\"{{ 'Refresh device resources' | translate }}\"\n      (click)=\"refresh()\"\n    >\n      <i class=\"fa fa-refresh\" [ngClass]=\"{ 'fa-spin': requestInProgress }\"></i>&nbsp;<span\n        translate\n        >Refresh device resources</span\n      >\n    </button>\n  </li>\n</ng-template>\n"
        })
    ], RefreshActionComponent);
    return RefreshActionComponent;
}());
export { RefreshActionComponent };
