// ngx-bootstrap/bundles/ngx-bootstrap.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('ngx-bootstrap/chronos'), require('ngx-bootstrap/utils'), require('ngx-bootstrap/accordion'), require('ngx-bootstrap/alert'), require('ngx-bootstrap/buttons'), require('ngx-bootstrap/carousel'), require('ngx-bootstrap/collapse'), require('ngx-bootstrap/datepicker'), require('ngx-bootstrap/modal'), require('ngx-bootstrap/dropdown'), require('ngx-bootstrap/pagination'), require('ngx-bootstrap/progressbar'), require('ngx-bootstrap/rating'), require('ngx-bootstrap/sortable'), require('ngx-bootstrap/tabs'), require('ngx-bootstrap/timepicker'), require('ngx-bootstrap/tooltip'), require('ngx-bootstrap/typeahead'), require('ngx-bootstrap/popover'), require('ngx-bootstrap/component-loader'), require('ngx-bootstrap/positioning'), require('ngx-bootstrap/mini-ngrx'), require('ngx-bootstrap/locale')) :
	typeof define === 'function' && define.amd ? define('ngx-bootstrap', ['exports', 'ngx-bootstrap/chronos', 'ngx-bootstrap/utils', 'ngx-bootstrap/accordion', 'ngx-bootstrap/alert', 'ngx-bootstrap/buttons', 'ngx-bootstrap/carousel', 'ngx-bootstrap/collapse', 'ngx-bootstrap/datepicker', 'ngx-bootstrap/modal', 'ngx-bootstrap/dropdown', 'ngx-bootstrap/pagination', 'ngx-bootstrap/progressbar', 'ngx-bootstrap/rating', 'ngx-bootstrap/sortable', 'ngx-bootstrap/tabs', 'ngx-bootstrap/timepicker', 'ngx-bootstrap/tooltip', 'ngx-bootstrap/typeahead', 'ngx-bootstrap/popover', 'ngx-bootstrap/component-loader', 'ngx-bootstrap/positioning', 'ngx-bootstrap/mini-ngrx', 'ngx-bootstrap/locale'], factory) :
	(global = global || self, factory(global['ngx-bootstrap'] = {}, global.chronos, global.utils, global.accordion, global.alert, global.buttons, global.carousel, global.collapse, global.datepicker, global.modal, global.dropdown, global.pagination, global.progressbar, global.rating, global.sortable, global.tabs, global.timepicker, global.tooltip, global.typeahead, global.popover, global.componentLoader, global.positioning, global.miniNgrx, global.locale));
}(this, function (exports, chronos, utils, accordion, alert, buttons, carousel, collapse, datepicker, modal, dropdown, pagination, progressbar, rating, sortable, tabs, timepicker, tooltip, typeahead, popover, componentLoader, positioning, miniNgrx, locale) { 'use strict';

	Object.defineProperty(exports, 'defineLocale', {
		enumerable: true,
		get: function () {
			return chronos.defineLocale;
		}
	});
	Object.defineProperty(exports, 'endOf', {
		enumerable: true,
		get: function () {
			return chronos.endOf;
		}
	});
	Object.defineProperty(exports, 'formatDate', {
		enumerable: true,
		get: function () {
			return chronos.formatDate;
		}
	});
	Object.defineProperty(exports, 'getDay', {
		enumerable: true,
		get: function () {
			return chronos.getDay;
		}
	});
	Object.defineProperty(exports, 'getFirstDayOfMonth', {
		enumerable: true,
		get: function () {
			return chronos.getFirstDayOfMonth;
		}
	});
	Object.defineProperty(exports, 'getFullYear', {
		enumerable: true,
		get: function () {
			return chronos.getFullYear;
		}
	});
	Object.defineProperty(exports, 'getLocale', {
		enumerable: true,
		get: function () {
			return chronos.getLocale;
		}
	});
	Object.defineProperty(exports, 'getMonth', {
		enumerable: true,
		get: function () {
			return chronos.getMonth;
		}
	});
	Object.defineProperty(exports, 'getSetGlobalLocale', {
		enumerable: true,
		get: function () {
			return chronos.getSetGlobalLocale;
		}
	});
	Object.defineProperty(exports, 'isAfter', {
		enumerable: true,
		get: function () {
			return chronos.isAfter;
		}
	});
	Object.defineProperty(exports, 'isArray', {
		enumerable: true,
		get: function () {
			return chronos.isArray;
		}
	});
	Object.defineProperty(exports, 'isBefore', {
		enumerable: true,
		get: function () {
			return chronos.isBefore;
		}
	});
	Object.defineProperty(exports, 'isDate', {
		enumerable: true,
		get: function () {
			return chronos.isDate;
		}
	});
	Object.defineProperty(exports, 'isDateValid', {
		enumerable: true,
		get: function () {
			return chronos.isDateValid;
		}
	});
	Object.defineProperty(exports, 'isFirstDayOfWeek', {
		enumerable: true,
		get: function () {
			return chronos.isFirstDayOfWeek;
		}
	});
	Object.defineProperty(exports, 'isSameDay', {
		enumerable: true,
		get: function () {
			return chronos.isSameDay;
		}
	});
	Object.defineProperty(exports, 'isSameMonth', {
		enumerable: true,
		get: function () {
			return chronos.isSameMonth;
		}
	});
	Object.defineProperty(exports, 'isSameYear', {
		enumerable: true,
		get: function () {
			return chronos.isSameYear;
		}
	});
	Object.defineProperty(exports, 'listLocales', {
		enumerable: true,
		get: function () {
			return chronos.listLocales;
		}
	});
	Object.defineProperty(exports, 'parseDate', {
		enumerable: true,
		get: function () {
			return chronos.parseDate;
		}
	});
	Object.defineProperty(exports, 'setFullDate', {
		enumerable: true,
		get: function () {
			return chronos.setFullDate;
		}
	});
	Object.defineProperty(exports, 'shiftDate', {
		enumerable: true,
		get: function () {
			return chronos.shiftDate;
		}
	});
	Object.defineProperty(exports, 'startOf', {
		enumerable: true,
		get: function () {
			return chronos.startOf;
		}
	});
	Object.defineProperty(exports, 'updateLocale', {
		enumerable: true,
		get: function () {
			return chronos.updateLocale;
		}
	});
	Object.defineProperty(exports, 'LinkedList', {
		enumerable: true,
		get: function () {
			return utils.LinkedList;
		}
	});
	Object.defineProperty(exports, 'OnChange', {
		enumerable: true,
		get: function () {
			return utils.OnChange;
		}
	});
	Object.defineProperty(exports, 'Trigger', {
		enumerable: true,
		get: function () {
			return utils.Trigger;
		}
	});
	Object.defineProperty(exports, 'Utils', {
		enumerable: true,
		get: function () {
			return utils.Utils;
		}
	});
	Object.defineProperty(exports, 'document', {
		enumerable: true,
		get: function () {
			return utils.document;
		}
	});
	Object.defineProperty(exports, 'isBs3', {
		enumerable: true,
		get: function () {
			return utils.isBs3;
		}
	});
	Object.defineProperty(exports, 'listenToTriggersV2', {
		enumerable: true,
		get: function () {
			return utils.listenToTriggersV2;
		}
	});
	Object.defineProperty(exports, 'parseTriggers', {
		enumerable: true,
		get: function () {
			return utils.parseTriggers;
		}
	});
	Object.defineProperty(exports, 'registerOutsideClick', {
		enumerable: true,
		get: function () {
			return utils.registerOutsideClick;
		}
	});
	Object.defineProperty(exports, 'setTheme', {
		enumerable: true,
		get: function () {
			return utils.setTheme;
		}
	});
	Object.defineProperty(exports, 'warnOnce', {
		enumerable: true,
		get: function () {
			return utils.warnOnce;
		}
	});
	Object.defineProperty(exports, 'window', {
		enumerable: true,
		get: function () {
			return utils.window;
		}
	});
	Object.defineProperty(exports, 'AccordionComponent', {
		enumerable: true,
		get: function () {
			return accordion.AccordionComponent;
		}
	});
	Object.defineProperty(exports, 'AccordionConfig', {
		enumerable: true,
		get: function () {
			return accordion.AccordionConfig;
		}
	});
	Object.defineProperty(exports, 'AccordionModule', {
		enumerable: true,
		get: function () {
			return accordion.AccordionModule;
		}
	});
	Object.defineProperty(exports, 'AccordionPanelComponent', {
		enumerable: true,
		get: function () {
			return accordion.AccordionPanelComponent;
		}
	});
	Object.defineProperty(exports, 'AlertComponent', {
		enumerable: true,
		get: function () {
			return alert.AlertComponent;
		}
	});
	Object.defineProperty(exports, 'AlertConfig', {
		enumerable: true,
		get: function () {
			return alert.AlertConfig;
		}
	});
	Object.defineProperty(exports, 'AlertModule', {
		enumerable: true,
		get: function () {
			return alert.AlertModule;
		}
	});
	Object.defineProperty(exports, 'ButtonCheckboxDirective', {
		enumerable: true,
		get: function () {
			return buttons.ButtonCheckboxDirective;
		}
	});
	Object.defineProperty(exports, 'ButtonRadioDirective', {
		enumerable: true,
		get: function () {
			return buttons.ButtonRadioDirective;
		}
	});
	Object.defineProperty(exports, 'ButtonRadioGroupDirective', {
		enumerable: true,
		get: function () {
			return buttons.ButtonRadioGroupDirective;
		}
	});
	Object.defineProperty(exports, 'ButtonsModule', {
		enumerable: true,
		get: function () {
			return buttons.ButtonsModule;
		}
	});
	Object.defineProperty(exports, 'CarouselComponent', {
		enumerable: true,
		get: function () {
			return carousel.CarouselComponent;
		}
	});
	Object.defineProperty(exports, 'CarouselConfig', {
		enumerable: true,
		get: function () {
			return carousel.CarouselConfig;
		}
	});
	Object.defineProperty(exports, 'CarouselModule', {
		enumerable: true,
		get: function () {
			return carousel.CarouselModule;
		}
	});
	Object.defineProperty(exports, 'SlideComponent', {
		enumerable: true,
		get: function () {
			return carousel.SlideComponent;
		}
	});
	Object.defineProperty(exports, 'CollapseDirective', {
		enumerable: true,
		get: function () {
			return collapse.CollapseDirective;
		}
	});
	Object.defineProperty(exports, 'CollapseModule', {
		enumerable: true,
		get: function () {
			return collapse.CollapseModule;
		}
	});
	Object.defineProperty(exports, 'BsDatepickerConfig', {
		enumerable: true,
		get: function () {
			return datepicker.BsDatepickerConfig;
		}
	});
	Object.defineProperty(exports, 'BsDatepickerContainerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.BsDatepickerContainerComponent;
		}
	});
	Object.defineProperty(exports, 'BsDatepickerDirective', {
		enumerable: true,
		get: function () {
			return datepicker.BsDatepickerDirective;
		}
	});
	Object.defineProperty(exports, 'BsDatepickerInlineContainerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.BsDatepickerInlineContainerComponent;
		}
	});
	Object.defineProperty(exports, 'BsDatepickerInlineDirective', {
		enumerable: true,
		get: function () {
			return datepicker.BsDatepickerInlineDirective;
		}
	});
	Object.defineProperty(exports, 'BsDatepickerInputDirective', {
		enumerable: true,
		get: function () {
			return datepicker.BsDatepickerInputDirective;
		}
	});
	Object.defineProperty(exports, 'BsDatepickerModule', {
		enumerable: true,
		get: function () {
			return datepicker.BsDatepickerModule;
		}
	});
	Object.defineProperty(exports, 'BsDaterangepickerConfig', {
		enumerable: true,
		get: function () {
			return datepicker.BsDaterangepickerConfig;
		}
	});
	Object.defineProperty(exports, 'BsDaterangepickerContainerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.BsDaterangepickerContainerComponent;
		}
	});
	Object.defineProperty(exports, 'BsDaterangepickerDirective', {
		enumerable: true,
		get: function () {
			return datepicker.BsDaterangepickerDirective;
		}
	});
	Object.defineProperty(exports, 'BsDaterangepickerInputDirective', {
		enumerable: true,
		get: function () {
			return datepicker.BsDaterangepickerInputDirective;
		}
	});
	Object.defineProperty(exports, 'BsLocaleService', {
		enumerable: true,
		get: function () {
			return datepicker.BsLocaleService;
		}
	});
	Object.defineProperty(exports, 'DateFormatter', {
		enumerable: true,
		get: function () {
			return datepicker.DateFormatter;
		}
	});
	Object.defineProperty(exports, 'DatePickerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.DatePickerComponent;
		}
	});
	Object.defineProperty(exports, 'DatePickerInnerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.DatePickerInnerComponent;
		}
	});
	Object.defineProperty(exports, 'DatepickerConfig', {
		enumerable: true,
		get: function () {
			return datepicker.DatepickerConfig;
		}
	});
	Object.defineProperty(exports, 'DatepickerModule', {
		enumerable: true,
		get: function () {
			return datepicker.DatepickerModule;
		}
	});
	Object.defineProperty(exports, 'DayPickerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.DayPickerComponent;
		}
	});
	Object.defineProperty(exports, 'MonthPickerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.MonthPickerComponent;
		}
	});
	Object.defineProperty(exports, 'YearPickerComponent', {
		enumerable: true,
		get: function () {
			return datepicker.YearPickerComponent;
		}
	});
	Object.defineProperty(exports, 'BsModalRef', {
		enumerable: true,
		get: function () {
			return modal.BsModalRef;
		}
	});
	Object.defineProperty(exports, 'BsModalService', {
		enumerable: true,
		get: function () {
			return modal.BsModalService;
		}
	});
	Object.defineProperty(exports, 'ModalBackdropComponent', {
		enumerable: true,
		get: function () {
			return modal.ModalBackdropComponent;
		}
	});
	Object.defineProperty(exports, 'ModalBackdropOptions', {
		enumerable: true,
		get: function () {
			return modal.ModalBackdropOptions;
		}
	});
	Object.defineProperty(exports, 'ModalDirective', {
		enumerable: true,
		get: function () {
			return modal.ModalDirective;
		}
	});
	Object.defineProperty(exports, 'ModalModule', {
		enumerable: true,
		get: function () {
			return modal.ModalModule;
		}
	});
	Object.defineProperty(exports, 'ModalOptions', {
		enumerable: true,
		get: function () {
			return modal.ModalOptions;
		}
	});
	Object.defineProperty(exports, 'BsDropdownConfig', {
		enumerable: true,
		get: function () {
			return dropdown.BsDropdownConfig;
		}
	});
	Object.defineProperty(exports, 'BsDropdownContainerComponent', {
		enumerable: true,
		get: function () {
			return dropdown.BsDropdownContainerComponent;
		}
	});
	Object.defineProperty(exports, 'BsDropdownDirective', {
		enumerable: true,
		get: function () {
			return dropdown.BsDropdownDirective;
		}
	});
	Object.defineProperty(exports, 'BsDropdownMenuDirective', {
		enumerable: true,
		get: function () {
			return dropdown.BsDropdownMenuDirective;
		}
	});
	Object.defineProperty(exports, 'BsDropdownModule', {
		enumerable: true,
		get: function () {
			return dropdown.BsDropdownModule;
		}
	});
	Object.defineProperty(exports, 'BsDropdownState', {
		enumerable: true,
		get: function () {
			return dropdown.BsDropdownState;
		}
	});
	Object.defineProperty(exports, 'BsDropdownToggleDirective', {
		enumerable: true,
		get: function () {
			return dropdown.BsDropdownToggleDirective;
		}
	});
	Object.defineProperty(exports, 'PagerComponent', {
		enumerable: true,
		get: function () {
			return pagination.PagerComponent;
		}
	});
	Object.defineProperty(exports, 'PaginationComponent', {
		enumerable: true,
		get: function () {
			return pagination.PaginationComponent;
		}
	});
	Object.defineProperty(exports, 'PaginationConfig', {
		enumerable: true,
		get: function () {
			return pagination.PaginationConfig;
		}
	});
	Object.defineProperty(exports, 'PaginationModule', {
		enumerable: true,
		get: function () {
			return pagination.PaginationModule;
		}
	});
	Object.defineProperty(exports, 'BarComponent', {
		enumerable: true,
		get: function () {
			return progressbar.BarComponent;
		}
	});
	Object.defineProperty(exports, 'ProgressbarComponent', {
		enumerable: true,
		get: function () {
			return progressbar.ProgressbarComponent;
		}
	});
	Object.defineProperty(exports, 'ProgressbarConfig', {
		enumerable: true,
		get: function () {
			return progressbar.ProgressbarConfig;
		}
	});
	Object.defineProperty(exports, 'ProgressbarModule', {
		enumerable: true,
		get: function () {
			return progressbar.ProgressbarModule;
		}
	});
	Object.defineProperty(exports, 'RatingComponent', {
		enumerable: true,
		get: function () {
			return rating.RatingComponent;
		}
	});
	Object.defineProperty(exports, 'RatingModule', {
		enumerable: true,
		get: function () {
			return rating.RatingModule;
		}
	});
	Object.defineProperty(exports, 'DraggableItemService', {
		enumerable: true,
		get: function () {
			return sortable.DraggableItemService;
		}
	});
	Object.defineProperty(exports, 'SortableComponent', {
		enumerable: true,
		get: function () {
			return sortable.SortableComponent;
		}
	});
	Object.defineProperty(exports, 'SortableModule', {
		enumerable: true,
		get: function () {
			return sortable.SortableModule;
		}
	});
	Object.defineProperty(exports, 'NgTranscludeDirective', {
		enumerable: true,
		get: function () {
			return tabs.NgTranscludeDirective;
		}
	});
	Object.defineProperty(exports, 'TabDirective', {
		enumerable: true,
		get: function () {
			return tabs.TabDirective;
		}
	});
	Object.defineProperty(exports, 'TabHeadingDirective', {
		enumerable: true,
		get: function () {
			return tabs.TabHeadingDirective;
		}
	});
	Object.defineProperty(exports, 'TabsModule', {
		enumerable: true,
		get: function () {
			return tabs.TabsModule;
		}
	});
	Object.defineProperty(exports, 'TabsetComponent', {
		enumerable: true,
		get: function () {
			return tabs.TabsetComponent;
		}
	});
	Object.defineProperty(exports, 'TabsetConfig', {
		enumerable: true,
		get: function () {
			return tabs.TabsetConfig;
		}
	});
	Object.defineProperty(exports, 'TimepickerComponent', {
		enumerable: true,
		get: function () {
			return timepicker.TimepickerComponent;
		}
	});
	Object.defineProperty(exports, 'TimepickerConfig', {
		enumerable: true,
		get: function () {
			return timepicker.TimepickerConfig;
		}
	});
	Object.defineProperty(exports, 'TimepickerModule', {
		enumerable: true,
		get: function () {
			return timepicker.TimepickerModule;
		}
	});
	Object.defineProperty(exports, 'TooltipConfig', {
		enumerable: true,
		get: function () {
			return tooltip.TooltipConfig;
		}
	});
	Object.defineProperty(exports, 'TooltipContainerComponent', {
		enumerable: true,
		get: function () {
			return tooltip.TooltipContainerComponent;
		}
	});
	Object.defineProperty(exports, 'TooltipDirective', {
		enumerable: true,
		get: function () {
			return tooltip.TooltipDirective;
		}
	});
	Object.defineProperty(exports, 'TooltipModule', {
		enumerable: true,
		get: function () {
			return tooltip.TooltipModule;
		}
	});
	Object.defineProperty(exports, 'TypeaheadContainerComponent', {
		enumerable: true,
		get: function () {
			return typeahead.TypeaheadContainerComponent;
		}
	});
	Object.defineProperty(exports, 'TypeaheadDirective', {
		enumerable: true,
		get: function () {
			return typeahead.TypeaheadDirective;
		}
	});
	Object.defineProperty(exports, 'TypeaheadMatch', {
		enumerable: true,
		get: function () {
			return typeahead.TypeaheadMatch;
		}
	});
	Object.defineProperty(exports, 'TypeaheadModule', {
		enumerable: true,
		get: function () {
			return typeahead.TypeaheadModule;
		}
	});
	Object.defineProperty(exports, 'TypeaheadOptions', {
		enumerable: true,
		get: function () {
			return typeahead.TypeaheadOptions;
		}
	});
	Object.defineProperty(exports, 'PopoverConfig', {
		enumerable: true,
		get: function () {
			return popover.PopoverConfig;
		}
	});
	Object.defineProperty(exports, 'PopoverContainerComponent', {
		enumerable: true,
		get: function () {
			return popover.PopoverContainerComponent;
		}
	});
	Object.defineProperty(exports, 'PopoverDirective', {
		enumerable: true,
		get: function () {
			return popover.PopoverDirective;
		}
	});
	Object.defineProperty(exports, 'PopoverModule', {
		enumerable: true,
		get: function () {
			return popover.PopoverModule;
		}
	});
	Object.defineProperty(exports, 'BsComponentRef', {
		enumerable: true,
		get: function () {
			return componentLoader.BsComponentRef;
		}
	});
	Object.defineProperty(exports, 'ComponentLoader', {
		enumerable: true,
		get: function () {
			return componentLoader.ComponentLoader;
		}
	});
	Object.defineProperty(exports, 'ComponentLoaderFactory', {
		enumerable: true,
		get: function () {
			return componentLoader.ComponentLoaderFactory;
		}
	});
	Object.defineProperty(exports, 'ContentRef', {
		enumerable: true,
		get: function () {
			return componentLoader.ContentRef;
		}
	});
	Object.defineProperty(exports, 'Positioning', {
		enumerable: true,
		get: function () {
			return positioning.Positioning;
		}
	});
	Object.defineProperty(exports, 'PositioningService', {
		enumerable: true,
		get: function () {
			return positioning.PositioningService;
		}
	});
	Object.defineProperty(exports, 'positionElements', {
		enumerable: true,
		get: function () {
			return positioning.positionElements;
		}
	});
	Object.defineProperty(exports, 'MiniState', {
		enumerable: true,
		get: function () {
			return miniNgrx.MiniState;
		}
	});
	Object.defineProperty(exports, 'MiniStore', {
		enumerable: true,
		get: function () {
			return miniNgrx.MiniStore;
		}
	});
	Object.defineProperty(exports, 'arLocale', {
		enumerable: true,
		get: function () {
			return locale.arLocale;
		}
	});
	Object.defineProperty(exports, 'bgLocale', {
		enumerable: true,
		get: function () {
			return locale.bgLocale;
		}
	});
	Object.defineProperty(exports, 'caLocale', {
		enumerable: true,
		get: function () {
			return locale.caLocale;
		}
	});
	Object.defineProperty(exports, 'csLocale', {
		enumerable: true,
		get: function () {
			return locale.csLocale;
		}
	});
	Object.defineProperty(exports, 'daLocale', {
		enumerable: true,
		get: function () {
			return locale.daLocale;
		}
	});
	Object.defineProperty(exports, 'deLocale', {
		enumerable: true,
		get: function () {
			return locale.deLocale;
		}
	});
	Object.defineProperty(exports, 'enGbLocale', {
		enumerable: true,
		get: function () {
			return locale.enGbLocale;
		}
	});
	Object.defineProperty(exports, 'esDoLocale', {
		enumerable: true,
		get: function () {
			return locale.esDoLocale;
		}
	});
	Object.defineProperty(exports, 'esLocale', {
		enumerable: true,
		get: function () {
			return locale.esLocale;
		}
	});
	Object.defineProperty(exports, 'esUsLocale', {
		enumerable: true,
		get: function () {
			return locale.esUsLocale;
		}
	});
	Object.defineProperty(exports, 'etLocale', {
		enumerable: true,
		get: function () {
			return locale.etLocale;
		}
	});
	Object.defineProperty(exports, 'fiLocale', {
		enumerable: true,
		get: function () {
			return locale.fiLocale;
		}
	});
	Object.defineProperty(exports, 'frLocale', {
		enumerable: true,
		get: function () {
			return locale.frLocale;
		}
	});
	Object.defineProperty(exports, 'glLocale', {
		enumerable: true,
		get: function () {
			return locale.glLocale;
		}
	});
	Object.defineProperty(exports, 'heLocale', {
		enumerable: true,
		get: function () {
			return locale.heLocale;
		}
	});
	Object.defineProperty(exports, 'hiLocale', {
		enumerable: true,
		get: function () {
			return locale.hiLocale;
		}
	});
	Object.defineProperty(exports, 'huLocale', {
		enumerable: true,
		get: function () {
			return locale.huLocale;
		}
	});
	Object.defineProperty(exports, 'idLocale', {
		enumerable: true,
		get: function () {
			return locale.idLocale;
		}
	});
	Object.defineProperty(exports, 'itLocale', {
		enumerable: true,
		get: function () {
			return locale.itLocale;
		}
	});
	Object.defineProperty(exports, 'jaLocale', {
		enumerable: true,
		get: function () {
			return locale.jaLocale;
		}
	});
	Object.defineProperty(exports, 'koLocale', {
		enumerable: true,
		get: function () {
			return locale.koLocale;
		}
	});
	Object.defineProperty(exports, 'ltLocale', {
		enumerable: true,
		get: function () {
			return locale.ltLocale;
		}
	});
	Object.defineProperty(exports, 'mnLocale', {
		enumerable: true,
		get: function () {
			return locale.mnLocale;
		}
	});
	Object.defineProperty(exports, 'nbLocale', {
		enumerable: true,
		get: function () {
			return locale.nbLocale;
		}
	});
	Object.defineProperty(exports, 'nlBeLocale', {
		enumerable: true,
		get: function () {
			return locale.nlBeLocale;
		}
	});
	Object.defineProperty(exports, 'nlLocale', {
		enumerable: true,
		get: function () {
			return locale.nlLocale;
		}
	});
	Object.defineProperty(exports, 'plLocale', {
		enumerable: true,
		get: function () {
			return locale.plLocale;
		}
	});
	Object.defineProperty(exports, 'ptBrLocale', {
		enumerable: true,
		get: function () {
			return locale.ptBrLocale;
		}
	});
	Object.defineProperty(exports, 'roLocale', {
		enumerable: true,
		get: function () {
			return locale.roLocale;
		}
	});
	Object.defineProperty(exports, 'ruLocale', {
		enumerable: true,
		get: function () {
			return locale.ruLocale;
		}
	});
	Object.defineProperty(exports, 'skLocale', {
		enumerable: true,
		get: function () {
			return locale.skLocale;
		}
	});
	Object.defineProperty(exports, 'slLocale', {
		enumerable: true,
		get: function () {
			return locale.slLocale;
		}
	});
	Object.defineProperty(exports, 'svLocale', {
		enumerable: true,
		get: function () {
			return locale.svLocale;
		}
	});
	Object.defineProperty(exports, 'thLocale', {
		enumerable: true,
		get: function () {
			return locale.thLocale;
		}
	});
	Object.defineProperty(exports, 'trLocale', {
		enumerable: true,
		get: function () {
			return locale.trLocale;
		}
	});
	Object.defineProperty(exports, 'viLocale', {
		enumerable: true,
		get: function () {
			return locale.viLocale;
		}
	});
	Object.defineProperty(exports, 'zhCnLocale', {
		enumerable: true,
		get: function () {
			return locale.zhCnLocale;
		}
	});

	Object.defineProperty(exports, '__esModule', { value: true });

}));
