// @c8y/ngx-components/esm5/core/stepper/stepper.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CdkStepperModule, CdkStep, STEPPER_GLOBAL_OPTIONS } from '@angular/cdk/stepper';
import { C8yStepper } from './stepper.component';
import { CommonModule } from '../common/common.module';
import { C8yStepperIcon } from './stepper-icon.directive';
import { C8yStepperProgress } from './stepper-progress.component';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { C8yStepperButtons } from './stepper-buttons.component';
var ɵ0 = { displayDefaultIndicatorType: false };
/**
 * The angular module definition for c8y-stepper (Wizard).
 * @exports C8yStepper
 * @exports C8yStepperIcon
 * @exports C8yStepperButtons
 * @exports CdkStep
 */
var StepperModule = /** @class */ (function () {
    function StepperModule() {
    }
    StepperModule = tslib_1.__decorate([
        NgModule({
            imports: [CommonModule, CdkStepperModule, TooltipModule],
            declarations: [C8yStepper, C8yStepperIcon, C8yStepperProgress, C8yStepperButtons],
            exports: [C8yStepper, CdkStep, C8yStepperIcon, C8yStepperButtons],
            providers: [
                {
                    /**
                     * This emables additional states, like: edit, done and error
                     * These states are used to display custom icons, if needed.
                     * Any change in here will most probably result in not showing custom icons anymore
                     */
                    provide: STEPPER_GLOBAL_OPTIONS,
                    useValue: ɵ0
                }
            ]
        })
    ], StepperModule);
    return StepperModule;
}());
export { StepperModule };
export { ɵ0 };
