// @c8y/ngx-components/esm5/device-profile/device-profile.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ONCE_ROUTE, gettext, ViewContext } from '@c8y/ngx-components';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { ReactiveFormsModule } from '@angular/forms';
import { DeviceProfileComponent } from './device-profile.component';
import { DeviceProfileListComponent } from './device-profile-list.component';
import { DeviceProfileNavigationFactory } from './device-profile-navigation.factory';
import { DeviceProfileService } from './device-profile.service';
import { AddDeviceProfileComponent } from './add-device-profile.component';
import { DeviceProfileGuard } from './device-profile.guard';
import { RepositoryModule } from '@c8y/ngx-components/repository';
import { SelectConfigurationModalComponent } from './select-configuration-modal.component';
import { DeviceTabProfileComponent } from './device-tab-profile/device-tab-profile.component';
import { DeviceTabProfileDetailComponent } from './device-tab-profile/device-tab-profile-detail.component';
import { DeviceProfileItemListComponent } from './device-tab-profile/device-profile-item-list.component';
var deviceProfilesRoutes = [
    {
        path: 'device-profiles/:id',
        component: DeviceProfileComponent
    },
    {
        path: 'device-profiles',
        component: DeviceProfileListComponent
    }
];
var deviceTabProfileRoutes = [
    {
        context: ViewContext.Device,
        path: 'device-profile',
        component: DeviceTabProfileComponent,
        label: gettext('Device profile'),
        icon: 'c8y-device-profile',
        canActivate: [DeviceProfileGuard]
    }
];
var ɵ0 = deviceTabProfileRoutes;
var DeviceProfileModule = /** @class */ (function () {
    function DeviceProfileModule() {
    }
    DeviceProfileModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                DeviceProfileComponent,
                DeviceProfileListComponent,
                AddDeviceProfileComponent,
                SelectConfigurationModalComponent,
                DeviceTabProfileComponent,
                DeviceTabProfileDetailComponent,
                DeviceProfileItemListComponent
            ],
            exports: [],
            imports: [
                CoreModule,
                CommonModule,
                RepositoryModule,
                RouterModule.forChild(deviceProfilesRoutes),
                BsDropdownModule.forRoot(),
                TooltipModule,
                ReactiveFormsModule,
                ButtonsModule,
                PopoverModule
            ],
            entryComponents: [
                AddDeviceProfileComponent,
                SelectConfigurationModalComponent,
                DeviceTabProfileComponent
            ],
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: DeviceProfileNavigationFactory,
                    multi: true
                },
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: ɵ0,
                    multi: true
                },
                DeviceProfileService,
                DeviceProfileGuard
            ]
        })
    ], DeviceProfileModule);
    return DeviceProfileModule;
}());
export { DeviceProfileModule };
export { ɵ0 };
