// @c8y/client/lib/esm2015/src/operation/OperationBulkGeneralStatus.js
import { gettext } from '../gettext';
/**
 * A bulk operation can have one of the following general statuses.
 */
export var OperationBulkGeneralStatus;
(function (OperationBulkGeneralStatus) {
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["SCHEDULED"] = gettext('SCHEDULED')] = "SCHEDULED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["EXECUTING"] = gettext('EXECUTING')] = "EXECUTING";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["EXECUTING_WITH_ERROR"] = gettext('EXECUTING_WITH_ERROR')] = "EXECUTING_WITH_ERROR";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["CANCELED"] = gettext('CANCELED')] = "CANCELED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["FAILED"] = gettext('FAILED')] = "FAILED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["SUCCESSFUL"] = gettext('SUCCESSFUL')] = "SUCCESSFUL";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["DELETED"] = gettext('DELETED')] = "DELETED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["INCORRECT"] = gettext('INCORRECT')] = "INCORRECT";
})(OperationBulkGeneralStatus || (OperationBulkGeneralStatus = {}));
