// @c8y/ngx-components/esm2015/core/tabs/tab.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild, Input } from '@angular/core';
import { TabsService } from './tabs.service';
let TabComponent = class TabComponent {
    constructor(tabs) {
        this.tabs = tabs;
        this.label = '';
    }
    ngAfterContentInit() {
        setTimeout(() => this.tabs.add(this));
    }
    ngOnDestroy() {
        this.tabs.remove(this);
    }
};
TabComponent.ctorParameters = () => [
    { type: TabsService }
];
tslib_1.__decorate([
    ViewChild('tpl', { static: false })
], TabComponent.prototype, "template", void 0);
tslib_1.__decorate([
    Input()
], TabComponent.prototype, "path", void 0);
tslib_1.__decorate([
    Input()
], TabComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Input()
], TabComponent.prototype, "icon", void 0);
tslib_1.__decorate([
    Input()
], TabComponent.prototype, "priority", void 0);
tslib_1.__decorate([
    Input()
], TabComponent.prototype, "orientation", void 0);
TabComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-tab',
        template: '<ng-template #tpl><ng-content></ng-content></ng-template>'
    })
], TabComponent);
export { TabComponent };
