// @c8y/ngx-components/esm2015/device-grid-example/device-grid-example.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, CommonModule, HOOK_NAVIGATOR_NODES } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { DeviceGridExampleNavigationFactory } from './device-grid-example-navigation.factory';
import { DeviceGridExampleGuard } from './device-grid-example-guard';
import { DeviceGridExampleComponent } from './device-grid-example.component';
const routes = [
    {
        path: 'grids/device-grid',
        component: DeviceGridExampleComponent,
        canActivate: [DeviceGridExampleGuard]
    }
];
let DeviceGridExampleModule = class DeviceGridExampleModule {
};
DeviceGridExampleModule = tslib_1.__decorate([
    NgModule({
        imports: [
            CoreModule,
            CommonModule,
            DeviceGridModule,
            RouterModule.forRoot(routes, { useHash: true })
        ],
        declarations: [DeviceGridExampleComponent],
        entryComponents: [DeviceGridExampleComponent],
        exports: [],
        providers: [
            {
                provide: HOOK_NAVIGATOR_NODES,
                useClass: DeviceGridExampleNavigationFactory,
                multi: true
            },
            DeviceGridExampleGuard
        ]
    })
], DeviceGridExampleModule);
export { DeviceGridExampleModule };
