// @c8y/ngx-components/esm2015/device-grid-example/device-grid-example.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { DeviceGridService, DeviceGridColumn, CustomDeviceGridColumn } from '@c8y/ngx-components/device-grid';
import { gettext } from '@c8y/ngx-components';
import { find } from 'lodash-es';
let DeviceGridExampleComponent = class DeviceGridExampleComponent {
    constructor(deviceGridService) {
        this.deviceGridService = deviceGridService;
        this.examples = [
            this.getExampleWithStandardColumns(),
            this.getExampleWithStandardColumnsAndSavedFilteringSorting(),
            this.getExampleWithCustomColumnAndPredefinedFiltering()
        ];
    }
    getExampleWithStandardColumns() {
        const columns = this.deviceGridService.getDefaultColumns();
        return {
            label: 'Standard columns',
            columns
        };
    }
    getExampleWithStandardColumnsAndSavedFilteringSorting() {
        const columns = this.deviceGridService.getDefaultColumns();
        const nameColumn = find(columns, { name: 'name' });
        nameColumn.externalFilterQuery = {
            names: ['*Temp*']
        };
        nameColumn.sortOrder = 'desc';
        return {
            label: 'Standard columns with saved filtering/sorting',
            columns
        };
    }
    getExampleWithCustomColumnAndPredefinedFiltering() {
        const columns = this.deviceGridService.getDefaultColumns();
        const typeColumn = new CustomDeviceGridColumn();
        typeColumn.name = 'type';
        typeColumn.path = 'type';
        typeColumn.header = gettext('Type');
        typeColumn.externalFilterQuery = {
            equals: ['c8y_MQTTDevice']
        };
        columns.splice(2, 0, typeColumn);
        return {
            label: 'Standard columns with custom one with predefined filtering',
            columns
        };
    }
    onExampleSelect(example) {
        this.columnsInput = example.columns;
    }
};
DeviceGridExampleComponent.ctorParameters = () => [
    { type: DeviceGridService }
];
DeviceGridExampleComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-device-grid-example',
        template: "<div class=\"row\">\n  <c8y-device-grid\n    [title]=\"'Select devices for bulk operation'\"\n    [columns]=\"columnsInput\"\n    [actionControls]=\"[]\"\n    (onColumnsChange)=\"columnsOutput = $event\"\n    (onDeviceQueryStringChange)=\"deviceQueryStringOutput = $event\"\n  ></c8y-device-grid>\n</div>\n\n<div class=\"row\">\n  <div class=\"card col-sm-6\">\n    <div class=\"card-header separator\">\n      <h4 class=\"card-title\">Input</h4>\n    </div>\n    <div class=\"card-block\">\n      <div class=\"c8y-select-wrapper\">\n        <select\n          class=\"form-control\"\n          [ngModel]=\"selectedExample\"\n          (ngModelChange)=\"onExampleSelect($event)\"\n        >\n          <option *ngFor=\"let example of examples\" [ngValue]=\"example\">{{\n            example.label\n          }}</option></select\n        ><span></span>\n      </div>\n      <pre\n        >{{ columnsInput | json }}\n      </pre>\n    </div>\n  </div>\n\n  <div class=\"card col-sm-6\">\n    <div class=\"card-header\">\n      <span>\n        <h4 class=\"card-title\">Output</h4>\n      </span>\n    </div>\n    <div class=\"card-block\">\n      <pre\n        >{{ deviceQueryStringOutput | json }}\n      </pre>\n      <pre\n        >{{ columnsOutput | json }}\n      </pre>\n    </div>\n  </div>\n</div>\n"
    })
], DeviceGridExampleComponent);
export { DeviceGridExampleComponent };
