// @c8y/ngx-components/esm2015/core/action/action-outlet.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { sortBy } from 'lodash-es';
let ActionOutletComponent = class ActionOutletComponent {
    constructor() {
        this.isPageToolbarExpanded = false;
        this.items = [];
    }
    getByPriority() {
        return sortBy(this.items, this.byPriority);
    }
    invoke(action) {
        if (action.action) {
            action.action.call();
        }
    }
    byPriority(item) {
        return -item.priority;
    }
};
tslib_1.__decorate([
    Input()
], ActionOutletComponent.prototype, "items", void 0);
ActionOutletComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-action-outlet',
        template: "<div class=\"dropdown\" dropdown *ngIf=\"items && items.length > 0\">\n  <button\n    type=\"button\"\n    class=\"main-header-button dropdown-toggle c8y-dropdown\"\n    dropdownToggle\n    aria-haspopup=\"true\"\n    aria-expanded=\"true\">\n    <i [c8yIcon]=\"'plus-circle'\" class=\"fa-2x\"></i>\n  </button>\n\n  <ul class=\"dropdown-menu cog-dropdown\" *dropdownMenu>\n    <ng-container *ngFor=\"let item of getByPriority()\">\n      <ng-container *c8yOutlet=\"item.template\"></ng-container>\n      <li *ngIf=\"item.label\">\n        <button title=\"{{ item.label | translate }}\" [attr.disabled]=\"item.disabled ? 'disabled' : null\" (click)=\"invoke(item)\">\n          <i *ngIf=\"item.icon\" [c8yIcon]=\"item.icon\"></i>{{item.label | translate}}\n        </button>\n      </li>\n    </ng-container>\n  </ul>\n</div>\n"
    })
], ActionOutletComponent);
export { ActionOutletComponent };
