// @c8y/cli/dist/webpack/config/production.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const OptimizeCSSAssetsPlugin = require("optimize-css-assets-webpack-plugin");
const TerserPlugin = require("terser-webpack-plugin");
function config(env) {
    return {
        devtool: false,
        output: {
            filename: '[name].[hash].js',
            chunkFilename: '[name].[hash].js',
        },
        cache: true,
        optimization: {
            splitChunks: {
                chunks(chunk) {
                    return chunk.name !== 'bootstrap';
                }
            },
            minimize: true,
            minimizer: [
                new TerserPlugin({
                    cache: true,
                    parallel: true,
                    sourceMap: false
                }),
                new OptimizeCSSAssetsPlugin({})
            ]
        },
        plugins: [
            new MiniCssExtractPlugin({ filename: '[name].[hash].css' })
        ]
    };
}
exports.config = config;
