// @c8y/ngx-components/esm2015/sms-gateway/sms-gateway.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FetchClient, IFetchOptions, TenantOptionsService } from '@c8y/client';
let SmsGatewayService = class SmsGatewayService {
    constructor(client, tenantOptionsService) {
        this.client = client;
        this.tenantOptionsService = tenantOptionsService;
        this.category = 'messaging';
        this.basePath = 'service/register/messaging';
        this.providerTemplates = {
            cepConfig: {
                'sms.senderAddress': 'cumulocity',
                'sms.senderName': 'cumulocity'
            },
            openit: {
                provider: 'openit',
                'openit.baseUrl': 'https://sms.plusserver.com/put.php'
            },
            sms77: {
                provider: 'sms77',
                'sms77.url': 'https://gateway.sms77.io/api/sms'
            }
        };
    }
    getProviderConfig() {
        return this.tenantOptionsService.detail({ category: this.category, key: '' });
    }
    saveProviderConfig(newConfig) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const providerTemplate = this.providerTemplates[newConfig.provider];
            const { cepConfig } = this.providerTemplates;
            Object.assign(newConfig, providerTemplate, cepConfig);
            const config = {
                method: 'PUT',
                headers: { 'content-type': 'application/json' },
                body: JSON.stringify(newConfig)
            };
            return this.client.fetch(this.basePath, config);
        });
    }
    deleteProviderConfig(config) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const providerTemplate = this.providerTemplates[config.provider];
            Object.assign(config, providerTemplate);
            return this.deleteProviderOptions(config);
        });
    }
    deleteProviderOptions(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            for (const optionKey of Object.keys(options)) {
                yield this.tenantOptionsService.delete({ category: this.category, key: optionKey });
            }
        });
    }
};
SmsGatewayService.ctorParameters = () => [
    { type: FetchClient },
    { type: TenantOptionsService }
];
SmsGatewayService = tslib_1.__decorate([
    Injectable()
], SmsGatewayService);
export { SmsGatewayService };
