// @c8y/ngx-components/esm2015/sms-gateway/sms-gateway.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ModalService, AlertService, gettext, Status } from '@c8y/ngx-components';
import { SmsGatewayService } from './sms-gateway.service';
import { pick } from 'lodash-es';
let SmsGatewayComponent = class SmsGatewayComponent {
    constructor(modalService, alertService, smsGatewayService) {
        this.modalService = modalService;
        this.alertService = alertService;
        this.smsGatewayService = smsGatewayService;
        this.smsProviders = [
            { id: 'openit', name: 'OpenIT' },
            { id: 'sms77', name: 'sms77' }
        ];
        this.supportedConfigOptions = [
            'provider',
            'openit.username',
            'credentials.openit.password',
            'credentials.sms77.api.key'
        ];
        this.pageTitle = gettext('SMS provider');
    }
    ngOnInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.smsConfig = this.getEmptyConfig();
            this.oldConfig = this.getEmptyConfig();
            const res = yield this.smsGatewayService.getProviderConfig();
            if (this.isSupportedProvider(res.data)) {
                this.smsConfig = this.getConfigurableProperties(res.data);
                this.oldConfig = this.smsConfig;
            }
        });
    }
    getConfigurableProperties(config) {
        return pick(config, this.supportedConfigOptions);
    }
    isSupportedProvider(config) {
        return this.smsProviders.some(prov => prov.id === config.provider);
    }
    saveSMSGatewayConfiguration(form) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const res = yield this.smsGatewayService.saveProviderConfig(this.smsConfig);
            if (res && res.status !== 200) {
                const data = res.json ? yield res.json() : undefined;
                this.alertService.addServerFailure({ data, res });
            }
            else {
                this.alertService.success(gettext('Credentials updated.'));
                Object.assign(this.oldConfig, this.smsConfig);
                form.pristine = true;
            }
        });
    }
    deleteSMSGatewayConfiguration() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.modalService.confirm(gettext('Delete credentials'), gettext(`You are about to delete SMS provider credentials. Deleting credentials will lock out any users with
      SMS-based two-factor authentication and deactivate the SMS-based smart rules and device communication.
      Do you want to proceed?`), Status.DANGER, { ok: gettext('Delete') });
            yield this.smsGatewayService.deleteProviderConfig(this.oldConfig);
            this.alertService.success(gettext('Credentials deleted.'));
            this.smsConfig = this.getEmptyConfig();
            this.oldConfig = this.getEmptyConfig();
        });
    }
    getEmptyConfig() {
        return {
            provider: undefined
        };
    }
};
SmsGatewayComponent.ctorParameters = () => [
    { type: ModalService },
    { type: AlertService },
    { type: SmsGatewayService }
];
SmsGatewayComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-sms-gateway',
        template: "<c8y-title>{{ pageTitle | translate }}</c8y-title>\n<div class=\"row\">\n  <div class=\"col-xs-12 col-sm-10 col-md-8 col-lg-6\">\n    <form\n      class=\"card\"\n      #smsGatewayForm=\"ngForm\"\n      (ngSubmit)=\"smsGatewayForm.form.valid && saveSMSGatewayConfiguration(smsGatewayForm.form)\"\n    >\n      <div class=\"card-header\">\n        <h4 class=\"card-title\" translate>Credentials</h4>\n      </div>\n      <div class=\"card-block\">\n        <p class=\"bottom-m\" translate>\n          Enter your credentials from your SMS provider. This enables platform features that utilize\n          SMS services (e.g. two-factor authentication and user notifications).\n        </p>\n        <label translate>SMS provider</label>\n        <div class=\"form-group\">\n          <ul class=\"list-unstyled\">\n            <li *ngFor=\"let provider of smsProviders\">\n              <label class=\"c8y-radio\">\n                <input\n                  class=\"form-control\"\n                  type=\"radio\"\n                  name=\"providerName\"\n                  [(ngModel)]=\"smsConfig.provider\"\n                  [value]=\"provider.id\"\n                />\n                <span></span>\n                {{ provider.name | translate }}\n              </label>\n            </li>\n          </ul>\n        </div>\n        <div class=\"form-group\" *ngIf=\"smsConfig.provider === 'openit'\">\n          <label translate for=\"userName\">Username</label>\n          <input\n            id=\"userName\"\n            class=\"form-control\"\n            type=\"text\"\n            autocomplete=\"off\"\n            name=\"userName\"\n            [(ngModel)]=\"smsConfig['openit.username']\"\n            placeholder=\"{{ 'e.g. joe`LOCALIZE`' | translate }}\"\n            required\n          />\n        </div>\n        <div class=\"form-group\" *ngIf=\"smsConfig.provider === 'openit'\">\n          <label translate for=\"password\">Password</label>\n          <input\n            id=\"password\"\n            class=\"form-control\"\n            type=\"password\"\n            autocomplete=\"off\"\n            name=\"password\"\n            [(ngModel)]=\"smsConfig['credentials.openit.password']\"\n            placeholder=\"{{ 'e.g. my_password' | translate }}\"\n            required\n          />\n        </div>\n        <div class=\"form-group\" *ngIf=\"smsConfig.provider === 'sms77'\">\n          <label translate for=\"apiKey\">API key</label>\n          <input\n            id=\"apiKey\"\n            class=\"form-control\"\n            type=\"text\"\n            autocomplete=\"off\"\n            name=\"apiKey\"\n            [(ngModel)]=\"smsConfig['credentials.sms77.api.key']\"\n            required\n          />\n        </div>\n      </div>\n      <div class=\"card-footer\">\n        <button\n          type=\"button\"\n          class=\"btn btn-default\"\n          (click)=\"deleteSMSGatewayConfiguration()\"\n          [disabled]=\"!(oldConfig && oldConfig.provider)\"\n          title=\"{{ 'Delete' | translate }}\"\n          translate\n        >\n          Delete\n        </button>\n        <button\n          type=\"submit\"\n          class=\"btn btn-primary\"\n          [disabled]=\"!smsGatewayForm.form.valid || smsGatewayForm.form.pristine\"\n          title=\"{{ 'Save' | translate }}\"\n          translate\n        >\n          Save\n        </button>\n      </div>\n    </form>\n  </div>\n</div>\n"
    })
], SmsGatewayComponent);
export { SmsGatewayComponent };
