// @c8y/ngx-components/esm2015/sms-gateway/sms-gateway-navigation.factory.js
var SmsGatewayNavigationFactory_1;
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { ApplicationService } from '@c8y/client';
let SmsGatewayNavigationFactory = SmsGatewayNavigationFactory_1 = class SmsGatewayNavigationFactory {
    constructor(applicationService) {
        this.applicationService = applicationService;
    }
    get() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.nodeItem) {
                const showSMSProvider = (yield this.applicationService.isAvailable(SmsGatewayNavigationFactory_1.applicationName)).data;
                if (showSMSProvider) {
                    this.nodeItem = new NavigatorNode({
                        label: gettext('SMS provider'),
                        path: 'smsgateway',
                        icon: 'envelope-o',
                        parent: gettext('Settings'),
                        priority: 1000
                    });
                }
            }
            return this.nodeItem;
        });
    }
};
SmsGatewayNavigationFactory.applicationName = 'sms-gateway';
SmsGatewayNavigationFactory.ctorParameters = () => [
    { type: ApplicationService }
];
SmsGatewayNavigationFactory = SmsGatewayNavigationFactory_1 = tslib_1.__decorate([
    Injectable()
], SmsGatewayNavigationFactory);
export { SmsGatewayNavigationFactory };
