// @c8y/ngx-components/esm2015/context-dashboard/widget.service.js
import * as tslib_1 from "tslib";
import { Injectable, Inject } from '@angular/core';
import { sortBy, cloneDeep } from 'lodash-es';
import { DynamicComponentService, DynamicComponentDefinition } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { CONTEXT_DASHBOARD_CONFIG } from './context-dashboard.model';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@ngx-translate/core";
import * as i3 from "./context-dashboard.model";
let WidgetService = class WidgetService {
    constructor(dynamicComponentService, translateService, moduleConfig) {
        this.dynamicComponentService = dynamicComponentService;
        this.translateService = translateService;
        this.moduleConfig = moduleConfig;
    }
    getWidgetDefinition(componentId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.dynamicComponentService.getById$(componentId).toPromise();
        });
    }
    getWidgetDefinitions() {
        const translatedComponents = Array.from(this.dynamicComponentService.state).map(cmp => (Object.assign({}, cmp, { label: this.translateService.instant(cmp.label) })));
        return cloneDeep(sortBy(translatedComponents, 'label').filter(this.moduleConfig.widgetFilter));
    }
};
WidgetService.ctorParameters = () => [
    { type: DynamicComponentService },
    { type: TranslateService },
    { type: undefined, decorators: [{ type: Inject, args: [CONTEXT_DASHBOARD_CONFIG,] }] }
];
WidgetService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function WidgetService_Factory() { return new WidgetService(i0.ɵɵinject(i1.DynamicComponentService), i0.ɵɵinject(i2.TranslateService), i0.ɵɵinject(i3.CONTEXT_DASHBOARD_CONFIG)); }, token: WidgetService, providedIn: "root" });
WidgetService = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    }),
    tslib_1.__param(2, Inject(CONTEXT_DASHBOARD_CONFIG))
], WidgetService);
export { WidgetService };
