// @c8y/ngx-components/esm5/core/user/user-menu-item.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, EventEmitter, Output } from '@angular/core';
import { UserMenuService } from './user-menu.service';
var UserMenuItemComponent = /** @class */ (function () {
    function UserMenuItemComponent(userService) {
        this.userService = userService;
        this.priority = 0;
        this.click = new EventEmitter();
    }
    UserMenuItemComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.viewInitTimeout = setTimeout(function () {
            return _this.userService.add(_this);
        });
    };
    UserMenuItemComponent.prototype.ngOnDestroy = function () {
        clearTimeout(this.viewInitTimeout);
        this.userService.remove(this);
    };
    UserMenuItemComponent.prototype.onClick = function () {
        this.click.emit(this);
    };
    UserMenuItemComponent.ctorParameters = function () { return [
        { type: UserMenuService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], UserMenuItemComponent.prototype, "icon", void 0);
    tslib_1.__decorate([
        Input()
    ], UserMenuItemComponent.prototype, "label", void 0);
    tslib_1.__decorate([
        Input()
    ], UserMenuItemComponent.prototype, "link", void 0);
    tslib_1.__decorate([
        Input()
    ], UserMenuItemComponent.prototype, "target", void 0);
    tslib_1.__decorate([
        Input()
    ], UserMenuItemComponent.prototype, "priority", void 0);
    tslib_1.__decorate([
        ViewChild('template', { static: false })
    ], UserMenuItemComponent.prototype, "template", void 0);
    tslib_1.__decorate([
        Output()
    ], UserMenuItemComponent.prototype, "click", void 0);
    UserMenuItemComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-user-menu-item',
            template: "<ng-template #template>\n  <li (click)=\"onClick()\">\n    <a style=\"cursor: pointer\" [attr.href]=\"link\" [attr.target]=\"target\">\n      <i [c8yIcon]=\"icon\"></i>\n      {{label | translate}}\n      <ng-content></ng-content>\n    </a>\n  </li>\n</ng-template>"
        })
    ], UserMenuItemComponent);
    return UserMenuItemComponent;
}());
export { UserMenuItemComponent };
