// @c8y/ngx-components/fesm2015/c8y-ngx-components-repository.js
import { __awaiter, __decorate } from 'tslib';
import { Injectable, Component, NgModule } from '@angular/core';
import { gettext, AlertService, NavigatorNode, OptionsService, CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ONCE_ROUTE } from '@c8y/ngx-components';
import { pipe, Subject, timer, of } from 'rxjs';
import { map, debounce, distinctUntilChanged } from 'rxjs/operators';
import { InventoryService, InventoryBinaryService } from '@c8y/client';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { uniqBy } from 'lodash';
import { TooltipModule } from 'ngx-bootstrap/tooltip';

var RepositoryType;
(function (RepositoryType) {
    RepositoryType["FIRMWARE"] = "c8y_Firmware";
    RepositoryType["SOFTWARE"] = "c8y_Software";
    RepositoryType["CONFIGURATION"] = "c8y_ConfigurationDump";
})(RepositoryType || (RepositoryType = {}));
const REPOSITORY_BINARY_TYPE = {
    [RepositoryType.SOFTWARE]: 'c8y_SoftwareBinary',
    [RepositoryType.FIRMWARE]: 'c8y_FirmwareBinary',
    [RepositoryType.CONFIGURATION]: 'c8y_ConfigurationDumpBinary'
};

let RepositoryService = class RepositoryService {
    constructor(inventory, inventoryBinary) {
        this.inventory = inventory;
        this.inventoryBinary = inventoryBinary;
    }
    listConfigurations() {
        return this.inventory.list({
            type: RepositoryType.CONFIGURATION,
            withTotalPages: true,
            pageSize: 50
        });
    }
    delete(entity) {
        // TODO: add cascading delete (will be done by Dawid)
        return this.inventory.delete(entity);
    }
    save(data, type, mo = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            switch (type) {
                case RepositoryType.CONFIGURATION: {
                    Object.assign(mo, {
                        type: RepositoryType.CONFIGURATION,
                        configurationType: data.selected ? data.selected.configurationType : undefined,
                        name: data.versionOrName,
                        description: data.description,
                        deviceType: data.deviceType,
                        c8y_Global: {}
                    });
                    break;
                }
            }
            const existingUrl = mo.url;
            if (data.binary.url) {
                mo.url = data.binary.url;
            }
            else if (data.binary.file) {
                const response = yield this.inventoryBinary.create(data.binary.file, { c8y_Global: {} });
                mo.url = response.data.self;
            }
            if (mo.id) {
                return this.updateEntry(mo, existingUrl);
            }
            return this.createEntry(mo);
        });
    }
    createEntry(mo) {
        return __awaiter(this, void 0, void 0, function* () {
            const binaryId = yield this.inventoryBinary.getIdFromUrl(mo.url);
            const newMo = yield this.inventory.create(mo);
            if (binaryId) {
                yield this.inventory.childAdditionsAdd(binaryId, newMo.data);
            }
            return newMo;
        });
    }
    updateEntry(mo, url) {
        return __awaiter(this, void 0, void 0, function* () {
            const existingBinaryId = yield this.inventoryBinary.getIdFromUrl(url);
            const newBinaryId = yield this.inventoryBinary.getIdFromUrl(mo.url);
            if (existingBinaryId && existingBinaryId !== newBinaryId) {
                const id = this.inventoryBinary.getIdFromUrl(url);
                yield this.inventoryBinary.delete(id);
            }
            if (newBinaryId) {
                yield this.inventory.childAdditionsAdd(newBinaryId, mo);
            }
            return this.inventory.update(mo);
        });
    }
};
RepositoryService.ctorParameters = () => [
    { type: InventoryService },
    { type: InventoryBinaryService }
];
RepositoryService = __decorate([
    Injectable()
], RepositoryService);

let ConfigurationNewModalComponent = class ConfigurationNewModalComponent {
    constructor(repositoryService, bsModalRef, alert) {
        this.repositoryService = repositoryService;
        this.bsModalRef = bsModalRef;
        this.alert = alert;
        this.binary = {
            file: undefined,
            url: undefined
        };
        this.pattern = '';
        this.mo = {};
        this.isLoading = false;
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.configs = yield this.repositoryService.listConfigurations();
            this.setPipe('');
        });
    }
    cancel() {
        this.bsModalRef.hide();
        this._cancel();
    }
    setPipe(filterStr) {
        this.pattern = filterStr;
        this.filterPipe = pipe(map((data) => uniqBy(data, 'configurationType')), map((data) => {
            return data.filter((mo) => mo.configurationType &&
                mo.configurationType.toLowerCase().indexOf(filterStr.toLowerCase()) > -1);
        }));
    }
    uploadFile(dropped, ngForm) {
        this.binary.file = dropped[0].file;
        this.binary.url = undefined;
        ngForm.form.markAsDirty();
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.isLoading = true;
                const { selected, versionOrName, description, binary, deviceType } = this;
                yield this.repositoryService.save({ selected, versionOrName, description, binary, deviceType }, RepositoryType.CONFIGURATION, this.mo);
                this.alert.success(this.mo.id ? gettext('Configuration updated.') : gettext('Configuration created.'));
                this.bsModalRef.hide();
                this._save();
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
                this._cancel();
            }
            finally {
                this.isLoading = false;
            }
        });
    }
};
ConfigurationNewModalComponent.ctorParameters = () => [
    { type: RepositoryService },
    { type: BsModalRef },
    { type: AlertService }
];
ConfigurationNewModalComponent = __decorate([
    Component({
        selector: 'c8y-configuration-new-modal',
        template: "<div class=\"modal-header bg-primary text-white text-center\">\n  <h1 c8yIcon=\"cogs\" style=\"font-size: 55px;\"></h1>\n  <h4 class=\"p-t-16 text-uppercase\" translate *ngIf=\"mo.id\">Update configuration</h4>\n  <h4 class=\"p-t-16 text-uppercase\" translate *ngIf=\"!mo.id\">Add configuration</h4>\n</div>\n\n<form #configurationForm=\"ngForm\" (ngSubmit)=\"configurationForm.form.valid && save()\">\n  <div class=\"modal-body\">\n    <c8y-form-group>\n      <label translate>Name</label>\n      <input\n        type=\"text\"\n        class=\"form-control\"\n        placeholder=\"e.g. hosts\"\n        autocomplete=\"off\"\n        required\n        maxlength=\"254\"\n        [(ngModel)]=\"versionOrName\"\n        name=\"versionOrName\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate>Device type</label>\n      <input\n        type=\"text\"\n        class=\"form-control\"\n        placeholder=\"e.g. Posix\"\n        maxlength=\"254\"\n        autocomplete=\"off\"\n        [(ngModel)]=\"deviceType\"\n        name=\"deviceType\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate>Description</label>\n      <input\n        type=\"text\"\n        class=\"form-control\"\n        placeholder=\"e.g. Host configuration Posix\"\n        maxlength=\"254\"\n        autocomplete=\"off\"\n        [(ngModel)]=\"description\"\n        name=\"description\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate>Configuration type</label>\n      <c8y-typeahead\n        [(selected)]=\"selected\"\n        placeholder=\"e.g. ssh\"\n        (onSearch)=\"setPipe($event)\"\n        displayProperty=\"configurationType\"\n      >\n        <c8y-li\n          *c8yFor=\"let config of configs; pipe: filterPipe; notFound: notFoundTemplate\"\n          class=\"p-l-8 p-r-8 c8y-list__item--link\"\n          (click)=\"selected = config; setPipe('')\"\n          [active]=\"selected === config\"\n        >\n          <c8y-highlight\n            [text]=\"config.configurationType || '--'\"\n            [pattern]=\"pattern\"\n          ></c8y-highlight>\n        </c8y-li>\n        <ng-template #notFoundTemplate>\n          <c8y-li class=\"bg-gray-lighter p-8\" *ngIf=\"pattern.length > 0\">\n            <span translate>No match found, add new`configuration`?</span>\n            <button type=\"button\" class=\"btn btn-primary btn-xs m-l-16\" translate>\n              Create new`configuration`\n            </button>\n          </c8y-li>\n        </ng-template>\n      </c8y-typeahead>\n    </c8y-form-group>\n\n    <c8y-drop-area\n      (dropped)=\"uploadFile($event, configurationForm)\"\n      [loadingMessage]=\"'Importing, please wait.' | translate\"\n      [title]=\"'Add a configuration file' | translate\"\n      [loading]=\"isLoading\"\n    >\n    </c8y-drop-area>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button (click)=\"cancel()\" type=\"button\" class=\"btn btn-default\" [disabled]=\"isLoading\">\n      <span translate>Cancel</span>\n    </button>\n    <button\n      type=\"button\"\n      class=\"btn btn-primary\"\n      type=\"submit\"\n      [disabled]=\"!configurationForm.valid || configurationForm.pristine || isLoading\"\n    >\n      <span translate *ngIf=\"!mo.id\">Add configuration</span>\n      <span translate *ngIf=\"mo.id\">Update configuration</span>\n    </button>\n  </div>\n</form>\n"
    })
], ConfigurationNewModalComponent);

let ConfigurationListComponent = class ConfigurationListComponent {
    constructor(alert, repositoryService, modalService) {
        this.alert = alert;
        this.repositoryService = repositoryService;
        this.modalService = modalService;
        this.filterChange$ = new Subject();
        this.filterTerm = '';
        this.CARRIAGE_RETURN_KEY = 13;
        this.CARRIAGE_RETURN_CODE = 'Enter';
        this.DELETED_SUCCESS_MSG = gettext('Configuration deleted.');
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.reset();
            this.filterChange$
                .pipe(debounce((event) => event.code === this.CARRIAGE_RETURN_CODE || event.keyCode === this.CARRIAGE_RETURN_KEY
                ? timer(10)
                : timer(300)), map((e) => e.target.value), distinctUntilChanged())
                .subscribe(filterTerm => {
                this.filterTerm = filterTerm;
                this.setPipe(filterTerm);
            });
        });
    }
    ngOnDestroy() {
        this.filterChange$.complete();
    }
    add() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modalService.show(ConfigurationNewModalComponent, {
                    class: 'modal-sm',
                    ignoreBackdropClick: true
                }).content.result;
                yield this.reset();
            }
            catch (ex) {
                // inteded empty
            }
        });
    }
    edit(configuration) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const modal = this.modalService.show(ConfigurationNewModalComponent, {
                    class: 'modal-sm',
                    ignoreBackdropClick: true,
                    initialState: {
                        selected: { id: configuration.id, configurationType: configuration.configurationType },
                        versionOrName: configuration.name,
                        deviceType: configuration.deviceType,
                        description: configuration.description,
                        binary: { url: configuration.url }
                    }
                }).content;
                modal.mo = configuration;
                yield modal.result;
                yield this.reset();
            }
            catch (ex) {
                // inteded empty
            }
        });
    }
    delete(configuration) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.repositoryService.delete(configuration);
                this.alert.success(this.DELETED_SUCCESS_MSG);
                yield this.reset();
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
            }
        });
    }
    setPipe(filterTerm) {
        this.filterPipe = pipe(map((data) => filterTerm.trim().length === 0
            ? data
            : data.filter((mo) => this.filterContainString(mo.name, filterTerm) ||
                this.filterContainString(mo.configurationType, filterTerm))));
    }
    filterContainString(name, filterTerm) {
        const term = filterTerm.toLowerCase().trim();
        return name && name.toLowerCase().indexOf(term) > -1;
    }
    reset() {
        return __awaiter(this, void 0, void 0, function* () {
            this.filterTerm = '';
            this.configurations$ = of(yield this.repositoryService.listConfigurations());
            this.setPipe('');
        });
    }
};
ConfigurationListComponent.ctorParameters = () => [
    { type: AlertService },
    { type: RepositoryService },
    { type: BsModalService }
];
ConfigurationListComponent = __decorate([
    Component({
        selector: 'c8y-configuration-list',
        template: "<c8y-title>\n  <span translate>Configuration snapshots repository</span>&nbsp;\n  <small *ngIf=\"(configurations$ | async)?.paging.totalPages === 1 && !filterTerm\"\n    >{{ (configurations$ | async).data.length }} <span translate>snapshots</span></small\n  >\n  <small\n    *ngIf=\"(configurations$ | async)?.paging.totalPages > 1 && !filterTerm\"\n    [tooltip]=\"'More data available. Scroll to the bottom of the list to load it.' | translate\"\n    container=\"body\"\n    >{{ (configurations$ | async).paging.pageSize }}+ <span translate>snapshots</span></small\n  >\n  <small *ngIf=\"filterTerm\"\n    ><span translate>Search results for</span>&nbsp;\"{{ this.filterTerm }}\"</small\n  >\n</c8y-title>\n\n<c8y-action-bar-item itemClass=\"navbar-form\">\n  <div class=\"input-group input-group-search\">\n    <input\n      class=\"form-control\"\n      placeholder=\"{{ 'Filter\u2026' | translate }}\"\n      type=\"text\"\n      [value]=\"filterTerm\"\n      (keyup)=\"filterChange$.next($event)\"\n    />\n    <span class=\"input-group-btn\">\n      <button class=\"btn btn-clean\" (click)=\"filterTerm = ''; setPipe('')\">\n        <i [c8yIcon]=\"filterTerm.length === 0 ? 'filter' : 'close'\"></i>\n      </button>\n    </span>\n  </div>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" (click)=\"add()\">\n    <i c8yIcon=\"plus-square\"></i> {{ 'Add configuration snapshot' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<div\n  class=\"c8y-empty-state text-center\"\n  *ngIf=\"!filterTerm && (configurations$ | async)?.data.length === 0\"\n>\n  <h1 c8yIcon=\"gears\"></h1>\n  <h3 translate>There are no configuration snapshots defined</h3>\n  <p translate>Add a configuration snapshot first.</p>\n  <div>\n    <button (click)=\"add()\" class=\"btn btn-primary\" translate>\n      Add configuration snapshot\n    </button>\n  </div>\n  <p c8y-guide-docs>\n    <small translate\n      >Find out more in the\n      <a c8y-guide-href=\"users-guide/device-management/#configuration-repository\"\n        >User guide`KEEP_ORIGINAL`</a\n      >.</small\n    >\n  </p>\n</div>\n\n<ng-template #notFound>\n  <c8y-li class=\"p-8 text-center\">\n    <p>\n      <span translate>No more entries found for filter:</span>&nbsp;<strong>{{\n        filterTerm\n      }}</strong>\n    </p>\n    <button class=\"btn btn-primary btn-xs m-t-8\" translate (click)=\"filterTerm = ''; setPipe('')\">\n      Reset filter\n    </button>\n  </c8y-li>\n</ng-template>\n<div class=\"p-b-32\">\n  <c8y-list-group>\n    <c8y-li\n      *c8yFor=\"\n        let configuration of configurations$;\n        pipe: filterPipe;\n        notFound: this.filterTerm ? notFound : undefined\n      \"\n    >\n      <c8y-li-icon icon=\"gears\"></c8y-li-icon>\n      <div class=\"row flex-row\">\n        <div class=\"col-xs-3\">\n          <c8y-highlight\n            [text]=\"configuration.name || '-'\"\n            elementClass=\"text-info\"\n            [pattern]=\"filterTerm\"\n          ></c8y-highlight>\n        </div>\n        <div class=\"col-xs-3\">\n          <span class=\"text-muted text-truncate\" [title]=\"configuration.description\">{{\n            configuration.description\n          }}</span>\n        </div>\n        <div class=\"col-xs-3\">\n          <small translate class=\"text-muted text-uppercase\">Device type:</small>&nbsp;\n          {{ configuration.deviceType || '-' }}\n        </div>\n        <div class=\"col-xs-3 text-right\">\n          <span class=\"label label-primary\" *ngIf=\"configuration.configurationType\">\n            <c8y-highlight\n              [text]=\"configuration.configurationType\"\n              elementClass=\"text-info\"\n              [pattern]=\"filterTerm\"\n            ></c8y-highlight>\n          </span>\n        </div>\n      </div>\n      <c8y-li-action (click)=\"edit(configuration)\" icon=\"pencil\">\n        Edit\n      </c8y-li-action>\n      <c8y-li-action (click)=\"delete(configuration)\" icon=\"trash-o\">\n        Delete\n      </c8y-li-action>\n    </c8y-li>\n  </c8y-list-group>\n</div>\n"
    })
], ConfigurationListComponent);

let RepositoryNavigationFactory = class RepositoryNavigationFactory {
    constructor(optionsService) {
        this.optionsService = optionsService;
        this.navs = [];
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            const betaFlag = localStorage.getItem('c8y_beta_status');
            if (betaFlag === 'all' && this.navs.length === 0) {
                this.navs.push(new NavigatorNode({
                    label: gettext('Configuration repository'),
                    path: 'configuration-new',
                    icon: 'gears',
                    parent: gettext('Management'),
                    priority: 800
                }));
            }
            return this.navs;
        });
    }
};
RepositoryNavigationFactory.ctorParameters = () => [
    { type: OptionsService }
];
RepositoryNavigationFactory = __decorate([
    Injectable()
], RepositoryNavigationFactory);

const ɵ0 = [
    {
        path: 'configuration-new',
        component: ConfigurationListComponent
    }
];
let RepositoryModule = class RepositoryModule {
};
RepositoryModule = __decorate([
    NgModule({
        imports: [CoreModule, TooltipModule],
        exports: [],
        entryComponents: [ConfigurationListComponent, ConfigurationNewModalComponent],
        declarations: [ConfigurationListComponent, ConfigurationNewModalComponent],
        providers: [
            RepositoryService,
            {
                provide: HOOK_NAVIGATOR_NODES,
                useClass: RepositoryNavigationFactory,
                multi: true
            },
            {
                provide: HOOK_ONCE_ROUTE,
                useValue: ɵ0,
                multi: true
            }
        ]
    })
], RepositoryModule);

/**
 * Generated bundle index. Do not edit.
 */

export { RepositoryModule, ɵ0, ConfigurationListComponent as ɵa, RepositoryService as ɵb, ConfigurationNewModalComponent as ɵc, RepositoryNavigationFactory as ɵd };
