// @c8y/ngx-components/fesm2015/c8y-ngx-components-context-dashboard.js
import { __awaiter, __decorate, __param } from 'tslib';
import { InjectionToken, Injectable, Inject, ViewChild, Component, ɵɵdefineInjectable, ɵɵinject, Renderer2, Input, HostBinding, EventEmitter, Output, NgModule, Host } from '@angular/core';
import { gettext, Status, getActivatedRoute, ViewContext, TabsService, ModalService, AppStateService, ICON_LIST, DynamicComponentService, DashboardChildChange, AlertService, HOOK_ONCE_ROUTE, HOOK_ACTION, CoreModule, DocsService, NavigatorService, DocsModule, QuickLinkModule, HOOK_NAVIGATOR_NODES, HOOK_COMPONENTS } from '@c8y/ngx-components';
import { Router, ActivatedRoute } from '@angular/router';
import { InventoryService, UserService, TenantService, ApplicationService } from '@c8y/client';
import { TranslateService, TranslateModule } from '@ngx-translate/core';
import { assign, pick, some, keys, keyBy, has, clone, cloneDeep, sortBy, omit, get, escapeRegExp, findIndex, values } from 'lodash-es';
import { of, from, Subject, iif, timer } from 'rxjs';
import { tap, map, catchError, throwIfEmpty, filter, toArray, mergeAll, mergeMap, switchMap } from 'rxjs/operators';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { ControlContainer, NgForm } from '@angular/forms';

const CONTEXT_DASHBOARD_CONFIG = new InjectionToken('ContextDashboardConfig');
var ContextDashboardType;
(function (ContextDashboardType) {
    ContextDashboardType["Device"] = "device";
    ContextDashboardType["DeviceType"] = "type";
    ContextDashboardType["Group"] = "group";
    ContextDashboardType["Named"] = "name";
})(ContextDashboardType || (ContextDashboardType = {}));
const WIDGET_HEADER_CLASSES = [
    {
        label: gettext('Regular`style`'),
        class: 'panel-title-regular',
        description: gettext('The widget has no border between header and content.')
    },
    {
        label: gettext('Border`style`'),
        class: 'panel-title-border',
        description: gettext('The widget has a small separation border between header and content.')
    },
    {
        label: gettext('Overlay`style`'),
        class: 'panel-title-overlay',
        description: gettext('The widget content overlays the header.')
    },
    {
        label: gettext('Hidden`style`'),
        class: 'panel-title-hidden',
        description: gettext('The widget header is not shown.')
    }
];
const WIDGET_CONTENT_CLASSES = [
    {
        label: gettext('Light`style`'),
        class: 'panel-content-light',
        description: gettext('The widget is styled with a light background and has low contrast.')
    },
    {
        label: gettext('Dark`style`'),
        class: 'panel-content-dark',
        description: gettext('The widget is styled with a dark background and has high contrast.')
    },
    {
        label: gettext('Transparent`style`'),
        class: 'panel-content-transparent',
        description: gettext('The widget is styled with a transparent background and therefore is border-less.')
    },
    {
        label: gettext('Branded`style`'),
        class: 'panel-content-branded',
        description: gettext('The widget is styled with the main brand color.')
    }
];
const DASHBOARD_THEME_CLASSES = [
    {
        label: gettext('Light`theme`'),
        class: 'dashboard-theme-light',
        description: gettext('The dashboard is styled with a light background and has low contrast.')
    },
    {
        label: gettext('Dark`theme`'),
        class: 'dashboard-theme-dark',
        description: gettext('The dashboard is styled with a dark background and has high contrast.')
    },
    {
        label: gettext('Transparent`theme`'),
        class: 'dashboard-theme-transparent',
        description: gettext('The dashboard is styled with a transparent background and therefore is border-less.')
    },
    {
        label: gettext('Branded`theme`'),
        class: 'dashboard-theme-branded',
        description: gettext('The dashboard is styled with the main brand color.')
    }
];

let ContextDashboardService = class ContextDashboardService {
    constructor(inventory, tabs, modal, translateService, router, user, appState) {
        this.inventory = inventory;
        this.tabs = tabs;
        this.modal = modal;
        this.translateService = translateService;
        this.router = router;
        this.user = user;
        this.appState = appState;
        this.cache = new Map();
        this.DEFAULT_PAGESIZE = 1000;
        this.FRAGMENT_NAME = 'c8y_Dashboard';
        this.DASHBOARD_ROUTE_PATH = 'dashboard';
        this.INDEX_SPLIT = '!';
        this._formDisabled = true;
    }
    get formDisabled() {
        return this._formDisabled;
    }
    set formDisabled(value) {
        this._formDisabled = value;
    }
    create(dashboardCfg, contextOrName) {
        return __awaiter(this, void 0, void 0, function* () {
            let id;
            let dashboardType;
            if (typeof contextOrName === 'string') {
                id = contextOrName;
                dashboardType = ContextDashboardType.Named;
            }
            else {
                id = contextOrName.contextData.id;
                dashboardType = this.getDashboardTypeFromViewContext(dashboardCfg, contextOrName);
            }
            const dashboard = {};
            assign(dashboard, { c8y_Dashboard: dashboardCfg });
            const value = dashboardType === ContextDashboardType.DeviceType ? dashboardCfg.deviceTypeValue : id;
            const fragmentKey = this.createFragmentKey(dashboardType, value);
            dashboard[fragmentKey] = {};
            if (this.shouldSetGlobal(dashboard)) {
                assign(dashboard, { c8y_Global: {} });
            }
            const { data } = dashboardType === ContextDashboardType.Group || dashboardType === ContextDashboardType.Device
                ? yield this.inventory.childAdditionsCreate(dashboard, id)
                : yield this.inventory.create(dashboard);
            return data;
        });
    }
    detail(dashboardMO) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.inventory.detail(dashboardMO);
            this.cache.set(dashboardMO.id, data);
            return data;
        });
    }
    update(dashboard) {
        return __awaiter(this, void 0, void 0, function* () {
            const cleanedDashboard = this.clean(pick(dashboard, [this.FRAGMENT_NAME, 'id']));
            cleanedDashboard.c8y_Global = this.shouldSetGlobal(dashboard);
            const { data } = yield this.inventory.update(cleanedDashboard);
            this.cache.set(dashboard.id, data);
            return data;
        });
    }
    delete(dashboard) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                let msg = gettext(`You are about to delete the dashboard "{{ dashboardName }}". Do you want to proceed?`);
                if (this.isDeviceType(dashboard)) {
                    msg = gettext(`You are about to delete the dashboard "{{ dashboardName }}" from all devices of the type "{{ deviceType }}".
           Do you want to proceed?`);
                }
                yield this.modal.confirm(gettext('Delete dashboard'), this.translateService.instant(msg, {
                    dashboardName: dashboard.c8y_Dashboard.name,
                    deviceType: dashboard.c8y_Dashboard.deviceTypeValue
                }), Status.DANGER, {
                    ok: gettext('Delete'),
                    cancel: gettext('Cancel')
                });
                yield this.inventory.delete(dashboard);
                const tabToRemove = Array.from(this.tabs.state).find(tab => tab.path.endsWith(`${this.DASHBOARD_ROUTE_PATH}/${dashboard.id}`));
                this.tabs.remove(tabToRemove);
                this.tabs.refresh();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    activateDashboards(route, types) {
        const { dashboardId } = route.params;
        if (dashboardId) {
            return this.getDashboard$(dashboardId, types, route.parent.data.contextData).pipe(tap(dashboard => {
                route.data = { dashboard };
            }), map(() => true), catchError(() => {
                return of(false);
            }));
        }
        return this.getTabs$(route.data.contextData, types);
    }
    getNamedDashboardOrCreate(name, defaultWidgets) {
        const children = this.mapWidgets(defaultWidgets);
        return this.getDashboard$(name, [ContextDashboardType.Named]).pipe(throwIfEmpty(), catchError(() => from(this.create({ children }, name))));
    }
    refreshTabs(dashboardMO) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.isNamed(dashboardMO)) {
                const tabToUpdate = Array.from(this.tabs.state).find(tab => tab.path.endsWith(`${this.DASHBOARD_ROUTE_PATH}/${dashboardMO.id}`));
                if (!tabToUpdate) {
                    this.addTab(dashboardMO);
                }
                else {
                    const data = yield this.detail(dashboardMO);
                    const { icon, priority, name } = data.c8y_Dashboard;
                    tabToUpdate.icon = icon;
                    tabToUpdate.priority = priority;
                    tabToUpdate.label = name;
                }
                this.tabs.refresh();
            }
        });
    }
    addTab(dashboard) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = yield this.detail(dashboard);
            const { icon, priority, name } = data.c8y_Dashboard;
            this.tabs.add({
                icon,
                priority,
                label: name,
                path: `${this.currentContextRoute}/${this.DASHBOARD_ROUTE_PATH}/${data.id}`
            });
        });
    }
    navigateToDashboard(dashboardMO) {
        return __awaiter(this, void 0, void 0, function* () {
            if (/dashboard/.test(this.router.url)) {
                this.router.navigate(['..', dashboardMO.id], {
                    relativeTo: getActivatedRoute(this.router)
                });
            }
            else {
                this.router.navigate(['..', 'dashboard', dashboardMO.id], {
                    relativeTo: getActivatedRoute(this.router)
                });
            }
        });
    }
    hasPermission() {
        return this.user.hasAnyRole(this.appState.currentUser.value, [
            'ROLE_INVENTORY_ADMIN',
            'ROLE_INVENTORY_CREATE'
        ]);
    }
    isNamed(dashboard) {
        return some(keys(dashboard), prop => new RegExp(`^${this.FRAGMENT_NAME}${this.INDEX_SPLIT}${ContextDashboardType.Named}${this.INDEX_SPLIT}`).test(prop));
    }
    isDeviceType(dashboard) {
        return some(keys(dashboard), prop => new RegExp(`^${this.FRAGMENT_NAME}${this.INDEX_SPLIT}${ContextDashboardType.DeviceType}${this.INDEX_SPLIT}`).test(prop));
    }
    getStyling(styleList, styleName, defaultValue) {
        const styling = styleList.find(style => style && new RegExp(styleName, 'i').test(style.class));
        return styling ? styling.class : defaultValue;
    }
    mapWidgets(widgets) {
        return keyBy(widgets.map(widget => {
            widget.id = String(Math.random()).substr(2);
            return widget;
        }), 'id');
    }
    getDashboard$(dashboardIdOrName, dashboardType, mo) {
        const cache = this.cache.get(dashboardIdOrName);
        const dashboards = mo
            ? this.getContextDashboards(mo, dashboardType)
            : [this.getNamedDashboard(dashboardIdOrName)];
        const cacheRefresh = this.cacheDashboards(dashboards).pipe(filter(dashboard => dashboard.id === dashboardIdOrName ||
            has(dashboard, `${this.FRAGMENT_NAME}${this.INDEX_SPLIT}${ContextDashboardType.Named}${this.INDEX_SPLIT}${dashboardIdOrName}`)));
        return cache ? of(cache) : cacheRefresh;
    }
    getTabs$(mo, dashboardType) {
        this.setBaseContextRoute(mo, dashboardType);
        return this.cacheDashboards(this.getContextDashboards(mo, dashboardType)).pipe(map(({ c8y_Dashboard: dashboard, id }) => ({
            icon: dashboard.icon,
            path: `${this.DASHBOARD_ROUTE_PATH}/${id}`,
            label: dashboard.name,
            priority: dashboard.priority
        })), toArray());
    }
    cacheDashboards(requests) {
        return from(requests).pipe(mergeAll(), mergeMap(response => response.data), tap(dashboard => this.cache.set(dashboard.id, dashboard)));
    }
    setBaseContextRoute(mo, dashboardType) {
        const type = dashboardType.includes(ContextDashboardType.Device)
            ? ContextDashboardType.Device
            : ContextDashboardType.Group;
        this.currentContextRoute = `${type}/${mo.id}`;
    }
    clean(dashboard) {
        const jsonString = JSON.stringify(dashboard, (key, value) => {
            if (key === '$$hashKey' || key === 'klasses') {
                return undefined;
            }
            return value;
        });
        return JSON.parse(jsonString);
    }
    getNamedDashboard(name) {
        return this.inventory.list({
            fragmentType: `${this.FRAGMENT_NAME}${this.INDEX_SPLIT}${ContextDashboardType.Named}${this.INDEX_SPLIT}${name}`,
            pageSize: 1
        });
    }
    getContextDashboards(mo, dashboardType) {
        return dashboardType.map((type) => this.inventory.list({
            fragmentType: `${this.FRAGMENT_NAME}${this.INDEX_SPLIT}${type}${this.INDEX_SPLIT}${type === ContextDashboardType.DeviceType ? mo.type : mo.id}`,
            pageSize: this.DEFAULT_PAGESIZE
        }));
    }
    getDashboardTypeFromViewContext(dashboardCfg, context) {
        let dashboardType;
        if (context.context === ViewContext.Device) {
            dashboardType = dashboardCfg.deviceType
                ? ContextDashboardType.DeviceType
                : ContextDashboardType.Device;
        }
        if (context.context === ViewContext.Group) {
            dashboardType = ContextDashboardType.Group;
        }
        return dashboardType;
    }
    createFragmentKey(contextDashboardType, value) {
        return [this.FRAGMENT_NAME, contextDashboardType, value].join(this.INDEX_SPLIT);
    }
    shouldSetGlobal(dashboard) {
        if (this.isNamed(dashboard) || this.isDeviceType(dashboard)) {
            return {};
        }
        return null;
    }
};
ContextDashboardService.ctorParameters = () => [
    { type: InventoryService },
    { type: TabsService },
    { type: ModalService },
    { type: TranslateService },
    { type: Router },
    { type: UserService },
    { type: AppStateService }
];
ContextDashboardService = __decorate([
    Injectable()
], ContextDashboardService);

let DeviceDashboardGuard = class DeviceDashboardGuard {
    constructor(contextDashboardService) {
        this.contextDashboardService = contextDashboardService;
    }
    canActivate(route) {
        return this.contextDashboardService.activateDashboards(route, [
            ContextDashboardType.Device,
            ContextDashboardType.DeviceType
        ]);
    }
};
DeviceDashboardGuard.ctorParameters = () => [
    { type: ContextDashboardService }
];
DeviceDashboardGuard = __decorate([
    Injectable()
], DeviceDashboardGuard);

let DashboardDetailComponent = class DashboardDetailComponent {
    constructor(modal, iconList, contextDashboardService) {
        this.modal = modal;
        this.contextDashboardService = contextDashboardService;
        this.styling = {
            themeClass: 'dashboard-theme-light',
            headerClass: 'panel-title-regular'
        };
        this.possibleStyling = { DASHBOARD_THEME_CLASSES, WIDGET_HEADER_CLASSES };
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
        this.DEFAULT_DASHBOARD_MARGIN = 12;
        this.DEFAULT_DASHBOARD_ICON = 'th';
        this.DEFAULT_DASHBOARD_PRIORITY = 10000;
        this.DEFAULT_DASHBOARD_NAME = 'Dashboard';
        this.icons = iconList;
        this.filteredIcons = iconList;
    }
    ngAfterContentInit() {
        const defaultDashboardCfg = {
            name: this.DEFAULT_DASHBOARD_NAME,
            priority: this.DEFAULT_DASHBOARD_PRIORITY,
            icon: this.DEFAULT_DASHBOARD_ICON,
            deviceTypeValue: this.deviceType
        };
        if (this.dashboard) {
            this.current = clone(this.dashboard);
            this.setDashboardStyle();
        }
        else {
            this.dashboard = defaultDashboardCfg;
            this.dashboardDetailForm.form.markAsDirty();
        }
    }
    save() {
        this.dashboard.classes = { [this.styling.themeClass]: true };
        this.dashboard.widgetClasses = { [this.styling.headerClass]: true };
        this._save(this.dashboard);
    }
    close() {
        this._cancel();
        this.modal.hide();
    }
    getDashboardPreviewStyle() {
        const cssClasses = {};
        cssClasses[this.styling.headerClass] = true;
        cssClasses[this.styling.themeClass] = true;
        return cssClasses;
    }
    selectIcon(icon) {
        this.dashboard.icon = icon;
        this.dashboardDetailForm.form.markAsDirty();
    }
    updateFiltered(term) {
        if (term) {
            const search = new RegExp(term, 'i');
            this.filteredIcons = this.icons.filter(val => search.test(val));
        }
        else {
            this.filteredIcons = this.icons;
        }
    }
    setDashboardStyle() {
        const allClasses = Object.assign({}, this.dashboard.classes, this.dashboard.widgetClasses);
        const styles = Object.keys(allClasses).map(c => c.split('-').pop());
        styles.forEach(styleName => {
            this.styling.themeClass = this.contextDashboardService.getStyling(DASHBOARD_THEME_CLASSES, styleName, this.styling.themeClass);
            this.styling.headerClass = this.contextDashboardService.getStyling(WIDGET_HEADER_CLASSES, styleName, this.styling.headerClass);
        });
    }
};
DashboardDetailComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: Array, decorators: [{ type: Inject, args: [ICON_LIST,] }] },
    { type: ContextDashboardService }
];
__decorate([
    ViewChild('dashboardDetailForm', { static: true })
], DashboardDetailComponent.prototype, "dashboardDetailForm", void 0);
DashboardDetailComponent = __decorate([
    Component({
        selector: 'c8y-dashboard-detail',
        template: "<div class=\"modal-header\">\n  <h3 *ngIf=\"!current\" translate>\n    Add dashboard\n  </h3>\n  <h3 *ngIf=\"current\" translate>Edit dashboard</h3>\n</div>\n\n<div class=\"modal-body\">\n  <form #dashboardDetailForm=\"ngForm\">\n    <div class=\"row\">\n      <div class=\"col-sm-6\">\n        <div *ngIf=\"!isNamedDashboard\">\n          <h6 class=\"legend form-block\"><span translate>Tab</span></h6>\n          <div class=\"input-group\">\n            <c8y-form-group>\n              <label class=\"m-r-8\" translate>Icon</label>\n              <div dropdown class=\"dropdown\">\n                <button\n                  title=\"{{ 'Icon' | translate }}\"\n                  class=\"btn-default btn btn-gray\"\n                  dropdownToggle\n                >\n                  <i c8yIcon=\"{{dashboard.icon}}\"></i>\n                  <span class=\"caret\"></span> \n                </button>                  \n                <ul\n                  *dropdownMenu\n                  class=\"dropdown-menu modal-inner-scroll dropdown-menu-grid-4\"\n                  style=\"max-height: 250px;margin-left: 0;\"\n                >\n                  <ng-container *ngFor=\"let icon of filteredIcons\">\n                    <li (click)=\"selectIcon(icon)\">\n                      <a style=\"cursor: pointer\" title=\"{{icon}}\" [ngClass]=\"{'active': dashboard.icon === icon}\">\n                        <i class=\"icon\" [c8yIcon]=\"icon\" ></i>\n                      </a>\n                    </li>\n                  </ng-container>\n                </ul>\n              </div>\n            </c8y-form-group>\n            <c8y-form-group class=\"flex-grow\">\n              <label>\n                <span class=\"m-r-4\" translate>Menu label</span>\n                <button class=\"btn btn-clean\"\n                  popover=\"{{\n                    'Menu label to display in submenu when dashboard is attached' | translate\n                  }}\"\n                  triggers=\"focus\"\n                  placement=\"right\"\n                  container=\"body\"\n                >\n                  <i\n                    [c8yIcon]=\"'question-circle-o'\"\n                    class=\"text-primary\"\n                  ></i>\n                </button>\n              </label>\n              <input\n                title=\"{{ 'Menu label' | translate }}\"\n                class=\"form-control\"\n                name=\"name\"\n                [(ngModel)]=\"dashboard.name\"\n                placeholder=\"{{ 'e.g. My dashboard' | translate }}\"\n                required\n              />\n            </c8y-form-group>\n          </div>\n          <div class=\"row\">\n            <div class=\"col-sm-6\">\n              <c8y-form-group>\n                <label>\n                  <span translate class=\"m-r-4\">Position in navigation</span>\n                  <button class=\"btn btn-clean\"\n                    popover=\"{{\n                      'Position in navigation menu (10000 first, -10000 last)' | translate\n                    }}\"\n                    triggers=\"focus\"\n                    placement=\"right\"\n                    container=\"body\"\n                  >\n                    <i \n                      [c8yIcon]=\"'question-circle-o'\"\n                      class=\"text-primary\"\n                    ></i>\n                  </button>\n                  \n                </label>\n                <input\n                  title=\"{{ 'Position in navigation' | translate }}\"\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"priority\"\n                  [(ngModel)]=\"dashboard.priority\"\n                  min=\"-10000\"\n                  max=\"10000\"\n                  placeholder=\"{{ 'e.g.' | translate }} 500\"\n                  required\n                />\n              </c8y-form-group>\n            </div>\n            <div class=\"col-sm-6\">\n             \n            </div>\n\n            <div style=\"position: relative;\">\n              <div class=\"col-sm-12\">\n                <div class=\"form-group m-b-24\" *ngIf=\"!current && deviceType\">\n                  <label class=\"c8y-checkbox\">\n                    <input type=\"checkbox\" name=\"deviceType\" [(ngModel)]=\"dashboard.deviceType\" />\n                    <span></span>\n                    {{ 'Apply dashboard to all devices of type ' | translate\n                    }}<i>{{ dashboard.deviceTypeValue }}</i>\n                  </label>\n                </div>\n\n                <div\n                  class=\"alert alert-info  m-b-24\"\n                  *ngIf=\"dashboard.deviceType && dashboard.deviceTypeValue\"\n                >\n                  <i c8y-icon=\"info\"></i>\n                  {{ 'This dashboard is shared between all devices of the type ' | translate }}\n                  <i>{{ dashboard.deviceTypeValue }}</i>\n                </div>\n              </div>\n            </div>\n          </div>\n        </div>\n\n        <div>\n          <c8y-appearance-settings\n            [(themeClass)]=\"styling.themeClass\"\n            [(headerClass)]=\"styling.headerClass\"\n          >\n          </c8y-appearance-settings>\n          <div class=\"row\">\n            <div class=\"col-sm-6\">\n              <c8y-form-group>\n                <label>{{ 'Widget margin' | translate }}</label>\n                <div class=\"input-group\">\n                  <input\n                    title=\"{{ 'Widget margin' | translate }}\"\n                    id=\"margin\"\n                    name=\"margin\"\n                    type=\"number\"\n                    class=\"form-control\"\n                    [(ngModel)]=\"dashboard.widgetMargin\"\n                    min=\"0\"\n                    max=\"50\"\n                    placeholder=\"{{ DEFAULT_DASHBOARD_MARGIN }}\"\n                  />\n                  <span class=\"input-group-addon\">px</span>\n                </div>\n              </c8y-form-group>\n            </div>\n            <div class=\"col-sm-6\">\n              <label translate>Widget titles</label>\n              <c8y-form-group>\n                <label\n                  title=\"{{ 'Translate widget titles if possible' | translate }}\"\n                  class=\"c8y-checkbox\"\n                >\n                  <input\n                    type=\"checkbox\"\n                    name=\"translateWidgetTitle\"\n                    [(ngModel)]=\"dashboard.translateWidgetTitle\"\n                  /><span></span>\n                  <span>{{ 'Translate if possible' | translate }}</span>\n                </label>\n              </c8y-form-group>\n            </div>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"col-sm-6\">\n        <c8y-widget-preview\n          [tab]=\"!isNamedDashboard ? dashboard : undefined\"\n          [previewClasses]=\"getDashboardPreviewStyle()\"\n        ></c8y-widget-preview>\n      </div>\n    </div>\n  </form>\n</div>\n\n<div class=\"modal-footer\">\n  <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\" translate>\n    Cancel\n  </button>\n  <button\n    title=\"{{ 'Save' | translate }}\"\n    class=\"btn btn-primary\"\n    (click)=\"save()\"\n    [disabled]=\"dashboardDetailForm.form.invalid || dashboardDetailForm.form.pristine\"\n    translate\n  >\n    Save\n  </button>\n</div>\n"
    }),
    __param(1, Inject(ICON_LIST))
], DashboardDetailComponent);

let WidgetService = class WidgetService {
    constructor(dynamicComponentService, translateService, moduleConfig) {
        this.dynamicComponentService = dynamicComponentService;
        this.translateService = translateService;
        this.moduleConfig = moduleConfig;
    }
    getWidgetDefinition(componentId) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.dynamicComponentService.getById$(componentId).toPromise();
        });
    }
    getWidgetDefinitions() {
        const translatedComponents = Array.from(this.dynamicComponentService.state).map(cmp => (Object.assign({}, cmp, { label: this.translateService.instant(cmp.label) })));
        return cloneDeep(sortBy(translatedComponents, 'label').filter(this.moduleConfig.widgetFilter));
    }
};
WidgetService.ctorParameters = () => [
    { type: DynamicComponentService },
    { type: TranslateService },
    { type: undefined, decorators: [{ type: Inject, args: [CONTEXT_DASHBOARD_CONFIG,] }] }
];
WidgetService.ngInjectableDef = ɵɵdefineInjectable({ factory: function WidgetService_Factory() { return new WidgetService(ɵɵinject(DynamicComponentService), ɵɵinject(TranslateService), ɵɵinject(CONTEXT_DASHBOARD_CONFIG)); }, token: WidgetService, providedIn: "root" });
WidgetService = __decorate([
    Injectable({
        providedIn: 'root'
    }),
    __param(2, Inject(CONTEXT_DASHBOARD_CONFIG))
], WidgetService);

let WidgetConfigComponent = class WidgetConfigComponent {
    constructor(widgetService, modal, contextDashboardService) {
        this.widgetService = widgetService;
        this.modal = modal;
        this.contextDashboardService = contextDashboardService;
        this.mode = 'select';
        this.searchChange$ = new Subject();
        this.searchTerm = '';
        this.styling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.defaultStyling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.possibleStyling = { WIDGET_HEADER_CLASSES, WIDGET_CONTENT_CLASSES };
        this.isUpgrade = false;
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
    }
    get isEdit() {
        return !!this.current;
    }
    ngAfterContentInit() {
        this.components = this.widgetService.getWidgetDefinitions();
        if (this.selected) {
            this.current = clone(this.selected);
            this.select(this.selected, this.isEdit ? 'config' : 'select');
        }
        this.searchSub = this.searchChange$
            .pipe(switchMap((event) => iif(() => event.which !== 13, timer(200), timer(10))))
            .subscribe(() => {
            this.search();
        });
        if (this.mo.c8y_Dashboard.classes) {
            this.styling = this.setDefaultStyle(Object.assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses, (this.isEdit ? this.current.data.classes : {})));
            this.defaultStyling = this.setDefaultStyle(Object.assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses));
        }
    }
    save() {
        const { _x, _y, _width, _height } = this.selected.data;
        if (this.widgetConfig && this.widgetConfig.device) {
            const { id, name } = this.widgetConfig.device;
            this.widgetConfig.device = { id, name };
        }
        const widget = Object.assign({ _x,
            _y,
            _width,
            _height, config: omit(this.widgetConfig, 'settings'), title: this.selected.data.title, componentId: this.selected.id, id: this.isEdit ? this.current.data.id : String(Math.random()).substr(2), classes: this.getStyle() }, (!this.isEdit ? this.widgetConfig.settings.widgetDefaults : {}));
        this._save(widget);
    }
    select(cmp, mode = 'config') {
        cmp.data = cmp.data || {};
        this.selected = cmp;
        this.isUpgrade = !!get(cmp, 'data.settings.upgrade');
        this.contextDashboardService.formDisabled = this.isUpgrade;
        if (this.isEdit) {
            const { _x, _y, _width, _height, classes, title } = this.current.data;
            this.selected.data = Object.assign({}, this.selected.data, { _x, _y, _width, _height, classes, title });
        }
        this.widgetConfig = cloneDeep(this.composeWidgetConfig(this.selected, this.context));
        this.selected.data.title = this.selected.data.title || cmp.label;
        this.componentLabel = cmp.label;
        this.mode = mode;
    }
    search() {
        if (this.searchTerm.length > 0) {
            this.searchResult = this.components.filter(cmp => new RegExp(escapeRegExp(this.searchTerm.trim()), 'i').test(cmp.label));
        }
        else {
            this.resetSearch();
        }
    }
    resetSearch() {
        this.searchTerm = '';
        this.searchResult = undefined;
    }
    changeMode(mode) {
        this.mode = mode;
    }
    close() {
        this._cancel();
        this.modal.hide();
    }
    getStyle(isPreview = false) {
        const cssClasses = {};
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.headerClass)) {
            cssClasses[this.styling.headerClass] = true;
        }
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.contentClass)) {
            cssClasses[this.styling.contentClass] = true;
        }
        if (isPreview) {
            cssClasses[`dashboard-theme-${this.defaultStyling.contentClass.split('-').pop()}`] = true;
        }
        return cssClasses;
    }
    ngOnDestroy() {
        this.contextDashboardService.formDisabled = true;
        if (this.searchSub) {
            this.searchSub.unsubscribe();
        }
    }
    isDashboardDefaultStyle(className) {
        const allClasses = Object.assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses);
        const styles = Object.keys(allClasses).map(cssClass => ({ class: cssClass }));
        const style = this.contextDashboardService.getStyling(styles, className.split('-').pop(), undefined);
        return !!style;
    }
    setDefaultStyle(setClasses) {
        let contentClass = this.styling.contentClass;
        let headerClass = this.styling.headerClass;
        const styles = Object.keys(setClasses).map(c => c.split('-').pop());
        styles.forEach(styleName => {
            contentClass = this.contextDashboardService.getStyling(WIDGET_CONTENT_CLASSES, styleName, contentClass);
            headerClass = this.contextDashboardService.getStyling(WIDGET_HEADER_CLASSES, styleName, headerClass);
        });
        return { headerClass, contentClass };
    }
    composeWidgetConfig(selectedComponent, context = {}) {
        const setting = Object.assign({ settings: Object.assign({}, selectedComponent.data.settings, get(selectedComponent.data.settings, 'ng1.options'), get(selectedComponent.data, 'ng1.options'), { context }) }, selectedComponent.data.config, { dashboardMo: this.mo.c8y_Dashboard });
        return this.applyTargetIfDeviceDashboard(setting);
    }
    applyTargetIfDeviceDashboard(widgetConfig) {
        const isDeviceType = this.contextDashboardService.isDeviceType(this.mo);
        if (isDeviceType) {
            widgetConfig.settings.hideTarget = isDeviceType;
            const noDeviceTarget = widgetConfig.settings.ng1
                ? widgetConfig.settings.ng1.options.noDeviceTarget
                : widgetConfig.settings.noDeviceTarget;
            if (!noDeviceTarget) {
                widgetConfig.device = {
                    id: this.context.id,
                    name: this.context.name
                };
            }
        }
        return widgetConfig;
    }
};
WidgetConfigComponent.ctorParameters = () => [
    { type: WidgetService },
    { type: BsModalRef },
    { type: ContextDashboardService }
];
WidgetConfigComponent = __decorate([
    Component({
        selector: 'c8y-widget-config',
        template: "<div class=\"modal-header separator\">\n  <h3 *ngIf=\"!current\" translate>Add widget</h3>\n  <h3 *ngIf=\"current\" translate>Edit widget</h3>\n</div>\n<form #configForm=\"ngForm\" name=\"form\">\n  <div class=\"c8y-modal-tabs\">\n    <div class=\"tabContainer\">\n      <ul class=\"nav nav-tabs nav-tabsc8y p-l-24\">\n        <li [class.active]=\"mode === 'select'\">\n          <button type=\"button\" class=\"btn\" (click)=\"changeMode('select'); (false)\">\n            <i c8yIcon=\"th-large\"></i> <span class=\"txt\" translate>Select widget</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'config'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('config'); (false)\"\n          >\n            <i c8yIcon=\"cog\"></i> <span class=\"txt\" translate>Configuration</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'style'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('style'); (false)\"\n          >\n            <i c8yIcon=\"paint-brush\"></i> <span class=\"txt\" translate>Appearance</span>\n          </button>\n        </li>\n      </ul>\n    </div>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <div\n      *ngIf=\"mode === 'select'\"\n      class=\"bg-white p-l-24 p-r-24 p-t-8 p-b-8 sticky-header-top-0\"\n      style=\"z-index: 2;\"\n    >\n      <div class=\"row\">\n        <div class=\"col-sm-6\">\n          <div class=\"input-group input-group-search\">\n            <input\n              class=\"form-control\"\n              placeholder=\"{{ 'Search\u2026' | translate }}\"\n              type=\"text\"\n              [(ngModel)]=\"searchTerm\"\n              [ngModelOptions]=\"{ standalone: true }\"\n              (keydown)=\"searchChange$.next($event)\"\n            />\n            <span class=\"input-group-btn\">\n              <button class=\"btn btn-clean\" (click)=\"resetSearch()\">\n                <i [c8yIcon]=\"searchTerm.length === 0 ? 'search' : 'close'\"></i>\n              </button>\n            </span>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div class=\"modal-body bg-gray-lighter\" *ngIf=\"mode === 'select'\">\n      <div class=\"card-group card-select\" style=\"margin-bottom:-24px;\">\n        <div\n          class=\"col-md-3 col-sm-4 col-xs-6\"\n          *ngFor=\"let cmp of searchResult || components\"\n          title=\"{{ cmp.description | translate }}\"\n        >\n          <div class=\"card p-8\" [class.active]=\"selected === cmp\" (click)=\"select(cmp)\">\n            <div\n              class=\"text-center p-8 m-b-8 flex-col flex-center\"\n              style=\"min-height: 170px; background-color: var(--body-background-color, #f2f3f4);\"\n            >\n              <ng-container *ngIf=\"!cmp.previewImage; else previewImage\">\n                <h1><i c8yIcon=\"file-image-o\"></i></h1>\n                <small translate>Preview not available</small>\n              </ng-container>\n              <ng-template #previewImage>\n                <img class=\"img-responsive\" [src]=\"cmp.previewImage\" />\n              </ng-template>\n            </div>\n            <p class=\"card-title text-truncate\">\n              <c8y-highlight\n                text=\"{{ cmp.label | translate }}\"\n                [pattern]=\"searchTerm\"\n              ></c8y-highlight>\n            </p>\n          </div>\n        </div>\n\n        <div class=\"c8y-empty-state text-center\" *ngIf=\"searchResult && searchResult.length === 0\">\n          <h1 class=\"c8y-icon c8y-icon-device c8y-icon-duocolor\"></h1>\n          <h3 translate>No widgets found.</h3>\n          <div>\n            <p translate>Rephrase your search term.</p>\n            <div>\n              <button class=\"btn btn-primary\" (click)=\"resetSearch()\" translate>\n                Reset search\n              </button>\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n\n    <div class=\"modal-body\" *ngIf=\"mode === 'config'\" style=\"min-height: calc(100vh - 290px);\">\n      <h4 style=\"text-align: left;\">\n        <strong>\n          {{ selected.label | translate }}\n        </strong>\n      </h4>\n      <p class=\"m-b-24\">\n        {{ selected.description | translate }}\n      </p>\n      <div class=\"row\">\n        <!-- change to col-sm-6 when preview is available -->\n        <div class=\"col-sm-12\">\n          <div class=\"legend form-block\" translate>Configuration</div>\n          <c8y-form-group>\n            <label for=\"widgetTitle\" translate>Title</label>\n            <input\n              id=\"widgetTitle\"\n              [(ngModel)]=\"selected.data.title\"\n              type=\"text\"\n              name=\"title\"\n              class=\"form-control\"\n              placeholder=\"{{ 'e.g.' | translate }} {{ componentLabel | translate }}\"\n              required\n            />\n          </c8y-form-group>\n\n          <!-- This is an upgraded component for the device selector and still needs to be migrated -->\n          <c8y-dynamic-component\n            componentId=\"device.selector.legacy\"\n            [config]=\"widgetConfig\"\n            [notFoundError]=\"false\"\n            *ngIf=\"!isUpgrade\"\n          ></c8y-dynamic-component>\n\n          <c8y-dynamic-component\n            [componentId]=\"selected.id\"\n            mode=\"config\"\n            [config]=\"widgetConfig\"\n            [notFoundError]=\"false\"\n          ></c8y-dynamic-component>\n        </div>\n        <!-- markup for the preview \n        \n        <div class=\"col-sm-6 sticky-header-top-0\" >\n          <div class=\"legend form-block\">\n            Preview\n          </div>\n          <div class=\"bg-gray-lighter p-16\">\n            <div class=\"card card-dashboard\" style=\"margin-bottom:0;\">\n              <div class=\"card-header-actions\">\n                <div class=\"card-title text-uppercase\">{{ selected.data.title }}</div>\n                <div class=\"header-actions\">\n                  <div class=\"optionsBtn dropdow\" class=\"btnIcon dropdown-toggle c8y-dropdown\">\n                    <i c8yIcon=\"cog\"></i>\n                  </div>\n                </div>\n              </div>\n              <div class=\"card-inner-scroll\">\n                <div class=\"card-block\" style=\"min-height: 240px;\">\n                  include here the widget\n                </div>\n              </div>\n            </div>\n          </div>\n        </div>\n      --></div>\n    </div>\n\n    <div *ngIf=\"mode === 'style'\" class=\"modal-body p-t-0\" style=\"min-height: calc(100vh - 290px);\">\n      <div class=\"row\">\n        <div class=\"col-xs-6\">\n          <c8y-appearance-settings\n            [(themeClass)]=\"styling.contentClass\"\n            [(headerClass)]=\"styling.headerClass\"\n            [possibleStylingTheme]=\"possibleStyling.WIDGET_CONTENT_CLASSES\"\n            [possibleStylingHeader]=\"possibleStyling.WIDGET_HEADER_CLASSES\"\n            [defaultThemeClass]=\"defaultStyling.contentClass\"\n            [defaultHeaderClass]=\"defaultStyling.headerClass\"\n          >\n          </c8y-appearance-settings>\n        </div>\n        <div class=\"col-xs-6 sticky-header-top-0\">\n          <c8y-widget-preview [previewClasses]=\"getStyle(true)\"></c8y-widget-preview>\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\" translate>\n      Cancel\n    </button>\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      translate\n      [disabled]=\"contextDashboardService.formDisabled || configForm.invalid\"\n    >\n      Save\n    </button>\n  </div>\n</form>\n"
    })
], WidgetConfigComponent);

/**
 * The context dashboard is a dashboard which resolves it data from the current context (device or group)
 * it is displayed on. It usually uses the route.data for it, but you can pass
 * a different managedObject to the [mo] input parameter to change that behavior.
 */
let ContextDashboardComponent = class ContextDashboardComponent {
    constructor(route, router, contextDashboardService, alert, renderer, moduleConfig, widgetService, bsModal) {
        this.route = route;
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.alert = alert;
        this.renderer = renderer;
        this.moduleConfig = moduleConfig;
        this.widgetService = widgetService;
        this.bsModal = bsModal;
        this.childrenClasses = '';
        this.setTitle = false;
        this.disabled = false;
        this.defaultWidgets = [];
        this.canDelete = true;
        this.isLoading = true;
        this.class = '';
        this.widgets = [];
    }
    ngOnInit() {
        if (!this.name) {
            this.loadContextDashboard();
            return;
        }
        this.loadNamedDashboard();
    }
    /**
     * Applies the current context to the widget
     * @param widget The widget to apply the context to.
     */
    applyDeviceTarget(widget) {
        if (widget.config.device) {
            widget.config.device = { id: this.context.id, name: this.context.name };
        }
    }
    /**
     * Removes the route listener.
     */
    ngOnDestroy() {
        if (this.dataSub) {
            this.dataSub.unsubscribe();
        }
    }
    /**
     * Restores the dashboard widgets to the default widgets.
     */
    restore() {
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            this.mo.c8y_Dashboard.children = this.contextDashboardService.mapWidgets(this.defaultWidgets);
            yield this.contextDashboardService.update(this.mo);
            this.onLoad();
        });
    }
    /**
     * Updates all dashboards children's. Useful for position changes on the dashboard.
     * @param child The child to change.
     */
    updateDashboardChildren(child) {
        return __awaiter(this, void 0, void 0, function* () {
            const { children } = child;
            const dashboardMO = this.mo;
            const mappedChildren = keyBy(children.map(c => this.componentToWidget(c)), 'id');
            dashboardMO.c8y_Dashboard.children = mappedChildren;
            return this.contextDashboardService.update(dashboardMO);
        });
    }
    /**
     * Remove the complete dashboard and navigate away.
     */
    deleteDashboard() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.contextDashboardService.delete(this.mo);
            if (this.route.parent) {
                const route = this.route.parent.snapshot.url.map(segment => segment.path).join('/');
                this.router.navigateByUrl(route);
            }
        });
    }
    /**
     * Edits the current dashboard
     */
    editDashboard() {
        return __awaiter(this, void 0, void 0, function* () {
            const initialState = {
                dashboard: this.dashboard,
                deviceType: this.context.type,
                isNamedDashboard: this.contextDashboardService.isNamed(this.mo)
            };
            const modal = this.bsModal.show(DashboardDetailComponent, {
                class: 'modal-lg',
                initialState,
                ignoreBackdropClick: true
            }).content;
            try {
                const dashboardMO = cloneDeep(this.mo);
                dashboardMO.c8y_Dashboard = yield modal.result;
                yield this.contextDashboardService.update(dashboardMO);
                yield this.contextDashboardService.refreshTabs(dashboardMO);
                this.onLoad();
                modal.close();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    /**
     * Edits a widget on the dashboard.
     * @param change The widget change event.
     */
    editWidget(change) {
        return __awaiter(this, void 0, void 0, function* () {
            const { x, y, width, height } = change.source;
            const component = yield this.widgetService.getWidgetDefinition(change.widget.name || change.widget.componentId);
            if (!component) {
                this.addWidget();
                return;
            }
            yield this.addWidget(Object.assign({}, component, { data: Object.assign({}, component.data, change.widget, { _x: x, _y: y, _width: width, _height: height }) }));
        });
    }
    /**
     * Adds a widget to the dashboard.
     * @param selected Define a selected component to switch to edit mode directly.
     */
    addWidget(selected) {
        return __awaiter(this, void 0, void 0, function* () {
            const initialState = {
                mo: this.mo,
                context: this.context,
                selected: cloneDeep(selected)
            };
            const modal = this.bsModal.show(WidgetConfigComponent, {
                class: 'modal-lg',
                initialState,
                ignoreBackdropClick: true
            }).content;
            try {
                const newWidget = yield modal.result;
                if (!this.mo.c8y_Dashboard.children) {
                    this.mo.c8y_Dashboard.children = {};
                }
                this.mo.c8y_Dashboard.children[newWidget.id] = newWidget;
                this.contextDashboardService.update(this.mo);
                newWidget.classes = this.mergeWidgetClasses(newWidget);
                yield this.updateWidget(newWidget);
                modal.close();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    /**
     * Updates a widget or adds a new one if it doesn't exist on
     * the dashboard.
     * @param widget The new widget
     */
    updateWidget(widget) {
        return __awaiter(this, void 0, void 0, function* () {
            const index = findIndex(this.widgets, { id: widget.id });
            const isNew = index === -1;
            const mappedWidget = yield this.mapLegacy(widget);
            if (isNew) {
                this.widgets.push(mappedWidget);
            }
            else {
                this.widgets.splice(index, 1, mappedWidget);
            }
        });
    }
    /**
     * Removes a widget and rearranges the remaining ones
     * if necessary.
     * @param change The change event.
     */
    deleteWidget(change) {
        const { widget, source } = change;
        delete this.mo.c8y_Dashboard.children[widget.id];
        const removed = this.widgets.find(({ id }) => id === widget.id);
        this.widgets.splice(this.widgets.indexOf(removed), 1);
        // using setTimeout to give the component the chance to remove it.
        setTimeout(() => {
            const child = new DashboardChildChange(source);
            child.collapseUpAll();
            this.updateDashboardChildren(child);
        });
    }
    /**
     * This is a workaround to ensure that the dragged-element
     * (which is attached to the body) has the right styling.
     */
    addDashboardClassToBody() {
        this.class.split(' ').forEach(cssClass => {
            this.renderer.addClass(document.body, cssClass);
        });
    }
    /**
     * This is a workaround to ensure that the dragged-element
     * (which is attached to the body) has the right styling.
     */
    removeDashboardClassFromBody() {
        this.class.split(' ').forEach(cssClass => {
            this.renderer.removeClass(document.body, cssClass);
        });
    }
    /**
     * Changes the dashboard settings to frozen or vice versa.
     * @param settings The current settings of the dashboard.
     */
    toggleFreeze(settings) {
        return __awaiter(this, void 0, void 0, function* () {
            this.toggleIsFrozenFlag(settings);
            try {
                yield this.contextDashboardService.update(this.mo);
                if (this.dashboard.isFrozen) {
                    this.alert.success(gettext('Your dashboard is locked now.'));
                }
                else {
                    this.alert.success(gettext('Your dashboard is unlocked now.'));
                }
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
                this.toggleIsFrozenFlag(settings);
            }
        });
    }
    toggleIsFrozenFlag(settings) {
        settings.isFrozen = !settings.isFrozen;
        this.dashboard.isFrozen = settings.isFrozen;
    }
    loadContextDashboard() {
        this.dataSub = this.route.data.subscribe(({ dashboard }) => {
            this.context = this.route.parent.snapshot.data.contextData;
            this.mo = dashboard;
            this.dashboard = this.mo.c8y_Dashboard;
            this.onLoad();
        });
    }
    loadNamedDashboard() {
        this.dataSub = this.contextDashboardService
            .getNamedDashboardOrCreate(this.name, this.defaultWidgets)
            .subscribe(mo => {
            this.context = this.context || {};
            this.mo = mo;
            this.dashboard = this.mo.c8y_Dashboard;
            this.onLoad();
        });
    }
    onLoad() {
        return __awaiter(this, void 0, void 0, function* () {
            this.disabled = !this.contextDashboardService.hasPermission();
            const dashboardChildren = cloneDeep(this.mo.c8y_Dashboard.children);
            const isDeviceType = this.contextDashboardService.isDeviceType(this.mo);
            const dashboardClasses = Object.assign({ 'c8y-grid-dashboard': true, dashboard: true }, this.dashboard.classes);
            this.widgets = yield Promise.all(values(dashboardChildren).map(widget => {
                widget.classes = this.mergeWidgetClasses(widget);
                if (isDeviceType) {
                    this.applyDeviceTarget(widget);
                }
                return this.mapLegacy(widget);
            }));
            this.class = Object.keys(dashboardClasses).join(' ');
            this.disabled = !this.contextDashboardService.hasPermission();
            this.isLoading = false;
        });
    }
    mergeWidgetClasses(widget) {
        const hasHeaderClass = WIDGET_HEADER_CLASSES.find(el => widget.classes && widget.classes[el.class]);
        const widgetClasses = hasHeaderClass
            ? Object.assign({}, widget.classes) : Object.assign({}, this.dashboard.widgetClasses, widget.classes);
        return Object.assign({ card: true, 'card-dashboard': true }, widgetClasses);
    }
    componentToWidget(child) {
        return Object.assign({}, omit(child.data, ['componentTransformConfigWithContext', 'transformConfigWithContext']), {
            _x: child.x,
            _y: child.y,
            _width: child.width,
            _height: child.height
        });
    }
    mapLegacy(widget) {
        return __awaiter(this, void 0, void 0, function* () {
            const cmp = yield this.widgetService.getWidgetDefinition(widget.componentId || widget.name);
            if (get(cmp, 'data.settings.upgrade')) {
                widget.widgetComponent = cmp.data.settings.widgetComponent;
                widget.configComponent = cmp.data.settings.configComponent;
                widget.templateUrl = cmp.data.settings.templateUrl;
                widget.configTemplateUrl = cmp.data.settings.configTemplateUrl;
                widget.transformConfigWithContext =
                    cmp.data.settings.componentTransformConfigWithContext ||
                        cmp.data.settings.transformConfigWithContext;
            }
            return widget;
        });
    }
};
ContextDashboardComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: Router },
    { type: ContextDashboardService },
    { type: AlertService },
    { type: Renderer2 },
    { type: undefined, decorators: [{ type: Inject, args: [CONTEXT_DASHBOARD_CONFIG,] }] },
    { type: WidgetService },
    { type: BsModalService }
];
__decorate([
    Input()
], ContextDashboardComponent.prototype, "name", void 0);
__decorate([
    Input()
], ContextDashboardComponent.prototype, "childrenClasses", void 0);
__decorate([
    Input()
], ContextDashboardComponent.prototype, "context", void 0);
__decorate([
    Input()
], ContextDashboardComponent.prototype, "setTitle", void 0);
__decorate([
    Input()
], ContextDashboardComponent.prototype, "disabled", void 0);
__decorate([
    Input()
], ContextDashboardComponent.prototype, "defaultWidgets", void 0);
__decorate([
    Input()
], ContextDashboardComponent.prototype, "canDelete", void 0);
__decorate([
    Input()
], ContextDashboardComponent.prototype, "isLoading", void 0);
__decorate([
    HostBinding('class')
], ContextDashboardComponent.prototype, "class", void 0);
ContextDashboardComponent = __decorate([
    Component({
        selector: 'c8y-context-dashboard',
        template: "<c8y-action-bar-item [placement]=\"'more'\" *ngIf=\"defaultWidgets.length > 0\">\n  <button (click)=\"restore()\" [disabled]=\"dashboard?.isFrozen || disabled\">\n    <i c8yIcon=\"undo\"></i>&nbsp;<span translate>Restore dashboard</span>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-widgets-dashboard\n  [context]=\"context\"\n  [contextDashboard]=\"dashboard\"\n  [widgets]=\"widgets\"\n  [settings]=\"{\n    isLoading: isLoading,\n    isFrozen: dashboard?.isFrozen,\n    isDisabled: disabled,\n    canDelete: canDelete,\n    translateWidgetTitle: dashboard?.translateWidgetTitle,\n    allowFullscreen: moduleConfig.allowFullscreen,\n    title: setTitle ? dashboard.name : undefined,\n    widgetMargin: dashboard?.widgetMargin\n  }\"\n  (onFreeze)=\"toggleFreeze($event)\"\n  (onChangeDashboard)=\"updateDashboardChildren($event)\"\n  (onAddWidget)=\"addWidget()\"\n  (onEditWidget)=\"editWidget($event)\"\n  (onDeleteWidget)=\"deleteWidget($event)\"\n  (onChangeStart)=\"addDashboardClassToBody()\"\n  (onChangeEnd)=\"removeDashboardClassFromBody()\"\n  (onEditDashboard)=\"editDashboard()\"\n  (onDeleteDashboard)=\"deleteDashboard()\"\n>\n</c8y-widgets-dashboard>\n",
        host: {
            style: `
      display: block;
    `,
            class: 'dashboard c8y-grid-dashboard'
        }
    }),
    __param(5, Inject(CONTEXT_DASHBOARD_CONFIG))
], ContextDashboardComponent);

let GroupDashboardGuard = class GroupDashboardGuard {
    constructor(contextDashboardService) {
        this.contextDashboardService = contextDashboardService;
    }
    canActivate(route) {
        return this.contextDashboardService.activateDashboards(route, [
            ContextDashboardType.Group
        ]);
    }
};
GroupDashboardGuard.ctorParameters = () => [
    { type: ContextDashboardService }
];
GroupDashboardGuard = __decorate([
    Injectable()
], GroupDashboardGuard);

let DashboardActionsFactory = class DashboardActionsFactory {
    constructor(router, contextDashboardService, bsModal) {
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
    }
    get(activatedRoute) {
        const context = !activatedRoute.parent || activatedRoute.snapshot.data.context
            ? activatedRoute.snapshot.data
            : activatedRoute.parent.snapshot.data;
        return [
            {
                label: gettext('Add dashboard'),
                priority: 2500,
                action: () => this.addDashboard(context),
                disabled: !this.isEnabled(context)
            }
        ];
    }
    addDashboard(context) {
        return __awaiter(this, void 0, void 0, function* () {
            let initialState;
            if (context.context === ViewContext.Device) {
                initialState = {
                    deviceType: context.contextData.type
                };
            }
            const modal = this.bsModal.show(DashboardDetailComponent, {
                class: 'modal-lg',
                initialState,
                ignoreBackdropClick: true
            }).content;
            try {
                const dashboardCfg = yield modal.result;
                const dashboardMO = yield this.contextDashboardService.create(dashboardCfg, context);
                yield this.contextDashboardService.navigateToDashboard(dashboardMO);
                yield this.contextDashboardService.refreshTabs(dashboardMO);
                modal.close();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    isEnabled(context) {
        return (this.contextDashboardService.hasPermission() &&
            [ViewContext.Device, ViewContext.Group].includes(context.context));
    }
};
DashboardActionsFactory.ctorParameters = () => [
    { type: Router },
    { type: ContextDashboardService },
    { type: BsModalService }
];
DashboardActionsFactory = __decorate([
    Injectable()
], DashboardActionsFactory);

let AppearanceSettingsComponent = class AppearanceSettingsComponent {
    constructor() {
        this.themeClass = 'dashboard-theme-light';
        this.headerClass = 'panel-title-regular';
        this.themeClassChange = new EventEmitter();
        this.headerClassChange = new EventEmitter();
        this.possibleStylingTheme = DASHBOARD_THEME_CLASSES;
        this.possibleStylingHeader = WIDGET_HEADER_CLASSES;
    }
    themeClassClick(value) {
        value = this.resetToDefault(value, this.themeClass, this.defaultThemeClass);
        this.themeClass = value;
        this.themeClassChange.emit(value);
    }
    headerClassClick(value) {
        value = this.resetToDefault(value, this.headerClass, this.defaultHeaderClass);
        this.headerClass = value;
        this.headerClassChange.emit(value);
    }
    resetToDefault(value, compareTo, defaultValue) {
        if (defaultValue && value === compareTo) {
            value = defaultValue;
        }
        return value;
    }
};
__decorate([
    Input()
], AppearanceSettingsComponent.prototype, "themeClass", void 0);
__decorate([
    Input()
], AppearanceSettingsComponent.prototype, "headerClass", void 0);
__decorate([
    Input()
], AppearanceSettingsComponent.prototype, "defaultThemeClass", void 0);
__decorate([
    Input()
], AppearanceSettingsComponent.prototype, "defaultHeaderClass", void 0);
__decorate([
    Output()
], AppearanceSettingsComponent.prototype, "themeClassChange", void 0);
__decorate([
    Output()
], AppearanceSettingsComponent.prototype, "headerClassChange", void 0);
__decorate([
    Input()
], AppearanceSettingsComponent.prototype, "possibleStylingTheme", void 0);
__decorate([
    Input()
], AppearanceSettingsComponent.prototype, "possibleStylingHeader", void 0);
AppearanceSettingsComponent = __decorate([
    Component({
        selector: 'c8y-appearance-settings',
        template: "<h6 class=\"legend form-block\"><span translate>Layout</span></h6>\n<div class=\"row \">\n  <div class=\"col-sm-6 m-b-32\">\n    <label translate>Theme</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row\"\n        style=\"align-items: center; box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let themeClassItem of possibleStylingTheme; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradiocontentclass' + i\"\n              name=\"content\"\n              [value]=\"themeClassItem.class\"\n              [ngModel]=\"themeClass\"\n              (click)=\"themeClassClick(themeClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradiocontentclass' + i\">\n          <span>{{ themeClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"themeClassItem.class === defaultThemeClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ themeClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\">\n          <i \n            [c8yIcon]=\"'question-circle-o'\"\n            class=\"text-primary\"\n          ></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-sm-6\">\n    <label translate>Widget header style</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row\"\n        style=\"align-items: center; box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let headerClassItem of possibleStylingHeader; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradioheaderclass' + i\"\n              name=\"header\"\n              [value]=\"headerClassItem.class\"\n              [ngModel]=\"headerClass\"\n              (click)=\"headerClassClick(headerClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradioheaderclass' + i\">\n          <span>{{ headerClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"headerClassItem.class === defaultHeaderClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ headerClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\">\n          <i \n            [c8yIcon]=\"'question-circle-o'\"\n            class=\"text-primary\"\n          ></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n</div>\n",
        viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
    })
], AppearanceSettingsComponent);

let WidgetPreviewComponent = class WidgetPreviewComponent {
    constructor() {
        this.previewClasses = {
            'dashboard-theme-light': true,
            'panel-title-regular': true
        };
    }
};
__decorate([
    Input()
], WidgetPreviewComponent.prototype, "previewClasses", void 0);
__decorate([
    Input()
], WidgetPreviewComponent.prototype, "tab", void 0);
WidgetPreviewComponent = __decorate([
    Component({
        selector: 'c8y-widget-preview',
        template: "<div class=\"legend form-block\" translate>Preview</div>\n<label *ngIf=\"tab\" translate>Tab and layout</label>\n<label *ngIf=\"!tab\" translate>Layout</label>\n<div class=\"dashboard-preview-slot\" [ngClass]=\"previewClasses\">\n  <div\n    class=\"page-tabs page-tabs-horizontal navigator-open\"\n    style=\" position: absolute; left: 0; margin: 0 15px; overflow: hidden;\"\n    *ngIf=\"tab\"\n  >\n    <div class=\"tabContainer hidden-xs\">\n      <ul class=\"nav nav-tabs nav-tabsc8y\">\n        <li class=\"active\" style=\"left: 16px;\">\n          <a routerlinkactive=\"\" title=\"{{ tab.name }}\">\n            <i class=\"text-primary\" [c8yIcon]=\"tab.icon\"></i>\n            <span class=\"txt\">{{ tab.name }}</span></a\n          >\n        </li>\n      </ul>\n    </div>\n    <div class=\"visible-xs mobile-tabs\"></div>\n  </div>\n  <div [ngClass]=\"{ 'm-t-64': !!tab }\">\n    <div class=\"card card-dashboard\" style=\"height: 360px;\">\n      <div class=\"card-header-actions\" style=\"z-index: 8;\">\n        <div class=\"card-title\">\n          <span translate>Widget title</span>\n        </div>\n        <div class=\"header-actions\">\n          <div class=\"optionsBtn dropdown\">\n            <a href=\"\" (click)=\"(false)\" class=\"dropdown-toggle c8y-dropdown\">\n              <i c8yIcon=\"cog\"></i>\n            </a>\n          </div>\n        </div>\n      </div>\n      <div class=\"card-inner-scroll p-t-16\">\n        <p translate>Widget example content.</p>\n      </div>\n    </div>\n  </div>\n</div>\n"
    })
], WidgetPreviewComponent);

var ContextDashboardModule_1;
const defaultConfig = {
    allowFullscreen: true,
    widgetFilter({ data }) {
        return !(data && data.settings && data.settings.noNewWidgets);
    },
    routes: [
        {
            path: 'dashboard/:dashboardId',
            component: ContextDashboardComponent,
            canActivate: [DeviceDashboardGuard],
            context: ViewContext.Device
        },
        {
            path: 'dashboard/:dashboardId',
            component: ContextDashboardComponent,
            canActivate: [GroupDashboardGuard],
            context: ViewContext.Group
        }
    ]
};
let ContextDashboardModule = ContextDashboardModule_1 = class ContextDashboardModule {
    static config(config = {}) {
        return {
            ngModule: ContextDashboardModule_1,
            providers: [
                {
                    provide: CONTEXT_DASHBOARD_CONFIG,
                    useValue: Object.assign({}, defaultConfig, config)
                },
                ContextDashboardService,
                DeviceDashboardGuard,
                GroupDashboardGuard,
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: Object.assign({}, defaultConfig, config).routes,
                    multi: true
                },
                {
                    provide: HOOK_ACTION,
                    useClass: DashboardActionsFactory,
                    multi: true
                }
            ]
        };
    }
};
ContextDashboardModule = ContextDashboardModule_1 = __decorate([
    NgModule({
        declarations: [
            AppearanceSettingsComponent,
            ContextDashboardComponent,
            WidgetConfigComponent,
            WidgetPreviewComponent,
            DashboardDetailComponent
        ],
        imports: [CoreModule, TooltipModule, PopoverModule, BsDropdownModule.forRoot()],
        exports: [ContextDashboardComponent, WidgetConfigComponent],
        entryComponents: [ContextDashboardComponent, WidgetConfigComponent, DashboardDetailComponent]
    })
], ContextDashboardModule);

let CockpitDashboardComponent = class CockpitDashboardComponent {
    constructor() {
        this.DEFAULT_HOME_WIDGETS = [
            {
                name: 'cockpit.welcome.widget',
                title: gettext(''),
                _x: 0,
                _y: 0,
                _height: 5,
                _width: 12,
                classes: { 'panel-title-hidden': true }
            },
            {
                name: 'Asset Alarms',
                title: gettext('Active, critical alarms'),
                _x: 0,
                _y: 5,
                _height: 4,
                _width: 6
            },
            {
                name: 'Recent Alarms',
                title: gettext('Recent alarms'),
                _x: 0,
                _y: 9,
                _height: 4,
                _width: 6
            },
            {
                name: 'Map',
                title: gettext('Map'),
                _x: 6,
                _y: 5,
                _height: 8,
                _width: 6
            }
        ];
    }
};
CockpitDashboardComponent = __decorate([
    Component({
        selector: 'c8y-cockpit-dashboard',
        template: "<c8y-title translate>\n  Home\n</c8y-title>\n\n<c8y-context-dashboard name=\"home-cockpit1\" [defaultWidgets]=\"DEFAULT_HOME_WIDGETS\" [canDelete]=\"false\"></c8y-context-dashboard>\n"
    })
], CockpitDashboardComponent);

let WelcomeToCockpit = class WelcomeToCockpit {
    constructor(tenantService, docs, router, navigator, applicationService) {
        this.tenantService = tenantService;
        this.docs = docs;
        this.router = router;
        this.navigator = navigator;
        this.applicationService = applicationService;
        this.quickLinks = [];
        this.REPORT_NODE = 'Reports';
        this.CONFIGURATION_NODE = 'Configuration';
        this.TRIAL = 'TRIAL';
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.tenantType = yield this.tenantService.currentTenantType();
            this.setMessage();
            this.navSubscription = this.navigator.items$.subscribe(nodes => {
                this.navNodes = nodes;
            });
            this.docsSubscription = this.docs.items$.subscribe(links => {
                this.links = links;
            });
            // <---TRIAL & REGULAR TENANT --->
            this.createConnectSmartphoneQuickLink();
            yield this.createQuicklinkRegisterDevice();
            // <--- TRIAL TENANT --->
            if (this.tenantType === this.TRIAL) {
                this.createQuicklinkUserGuide();
                return;
            }
            // <--- REGULAR TENANT --->
            this.createQuicklinkAddGroup();
            this.createQuickLinkReports();
            this.createQuicklinkDataPointLibrary();
            this.createQuicklinkSmartRules();
        });
    }
    ngOnDestroy() {
        if (this.docsSubscription && !this.docsSubscription.closed) {
            this.docsSubscription.unsubscribe();
        }
        if (this.navSubscription && !this.navSubscription.closed) {
            this.navSubscription.unsubscribe();
        }
    }
    setMessage() {
        if (this.tenantType === this.TRIAL) {
            this.welcomeMessage = gettext(`
        The Cockpit application allows you to build IoT applications in minutes.
        To get started, connect any device to the platform.
        If you do not have an IoT device to hand, you can start by connecting your smartphone.
        Click below to be guided through the process.
      `);
        }
        else {
            this.welcomeMessage = gettext(`
        The Cockpit application provides you with options to manage
        and monitor Internet of Things assets and data from business perspective.
      `);
        }
    }
    createQuicklinkAddGroup() {
        // comes from angularJS factory c8yQuickLinks
        const addGroup = this.links.find(link => link.label === 'Add group');
        if (addGroup) {
            this.quickLinks.push(addGroup);
        }
    }
    createConnectSmartphoneQuickLink() {
        // Provider in SensorPhoneModule defines the
        // 'Connect smartphone' quicklink.
        const connectSmartphone = this.links.find(link => link.label === 'Connect smartphone');
        if (connectSmartphone) {
            this.quickLinks.push(connectSmartphone);
        }
    }
    createQuickLinkReports() {
        const reports = {
            icon: 'c8y-reports',
            label: gettext('Reports'),
            url: '/export'
        };
        const reportsNode = this.findNavigatorNode(this.REPORT_NODE);
        if (reportsNode && reportsNode.show) {
            reports.click = () => {
                reportsNode.open = true;
                this.router.navigateByUrl(reports.url);
            };
            this.quickLinks.push(reports);
        }
    }
    createQuicklinkDataPointLibrary() {
        const dataPointLib = {
            icon: 'c8y-data-points',
            label: gettext('Data point library'),
            url: '/datapointlibrary'
        };
        const configurationNode = this.findNavigatorNode(this.CONFIGURATION_NODE);
        if (configurationNode && configurationNode.show) {
            dataPointLib.click = () => {
                configurationNode.open = true;
                this.router.navigateByUrl(dataPointLib.url);
            };
            this.quickLinks.push(dataPointLib);
        }
    }
    createQuicklinkSmartRules() {
        const smartRules = {
            icon: 'c8y-smart-rules',
            label: gettext('Smart rules'),
            url: '/rules'
        };
        const configurationNode = this.findNavigatorNode(this.CONFIGURATION_NODE);
        if (configurationNode && configurationNode.show) {
            smartRules.click = () => {
                configurationNode.open = true;
                this.router.navigateByUrl(smartRules.url);
            };
            this.quickLinks.push(smartRules);
        }
    }
    createQuicklinkRegisterDevice() {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.applicationService.listByUser();
            if (data) {
                const deviceManagement = data.find(app => app.contextPath === 'devicemanagement');
                if (deviceManagement) {
                    const deviceMgmtUrl = this.applicationService.getHref(deviceManagement);
                    const registerDevice = {
                        icon: 'c8y-device-connect',
                        label: gettext('Register device'),
                        click: () => window.open(`${deviceMgmtUrl}/#/deviceregistration`, '_self')
                    };
                    this.quickLinks.push(registerDevice);
                }
            }
        });
    }
    createQuicklinkUserGuide() {
        const userGuide = {
            icon: 'c8y-user',
            label: gettext('User guide'),
            url: '/users-guide/getting-started',
            click: () => {
                const userGuideURL = this.docs.getUserGuideLink(userGuide.url);
                window.open(userGuideURL);
            }
        };
        this.quickLinks.push(userGuide);
    }
    findNavigatorNode(nodeRealName) {
        if (this.navNodes && this.navNodes.length > 0) {
            return this.navNodes.find((node) => node.realName === nodeRealName);
        }
        return undefined;
    }
};
WelcomeToCockpit.ctorParameters = () => [
    { type: TenantService },
    { type: DocsService },
    { type: Router },
    { type: NavigatorService },
    { type: ApplicationService }
];
__decorate([
    Input()
], WelcomeToCockpit.prototype, "config", void 0);
WelcomeToCockpit = __decorate([
    Component({
        selector: 'c8y-welcome-to-cockpit',
        template: "<div class=\"welcome-widget welcome-cockpit\">\r\n  <div class=\"flex-row\">\r\n    <div class=\"col-xs-12 col-md-5 flex-item-v-stretch p-24\">\r\n      <h2 class=\"text-light\">{{ 'Welcome to Cockpit' | translate }}</h2>\r\n      <p class=\"text-16 text-light p-t-16 p-b-24\">{{ welcomeMessage | translate }}</p>\r\n      <div class=\"card-group interact-grid tight-grid\">\r\n        <div *ngFor=\"let link of quickLinks\" class=\"col-sm-4 col-xs-6\">\r\n          <c8y-quick-link\r\n            (click)=\"link.click ? link.click() : false\"\r\n            [icon]=\"link.icon\"\r\n            [label]=\"link.label\"\r\n            class=\"card\"\r\n          >\r\n          </c8y-quick-link>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <!-- <div class=\"col-sm-6 welcome-illustration flex-item-v-stretch\"></div> -->\r\n  </div>\r\n</div>\r\n"
    })
], WelcomeToCockpit);

let LegacyWelcomeComponent = class LegacyWelcomeComponent {
    constructor(dashboard) {
        this.dashboard = dashboard;
    }
    restore() {
        this.dashboard.restore();
    }
};
LegacyWelcomeComponent.ctorParameters = () => [
    { type: ContextDashboardComponent, decorators: [{ type: Host }] }
];
LegacyWelcomeComponent = __decorate([
    Component({
        selector: 'c8y-legacy-welcome-widget',
        template: "<div class=\"card-block\">\r\n  <p class=\"lead\" c8y-guide-docs>\r\n    <span>{{ 'Welcome to Cockpit' | translate }}.</span>\r\n    <span title=\"{{ 'Browse the manual' | translate }}\" class=\"cockpit-welcome-manual\" translate\r\n      >Browse the manual <a c8y-guide-href=\"users-guide/cockpit/\">here</a>.</span\r\n    >\r\n  </p>\r\n  <p class=\"m-t-16\">\r\n    <em>{{ 'New default dashboard available.' | translate }} </em>\r\n  </p>\r\n  <button class=\"btn btn-primary btn-sm m-t-16\" (click)=\"restore(); (false)\" translate>\r\n    Get new dashboard\r\n  </button>\r\n</div>\r\n"
    }),
    __param(0, Host())
], LegacyWelcomeComponent);

// import for the preview img to get into the build.
let previewImage = '';
try {
    // tslint:disable-next-line: no-var-requires
    previewImage = require('./welcome/img/welcome-widget-pr.png');
}
catch (ex) {
    // intended empty
}
const ɵ0 = [{ label: gettext('Home'), icon: 'home', priority: 10000, path: '/' }], ɵ1 = [
    {
        id: 'cockpit.welcome.widget',
        label: gettext('Welcome to Cockpit'),
        description: gettext('Displays a welcome message for Cockpit'),
        component: WelcomeToCockpit,
        previewImage,
        data: {
            settings: {
                noNewWidgets: false,
                widgetDefaults: {
                    _width: 12,
                    _height: 6
                },
                ng1: {
                    options: {
                        noDeviceTarget: true,
                        groupsSelectable: false
                    }
                }
            }
        }
    }
], ɵ2 = [
    {
        id: 'Cockpit Welcome',
        label: gettext('Welcome to Cockpit'),
        description: gettext('Displays a welcome message for Cockpit'),
        component: LegacyWelcomeComponent,
        data: {
            settings: {
                noNewWidgets: true,
                ng1: {
                    options: {
                        noDeviceTarget: true,
                        groupsSelectable: false
                    }
                }
            }
        }
    }
];
let CockpitDashboardModule = class CockpitDashboardModule {
};
CockpitDashboardModule = __decorate([
    NgModule({
        entryComponents: [CockpitDashboardComponent, WelcomeToCockpit, LegacyWelcomeComponent],
        imports: [
            DocsModule,
            CoreModule,
            ContextDashboardModule.config({
                routes: [
                    {
                        path: 'dashboard/:dashboardId',
                        component: ContextDashboardComponent,
                        canActivate: [DeviceDashboardGuard],
                        context: ViewContext.Device
                    },
                    {
                        path: 'dashboard/:dashboardId',
                        component: ContextDashboardComponent,
                        canActivate: [GroupDashboardGuard],
                        context: ViewContext.Group
                    },
                    {
                        path: '',
                        component: CockpitDashboardComponent
                    }
                ]
            }),
            QuickLinkModule,
            TranslateModule
        ],
        declarations: [CockpitDashboardComponent, WelcomeToCockpit, LegacyWelcomeComponent],
        providers: [
            {
                provide: HOOK_NAVIGATOR_NODES,
                useValue: ɵ0,
                multi: true
            },
            {
                provide: HOOK_COMPONENTS,
                multi: true,
                useValue: ɵ1
            },
            {
                provide: HOOK_COMPONENTS,
                multi: true,
                useValue: ɵ2
            }
        ]
    })
], CockpitDashboardModule);

/**
 * Generated bundle index. Do not edit.
 */

export { CONTEXT_DASHBOARD_CONFIG, CockpitDashboardModule, ContextDashboardModule, ContextDashboardService, ContextDashboardType, DASHBOARD_THEME_CLASSES, DeviceDashboardGuard, GroupDashboardGuard, WIDGET_CONTENT_CLASSES, WIDGET_HEADER_CLASSES, ɵ0, ɵ1, ɵ2, AppearanceSettingsComponent as ɵa, ContextDashboardComponent as ɵb, WidgetService as ɵc, WidgetConfigComponent as ɵd, WidgetPreviewComponent as ɵe, DashboardDetailComponent as ɵf, DashboardActionsFactory as ɵg, CockpitDashboardComponent as ɵh, WelcomeToCockpit as ɵi, LegacyWelcomeComponent as ɵj };
