// @c8y/ngx-components/esm5/core/breadcrumb/breadcrumb-outlet.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var BreadcrumbOutletComponent = /** @class */ (function () {
    function BreadcrumbOutletComponent() {
        this.showAll = false;
        this.breadcrumbs = [];
    }
    /**
     * For upgrade only. Old angularjs routes start with hash, new ones not.
     */
    BreadcrumbOutletComponent.prototype.normalizePath = function (path) {
        return path.replace(/^#\/?/, '');
    };
    tslib_1.__decorate([
        Input()
    ], BreadcrumbOutletComponent.prototype, "breadcrumbs", void 0);
    BreadcrumbOutletComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-breadcrumb-outlet',
            template: "<div\n  class=\"breadcrumbs-container\"\n  [ngClass]=\"{\n    multiple: breadcrumbs.length > 1,\n    open: showAll\n  }\"\n  *ngIf=\"breadcrumbs && breadcrumbs.length > 0\"\n>\n  <button\n    *ngIf=\"breadcrumbs.length > 1\"\n    class=\"btn-show-all-breadcrumbs pull-left\"\n    (click)=\"showAll = !showAll\"\n  >\n    <i\n      *ngIf=\"!showAll\"\n      [c8yIcon]=\"'caret-right'\"\n      tooltip=\"{{ 'Show all breadcrumbs' | translate }}\"\n      container=\"body\"\n      placement=\"top\"\n    ></i>\n    <i\n      *ngIf=\"showAll\"\n      [c8yIcon]=\"'caret-down'\"\n      tooltip=\"{{ 'Collapse breadcrumbs' | translate }}\"\n      container=\"body\"\n      placement=\"top\"\n    ></i>\n  </button>\n\n  <ul class=\"breadcrumbs text-muted\" *ngFor=\"let breadcrumb of breadcrumbs; let first = first\">\n    <ng-container *ngIf=\"first || showAll\">\n      <li class=\"text-muted\" *ngFor=\"let item of breadcrumb.items; let firstItem = first\">\n        <span *ngIf=\"!firstItem\"> > </span>\n        <ng-container *c8yOutlet=\"item.template\"></ng-container>\n        <ng-container *ngIf=\"item.label\">\n          <a\n            [routerLink]=\"normalizePath(item.path)\"\n            class=\"word-break\"\n            title=\"{{ item.label | translate }}\"\n          >\n            <i [c8yIcon]=\"item.icon\" *ngIf=\"firstItem\" class=\"right-m-xs\"></i>\n            <span>{{ item.label | translate }}</span>\n          </a>\n        </ng-container>\n      </li>\n    </ng-container>\n  </ul>\n</div>\n"
        })
    ], BreadcrumbOutletComponent);
    return BreadcrumbOutletComponent;
}());
export { BreadcrumbOutletComponent };
