// @c8y/ngx-components/esm2015/repository/repository-navigation-factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, NavigatorNodeFactory, OptionsService } from '@c8y/ngx-components';
let RepositoryNavigationFactory = class RepositoryNavigationFactory {
    constructor(optionsService) {
        this.optionsService = optionsService;
        this.navs = [];
    }
    get() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const betaFlag = localStorage.getItem('c8y_beta_status');
            if (betaFlag === 'all' && this.navs.length === 0) {
                this.navs.push(new NavigatorNode({
                    label: gettext('Configuration repository'),
                    path: 'configuration-new',
                    icon: 'gears',
                    parent: gettext('Management'),
                    priority: 800
                }));
            }
            return this.navs;
        });
    }
};
RepositoryNavigationFactory.ctorParameters = () => [
    { type: OptionsService }
];
RepositoryNavigationFactory = tslib_1.__decorate([
    Injectable()
], RepositoryNavigationFactory);
export { RepositoryNavigationFactory };
