// @c8y/ngx-components/esm2015/repository/repository.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { InventoryService, IIdentified, InventoryBinaryService, IManagedObject } from '@c8y/client';
import { RepositoryType } from './repository.model';
let RepositoryService = class RepositoryService {
    constructor(inventory, inventoryBinary) {
        this.inventory = inventory;
        this.inventoryBinary = inventoryBinary;
    }
    listConfigurations() {
        return this.inventory.list({
            type: RepositoryType.CONFIGURATION,
            withTotalPages: true,
            pageSize: 50
        });
    }
    delete(entity) {
        // TODO: add cascading delete (will be done by Dawid)
        return this.inventory.delete(entity);
    }
    save(data, type, mo = {}) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            switch (type) {
                case RepositoryType.CONFIGURATION: {
                    Object.assign(mo, {
                        type: RepositoryType.CONFIGURATION,
                        configurationType: data.selected ? data.selected.configurationType : undefined,
                        name: data.versionOrName,
                        description: data.description,
                        deviceType: data.deviceType,
                        c8y_Global: {}
                    });
                    break;
                }
            }
            const existingUrl = mo.url;
            if (data.binary.url) {
                mo.url = data.binary.url;
            }
            else if (data.binary.file) {
                const response = yield this.inventoryBinary.create(data.binary.file, { c8y_Global: {} });
                mo.url = response.data.self;
            }
            if (mo.id) {
                return this.updateEntry(mo, existingUrl);
            }
            return this.createEntry(mo);
        });
    }
    createEntry(mo) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const binaryId = yield this.inventoryBinary.getIdFromUrl(mo.url);
            const newMo = yield this.inventory.create(mo);
            if (binaryId) {
                yield this.inventory.childAdditionsAdd(binaryId, newMo.data);
            }
            return newMo;
        });
    }
    updateEntry(mo, url) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const existingBinaryId = yield this.inventoryBinary.getIdFromUrl(url);
            const newBinaryId = yield this.inventoryBinary.getIdFromUrl(mo.url);
            if (existingBinaryId && existingBinaryId !== newBinaryId) {
                const id = this.inventoryBinary.getIdFromUrl(url);
                yield this.inventoryBinary.delete(id);
            }
            if (newBinaryId) {
                yield this.inventory.childAdditionsAdd(newBinaryId, mo);
            }
            return this.inventory.update(mo);
        });
    }
};
RepositoryService.ctorParameters = () => [
    { type: InventoryService },
    { type: InventoryBinaryService }
];
RepositoryService = tslib_1.__decorate([
    Injectable()
], RepositoryService);
export { RepositoryService };
