// @c8y/client/lib/esm5/src/services.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("./core");
exports.FetchClient = core_1.FetchClient;
exports.BasicAuth = core_1.BasicAuth;
var event_1 = require("./event");
exports.EventService = event_1.EventService;
exports.EventBinaryService = event_1.EventBinaryService;
var inventory_1 = require("./inventory");
exports.InventoryService = inventory_1.InventoryService;
exports.InventoryBinaryService = inventory_1.InventoryBinaryService;
var measurement_1 = require("./measurement");
exports.MeasurementService = measurement_1.MeasurementService;
var alarm_1 = require("./alarm");
exports.AlarmService = alarm_1.AlarmService;
var operation_1 = require("./operation");
exports.OperationBulkService = operation_1.OperationBulkService;
exports.OperationService = operation_1.OperationService;
var application_1 = require("./application");
exports.ApplicationService = application_1.ApplicationService;
var user_1 = require("./user");
exports.UserService = user_1.UserService;
var tenant_1 = require("./tenant");
exports.TenantService = tenant_1.TenantService;
var system_options_1 = require("./system-options");
exports.SystemOptionsService = system_options_1.SystemOptionsService;
var tenant_options_1 = require("./tenant-options");
exports.TenantOptionsService = tenant_options_1.TenantOptionsService;
var tenant_security_options_1 = require("./tenant-security-options");
exports.TenantSecurityOptionsService = tenant_security_options_1.TenantSecurityOptionsService;
var audit_1 = require("./audit");
exports.AuditService = audit_1.AuditService;
var inventory_role_1 = require("./inventory-role");
exports.InventoryRoleService = inventory_role_1.InventoryRoleService;
var device_registration_1 = require("./device-registration");
exports.DeviceRegistrationService = device_registration_1.DeviceRegistrationService;
exports.DeviceRegistrationBulkService = device_registration_1.DeviceRegistrationBulkService;
var user_role_1 = require("./user-role");
exports.UserRoleService = user_role_1.UserRoleService;
var user_group_1 = require("./user-group");
exports.UserGroupService = user_group_1.UserGroupService;
var realtime_1 = require("./realtime");
exports.Realtime = realtime_1.Realtime;
var identity_1 = require("./identity");
exports.IdentityService = identity_1.IdentityService;
var trusted_certificates_1 = require("./trusted-certificates");
exports.TrustedCertificateService = trusted_certificates_1.TrustedCertificateService;
