// @c8y/ngx-components/esm5/core/authentication/authentication.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { PasswordCheckListComponent } from './password-check-list.component';
import { PasswordConfirm } from './password-confirm.directive';
import { PasswordStrengthCheckerService } from './password-strength-checker.service';
import { PasswordStrengthComponent } from './password-strength.component';
import { NewPasswordComponent } from './new-password.component';
import { PasswordConfirmModalComponent } from './password-confirm-modal.component';
import { ModalModule } from '../modal/modal.module';
import { FormsModule } from '../forms/forms.module';
import { TotpChallengeComponent } from './totp-challenge.component';
import { SmsChallengeComponent } from './sms-challenge.component';
import { TotpSetupComponent } from './totp-setup.component';
import { PasswordService } from './password.service';
import { ProvidePhoneNumberComponent } from './provide-phone-number.component';
/**
 * The angular module definition for new password.
 * @exports PasswordStrengthComponent
 * @exports PasswordCheckListComponent
 * @exports PasswordConfirm
 * @exports NewPasswordComponent
 * @exports PasswordConfirmModalComponent
 */
var AuthenticationModule = /** @class */ (function () {
    function AuthenticationModule() {
    }
    AuthenticationModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                TotpChallengeComponent,
                PasswordConfirm,
                PasswordStrengthComponent,
                PasswordCheckListComponent,
                NewPasswordComponent,
                PasswordConfirmModalComponent,
                SmsChallengeComponent,
                TotpSetupComponent,
                ProvidePhoneNumberComponent
            ],
            imports: [
                FormsModule,
                TooltipModule,
                CommonModule,
                ModalModule
            ],
            exports: [
                TotpChallengeComponent,
                PasswordStrengthComponent,
                PasswordCheckListComponent,
                PasswordConfirm,
                NewPasswordComponent,
                PasswordConfirmModalComponent,
                SmsChallengeComponent,
                TotpSetupComponent,
                ProvidePhoneNumberComponent
            ],
            entryComponents: [
                PasswordConfirmModalComponent
            ],
            providers: [
                PasswordStrengthCheckerService,
                PasswordService
            ]
        })
    ], AuthenticationModule);
    return AuthenticationModule;
}());
export { AuthenticationModule };
