// @c8y/ngx-components/esm5/core/authentication/password-confirm-modal.component.js
import * as tslib_1 from "tslib";
import { Component, Output, EventEmitter, ViewChild } from '@angular/core';
import { UserService, ICredentials, FetchClient, Client, IAuthentication } from '@c8y/client';
import { AppStateService } from '../common/ui-state.service';
import { AlertService } from '../alert/alert.service';
import { LoginService } from '../login/public-api';
import { clone } from 'lodash-es';
import { gettext } from '../i18n/gettext';
var PasswordConfirmModalComponent = /** @class */ (function () {
    function PasswordConfirmModalComponent(user, ui, client, alert, loginService) {
        this.user = user;
        this.ui = ui;
        this.client = client;
        this.alert = alert;
        this.loginService = loginService;
        this.passwordConfirmedEmitter = new EventEmitter();
        this.loading = false;
    }
    PasswordConfirmModalComponent.prototype.passwordConfirm = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var supportUserName, userId, credentials, strategy, newClient, e_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.password) return [3 /*break*/, 5];
                        supportUserName = this.ui.currentSupportUserName.value;
                        userId = this.ui.currentUser.value.id;
                        credentials = {
                            password: this.password,
                            user: "" + (supportUserName ? supportUserName + "$" : '') + userId,
                            tenant: this.client.tenant
                        };
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        this.loading = true;
                        strategy = clone(this.loginService.useBasicAuth({}));
                        delete credentials.tfa;
                        strategy.updateCredentials(credentials);
                        newClient = this.createNewClient(strategy, this.client.baseUrl);
                        return [4 /*yield*/, newClient.user.current()];
                    case 2:
                        _a.sent();
                        this.emitSuccessAndClose();
                        return [3 /*break*/, 5];
                    case 3:
                        e_1 = _a.sent();
                        if (e_1.res && e_1.res.status === 401 && e_1.data && /pin|totp/i.test(e_1.data.message)) {
                            this.emitSuccessAndClose();
                        }
                        else {
                            this.alert.danger(gettext("Provided password doesn't match your current one."));
                        }
                        return [3 /*break*/, 5];
                    case 4:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    PasswordConfirmModalComponent.prototype.cancel = function () {
        this.passwordConfirmedEmitter.emit(false);
        this.modal._dismiss();
    };
    PasswordConfirmModalComponent.prototype.createNewClient = function (strategy, baseUrl) {
        return new Client(strategy, baseUrl);
    };
    PasswordConfirmModalComponent.prototype.emitSuccessAndClose = function () {
        this.passwordConfirmedEmitter.emit(true);
        this.modal._dismiss();
    };
    PasswordConfirmModalComponent.ctorParameters = function () { return [
        { type: UserService },
        { type: AppStateService },
        { type: FetchClient },
        { type: AlertService },
        { type: LoginService }
    ]; };
    tslib_1.__decorate([
        Output()
    ], PasswordConfirmModalComponent.prototype, "passwordConfirmedEmitter", void 0);
    tslib_1.__decorate([
        ViewChild('modal', { static: false })
    ], PasswordConfirmModalComponent.prototype, "modal", void 0);
    PasswordConfirmModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-password-confirm-modal',
            template: "<c8y-modal [customFooter]=\"true\" [title]=\"'Confirm password' | translate\" #modal>\r\n  <form #confirmForm=\"ngForm\" (ngSubmit)=\"confirmForm.form.valid && passwordConfirm()\">\r\n    <c8y-form-group>\r\n      <label translate for=\"currentPassword\">Current password</label>\r\n      <input\r\n        id=\"currentPassword\"\r\n        [(ngModel)]=\"password\"\r\n        type=\"password\"\r\n        name=\"password\"\r\n        class=\"form-control\"\r\n        placeholder=\"{{ 'Current password' | translate }}\"\r\n        required\r\n      />\r\n    </c8y-form-group>\r\n\r\n    <div class=\"modal-footer\">\r\n      <button\r\n        title=\"{{ 'Cancel' | translate }}\"\r\n        class=\"btn btn-default\"\r\n        type=\"button\"\r\n        (click)=\"cancel()\"\r\n        translate\r\n      >\r\n        Cancel\r\n      </button>\r\n      <button\r\n        title=\"{{ 'Save' | translate }}\"\r\n        class=\"btn btn-primary\"\r\n        type=\"submit\"\r\n        [disabled]=\"!confirmForm.form.valid || loading\"\r\n        translate\r\n      >\r\n        Save\r\n      </button>\r\n    </div>\r\n  </form>\r\n</c8y-modal>\r\n"
        })
    ], PasswordConfirmModalComponent);
    return PasswordConfirmModalComponent;
}());
export { PasswordConfirmModalComponent };
