// @c8y/client/lib/cjs/src/application/ApplicationBinaryService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("../core");
var form_data_1 = __importDefault(require("form-data"));
var ApplicationBinaryService = /** @class */ (function (_super) {
    __extends(ApplicationBinaryService, _super);
    function ApplicationBinaryService(client, applicationOrId) {
        var _this = _super.call(this, client) || this;
        _this.listUrl = 'binaries';
        _this.propertyName = 'attachments';
        _this.baseUrl = "application/applications/" + _this.getIdString(applicationOrId);
        return _this;
    }
    ApplicationBinaryService.prototype.upload = function (binary, fileName) {
        return __awaiter(this, void 0, void 0, function () {
            var url, method, body, bufferOrStream, uploadFileName, bodyHeaders, headers, res, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = this.listUrl;
                        method = 'POST';
                        body = new form_data_1.default();
                        bufferOrStream = binary instanceof ArrayBuffer ? Buffer.from(binary) : binary;
                        uploadFileName = fileName;
                        if (typeof (File) !== 'undefined' && binary instanceof File) {
                            uploadFileName = binary.name;
                        }
                        body.append('file', bufferOrStream, uploadFileName);
                        body.append('fileName', uploadFileName);
                        if (typeof body.getHeaders === 'function') {
                            bodyHeaders = body.getHeaders();
                        }
                        headers = Object.assign({
                            accept: 'application/json'
                        }, bodyHeaders);
                        return [4 /*yield*/, this.fetch(url, { method: method, body: body, headers: headers })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    ApplicationBinaryService.prototype.list = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, _super.prototype.list.call(this)];
            });
        });
    };
    ApplicationBinaryService.prototype.list$ = function (filter, options) {
        if (filter === void 0) { filter = {}; }
        if (options === void 0) { options = {}; }
        return _super.prototype.list$.call(this, filter, options);
    };
    ApplicationBinaryService.prototype.delete = function (binaryOrId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, _super.prototype.delete.call(this, binaryOrId)];
            });
        });
    };
    ApplicationBinaryService.prototype.listPlugins = function () {
        return __awaiter(this, void 0, void 0, function () {
            var headers, url, res, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        headers = { accept: 'application/json' };
                        url = this.listUrl + "/plugins";
                        return [4 /*yield*/, this.fetch(url, { headers: headers })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    ApplicationBinaryService.prototype.addPlugin = function (pluginName, pluginFile) {
        return __awaiter(this, void 0, void 0, function () {
            var url, method, body, bufferOrStream, bodyHeaders, headers, res, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = this.listUrl + "/plugins/" + encodeURIComponent(pluginName);
                        method = 'POST';
                        body = new form_data_1.default();
                        bufferOrStream = pluginFile instanceof ArrayBuffer ? Buffer.from(pluginFile) : pluginFile;
                        body.append('file', bufferOrStream);
                        if (typeof body.getHeaders === 'function') {
                            bodyHeaders = body.getHeaders();
                        }
                        headers = Object.assign({
                            accept: 'application/json'
                        }, bodyHeaders);
                        return [4 /*yield*/, this.fetch(url, { method: method, body: body, headers: headers })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    ApplicationBinaryService.prototype.removePlugin = function (pluginName) {
        return __awaiter(this, void 0, void 0, function () {
            var method, headers, url, res, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        method = 'DELETE';
                        headers = { accept: 'application/json' };
                        url = this.listUrl + "/plugins/" + pluginName;
                        return [4 /*yield*/, this.fetch(url, { method: method, headers: headers })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    ApplicationBinaryService.prototype.updateFiles = function (files) {
        return __awaiter(this, void 0, void 0, function () {
            var url, method, body, bodyHeaders, headers, res, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = this.listUrl + "/files";
                        method = 'POST';
                        body = new form_data_1.default();
                        files.forEach(function (file) {
                            var bufferOrStream = file.contents instanceof ArrayBuffer ? Buffer.from(file.contents) : file.contents;
                            body.append(file.path, bufferOrStream);
                        });
                        if (typeof body.getHeaders === 'function') {
                            bodyHeaders = body.getHeaders();
                        }
                        headers = Object.assign({
                            accept: 'application/json'
                        }, bodyHeaders);
                        return [4 /*yield*/, this.fetch(url, { method: method, body: body, headers: headers })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    return ApplicationBinaryService;
}(core_1.Service));
exports.ApplicationBinaryService = ApplicationBinaryService;
