// @c8y/ngx-components/esm5/core/audit-log/audit-log.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { AuditService, IAuditRecord, IResultList } from '@c8y/client';
var AuditLogComponent = /** @class */ (function () {
    function AuditLogComponent(audit) {
        this.audit = audit;
    }
    AuditLogComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.source) return [3 /*break*/, 2];
                        _a = this;
                        return [4 /*yield*/, this.audit.list({
                                withTotalPages: true,
                                source: this.source
                            })];
                    case 1:
                        _a.records = _b.sent();
                        _b.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    AuditLogComponent.ctorParameters = function () { return [
        { type: AuditService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], AuditLogComponent.prototype, "source", void 0);
    AuditLogComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-audit-log',
            template: "<div class=\"inner-scroll\">\n  <ul class=\"timeline-list list-condensed m-t-16\">\n    <li class=\"timeline-list-item flex-row\" *c8yFor=\"let record of records\">\n      <small class=\"timeline-item-date text-right\">\n        <span>{{ record.time | c8yDate }}</span>\n      </small>\n      <div class=\"timeline-item-content flex-row\">\n        <div class=\"list-item-body\">\n          <div class=\"row text-break-word\">\n            <div class=\"col-md-5\">\n              <span class=\"text-truncate\">{{ record.activity | translate }}</span>\n              <small class=\"text-muted\">{{ record.user }}</small>\n            </div>\n            <div class=\"col-md-7\">\n              <div class=\"small text-truncate-wrap\">{{ record.text | translate }}</div>\n            </div>\n          </div>\n        </div>\n      </div>\n    </li>\n  </ul>\n</div>\n"
        })
    ], AuditLogComponent);
    return AuditLogComponent;
}());
export { AuditLogComponent };
