// @c8y/ngx-components/esm5/core/device-status/device-availability.js
import { gettext } from '../i18n/gettext';
import { formatDate } from '@angular/common';
// TODO remove  when c8y-icon will works
export var deviceAvailabilityIconMap = {
    sendData: {
        AVAILABLE: {
            icon: 'long-arrow-right',
            class: 'statusOk'
        },
        UNAVAILABLE: {
            icon: 'long-arrow-right',
            class: 'statusNok'
        },
        UNKNOWN: {
            icon: 'long-arrow-right',
            class: 'statusUnknown'
        },
        NOT_MONITORED: {
            icon: 'long-arrow-right',
            class: 'statusUnknown'
        }
    },
    push: {
        CONNECTED: {
            icon: 'long-arrow-left',
            class: 'statusOk'
        },
        DISCONNECTED: {
            icon: 'long-arrow-left',
            class: 'statusUnknown'
        },
        UNKNOWN: {
            icon: 'long-arrow-left',
            class: 'statusUnknown'
        },
        MAINTENANCE: {
            icon: 'wrench',
            class: 'statusAlert'
        }
    },
    device: {
        MAINTENANCE: {
            icon: 'wrench',
            class: 'statusAlert'
        }
    }
};
var ɵ0 = function (lastMessage, translateService) {
    var date = formatDate(new Date(lastMessage), 'medium', translateService.currentLang);
    return translateService.instant(gettext('Device last request:\n{{date}}'), { date: date });
};
export var tooltips = {
    sendStatusConnected: ɵ0,
    sendStatusDisconnected: gettext('Connection not monitored'),
    pushStatusConnected: gettext('Connection to device established'),
    pushStatusDisconnected: gettext('Connection to device not established (not supported by device or no connectivity)'),
    maintenance: gettext('Device is under maintenance')
};
export var PushStatusLabels;
(function (PushStatusLabels) {
    PushStatusLabels[PushStatusLabels["CONNECTED"] = gettext('Push connection: active')] = "CONNECTED";
    PushStatusLabels[PushStatusLabels["DISCONNECTED"] = gettext('Push connection: inactive')] = "DISCONNECTED";
    PushStatusLabels[PushStatusLabels["UNKNOWN"] = gettext('Push connection: inactive')] = "UNKNOWN";
})(PushStatusLabels || (PushStatusLabels = {}));
export var SendStatusLabels;
(function (SendStatusLabels) {
    SendStatusLabels[SendStatusLabels["AVAILABLE"] = gettext('Send connection: online')] = "AVAILABLE";
    SendStatusLabels[SendStatusLabels["NOT_MONITORED"] = gettext('Send connection: not monitored')] = "NOT_MONITORED";
    SendStatusLabels[SendStatusLabels["UNAVAILABLE"] = gettext('Send connection: offline')] = "UNAVAILABLE";
    SendStatusLabels[SendStatusLabels["UNKNOWN"] = gettext('Send connection status unknown')] = "UNKNOWN";
    SendStatusLabels[SendStatusLabels["MAINTENANCE"] = gettext('Device is under maintenance')] = "MAINTENANCE";
})(SendStatusLabels || (SendStatusLabels = {}));
export { ɵ0 };
