// @c8y/ngx-components/esm5/core/modal/modal.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Optional, Output, EventEmitter, HostListener } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { gettext } from '../i18n/gettext';
import { isEmpty } from 'lodash';
/**
 * C8y modal component.
 * ## Example:
 *
 * Creating custom modal:
 * ```typescript
 *  import { Component } from "@angular/core";
 *  import { Subject } from 'rxjs';
 *
 * @Component({
 *  selector: "my-modal",
 *  template: `
 *    <c8y-modal title="CustomTitle"
 *      (onClose)="onClose($event)"
 *      (onDismiss)="onDismiss($event)"
 *      [labels]="labels"
 *      [disabled]="true" <- will disable ok button
 *      >
 *        <span>I am body of modal</span>
 *        <span>For simple string use body="string"</span>
 *    </c8y-modal>`
 * })
 *
 * export class MyModalComponent {
 *
 *  closeSubject: Subject<boolean> = new Subject();
 *  labels : ModalLabels = {ok: "customOK", cancel: "customCancel"};
 *
 *  onDismiss(event){
 *    this.closeSubject.next(false);
 *  }
 *
 *  onClose(event) {
 *    this.closeSubject.next(true);
 *  }
 * }
 * ```
 * Showing modal:
 * ```typescript
 * import { BsModalService } from "ngx-bootstrap/modal";
 *
 * constructor(
 *  public bsModalService: BsModalService,
 * ) {}
 *
 * showModal() {
 *  const modalRef = this.bsModalService.show(MyModalComponent);
 *  modalRef.content.closeSubject.subscribe(result => {
 *    console.log('results:', result);
 *  });
 * }
 * ```
 */
var ModalComponent = /** @class */ (function () {
    function ModalComponent(modal) {
        this.modal = modal;
        this.onDismiss = new EventEmitter();
        this.onClose = new EventEmitter();
        this.disabled = false;
        this.customFooter = false;
        this._labels = { ok: gettext('OK'), cancel: gettext('Cancel') };
    }
    Object.defineProperty(ModalComponent.prototype, "labels", {
        get: function () { return this._labels; },
        set: function (_a) {
            var ok = _a.ok, cancel = _a.cancel;
            this._labels = { ok: ok, cancel: cancel };
        },
        enumerable: true,
        configurable: true
    });
    ModalComponent.prototype._dismiss = function () {
        var _this = this;
        var fn = this.dismiss;
        if (!fn && this.modal) {
            fn = function () { return _this.modal.hide(); };
        }
        if (fn) {
            fn();
        }
        this.onDismiss.emit(true);
    };
    ModalComponent.prototype._close = function () {
        var _this = this;
        var fn = this.close;
        if (!fn && this.modal) {
            fn = function () { return _this.modal.hide(); };
        }
        if (fn) {
            fn();
        }
        this.onClose.emit(true);
    };
    ModalComponent.prototype.onEnterKeyDown = function (event) {
        var hasOnlyOneButton = isEmpty(this.labels.ok) !== isEmpty(this.labels.cancel);
        var isCancelClickable = !this.customFooter && this.labels.cancel;
        var isOkClickable = !this.customFooter && !this.disabled && this.labels.ok;
        if (hasOnlyOneButton && isCancelClickable) {
            this._dismiss();
        }
        if (hasOnlyOneButton && isOkClickable) {
            this._close();
        }
    };
    ModalComponent.ctorParameters = function () { return [
        { type: BsModalRef, decorators: [{ type: Optional }] }
    ]; };
    tslib_1.__decorate([
        Output()
    ], ModalComponent.prototype, "onDismiss", void 0);
    tslib_1.__decorate([
        Output()
    ], ModalComponent.prototype, "onClose", void 0);
    tslib_1.__decorate([
        Input()
    ], ModalComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        Input()
    ], ModalComponent.prototype, "close", void 0);
    tslib_1.__decorate([
        Input()
    ], ModalComponent.prototype, "dismiss", void 0);
    tslib_1.__decorate([
        Input()
    ], ModalComponent.prototype, "title", void 0);
    tslib_1.__decorate([
        Input()
    ], ModalComponent.prototype, "body", void 0);
    tslib_1.__decorate([
        Input()
    ], ModalComponent.prototype, "customFooter", void 0);
    tslib_1.__decorate([
        Input()
    ], ModalComponent.prototype, "labels", null);
    tslib_1.__decorate([
        HostListener('document:keydown.enter', ['$event'])
    ], ModalComponent.prototype, "onEnterKeyDown", null);
    ModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-modal',
            template: "<div class=\"modal-header\">\n  <ng-content select=\"[c8y-modal-title]\"></ng-content>\n  <h3 [hidden]=\"!title\">{{title | translate}}</h3>\n</div>\n<div class=\"modal-body\">\n  <p\n    [hidden]=\"!title\"\n    class=\"text-center text-wrap\"\n    style=\"word-wrap:break-word;\"\n  >{{body}}</p>\n  <ng-content></ng-content>\n</div>\n<div class=\"modal-footer\" *ngIf=\"!customFooter\">\n  <ng-content select=\"[c8y-modal-footer]\"></ng-content>\n  <button\n    title=\"{{ labels.cancel | translate }}\" \n    *ngIf=\"labels.cancel\"\n    class=\"btn btn-default\"\n    (click)=\"_dismiss()\"\n  >{{labels.cancel | translate}}</button>\n  <button\n    title=\"{{ labels.ok | translate }}\" \n    *ngIf=\"labels.ok\"\n    class=\"btn btn-primary\"\n    (click)=\"_close()\"\n    [disabled]=\"disabled\"\n  >{{labels.ok | translate}}</button>\n</div>\n"
        }),
        tslib_1.__param(0, Optional())
    ], ModalComponent);
    return ModalComponent;
}());
export { ModalComponent };
