// @c8y/ngx-components/esm2015/repository/software-device-tab/device-software-list.component.js
import * as tslib_1 from "tslib";
import { Component, Output, Input, EventEmitter } from '@angular/core';
import { filter } from 'lodash-es';
let DeviceSoftwareListComponent = class DeviceSoftwareListComponent {
    constructor() {
        this.update = new EventEmitter();
        this.remove = new EventEmitter();
    }
    ngAfterContentInit() {
        this.showUpdate = this.update.observers.length > 0;
        this.showRemove = this.remove.observers.length > 0;
    }
    isSoftwareGoingToBeChanged(software) {
        const relevantChanges = filter(this.deviceSoftwareChanges, software);
        return relevantChanges.length > 0;
    }
};
tslib_1.__decorate([
    Input()
], DeviceSoftwareListComponent.prototype, "softwareList", void 0);
tslib_1.__decorate([
    Input()
], DeviceSoftwareListComponent.prototype, "deviceSoftwareChanges", void 0);
tslib_1.__decorate([
    Output()
], DeviceSoftwareListComponent.prototype, "update", void 0);
tslib_1.__decorate([
    Output()
], DeviceSoftwareListComponent.prototype, "remove", void 0);
DeviceSoftwareListComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-device-software-list',
        template: "<c8y-list-group>\n  <c8y-li\n    [ngClass]=\"{ disabled: isSoftwareGoingToBeChanged(software) }\"\n    *ngFor=\"let software of softwareList\"\n  >\n    <!-- SOFTWARE ICON -->\n    <c8y-li-icon>\n      <i c8yIcon=\"c8y-tools\"></i>\n    </c8y-li-icon>\n\n    <c8y-li-body class=\"content-flex-20\">\n      <!-- SOFTWARE NAME -->\n      <div title=\"{{ software.name }}\" class=\"col-6\">\n        <p class=\"text-truncate\">{{ software.name }}</p>\n      </div>\n\n      <!-- SOFTWARE VERSION -->\n      <div [ngClass]=\"{ 'col-3': showUpdate || showRemove, 'col-5': !showUpdate && showRemove }\">\n        <p class=\"text-truncate\">\n          <span class=\"text-label-small m-r-4\" translate>\n            Version\n          </span>\n          <span title=\"{{ software.version }}\">\n            {{ software.version }}\n          </span>\n        </p>\n      </div>\n\n      <div *ngIf=\"showUpdate || showRemove\" class=\"col-3 p-r-16\">\n        <!-- UPDATE SOFTWARE -->\n        <button\n          *ngIf=\"showUpdate && !isSoftwareGoingToBeChanged(software)\"\n          class=\"btn btn-default btn-xs showOnHover\"\n          (click)=\"update.emit(software)\"\n          title=\"{{ 'Update`software,verb`' | translate }}\"\n          translate\n        >\n          Update\n        </button>\n\n        <!-- REMOVE SOFTWARE -->\n        <button\n          *ngIf=\"showRemove && !isSoftwareGoingToBeChanged(software)\"\n          title=\"{{ 'Remove`software`' | translate }}\"\n          class=\"showOnHover btn btn-dot pull-right\"\n          (click)=\"remove.emit(software)\"\n        >\n          <i c8yIcon=\"minus-circle\" class=\"text-danger\"></i>\n        </button>\n      </div>\n    </c8y-li-body>\n  </c8y-li>\n</c8y-list-group>\n"
    })
], DeviceSoftwareListComponent);
export { DeviceSoftwareListComponent };
