// devextreme-schematics/src/add-sample-views/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const ast_utils_1 = require("@schematics/angular/utility/ast-utils");
const add_view_1 = require("../add-view");
const project_1 = require("../utility/project");
const string_1 = require("../utility/string");
const change_1 = require("../utility/change");
const source_1 = require("../utility/source");
const sampleViewOptions = [
    {
        name: 'home',
        componentName: 'HomeComponent',
        relativePath: './pages/home/home.component'
    }, {
        name: 'profile',
        componentName: 'ProfileComponent',
        relativePath: './pages/profile/profile.component'
    }, {
        name: 'display-data',
        componentName: 'DisplayDataComponent',
        relativePath: './pages/display-data/display-data.component'
    }
];
const devextremeOptions = [
    {
        componentName: 'DxDataGridModule',
        relativePath: 'devextreme-angular'
    }, {
        componentName: 'DxFormModule',
        relativePath: 'devextreme-angular'
    }
];
const navigations = [
    `  {
    text: 'Home',
    path: '/home',
    icon: 'home'
  }`,
    `  {
    text: 'Examples',
    icon: 'folder',
    items: [
      {
        text: 'Profile',
        path: '/profile'
      },
      {
        text: 'Display Data',
        path: '/display-data'
      }
    ]
  }`
];
function addImportsToRoutingModule(isView, routingPath, options) {
    return (host) => {
        const source = source_1.getSourceFile(host, routingPath);
        if (!source) {
            return host;
        }
        let changes;
        if (isView) {
            changes = ast_utils_1.addDeclarationToModule(source, routingPath, options.componentName, options.relativePath);
        }
        else {
            changes = ast_utils_1.addImportToModule(source, routingPath, options.componentName, options.relativePath);
        }
        return change_1.applyChanges(host, changes, routingPath);
    };
}
function addDefaultNavigation(rootPath) {
    return (host) => {
        const navigationPath = rootPath + 'app-navigation.ts';
        navigations.forEach((navigation) => {
            const navigationSource = source_1.getSourceFile(host, navigationPath);
            change_1.insertItemToArray(host, navigationPath, navigationSource, navigation, { location: 'end' });
        });
        return host;
    };
}
function default_1(options) {
    return (host) => {
        const project = project_1.getProjectName(host, options.project);
        const rootPath = project_1.getApplicationPath(host, project);
        const routingPath = rootPath + 'app-routing.module.ts';
        const rules = [];
        const templateSource = schematics_1.apply(schematics_1.url('./files'), [
            schematics_1.template({
                project: string_1.humanize(project)
            }),
            schematics_1.move(rootPath)
        ]);
        rules.push(schematics_1.mergeWith(templateSource));
        sampleViewOptions.forEach((viewOptions) => {
            rules.push(add_view_1.addViewToRouting({ name: viewOptions.name, project, module: 'app-routing' }));
            rules.push(addImportsToRoutingModule(true, routingPath, viewOptions));
        });
        devextremeOptions.forEach((moduleOptions) => {
            rules.push(addImportsToRoutingModule(false, routingPath, moduleOptions));
        });
        rules.push(addDefaultNavigation(rootPath));
        return schematics_1.chain(rules);
    };
}
exports.default = default_1;
