// @c8y/ngx-components/esm2015/core/navigator/navigator-entry.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild, Input } from '@angular/core';
import { NavigatorNode } from './navigator-node';
import { NavigatorService } from './navigator.service';
let NavigatorItemComponent = class NavigatorItemComponent {
    constructor(service) {
        this.service = service;
        this.priority = 0;
        this.nodeData = new NavigatorNode();
    }
    ngOnInit() {
        this.ngOnChanges();
        setTimeout(() => this.service.add(this.nodeData));
    }
    ngOnChanges() {
        this.nodeData.icon = this.icon;
        this.nodeData.label = this.label;
        this.nodeData.path = this.path;
        this.nodeData.priority = this.priority;
        this.nodeData.iconTemplate = !this.icon ? this.template : undefined;
        this.nodeData.icon = this.icon;
    }
    ngOnDestroy() {
        this.service.remove(this.nodeData);
    }
};
NavigatorItemComponent.ctorParameters = () => [
    { type: NavigatorService }
];
tslib_1.__decorate([
    ViewChild('iconTpl', { static: true })
], NavigatorItemComponent.prototype, "template", void 0);
tslib_1.__decorate([
    Input()
], NavigatorItemComponent.prototype, "label", void 0);
tslib_1.__decorate([
    Input()
], NavigatorItemComponent.prototype, "icon", void 0);
tslib_1.__decorate([
    Input()
], NavigatorItemComponent.prototype, "path", void 0);
tslib_1.__decorate([
    Input()
], NavigatorItemComponent.prototype, "priority", void 0);
NavigatorItemComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-navigator-item',
        template: '<ng-template #iconTpl><ng-content></ng-content></ng-template>'
    })
], NavigatorItemComponent);
export { NavigatorItemComponent };
