// @c8y/ngx-components/esm2015/context-dashboard/welcome/welcome.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { gettext, DocsService, DocLink, NavigatorService } from '@c8y/ngx-components';
import { Router } from '@angular/router';
import { TenantService, ApplicationService, IApplication } from '@c8y/client';
let WelcomeToCockpit = class WelcomeToCockpit {
    constructor(tenantService, docs, router, navigator, applicationService) {
        this.tenantService = tenantService;
        this.docs = docs;
        this.router = router;
        this.navigator = navigator;
        this.applicationService = applicationService;
        this.quickLinks = [];
        this.REPORT_NODE = 'Reports';
        this.CONFIGURATION_NODE = 'Configuration';
        this.TRIAL = 'TRIAL';
    }
    ngOnInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.tenantType = yield this.tenantService.currentTenantType();
            this.setMessage();
            this.navSubscription = this.navigator.items$.subscribe(nodes => {
                this.navNodes = nodes;
            });
            this.docsSubscription = this.docs.items$.subscribe(links => {
                this.links = links;
            });
            // <---TRIAL & REGULAR TENANT --->
            this.createConnectSmartphoneQuickLink();
            yield this.createQuicklinkRegisterDevice();
            // <--- TRIAL TENANT --->
            if (this.tenantType === this.TRIAL) {
                this.createQuicklinkUserGuide();
                return;
            }
            // <--- REGULAR TENANT --->
            this.createQuicklinkAddGroup();
            this.createQuickLinkReports();
            this.createQuicklinkDataPointLibrary();
            this.createQuicklinkSmartRules();
        });
    }
    ngOnDestroy() {
        if (this.docsSubscription && !this.docsSubscription.closed) {
            this.docsSubscription.unsubscribe();
        }
        if (this.navSubscription && !this.navSubscription.closed) {
            this.navSubscription.unsubscribe();
        }
    }
    setMessage() {
        if (this.tenantType === this.TRIAL) {
            this.welcomeMessage = gettext(`
        The Cockpit application allows you to build IoT applications in minutes.
        To get started, connect any device to the platform.
        If you do not have an IoT device to hand, you can start by connecting your smartphone.
        Click below to be guided through the process.
      `);
        }
        else {
            this.welcomeMessage = gettext(`
        The Cockpit application provides you with options to manage
        and monitor Internet of Things assets and data from business perspective.
      `);
        }
    }
    createQuicklinkAddGroup() {
        // comes from angularJS factory c8yQuickLinks
        const addGroup = this.links.find(link => link.label === 'Add group');
        if (addGroup) {
            this.quickLinks.push(addGroup);
        }
    }
    createConnectSmartphoneQuickLink() {
        // Provider in SensorPhoneModule defines the
        // 'Connect smartphone' quicklink.
        const connectSmartphone = this.links.find(link => link.label === 'Connect smartphone');
        if (connectSmartphone) {
            this.quickLinks.push(connectSmartphone);
        }
    }
    createQuickLinkReports() {
        const reports = {
            icon: 'c8y-reports',
            label: gettext('Reports'),
            url: '/export'
        };
        const reportsNode = this.findNavigatorNode(this.REPORT_NODE);
        if (reportsNode && reportsNode.show) {
            reports.click = () => {
                reportsNode.open = true;
                this.router.navigateByUrl(reports.url);
            };
            this.quickLinks.push(reports);
        }
    }
    createQuicklinkDataPointLibrary() {
        const dataPointLib = {
            icon: 'c8y-data-points',
            label: gettext('Data point library'),
            url: '/datapointlibrary'
        };
        const configurationNode = this.findNavigatorNode(this.CONFIGURATION_NODE);
        if (configurationNode && configurationNode.show) {
            dataPointLib.click = () => {
                configurationNode.open = true;
                this.router.navigateByUrl(dataPointLib.url);
            };
            this.quickLinks.push(dataPointLib);
        }
    }
    createQuicklinkSmartRules() {
        const smartRules = {
            icon: 'c8y-smart-rules',
            label: gettext('Smart rules'),
            url: '/rules'
        };
        const configurationNode = this.findNavigatorNode(this.CONFIGURATION_NODE);
        if (configurationNode && configurationNode.show) {
            smartRules.click = () => {
                configurationNode.open = true;
                this.router.navigateByUrl(smartRules.url);
            };
            this.quickLinks.push(smartRules);
        }
    }
    createQuicklinkRegisterDevice() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.applicationService.listByUser();
            if (data) {
                const deviceManagement = data.find(app => app.contextPath === 'devicemanagement');
                if (deviceManagement) {
                    const deviceMgmtUrl = this.applicationService.getHref(deviceManagement);
                    const registerDevice = {
                        icon: 'c8y-device-connect',
                        label: gettext('Register device'),
                        click: () => window.open(`${deviceMgmtUrl}/#/deviceregistration`, '_self')
                    };
                    this.quickLinks.push(registerDevice);
                }
            }
        });
    }
    createQuicklinkUserGuide() {
        const userGuide = {
            icon: 'c8y-user',
            label: gettext('User guide'),
            url: '/users-guide/getting-started',
            click: () => {
                const userGuideURL = this.docs.getUserGuideLink(userGuide.url);
                window.open(userGuideURL);
            }
        };
        this.quickLinks.push(userGuide);
    }
    findNavigatorNode(nodeRealName) {
        if (this.navNodes && this.navNodes.length > 0) {
            return this.navNodes.find((node) => node.realName === nodeRealName);
        }
        return undefined;
    }
};
WelcomeToCockpit.ctorParameters = () => [
    { type: TenantService },
    { type: DocsService },
    { type: Router },
    { type: NavigatorService },
    { type: ApplicationService }
];
tslib_1.__decorate([
    Input()
], WelcomeToCockpit.prototype, "config", void 0);
WelcomeToCockpit = tslib_1.__decorate([
    Component({
        selector: 'c8y-welcome-to-cockpit',
        template: "<div class=\"welcome-widget welcome-cockpit\">\r\n  <div class=\"flex-row\">\r\n    <div class=\"col-xs-12 col-md-5 flex-item-v-stretch p-24\">\r\n      <h2 class=\"text-light\">{{ 'Welcome to Cockpit' | translate }}</h2>\r\n      <p class=\"text-16 text-light p-t-16 p-b-24\">{{ welcomeMessage | translate }}</p>\r\n      <div class=\"card-group interact-grid tight-grid\">\r\n        <div *ngFor=\"let link of quickLinks\" class=\"col-sm-4 col-xs-6\">\r\n          <c8y-quick-link\r\n            (click)=\"link.click ? link.click() : false\"\r\n            [icon]=\"link.icon\"\r\n            [label]=\"link.label\"\r\n            class=\"card\"\r\n          >\r\n          </c8y-quick-link>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <!-- <div class=\"col-sm-6 welcome-illustration flex-item-v-stretch\"></div> -->\r\n  </div>\r\n</div>\r\n"
    })
], WelcomeToCockpit);
export { WelcomeToCockpit };
