// @c8y/ngx-components/esm5/core/user/user-totp-revoke.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { IUser, UserService } from '@c8y/client';
import { AlertService } from '../alert/public-api';
import { gettext } from '../i18n/gettext';
var UserTotpRevokeComponent = /** @class */ (function () {
    function UserTotpRevokeComponent(users, alert) {
        this.users = users;
        this.alert = alert;
        this.revokeEmitter = new EventEmitter();
    }
    UserTotpRevokeComponent.prototype.revoke = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var e_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.users.totpRevokeSecret(this.user)];
                    case 1:
                        _a.sent();
                        this.alert.success(gettext('User TOTP secret revoked.'));
                        this.revokeEmitter.next();
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _a.sent();
                        this.alert.addServerFailure(e_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    UserTotpRevokeComponent.ctorParameters = function () { return [
        { type: UserService },
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], UserTotpRevokeComponent.prototype, "user", void 0);
    tslib_1.__decorate([
        Output('revoke')
    ], UserTotpRevokeComponent.prototype, "revokeEmitter", void 0);
    UserTotpRevokeComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-user-totp-revoke',
            template: "<button\n  type=\"button\"\n  class=\"btn btn-default\"\n  [popover]=\"confirm\"\n  [popoverTitle]=\"'Revoke TOTP secret' | translate\"\n  #popover=\"bs-popover\"\n  placement=\"right\"\n>\n  <i c8yIcon=\"key\"></i> {{ 'Revoke TOTP secret' | translate }}\n</button>\n\n<ng-template #confirm>\n  <p class=\"m-b-16\" translate>\n    This action will revoke the user's stored TOTP secret which will require a new setup.\n  </p>\n  <button\n    title=\"{{ 'Revoke' | translate }}\"\n    type=\"button\"\n    class=\"btn btn-danger\"\n    (click)=\"revoke(); popover.hide()\"\n    translate\n  >\n    Revoke\n  </button>\n  <button\n    title=\"{{ 'Cancel' | translate }}\"\n    type=\"button\"\n    class=\"btn btn-default\"\n    (click)=\"popover.hide()\"\n    translate\n  >\n    Cancel\n  </button>\n</ng-template>\n"
        })
    ], UserTotpRevokeComponent);
    return UserTotpRevokeComponent;
}());
export { UserTotpRevokeComponent };
