// @c8y/ngx-components/esm5/context-dashboard/appearance-settings.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { DASHBOARD_THEME_CLASSES, WIDGET_HEADER_CLASSES } from './context-dashboard.model';
import { ControlContainer, NgForm } from '@angular/forms';
var AppearanceSettingsComponent = /** @class */ (function () {
    function AppearanceSettingsComponent() {
        this.themeClass = 'dashboard-theme-light';
        this.headerClass = 'panel-title-regular';
        this.themeClassChange = new EventEmitter();
        this.headerClassChange = new EventEmitter();
        this.possibleStylingTheme = DASHBOARD_THEME_CLASSES;
        this.possibleStylingHeader = WIDGET_HEADER_CLASSES;
    }
    AppearanceSettingsComponent.prototype.themeClassClick = function (value) {
        value = this.resetToDefault(value, this.themeClass, this.defaultThemeClass);
        this.themeClass = value;
        this.themeClassChange.emit(value);
    };
    AppearanceSettingsComponent.prototype.headerClassClick = function (value) {
        value = this.resetToDefault(value, this.headerClass, this.defaultHeaderClass);
        this.headerClass = value;
        this.headerClassChange.emit(value);
    };
    AppearanceSettingsComponent.prototype.resetToDefault = function (value, compareTo, defaultValue) {
        if (defaultValue && value === compareTo) {
            value = defaultValue;
        }
        return value;
    };
    tslib_1.__decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "themeClass", void 0);
    tslib_1.__decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "headerClass", void 0);
    tslib_1.__decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "defaultThemeClass", void 0);
    tslib_1.__decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "defaultHeaderClass", void 0);
    tslib_1.__decorate([
        Output()
    ], AppearanceSettingsComponent.prototype, "themeClassChange", void 0);
    tslib_1.__decorate([
        Output()
    ], AppearanceSettingsComponent.prototype, "headerClassChange", void 0);
    tslib_1.__decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "possibleStylingTheme", void 0);
    tslib_1.__decorate([
        Input()
    ], AppearanceSettingsComponent.prototype, "possibleStylingHeader", void 0);
    AppearanceSettingsComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-appearance-settings',
            template: "<h6 class=\"legend form-block\"><span translate>Layout</span></h6>\n<div class=\"row \">\n  <div class=\"col-sm-6 m-b-32\">\n    <label translate>Theme</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row\"\n        style=\"align-items: center; box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let themeClassItem of possibleStylingTheme; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradiocontentclass' + i\"\n              name=\"content\"\n              [value]=\"themeClassItem.class\"\n              [ngModel]=\"themeClass\"\n              (click)=\"themeClassClick(themeClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradiocontentclass' + i\">\n          <span>{{ themeClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"themeClassItem.class === defaultThemeClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ themeClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\">\n          <i \n            [c8yIcon]=\"'question-circle-o'\"\n            class=\"text-primary\"\n          ></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-sm-6\">\n    <label translate>Widget header style</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row\"\n        style=\"align-items: center; box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let headerClassItem of possibleStylingHeader; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradioheaderclass' + i\"\n              name=\"header\"\n              [value]=\"headerClassItem.class\"\n              [ngModel]=\"headerClass\"\n              (click)=\"headerClassClick(headerClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradioheaderclass' + i\">\n          <span>{{ headerClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"headerClassItem.class === defaultHeaderClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ headerClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\">\n          <i \n            [c8yIcon]=\"'question-circle-o'\"\n            class=\"text-primary\"\n          ></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n</div>\n",
            viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
        })
    ], AppearanceSettingsComponent);
    return AppearanceSettingsComponent;
}());
export { AppearanceSettingsComponent };
