// @c8y/ngx-components/esm5/context-dashboard/cockpit-dashboard.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
var CockpitDashboardComponent = /** @class */ (function () {
    function CockpitDashboardComponent() {
        this.DEFAULT_HOME_WIDGETS = [
            {
                name: 'cockpit.welcome.widget',
                title: gettext(''),
                _x: 0,
                _y: 0,
                _height: 5,
                _width: 12,
                classes: { 'panel-title-hidden': true }
            },
            {
                name: 'Asset Alarms',
                title: gettext('Active, critical alarms'),
                _x: 0,
                _y: 5,
                _height: 4,
                _width: 6
            },
            {
                name: 'Recent Alarms',
                title: gettext('Recent alarms'),
                _x: 0,
                _y: 9,
                _height: 4,
                _width: 6
            },
            {
                name: 'Map',
                title: gettext('Map'),
                _x: 6,
                _y: 5,
                _height: 8,
                _width: 6
            }
        ];
    }
    CockpitDashboardComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-cockpit-dashboard',
            template: "<c8y-title translate>\n  Home\n</c8y-title>\n\n<c8y-context-dashboard name=\"home-cockpit1\" [defaultWidgets]=\"DEFAULT_HOME_WIDGETS\" [canDelete]=\"false\"></c8y-context-dashboard>\n"
        })
    ], CockpitDashboardComponent);
    return CockpitDashboardComponent;
}());
export { CockpitDashboardComponent };
