// @c8y/ngx-components/esm5/protocol-opcua/opcua-server-list.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
var OpcuaServerListComponent = /** @class */ (function () {
    function OpcuaServerListComponent() {
        this.serverList = [];
        this.present = new EventEmitter();
    }
    OpcuaServerListComponent.prototype.presentConfig = function (server) {
        server.active = true;
        this.present.emit(server);
    };
    tslib_1.__decorate([
        Input()
    ], OpcuaServerListComponent.prototype, "serverList", void 0);
    tslib_1.__decorate([
        Output()
    ], OpcuaServerListComponent.prototype, "present", void 0);
    OpcuaServerListComponent = tslib_1.__decorate([
        Component({
            selector: 'opcua-server-list',
            template: "<!-- add active class to c8y-stacked-item when clicked -->\n<div\n  class=\"c8y-stacked-item\"\n  [class.active]=\"server.active\"\n  *ngFor=\"let server of serverList\"\n  (click)=\"presentConfig(server)\"\n>\n  <i [c8yIcon]=\"'server'\" style=\"font-size:1.4em;\" class=\"text-gray-darker\"></i>\n  <div class=\"left-p-md text-truncate\">\n    <p id=\"serverName\" class=\"text-truncate\">{{ server.name }}</p>\n    <small id=\"serverUrl\" class=\"text-truncate text-muted\">{{ server.config.serverUrl }}</small>\n  </div>\n  <div class=\"flex-item-right left-p-sm\" style=\"display: flex;\">\n    <device-status class=\"right-p-sm\" [mo]=\"server\"></device-status>\n    <a\n      title=\"{{ server.quickInfo.padlockMsg | translate }}\"\n      placement=\"left\"\n      class=\"text-gray-darker\"\n    >\n      <i [c8yIcon]=\"server.quickInfo.padlock\"></i>\n    </a>\n  </div>\n</div>\n"
        })
    ], OpcuaServerListComponent);
    return OpcuaServerListComponent;
}());
export { OpcuaServerListComponent };
