// @c8y/ngx-components/bundles/c8y-ngx-components-device-lists.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/router'), require('@c8y/ngx-components'), require('lodash-es')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/device-lists', ['exports', '@angular/core', '@angular/router', '@c8y/ngx-components', 'lodash-es'], factory) :
    (global = global || self, factory((global.c8y = global.c8y || {}, global.c8y['ngx-components'] = global.c8y['ngx-components'] || {}, global.c8y['ngx-components']['device-lists'] = {}), global.ng.core, global.ng.router, global.c8y['ngx-components'], global.lodashEs));
}(this, (function (exports, core, router, ngxComponents, lodashEs) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    var DeviceGridExampleGuard = /** @class */ (function () {
        function DeviceGridExampleGuard(optionsService) {
            this.optionsService = optionsService;
        }
        DeviceGridExampleGuard.prototype.canActivate = function (route, state) {
            // FIXME: not entirely correct but enough for testing (see beta provider isActivatedFor)
            return Boolean(this.optionsService.get('beta', false));
        };
        DeviceGridExampleGuard.ctorParameters = function () { return [
            { type: ngxComponents.OptionsService }
        ]; };
        DeviceGridExampleGuard = __decorate([
            core.Injectable()
        ], DeviceGridExampleGuard);
        return DeviceGridExampleGuard;
    }());

    var DeviceListsNavigationFactory = /** @class */ (function () {
        function DeviceListsNavigationFactory(deviceGridExampleGuard) {
            this.deviceGridExampleGuard = deviceGridExampleGuard;
        }
        DeviceListsNavigationFactory.prototype.get = function () {
            return __awaiter(this, void 0, void 0, function () {
                var navs;
                return __generator(this, function (_a) {
                    if (!this.deviceGridExampleGuard.canActivate(null, null)) {
                        return [2 /*return*/, []];
                    }
                    navs = new ngxComponents.NavigatorNode({
                        label: ngxComponents.gettext('Grids'),
                        icon: 'table'
                    });
                    navs.add(new ngxComponents.NavigatorNode({
                        label: ngxComponents.gettext('Device grid'),
                        path: '/grids/device-grid',
                        icon: 'table'
                    }));
                    return [2 /*return*/, navs];
                });
            });
        };
        DeviceListsNavigationFactory.ctorParameters = function () { return [
            { type: DeviceGridExampleGuard }
        ]; };
        DeviceListsNavigationFactory = __decorate([
            core.Injectable()
        ], DeviceListsNavigationFactory);
        return DeviceListsNavigationFactory;
    }());

    var DeviceGridExampleComponent = /** @class */ (function () {
        function DeviceGridExampleComponent(deviceGridService) {
            this.deviceGridService = deviceGridService;
            this.examples = [
                this.getExampleWithStandardColumns(),
                this.getExampleWithStandardColumnsAndSavedFilteringSorting(),
                this.getExampleWithCustomColumnAndPredefinedFiltering()
            ];
        }
        DeviceGridExampleComponent.prototype.getExampleWithStandardColumns = function () {
            var columns = this.deviceGridService.getDefaultColumns();
            return {
                label: 'Standard columns',
                columns: columns
            };
        };
        DeviceGridExampleComponent.prototype.getExampleWithStandardColumnsAndSavedFilteringSorting = function () {
            var columns = this.deviceGridService.getDefaultColumns();
            var nameColumn = lodashEs.find(columns, { name: 'name' });
            nameColumn.externalFilterQuery = {
                names: ['*Temp*']
            };
            nameColumn.sortOrder = 'desc';
            return {
                label: 'Standard columns with saved filtering/sorting',
                columns: columns
            };
        };
        DeviceGridExampleComponent.prototype.getExampleWithCustomColumnAndPredefinedFiltering = function () {
            var columns = this.deviceGridService.getDefaultColumns();
            var typeColumn = new ngxComponents.CustomDeviceGridColumn();
            typeColumn.name = 'type';
            typeColumn.path = 'type';
            typeColumn.header = ngxComponents.gettext('Type');
            typeColumn.externalFilterQuery = {
                equals: ['c8y_MQTTDevice']
            };
            columns.splice(2, 0, typeColumn);
            return {
                label: 'Standard columns with custom one with predefined filtering',
                columns: columns
            };
        };
        DeviceGridExampleComponent.prototype.onExampleSelect = function (example) {
            this.columnsInput = example.columns;
        };
        DeviceGridExampleComponent.ctorParameters = function () { return [
            { type: ngxComponents.DeviceGridService }
        ]; };
        DeviceGridExampleComponent = __decorate([
            core.Component({
                selector: 'c8y-device-grid-example',
                template: "<div class=\"row\">\n  <c8y-device-grid\n    [title]=\"'Select devices for bulk operation'\"\n    [columns]=\"columnsInput\"\n    [actionControls]=\"[]\"\n    (onColumnsChange)=\"columnsOutput = $event\"\n    (onDeviceQueryStringChange)=\"deviceQueryStringOutput = $event\"\n  ></c8y-device-grid>\n</div>\n\n<div class=\"row\">\n  <div class=\"card col-sm-6\">\n    <div class=\"card-header separator\">\n      <h4 class=\"card-title\">Input</h4>\n    </div>\n    <div class=\"card-block\">\n      <div class=\"c8y-select-wrapper\">\n        <select\n          class=\"form-control\"\n          [ngModel]=\"selectedExample\"\n          (ngModelChange)=\"onExampleSelect($event)\"\n        >\n          <option *ngFor=\"let example of examples\" [ngValue]=\"example\">{{\n            example.label\n          }}</option></select\n        ><span></span>\n      </div>\n      <pre\n        >{{ columnsInput | json }}\n      </pre>\n    </div>\n  </div>\n\n  <div class=\"card col-sm-6\">\n    <div class=\"card-header\">\n      <span>\n        <h4 class=\"card-title\">Output</h4>\n      </span>\n    </div>\n    <div class=\"card-block\">\n      <pre\n        >{{ deviceQueryStringOutput | json }}\n      </pre>\n      <pre\n        >{{ columnsOutput | json }}\n      </pre>\n    </div>\n  </div>\n</div>\n"
            })
        ], DeviceGridExampleComponent);
        return DeviceGridExampleComponent;
    }());

    var routes = [
        {
            path: 'grids/device-grid',
            component: DeviceGridExampleComponent,
            canActivate: [DeviceGridExampleGuard]
        }
    ];
    var DeviceListsModule = /** @class */ (function () {
        function DeviceListsModule() {
        }
        DeviceListsModule = __decorate([
            core.NgModule({
                imports: [ngxComponents.CoreModule, ngxComponents.CommonModule, router.RouterModule.forRoot(routes, { useHash: true })],
                declarations: [DeviceGridExampleComponent],
                entryComponents: [DeviceGridExampleComponent],
                exports: [],
                providers: [
                    {
                        provide: ngxComponents.HOOK_NAVIGATOR_NODES,
                        useClass: DeviceListsNavigationFactory,
                        multi: true
                    },
                    DeviceGridExampleGuard
                ]
            })
        ], DeviceListsModule);
        return DeviceListsModule;
    }());

    exports.DeviceListsModule = DeviceListsModule;
    exports.ɵa = DeviceGridExampleComponent;
    exports.ɵb = DeviceGridExampleGuard;
    exports.ɵc = DeviceListsNavigationFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
