// @c8y/ngx-components/esm5/repository/repository.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ONCE_ROUTE } from '@c8y/ngx-components';
import { ConfigurationListComponent } from './configuration/configuration-list.component';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { RepositoryService } from './repository.service';
import { ConfigurationNewModalComponent } from './configuration/configuration-new-modal.component';
import { RepositoryNavigationFactory } from './repository-navigation-factory';
var ɵ0 = [
    {
        path: 'configuration-new',
        component: ConfigurationListComponent
    }
];
var RepositoryModule = /** @class */ (function () {
    function RepositoryModule() {
    }
    RepositoryModule = tslib_1.__decorate([
        NgModule({
            imports: [CoreModule, TooltipModule],
            exports: [],
            entryComponents: [ConfigurationListComponent, ConfigurationNewModalComponent],
            declarations: [ConfigurationListComponent, ConfigurationNewModalComponent],
            providers: [
                RepositoryService,
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: RepositoryNavigationFactory,
                    multi: true
                },
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: ɵ0,
                    multi: true
                }
            ]
        })
    ], RepositoryModule);
    return RepositoryModule;
}());
export { RepositoryModule };
export { ɵ0 };
