// @c8y/ngx-components/esm5/core/common/humanize-app-name.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { gettext } from '../i18n/gettext';
import { get, find, negate, isUndefined } from 'lodash';
/*
 * Humanize an application name and translates it.
 * Usage:
 *   appNameOrApp | humanizeAppName | async
*/
var HumanizeAppNamePipe = /** @class */ (function () {
    function HumanizeAppNamePipe(translateService) {
        this.translateService = translateService;
        this.internalApps = {
            administration: 'Administration',
            cockpit: 'Cockpit',
            devicemanagement: 'Device management'
        };
    }
    HumanizeAppNamePipe.prototype.transform = function (appNameOrApp) {
        if (appNameOrApp === void 0) { appNameOrApp = 'Cumulocity'; }
        var appName = this.getAppName(appNameOrApp);
        var text;
        if (this.isPublicApp(appNameOrApp)) {
            text = gettext('{{ appName | translate }} (public)');
        }
        else {
            text = appName;
        }
        return this.translateService.get(text, { appName: appName });
    };
    HumanizeAppNamePipe.prototype.getAppName = function (appNameOrApp) {
        var appName = this.getFirstDefined([
            get(appNameOrApp, 'name'),
            get(appNameOrApp, 'contextPath'),
            appNameOrApp
        ]) || '';
        return this.internalApps[appName] || appName;
    };
    HumanizeAppNamePipe.prototype.getFirstDefined = function (values) {
        return find(values, negate(isUndefined));
    };
    HumanizeAppNamePipe.prototype.isPublicApp = function (appNameOrApp) {
        return Boolean(get(appNameOrApp, 'public'));
    };
    HumanizeAppNamePipe.ctorParameters = function () { return [
        { type: TranslateService }
    ]; };
    HumanizeAppNamePipe = tslib_1.__decorate([
        Pipe({ name: 'humanizeAppName' })
    ], HumanizeAppNamePipe);
    return HumanizeAppNamePipe;
}());
export { HumanizeAppNamePipe };
