// @c8y/ngx-components/esm5/core/common/single-operation.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { operationStatusClasses, operationStatusIcons } from './status.model';
/**
 * Single operation component displays a single operation status, description and creationTime.
 *
 * Example:
 *
 * ```html
 * <c8y-single-operation [operation]="operation"></c8y-single-operation>
 * ```
 */
var SingleOperationComponent = /** @class */ (function () {
    function SingleOperationComponent() {
    }
    SingleOperationComponent.prototype.ngOnInit = function () {
        this.statusIcons = operationStatusIcons;
        this.statusClasses = operationStatusClasses;
    };
    tslib_1.__decorate([
        Input()
    ], SingleOperationComponent.prototype, "operation", void 0);
    SingleOperationComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-single-operation',
            template: "<div class=\"c8y-list--group\" *ngIf=\"operation\">\n  <a href=\"#/device/{{ operation.deviceId }}/control\" class=\"c8y-list__item\">\n    <div class=\"c8y-list__item__block\">\n      <div class=\"c8y-list__item__icon\">\n        <i\n          [c8yIcon]=\"statusIcons[operation.status]\"\n          [ngClass]=\"statusClasses[operation.status]\"\n          [tooltip]=\"operation.status?.toString() | translate\"\n          placement=\"right\"\n          container=\"body\"\n        ></i>\n      </div>\n      <div class=\"c8y-list__item__body p-r-16\">\n        <div class=\"content-flex-40\">\n          <div class=\"col-8 text-break-word\">{{ operation.description | translate }}</div>\n          <div class=\"col-4 text-muted\">\n            <i c8yIcon=\"calendar\" class=\"m-r-4\"></i>\n            <small>{{ operation.creationTime | c8yDate }}</small>\n          </div>\n        </div>\n      </div>\n    </div>\n  </a>\n</div>\n"
        })
    ], SingleOperationComponent);
    return SingleOperationComponent;
}());
export { SingleOperationComponent };
