// @c8y/ngx-components/esm5/core/common/status.model.js
var _a, _b, _c, _d, _e;
import { OperationStatus } from '@c8y/client';
export var Status;
(function (Status) {
    Status["SUCCESS"] = "success";
    Status["WARNING"] = "warning";
    Status["DANGER"] = "danger";
    Status["INFO"] = "info";
})(Status || (Status = {}));
export var statusIcons = (_a = {},
    _a[Status.SUCCESS] = 'check-circle',
    _a[Status.WARNING] = 'exclamation-triangle',
    _a[Status.DANGER] = 'exclamation-circle',
    _a[Status.INFO] = 'info-circle',
    _a);
export var statusClasses = (_b = {},
    _b[Status.SUCCESS] = "text-" + Status.SUCCESS,
    _b[Status.WARNING] = "text-" + Status.WARNING,
    _b[Status.DANGER] = "text-" + Status.DANGER,
    _b[Status.INFO] = "text-" + Status.INFO,
    _b);
export var statusAlert = (_c = {},
    _c[Status.SUCCESS] = "alert-" + Status.SUCCESS,
    _c[Status.WARNING] = "alert-" + Status.WARNING,
    _c[Status.DANGER] = "alert-" + Status.DANGER,
    _c[Status.INFO] = "alert-" + Status.INFO,
    _c);
export var operationStatusIcons = (_d = {},
    _d[OperationStatus.PENDING] = 'clock-o',
    _d[OperationStatus.EXECUTING] = 'refresh',
    _d[OperationStatus.SUCCESSFUL] = 'check-circle',
    _d[OperationStatus.FAILED] = 'exclamation-circle',
    _d);
export var operationStatusClasses = (_e = {},
    _e[OperationStatus.PENDING] = 'text-info',
    _e[OperationStatus.EXECUTING] = 'text-info',
    _e[OperationStatus.SUCCESSFUL] = 'text-success',
    _e[OperationStatus.FAILED] = 'text-danger',
    _e);
