// @c8y/ngx-components/esm5/core/forms/filter-input.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Subject, timer } from 'rxjs';
import { debounce, distinctUntilChanged, map } from 'rxjs/operators';
var FilterInputComponent = /** @class */ (function () {
    function FilterInputComponent() {
        this.onSearch = new EventEmitter();
        this.filterChange$ = new Subject();
        this.filterTerm = '';
        this.CARRIAGE_RETURN_KEY = 13;
        this.CARRIAGE_RETURN_CODE = 'Enter';
    }
    FilterInputComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.reset();
        this.filterChange$
            .pipe(debounce(function (event) {
            return event.code === _this.CARRIAGE_RETURN_CODE || event.keyCode === _this.CARRIAGE_RETURN_KEY
                ? timer(10)
                : timer(300);
        }), map(function (e) { return e.target.value; }), distinctUntilChanged())
            .subscribe(function (filterTerm) {
            _this.filterTerm = filterTerm;
            _this.onSearch.emit(filterTerm);
        });
    };
    FilterInputComponent.prototype.reset = function () {
        this.filterTerm = '';
        this.onSearch.emit('');
    };
    FilterInputComponent.prototype.ngOnDestroy = function () {
        this.filterChange$.complete();
    };
    tslib_1.__decorate([
        Input()
    ], FilterInputComponent.prototype, "icon", void 0);
    tslib_1.__decorate([
        Output()
    ], FilterInputComponent.prototype, "onSearch", void 0);
    FilterInputComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-filter',
            template: "<div class=\"input-group input-group-search\">\n  <input\n    class=\"form-control\"\n    placeholder=\"{{ 'Filter\u2026' | translate }}\"\n    type=\"text\"\n    [value]=\"filterTerm\"\n    (keyup)=\"filterChange$.next($event)\"\n  />\n  <span class=\"input-group-btn\">\n    <button class=\"btn btn-clean\" (click)=\"reset()\">\n      <i [c8yIcon]=\"filterTerm.length === 0 ? icon : 'close'\"></i>\n    </button>\n  </span>\n</div>\n"
        })
    ], FilterInputComponent);
    return FilterInputComponent;
}());
export { FilterInputComponent };
