// @c8y/ngx-components/esm5/core/forms/input-group-list.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
var InputGroupListComponent = /** @class */ (function () {
    function InputGroupListComponent() {
        this.plus = true;
        this.minus = true;
        this.onRemove = new EventEmitter();
        this.onAdd = new EventEmitter();
    }
    InputGroupListComponent.prototype.add = function () {
        this.onAdd.emit(this.index);
    };
    InputGroupListComponent.prototype.remove = function () {
        this.onRemove.emit(this.index);
    };
    tslib_1.__decorate([
        Input()
    ], InputGroupListComponent.prototype, "plus", void 0);
    tslib_1.__decorate([
        Input()
    ], InputGroupListComponent.prototype, "minus", void 0);
    tslib_1.__decorate([
        Input()
    ], InputGroupListComponent.prototype, "index", void 0);
    tslib_1.__decorate([
        Output()
    ], InputGroupListComponent.prototype, "onRemove", void 0);
    tslib_1.__decorate([
        Output()
    ], InputGroupListComponent.prototype, "onAdd", void 0);
    InputGroupListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-input-group-list',
            template: "<ng-container>\n  <div class=\"input-group\">\n    <ng-content></ng-content>\n    <div class=\"input-group-btn\">\n      <button\n        *ngIf=\"minus\"\n        title=\"{{ 'Remove' | translate }}\"\n        class=\"btn btn-clean text-primary\"\n        (click)=\"remove()\"\n      >\n        <i c8yIcon=\"minus-circle\" class=\"text-danger\"></i>\n      </button>\n      <button\n        *ngIf=\"plus\"\n        title=\"{{ 'Add' | translate }}\"\n        class=\"btn btn-clean text-primary\"\n        (click)=\"add()\"\n      >\n        <i class=\"fa fa-plus-circle\"></i>\n      </button>\n    </div>\n  </div>\n</ng-container>\n"
        })
    ], InputGroupListComponent);
    return InputGroupListComponent;
}());
export { InputGroupListComponent };
