// @c8y/ngx-components/esm5/core/forms/text-area-row-height.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, HostBinding, Renderer2 } from '@angular/core';
var TextAreaRowHeightDirective = /** @class */ (function () {
    function TextAreaRowHeightDirective(elem, renderer) {
        this.elem = elem;
        this.renderer = renderer;
        this.rows = 1;
        this.minHeightPerRow = 20;
    }
    TextAreaRowHeightDirective.prototype.ngAfterViewInit = function () {
        var convertedValue = Number(this.heightPerRow);
        if (this.heightPerRow && convertedValue && convertedValue > this.minHeightPerRow) {
            this.setHeight(this.rows * convertedValue + "px");
        }
        else {
            this.setHeight(this.rows * this.minHeightPerRow + "px");
        }
    };
    TextAreaRowHeightDirective.prototype.setHeight = function (value) {
        this.renderer.setStyle(this.elem.nativeElement, 'height', value);
    };
    TextAreaRowHeightDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    tslib_1.__decorate([
        Input(),
        HostBinding('rows')
    ], TextAreaRowHeightDirective.prototype, "rows", void 0);
    tslib_1.__decorate([
        Input('heightPerRow')
    ], TextAreaRowHeightDirective.prototype, "heightPerRow", void 0);
    TextAreaRowHeightDirective = tslib_1.__decorate([
        Directive({
            selector: 'textarea[heightPerRow]'
        })
    ], TextAreaRowHeightDirective);
    return TextAreaRowHeightDirective;
}());
export { TextAreaRowHeightDirective };
