// @c8y/ngx-components/esm5/repository/select-modal/repository-select-modal.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { gettext, ModalSelectionMode } from '@c8y/ngx-components';
import { get } from 'lodash-es';
import { QueriesUtil } from '@c8y/client';
import { Subject, from, BehaviorSubject } from 'rxjs';
import { switchMap, map, mergeMap, shareReplay, tap } from 'rxjs/operators';
import { RepositoryService } from '../repository.service';
// MODAL STRUCTURE
// - selectModalObject (repository entry (repositoryCategory) -> type c8y_Firmware/c8y_Software)
//   -- ISelectModalOption (repository binary entry (repositoryBinary) => type c8y_FirmwareBinary/c8y_SoftwareBinary)
//   -- ISelectModalOption...
// - selectModalObject...
/**
 * RepositorySelectModalComponent displays repository entries options and allows to select them.
 *
 * @example
 * ```
 * import { take } from 'rxjs/operators';
 * import { RepositorySelectModalComponent, ModalSelectionMode, RepositoryType } from '@c8y/ngx-components/repository';
 *
 * const initialState = {
 *   repositoryType: RepositoryType.FIRMWARE,
 *   title: gettext('Install firmware'),
 *   subTitle: gettext('Available firmwares matching the device type'),
 *   icon: 'c8y-firmware',
 *   mode: ModalSelectionMode.SINGLE,
 *   labels: { ok: gettext('Install') },
 *   disableSelected: false
 * };
 *
 * const modal = this.bsModal.show(RepositorySelectModalComponent, {
 *   ignoreBackdropClick: true,
 *   initialState
 * });
 *
 * modal.content.load.next();
 * modal.content.resultEmitter.pipe(take(1)).subscribe((firmware) => {
 *   ...
 * })
 * ```
 */
var RepositorySelectModalComponent = /** @class */ (function () {
    function RepositorySelectModalComponent(repositoryService, translateService) {
        var _this = this;
        this.repositoryService = repositoryService;
        this.translateService = translateService;
        /**
         * Optional
         * Allows to provide custom data.
         * @example
         * ```
         * import { from } from 'rxjs';
         *
         * const repositoryEntry = { name: 'ExampleEntry', type: 'c8y_Firmware' };
         * const versions = [{ c8y_Firmware: { version: '1.0.0', url: 'http://example.com' } }];
         *
         * const initialState = {repositoryEntriesWithVersions$: from({ ...repositoryEntry, versions })};
         * ```
         */
        this.repositoryEntriesWithVersions$ = undefined;
        /**
         * Optional
         * Allows to use custom badges templates.
         * @example
         * ```
         * import { gettext } from '@c8y/ngx-components';
         *
         * const badgeTemplates = { '=1': gettext('{{count}} version'), other: gettext('{{count}} versions') };
         * const initialState = { badgeTemplates };
         * ```
         */
        this.badgeTemplates = { '=1': gettext('{{count}} version'), other: gettext('{{count}} versions') };
        /**
         * Optional
         * Allows to provide custom modal title.
         */
        this.title = gettext('Select repository entry');
        /**
         * Loads the content of the modal.
         * Must be invoked by the modal's caller.
         */
        this.load = new Subject();
        /**
         * Optional
         * Emits a search string currently entered in the filter input.
         * Use it to filter the items if you use custom repositoryEntriesWithVersions$.
         */
        this.searchTerm = new BehaviorSubject('');
        /**
         * Optional
         * Allows to provide device type query to restrict search criteria.
         * Only takes effect when repositoryEntriesWithVersions$ is not provided,
         * otherwise modal's caller have to provide already filtered data in the repositoryEntriesWithVersions$.
         */
        this.deviceTypeQuery = {};
        /**
         * Optional
         * Allows to provide query to restrict search criteria.
         * Only takes effect when repositoryEntriesWithVersions$ is not provided,
         * otherwise modal's caller have to provide already filtered data in the repositoryEntriesWithVersions$.
         */
        this.searchQuery = {};
        /**
         * Optional
         * Allows to provide custom labels for the buttons responsible for confirm/dismiss modal actions.
         */
        this.labels = { ok: gettext('Save') };
        /**
         * Optional
         * Allows to hide the filter input field.
         * By default, the filter input field is displayed.
         */
        this.showFilter = true;
        /**
         * Optional
         * Allows to show a warning that the search criteria should be narrowed down.
         * By default, this warning is hidden.
         */
        this.areMoreEntries = false;
        /**
         * Emits the list of selected options.
         */
        this.resultEmitter = new EventEmitter();
        /**
         * Optional
         * Allows to change selection mode.
         * Supported options:
         *   * single: only single option can be selected.
         *   * multiple: multiple options can be selected.
         */
        this.mode = ModalSelectionMode.SINGLE;
        /**
         * Allows to block selection of the other versions from the same repository entry.
         */
        this.disableSelected = true;
        this.modalEntries = this.load.pipe(switchMap(function () { return _this.repositoryEntriesWithVersions$; }), mergeMap(function (mos) { return _this.aggregate(mos); }), tap(function (items) {
            _this.areMoreEntries = items.length >= _this.PAGE_SIZE ? true : false;
        }));
        this.PAGE_SIZE = 100;
        this.queriesUtil = new QueriesUtil();
    }
    RepositorySelectModalComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.repositoryType) {
            throw new Error('Repository type must be defined');
        }
        if (!this.repositoryEntriesWithVersions$) {
            this.repositoryEntriesWithVersions$ = from(this.repositoryService.listRepositoryEntries(this.repositoryType, {
                query: this.queriesUtil.addOrFilter(this.deviceTypeQuery, this.searchQuery),
                params: { pageSize: this.PAGE_SIZE }
            })).pipe(map(function (_a) {
                var data = _a.data;
                return data;
            }), map(function (mos) { return _this.getAndAssignRepositoryBinaries(mos); }), shareReplay(1));
        }
    };
    RepositorySelectModalComponent.prototype.getAndAssignRepositoryBinaries = function (mos) {
        var _this = this;
        mos.forEach(function (mo) {
            mo.versions = _this.repositoryService.listAllVersions(mo);
        });
        return mos;
    };
    RepositorySelectModalComponent.prototype.search = function (searchTerm) {
        this.searchTerm.next(searchTerm);
        if (!searchTerm) {
            this.searchQuery = {};
        }
        else {
            this.searchQuery = { name: "*" + searchTerm + "*" };
        }
        this.load.next();
    };
    RepositorySelectModalComponent.prototype.result = function (selectedItems) {
        this.resultEmitter.emit(selectedItems);
    };
    RepositorySelectModalComponent.prototype.aggregate = function (mos) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var repositoryType, selectedItems;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                repositoryType = this.repositoryType;
                selectedItems = this.selected;
                return [2 /*return*/, Promise.all(mos.map(function (repositoryEntry) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                        var options, _a, selectModalObject;
                        return tslib_1.__generator(this, function (_b) {
                            switch (_b.label) {
                                case 0:
                                    _a = this.getSelectModalOptions;
                                    return [4 /*yield*/, this.repositoryService.fetchAllItemsFromList(repositoryEntry.versions)];
                                case 1:
                                    options = _a.apply(this, [_b.sent(),
                                        selectedItems,
                                        repositoryEntry,
                                        repositoryType]);
                                    selectModalObject = this.getSelectModalObject(repositoryEntry, options);
                                    return [2 /*return*/, selectModalObject];
                            }
                        });
                    }); }))];
            });
        });
    };
    RepositorySelectModalComponent.prototype.getSelectModalOptions = function (versions, selectedItems, repositoryEntry, repositoryType) {
        var _this = this;
        var selectModalOptions = [];
        versions.forEach(function (repositoryBinary) {
            var isSelected = _this.isBinaryRepositorySelected(selectedItems, repositoryEntry, repositoryBinary, repositoryType);
            var version = repositoryBinary["" + repositoryType].version;
            var bodyValue = version || "(" + _this.translateService.instant(gettext('not specified`version`')) + ")";
            var bodyClass = version ? '' : 'text-muted';
            selectModalOptions.push({
                body: [
                    {
                        value: bodyValue,
                        class: bodyClass
                    }
                ],
                obj: tslib_1.__assign({ id: repositoryBinary.id, name: repositoryEntry.name, version: version }, (get(repositoryBinary, 'c8y_Patch.dependency') && {
                    dependency: get(repositoryBinary, 'c8y_Patch.dependency')
                }), (get(repositoryBinary, 'c8y_Patch') && { c8y_Patch: true }), { url: repositoryBinary["" + repositoryType].url }),
                selected: isSelected
            });
        });
        return selectModalOptions;
    };
    RepositorySelectModalComponent.prototype.isBinaryRepositorySelected = function (selectedItems, repositoryEntry, repositoryBinary, repositoryType) {
        var isSelected = selectedItems
            ? selectedItems.filter(function (repositoryFragment) {
                return repositoryFragment.name === repositoryEntry.name &&
                    repositoryFragment.version === repositoryBinary["" + repositoryType].version;
            }).length > 0
            : false;
        return isSelected;
    };
    RepositorySelectModalComponent.prototype.getSelectModalObject = function (repositoryEntry, options) {
        var label = options.length === 1
            ? this.translateService.instant(this.badgeTemplates['=1'], { count: options.length })
            : this.translateService.instant(this.badgeTemplates.other, { count: options.length });
        var selectModalObject = {
            groupId: repositoryEntry.id,
            body: [
                { value: repositoryEntry.name, class: 'text-truncate' },
                { value: repositoryEntry.description, class: 'text-truncate text-muted' }
            ],
            additionalInformation: { value: label, class: 'label label-info' },
            options: options
        };
        return selectModalObject;
    };
    RepositorySelectModalComponent.ctorParameters = function () { return [
        { type: RepositoryService },
        { type: TranslateService }
    ]; };
    RepositorySelectModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-repository-select-modal',
            template: "<c8y-select-modal\n  [icon]=\"icon\"\n  [title]=\"title\"\n  [subTitle]=\"subTitle\"\n  [items]=\"modalEntries | async\"\n  [mode]=\"mode\"\n  [disableSelected]=\"disableSelected\"\n  [labels]=\"labels\"\n  [showFilter]=\"showFilter\"\n  [areMoreEntries]=\"areMoreEntries\"\n  (search)=\"search($event)\"\n  (result)=\"result($event)\"\n>\n</c8y-select-modal>\n"
        })
    ], RepositorySelectModalComponent);
    return RepositorySelectModalComponent;
}());
export { RepositorySelectModalComponent };
