// @c8y/client/lib/cjs/src/identity/IdentityService.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This class allwos for managing external identifiers.
 */
var IdentityService = /** @class */ (function () {
    function IdentityService(client) {
        this.baseUrl = 'identity';
        this.propertyName = 'externalIds';
        this.client = client;
    }
    /**
     * Gets the list of identities filtered by parameters.
     *
     * @returns Response wrapped in [[IResultList]]
     *
     * @param {object} filter Object containing filters for querying identity.
     *
     * @example
     * ```typescript
     *
     *  const filter: object = {
     *     pageSize: 100,
     *     withTotalPages: true
     *   };
     *
     *   (async () => {
     *     const {data, res, paging} = await identityService.list(filter);
     *   })();
     * ```
     */
    IdentityService.prototype.list = function (managedObjectId) {
        return __awaiter(this, void 0, void 0, function () {
            var headers, url, res, json, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        headers = { accept: 'application/json' };
                        url = this.getExternalIdsOfGlobalIdUrl(managedObjectId);
                        return [4 /*yield*/, this.fetch(url, { headers: headers })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        json = _a.sent();
                        data = json[this.propertyName];
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    /**
     * Creates a new identity.
     *
     * @param {IExternalIdentity} identity Identity object with mandantory fragments.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *   const linkedManagedObjectId = '123';
     *   const identity: IExternalIdentity = {
     *      type: 'type',
     *      externalId: '1',
     *      managedObject: {
     *        id: linkedManagedObjectId
     *      }
     *    };
     *
     *    (async () => {
     *      const {data, res} = await identityService.create(identity);
     *   })();
     * ```
     */
    IdentityService.prototype.create = function (identity) {
        return __awaiter(this, void 0, void 0, function () {
            var headers, method, body, url, res, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        headers = { 'content-type': 'application/json', accept: 'application/json' };
                        method = 'POST';
                        body = JSON.stringify(identity);
                        url = this.getExternalIdsOfGlobalIdUrl(identity.managedObject.id);
                        return [4 /*yield*/, this.fetch(url, { headers: headers, method: method, body: body })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    /**
     * Gets the details of an identity.
     *
     * @param {IExternalIdentity} identity Identity object with mandantory fragments.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @example
     * ```typescript
     *
     *    const identity: IExternalIdentity = {
     *      type: 'type',
     *      externalId: '1'
     *    };
     *
     *    (async () => {
     *      const {data, res} = await identityService.detail(identity);
     *   })();
     * ```
     */
    IdentityService.prototype.detail = function (identity) {
        return __awaiter(this, void 0, void 0, function () {
            var headers, url, res, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        headers = { accept: 'application/json' };
                        url = this.getExternalIdUrl(identity);
                        return [4 /*yield*/, this.fetch(url, { headers: headers })];
                    case 1:
                        res = _a.sent();
                        return [4 /*yield*/, res.json()];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, { res: res, data: data }];
                }
            });
        });
    };
    /**
     * Removes an identity with given id.
     *
     * @returns Response wrapped in [[IResult]]
     *
     * @param {IExternalIdentity} identity Identity object with mandantory fragments.
     *
     * @example
     * ```typescript
     *
     *   const identity: IExternalIdentity = {
     *      type: 'type',
     *      externalId: '1'
     *    };
     *
     *    (async () => {
     *      const {data, res} = await identityService.delete(identity);
     *   })();
     * ```
     */
    IdentityService.prototype.delete = function (identity) {
        return __awaiter(this, void 0, void 0, function () {
            var headers, method, url, res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        headers = { accept: 'application/json' };
                        method = 'DELETE';
                        url = this.getExternalIdUrl(identity);
                        return [4 /*yield*/, this.fetch(url, { headers: headers, method: method })];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, { res: res, data: null }];
                }
            });
        });
    };
    IdentityService.prototype.fetch = function (url, init) {
        return __awaiter(this, void 0, void 0, function () {
            var res, data, ex_1, ex_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.fetch(url, init)];
                    case 1:
                        res = _a.sent();
                        if (!(res.status >= 400)) return [3 /*break*/, 10];
                        data = null;
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, , 9]);
                        return [4 /*yield*/, res.json()];
                    case 3:
                        data = _a.sent();
                        return [3 /*break*/, 9];
                    case 4:
                        ex_1 = _a.sent();
                        _a.label = 5;
                    case 5:
                        _a.trys.push([5, 7, , 8]);
                        return [4 /*yield*/, res.text()];
                    case 6:
                        data = _a.sent();
                        return [3 /*break*/, 8];
                    case 7:
                        ex_2 = _a.sent();
                        return [3 /*break*/, 8];
                    case 8: return [3 /*break*/, 9];
                    case 9: throw { res: res, data: data };
                    case 10: return [2 /*return*/, res];
                }
            });
        });
    };
    IdentityService.prototype.getExternalIdsOfGlobalIdUrl = function (managedObjectId) {
        return "/" + this.baseUrl + "/globalIds/" + managedObjectId + "/externalIds";
    };
    IdentityService.prototype.getExternalIdUrl = function (identity) {
        return "/" + this.baseUrl + "/externalIds/" + identity.type + "/" + identity.externalId;
    };
    return IdentityService;
}());
exports.IdentityService = IdentityService;
