// @c8y/ngx-components/esm2015/repository/repository-navigation-factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
let RepositoryNavigationFactory = class RepositoryNavigationFactory {
    constructor() {
        this.navs = [];
    }
    get() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.navs.length === 0) {
                this.navs.push(new NavigatorNode({
                    label: gettext('Firmware repository'),
                    path: 'firmware',
                    icon: 'c8y-firmware',
                    parent: gettext('Management'),
                    priority: 1000
                }));
                this.navs.push(new NavigatorNode({
                    label: gettext('Software repository'),
                    path: 'software',
                    icon: 'c8y-tools',
                    parent: gettext('Management'),
                    priority: 900
                }));
                this.navs.push(new NavigatorNode({
                    label: gettext('Configuration repository'),
                    path: 'configuration',
                    icon: 'gears',
                    parent: gettext('Management'),
                    priority: 800
                }));
            }
            return this.navs;
        });
    }
};
RepositoryNavigationFactory = tslib_1.__decorate([
    Injectable()
], RepositoryNavigationFactory);
export { RepositoryNavigationFactory };
