// @c8y/ngx-components/esm2015/app-logs/app-logs.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FetchClient, IFetchOptions } from '@c8y/client';
import { from } from 'rxjs';
import { flatMap } from 'rxjs/operators';
let AppLogsService = class AppLogsService {
    constructor(client) {
        this.client = client;
        this.BASE_PATH_APPLICATION = 'application/applications';
        this.header = {
            Accept: 'application/vnd.com.nsn.cumulocity.applicationLogs+json;charset=UTF-8;ver=0.9'
        };
    }
    getLogs$(appId, instanceName, params = {}) {
        const url = `${this.BASE_PATH_APPLICATION}/${appId}/logs/${instanceName}`;
        const defaultConfig = {
            headers: this.header,
            method: 'GET'
        };
        const formattedDateFrom = params.dateFrom ? this.formatDate(params.dateFrom) : null;
        const formattedDateTo = params.dateTo ? this.formatDate(params.dateTo) : null;
        const modifiedParams = Object.assign({}, params, (formattedDateFrom && { dateFrom: formattedDateFrom }), (formattedDateTo && { dateTo: formattedDateTo }));
        const config = params ? Object.assign({}, defaultConfig, { params: modifiedParams }) : Object.assign({}, defaultConfig);
        return from(this.client.fetch(url, config)).pipe(flatMap(res => res.json()));
    }
    formatDate(dateToFormat) {
        return new Date(dateToFormat).toISOString();
    }
};
AppLogsService.ctorParameters = () => [
    { type: FetchClient }
];
AppLogsService = tslib_1.__decorate([
    Injectable()
], AppLogsService);
export { AppLogsService };
