// @c8y/ngx-components/esm2015/upgrade/c8y-ngx-components-upgrade.js
/**
 * Generated bundle index. Do not edit.
 */
export * from './index';
export { Ng1ActionBarFactoryService as ɵf } from './action-bar.factory';
export { Ng1ActionFactoryService as ɵg } from './action.factory';
export { Ng1BreadcrumbFactoryService as ɵh } from './breadcrumb.factory';
export { bridgeServiceFactory as ɵa, bridgeServiceProvider as ɵb } from './bridge.service';
export { Ng1ComponentFactoryService as ɵo } from './dashboard/component.factory';
export { CopyActionComponent as ɵm } from './dashboard/copy-action.component';
export { Ng1DashboardActionFactoryService as ɵp } from './dashboard/dashboard-actions.factory';
export { dashboardBridgeFactory as ɵc } from './dashboard/dashboard-upgrade.module';
export { DeviceSelectorComponent as ɵk } from './dashboard/device-selector.component';
export { PasteActionComponent as ɵn } from './dashboard/paste-action.component';
export { WidgetComponent as ɵl } from './dashboard/widget.component';
export { Ng1DocsFactoryService as ɵj } from './docs.factory';
export { Ng1NodesFactoryService as ɵd } from './nodes.factory';
export { Ng1SearchFactoryService as ɵi } from './search.factory';
export { Ng1TabsFactoryService as ɵe } from './tabs.factory';
