// @c8y/ngx-components/esm2015/core/modal/popover-confirm.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { PopoverDirective } from 'ngx-bootstrap/popover';
import { gettext } from '../i18n/gettext';
/**
 * A component to show a popover right next to the
 * user action. Add this component where the confirm modal
 * should be shown and toggle isOpen to display it. You
 * can use `await isShow()` to wait for a result on the
 * click action.
 */
let PopoverConfirmComponent = class PopoverConfirmComponent {
    /**
     * A component to show a popover right next to the
     * user action. Add this component where the confirm modal
     * should be shown and toggle isOpen to display it. You
     * can use `await isShow()` to wait for a result on the
     * click action.
     */
    constructor() {
        this.buttons = [{ label: gettext('cancel'), action: () => Promise.reject() }];
        this.isOpen = false;
        this.containerClass = '';
        this.placement = 'right';
        this.outsideClick = false;
    }
    /**
     * Show the popover. Wait until an action is performed on it.
     */
    show(buttons) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.buttons = buttons || this.buttons;
            this.isOpen = true;
            let result;
            try {
                result = yield new Promise((resolve, reject) => {
                    return this.click = (action) => action().then(resolve).catch(reject);
                });
            }
            catch (ex) {
                throw (ex);
            }
            finally {
                this.isOpen = false;
            }
            return result;
        });
    }
    buttonStatus(status) {
        if (status) {
            return `btn-${status}`;
        }
        return 'btn-link btn-clean';
    }
    handleOutsideClick() {
        this.click(() => Promise.reject());
    }
};
tslib_1.__decorate([
    ViewChild(PopoverDirective, { static: false })
], PopoverConfirmComponent.prototype, "popover", void 0);
tslib_1.__decorate([
    Input()
], PopoverConfirmComponent.prototype, "buttons", void 0);
tslib_1.__decorate([
    Input()
], PopoverConfirmComponent.prototype, "message", void 0);
tslib_1.__decorate([
    Input()
], PopoverConfirmComponent.prototype, "title", void 0);
tslib_1.__decorate([
    Input()
], PopoverConfirmComponent.prototype, "isOpen", void 0);
tslib_1.__decorate([
    Input()
], PopoverConfirmComponent.prototype, "containerClass", void 0);
tslib_1.__decorate([
    Input()
], PopoverConfirmComponent.prototype, "placement", void 0);
tslib_1.__decorate([
    Input()
], PopoverConfirmComponent.prototype, "outsideClick", void 0);
PopoverConfirmComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-popover-confirm',
        template: "<span\n  [popover]=\"popover\"\n  container=\"body\"\n  [isOpen]=\"isOpen\"\n  [popoverTitle]=\"title | translate\"\n  [placement]=\"placement\"\n  [containerClass]=\"containerClass\"\n  [outsideClick]=\"outsideClick\"\n  (onHidden)=\"handleOutsideClick()\"\n></span>\n<ng-template #popover>\n  <p class=\"bottom-m\">\n    {{ message | translate }}\n  </p>\n  <div class=\"popover-footer\">\n    <ng-container *ngFor=\"let button of buttons\">\n      <button\n        title=\"{{ button.label | translate }}\"\n        class=\"btn btn-sm right-m-md\"\n        [ngClass]=\"buttonStatus(button.status)\"\n        (click)=\"click(button.action)\"\n      >\n        {{ button.label | translate }}\n      </button>\n    </ng-container>\n  </div>\n</ng-template>\n"
    })
], PopoverConfirmComponent);
export { PopoverConfirmComponent };
