// @c8y/ngx-components/esm5/operations/details/operation-details.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { gettext } from '@c8y/ngx-components';
import { get } from 'lodash-es';
import { BulkOperationType } from '../bulk/bulk-operation.model';
var OperationDetailsComponent = /** @class */ (function () {
    function OperationDetailsComponent(formBuilder) {
        var _a;
        this.formBuilder = formBuilder;
        this.titlePlaceholder = gettext('e.g. Reset factory settings');
        this.descriptionPlaceholder = gettext('e.g. Reset factory settings of all devices of model XYZ');
        this.placeholders = (_a = {},
            _a[BulkOperationType.SOFTWARE] = {
                title: gettext('e.g. Cloud connectivity software'),
                description: gettext('e.g. Cloud connectivity software applied to devices with type c8y_Linux`KEEP_ORIGINAL`')
            },
            _a[BulkOperationType.FIRMWARE] = {
                title: gettext('e.g. Firmware for hardware revision b'),
                description: gettext('e.g. Firmware for hardware revision b applied to devices with type c8y_Linux`KEEP_ORIGINAL`')
            },
            _a[BulkOperationType.CONFIGURATION] = {
                title: gettext('e.g. Host configuration c8y_Linux`KEEP_ORIGINAL`'),
                description: gettext('e.g. Host configuration applied to devices with type c8y_Linux`KEEP_ORIGINAL`')
            },
            _a[BulkOperationType.DEVICE_PROFILE] = {
                title: gettext('e.g. Device profile for c8y_Linux`KEEP_ORIGINAL` devices'),
                description: gettext('e.g. Device profile applied to devices with type c8y_Linux`KEEP_ORIGINAL`')
            },
            _a);
    }
    Object.defineProperty(OperationDetailsComponent.prototype, "bulkOperationType", {
        set: function (bulkOperationType) {
            this.titlePlaceholder = get(this.placeholders[bulkOperationType], 'title');
            this.descriptionPlaceholder = get(this.placeholders[bulkOperationType], 'description');
        },
        enumerable: true,
        configurable: true
    });
    OperationDetailsComponent.prototype.ngOnInit = function () {
        var config = {
            description: ['', Validators.required],
            note: [''],
            schedule: []
        };
        this.fgOperationDescription = this.formBuilder.group(config);
    };
    OperationDetailsComponent.ctorParameters = function () { return [
        { type: FormBuilder }
    ]; };
    tslib_1.__decorate([
        Input()
    ], OperationDetailsComponent.prototype, "bulkOperationType", null);
    OperationDetailsComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-operation-details',
            template: "<div [formGroup]=\"fgOperationDescription\">\n  <c8y-form-group>\n    <label translate>Title</label>\n    <input\n      formControlName=\"description\"\n      class=\"form-control\"\n      [placeholder]=\"titlePlaceholder | translate\"\n      required\n    />\n    <c8y-messages *ngIf=\"fgOperationDescription?.controls?.description?.errors\">\n      <c8y-message *ngIf=\"fgOperationDescription.controls.description.errors.required\" translate>\n        This field is required.\n      </c8y-message>\n    </c8y-messages>\n  </c8y-form-group>\n  <c8y-form-group>\n    <label translate>Description</label>\n    <textarea\n      formControlName=\"note\"\n      class=\"form-control\"\n      [placeholder]=\"descriptionPlaceholder | translate\"\n      rows=\"3\"\n    ></textarea>\n  </c8y-form-group>\n  <c8y-operation-scheduler formControlName=\"schedule\"></c8y-operation-scheduler>\n</div>\n"
        })
    ], OperationDetailsComponent);
    return OperationDetailsComponent;
}());
export { OperationDetailsComponent };
