// @c8y/ngx-components/esm5/core/modal/modal.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { ConfirmModalComponent } from './modal-confirm.component';
import { Status } from '../common/status.model';
import { gettext } from '../i18n/gettext';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
/**
 * Service to shiow a modal.
 */
var ModalService = /** @class */ (function () {
    function ModalService(modalService) {
        this.modalService = modalService;
    }
    /**
     * Shows a quick confirm message modal.
     * @param title The title of that modal.
     * @param body The text body to display.
     * @param status The status.
     * @param labels The labels to use. Default: { ok: 'Confirm', cancel: 'Cancel'}
     */
    ModalService.prototype.confirm = function (title, body, status, labels) {
        if (status === void 0) { status = Status.INFO; }
        if (labels === void 0) { labels = {}; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var modalLabels, modalRef;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        modalLabels = { ok: labels.ok || gettext('Confirm'), cancel: labels.cancel || gettext('Cancel') };
                        modalRef = this.modalService.show(ConfirmModalComponent, { initialState: { title: title, body: body, labels: modalLabels, status: status } });
                        return [4 /*yield*/, modalRef.content.result];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Shows a quick acknowledge message modal.
     * @param title The title of that modal.
     * @param body The text body to display.
     * @param status The status.
     * @param acknowledgeLabel The label to use.
     */
    ModalService.prototype.acknowledge = function (title, body, status, acknowledgeLabel) {
        if (status === void 0) { status = Status.INFO; }
        if (acknowledgeLabel === void 0) { acknowledgeLabel = gettext('Confirm'); }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var labels, modalRef;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        labels = { ok: acknowledgeLabel, cancel: null };
                        modalRef = this.modalService.show(ConfirmModalComponent, { initialState: { title: title, body: body, labels: labels, status: status } });
                        return [4 /*yield*/, modalRef.content.result];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    ModalService.ctorParameters = function () { return [
        { type: BsModalService }
    ]; };
    ModalService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function ModalService_Factory() { return new ModalService(i0.ɵɵinject(i1.BsModalService)); }, token: ModalService, providedIn: "root" });
    ModalService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        })
    ], ModalService);
    return ModalService;
}());
export { ModalService };
