// @c8y/ngx-components/esm5/core/modal/modal.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ModalModule as BsModalModule } from 'ngx-bootstrap/modal';
import { ConfirmModalComponent } from './modal-confirm.component';
import { ModalComponent } from './modal.component';
import { ModalService } from './modal.service';
import { CommonModule } from '../common/common.module';
import { PopoverConfirmComponent } from './popover-confirm.component';
import { PopoverModule } from 'ngx-bootstrap/popover';
/**
 * The angular module definition for modal.
 * @exports ConfirmModalComponent
 * @exports ModalComponent
 * @exports PopoverConfirmComponent
 */
var ModalModule = /** @class */ (function () {
    function ModalModule() {
    }
    ModalModule = tslib_1.__decorate([
        NgModule({
            imports: [
                BsModalModule.forRoot(),
                CommonModule,
                PopoverModule.forRoot()
            ],
            exports: [
                ConfirmModalComponent,
                ModalComponent,
                PopoverConfirmComponent
            ],
            declarations: [
                ConfirmModalComponent,
                ModalComponent,
                PopoverConfirmComponent
            ],
            providers: [
                ModalService
            ],
            entryComponents: [
                ConfirmModalComponent
            ]
        })
    ], ModalModule);
    return ModalModule;
}());
export { ModalModule };
