// @c8y/ngx-components/esm2015/core/router/router.module.js
var RouterModule_1;
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule as NgRouterModule } from '@angular/router';
import { RouterService } from './router.service';
import { ContextRouteComponent } from './context-route.component';
import { HeaderModule } from '../header/header.module';
import { CommonModule } from '../common/common.module';
import { ContextRouteGuard } from './context-route.guard';
import { ViewContext } from './router.models';
import { RouterTabsResolver } from './router-tabs.resolver';
import { EmptyComponent } from './empty.component';
export const routes = [{
        path: 'users/new',
        component: EmptyComponent
    },
    {
        path: 'tenants/new',
        component: EmptyComponent
    }];
// tslint:disable-next-line: forin
for (const key in ViewContext) {
    const context = ViewContext[key];
    routes.push({
        path: context,
        component: ContextRouteComponent,
        canActivate: [ContextRouteGuard],
        data: { context, contextData: {} },
        resolve: {
            tabs: RouterTabsResolver,
        }
    });
}
let RouterModule = RouterModule_1 = class RouterModule {
    constructor(routerService) {
        // Declared but never used is intended: to always have a instance of RouterService.
    }
    static providers() {
        return [
            ContextRouteGuard,
            RouterTabsResolver,
            RouterService
        ];
    }
    static forRoot() {
        return {
            ngModule: RouterModule_1,
            providers: RouterModule_1.providers()
        };
    }
};
RouterModule.ctorParameters = () => [
    { type: RouterService }
];
RouterModule = RouterModule_1 = tslib_1.__decorate([
    NgModule({
        declarations: [ContextRouteComponent, EmptyComponent],
        imports: [
            NgRouterModule.forRoot(routes),
            HeaderModule,
            CommonModule
        ],
        entryComponents: [ContextRouteComponent, EmptyComponent]
    })
], RouterModule);
export { RouterModule };
