// @c8y/ngx-components/esm5/api/api.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FetchClient, IFetchOptions } from '@c8y/client';
import { Subject } from 'rxjs';
import { filter } from 'rxjs/operators';
var ApiService = /** @class */ (function () {
    function ApiService(client) {
        this.client = client;
        this.callsSubject = new Subject();
        this.calls = this.callsSubject.asObservable();
        this.hookIntoClientFetch();
    }
    ApiService.prototype.hookResponse = function (hookFilter) {
        return this.callsSubject.pipe(filter(function (_a) {
            var phase = _a.phase;
            return phase === 'finish';
        }), filter(hookFilter));
    };
    ApiService.prototype.hookRequest = function (hookFilter) {
        return this.callsSubject.pipe(filter(function (_a) {
            var phase = _a.phase;
            return phase === 'start';
        }), filter(hookFilter));
    };
    ApiService.prototype.onFinish = function (call) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.callsSubject.next(tslib_1.__assign({ phase: 'finish' }, call));
                return [2 /*return*/];
            });
        });
    };
    ApiService.prototype.onStart = function (call) {
        this.callsSubject.next(tslib_1.__assign({ phase: 'start' }, call));
    };
    ApiService.prototype.resolveData = function (call) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var response, method, cb;
            return tslib_1.__generator(this, function (_a) {
                response = call.response, method = call.method;
                if ('data' in response) {
                    return [2 /*return*/, Promise.resolve({ data: response.data, method: method })];
                }
                else {
                    cb = function (data) { return ({ data: data, method: method }); };
                    return [2 /*return*/, response.clone().json().then(cb, cb)];
                }
                return [2 /*return*/];
            });
        });
    };
    ApiService.prototype.hookIntoClientFetch = function () {
        var _this = this;
        var fetch = this.client.fetch.bind(this.client);
        this.client.fetch = function (url, options) {
            if (options === void 0) { options = { method: 'GET' }; }
            return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var method, fetchPromise;
                var _this = this;
                return tslib_1.__generator(this, function (_a) {
                    method = options.method;
                    this.onStart({ options: options, method: method, url: url });
                    fetchPromise = fetch(url, options);
                    if (typeof options.responseInterceptor === 'function') {
                        fetchPromise = fetchPromise.then(options.responseInterceptor);
                    }
                    fetchPromise.then(function (response) { return _this.onFinish({ response: response, url: url, options: options, method: method }); }, function (response) { return _this.onFinish({ response: response, url: url, options: options, method: method }); });
                    return [2 /*return*/, fetchPromise];
                });
            });
        };
    };
    ApiService.ctorParameters = function () { return [
        { type: FetchClient }
    ]; };
    ApiService = tslib_1.__decorate([
        Injectable()
    ], ApiService);
    return ApiService;
}());
export { ApiService };
