// @c8y/ngx-components/esm5/core/device-grid/columns/group.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { GroupCellRendererComponent } from './group.cell-renderer.component';
import { gettext } from '../../i18n/gettext';
var GroupDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(GroupDeviceGridColumn, _super);
    function GroupDeviceGridColumn() {
        var _this = _super.call(this) || this;
        _this.name = 'group';
        _this.header = gettext('Group');
        _this.cellRendererComponent = GroupCellRendererComponent;
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    deviceGroupId: {
                        type: 'integer',
                        format: 'deviceGroupId'
                    }
                }
            },
            form: [
                {
                    key: 'deviceGroupId',
                    title: gettext('Show devices in group')
                }
            ],
            getFilter: function (model) {
                var filter = {};
                if (model.deviceGroupId) {
                    filter.__bygroupid = model.deviceGroupId;
                }
                return filter;
            }
        };
        _this.sortable = false;
        return _this;
    }
    return GroupDeviceGridColumn;
}(BaseDeviceGridColumn));
export { GroupDeviceGridColumn };
