// @c8y/ngx-components/esm5/core/device-grid/columns/name.device-grid-column.js
import * as tslib_1 from "tslib";
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { NameCellRendererComponent } from './name.cell-renderer.component';
import { gettext } from '../../i18n/gettext';
var NameDeviceGridColumn = /** @class */ (function (_super) {
    tslib_1.__extends(NameDeviceGridColumn, _super);
    function NameDeviceGridColumn() {
        var _this = _super.call(this) || this;
        _this.name = 'name';
        _this.path = 'name';
        _this.header = gettext('Name');
        _this.cellCSSClassName = "data-record-header" /* Header */;
        _this.cellRendererComponent = NameCellRendererComponent;
        _this.filterable = true;
        _this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    names: {
                        type: 'array',
                        items: {
                            type: 'string'
                        }
                    }
                }
            },
            form: [
                {
                    key: 'names',
                    title: gettext('Show devices with name'),
                    add: gettext('Or'),
                    items: [
                        {
                            key: 'names[]',
                            notitle: true,
                            placeholder: gettext('e.g. My Device, you can use an asterisk (*) as wildcard character')
                        }
                    ]
                }
            ],
            getFilter: function (model) {
                var filter = {};
                if (model.names.length) {
                    filter.name = { __in: model.names };
                }
                return filter;
            }
        };
        _this.sortable = true;
        _this.sortingConfig = {
            pathSortingConfigs: [{ path: _this.path }]
        };
        return _this;
    }
    return NameDeviceGridColumn;
}(BaseDeviceGridColumn));
export { NameDeviceGridColumn };
