// @c8y/ngx-components/esm5/device-configuration/configuration-list.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var ConfigurationListComponent = /** @class */ (function () {
    function ConfigurationListComponent() {
        this.configSelected = new EventEmitter();
        this.filterTerm = '';
    }
    ConfigurationListComponent.prototype.showConfigurationTypePreview = function (config) {
        this.selectedConfig = config;
        this.configSelected.emit(config);
    };
    ConfigurationListComponent.prototype.updatePipe = function (filterTerm) {
        this.filterTerm = filterTerm;
    };
    tslib_1.__decorate([
        Input()
    ], ConfigurationListComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        Input()
    ], ConfigurationListComponent.prototype, "itemIcon", void 0);
    tslib_1.__decorate([
        Input()
    ], ConfigurationListComponent.prototype, "emptyState", void 0);
    tslib_1.__decorate([
        Input()
    ], ConfigurationListComponent.prototype, "isFilterEnabled", void 0);
    tslib_1.__decorate([
        Output()
    ], ConfigurationListComponent.prototype, "configSelected", void 0);
    ConfigurationListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-configuration-list',
            template: "<div class=\"left-p-md bottom-p-xs\" *ngIf=\"isFilterEnabled\">\n  <c8y-filter [icon]=\"'search'\" (onSearch)=\"updatePipe($event)\"></c8y-filter>\n</div>\n\n<!-- EMPTY STATE -->\n<div class=\"c8y-empty-state text-left\" *ngIf=\"items?.length === 0\">\n  <h1 [c8yIcon]=\"emptyState.icon\"></h1>\n  <p>\n    <strong>{{ emptyState.title | translate }}</strong\n    ><br />\n    <small>{{ emptyState.text | translate }}</small>\n  </p>\n</div>\n\n<!-- CONFIGURATIONS AVAILABLE -->\n<div class=\"c8y-nav-stacked\">\n  <div\n    class=\"c8y-stacked-item flex-row\"\n    [class.active]=\"config === selectedConfig\"\n    *ngFor=\"let config of items | configurationFilterPipe: filterTerm\"\n    (click)=\"showConfigurationTypePreview(config)\"\n  >\n    <div class=\"list-item-icon\">\n      <i [c8yIcon]=\"itemIcon\"></i>\n    </div>\n    <div class=\"list-item-body text-truncate\">\n      <span>{{ config.name }}</span>\n      <span class=\"text-uppercase text-gray-light pull-right\">{{ config.deviceType }}</span>\n    </div>\n  </div>\n</div>\n\n<!-- for Carlos: config.configurationType to differentiate whether a config matches configuration type. -->\n"
        })
    ], ConfigurationListComponent);
    return ConfigurationListComponent;
}());
export { ConfigurationListComponent };
