// @c8y/ngx-components/esm5/device-configuration/configuration-operation.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { operationStatusIcons, operationStatusClasses } from '@c8y/ngx-components';
var ConfigurationOperationComponent = /** @class */ (function () {
    function ConfigurationOperationComponent() {
    }
    ConfigurationOperationComponent.prototype.ngOnInit = function () {
        this.statusIcons = operationStatusIcons;
        this.statusClasses = operationStatusClasses;
    };
    tslib_1.__decorate([
        Input()
    ], ConfigurationOperationComponent.prototype, "operation", void 0);
    ConfigurationOperationComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-configuration-operation',
            template: "<div class=\"card\">\n  <div class=\"list-group list-group-links\">\n    <a class=\"list-group-item flex-row\" href=\"#/device/{{ operation.deviceId }}/control\">\n      <div class=\"list-item-icon\">\n        <i\n          [c8yIcon]=\"statusIcons[operation.status]\"\n          [ngClass]=\"statusClasses[operation.status]\"\n          [tooltip]=\"operation.status | translate\"\n          tooltip-placement=\"right\"\n        ></i>\n      </div>\n      <div class=\"list-item-body\">\n        <div class=\"flex-row row\">\n          <span class=\"col-sm-8\">{{ operation.description | translate }}</span>\n          <div class=\"col-sm-4 text-muted\">\n            <i c8yIcon=\"calendar\"></i>&nbsp;\n            <small>{{ operation.creationTime | date: 'medium' }}</small>\n          </div>\n        </div>\n      </div>\n    </a>\n  </div>\n</div>\n"
        })
    ], ConfigurationOperationComponent);
    return ConfigurationOperationComponent;
}());
export { ConfigurationOperationComponent };
