// @c8y/ngx-components/esm5/device-configuration/device-configuration.service.js
import * as tslib_1 from "tslib";
import { EventEmitter, Injectable } from '@angular/core';
import { EventBinaryService, EventService, IEvent, IManagedObject, InventoryBinaryService, InventoryService, IOperation, OperationService } from '@c8y/client';
import { AlertService, gettext } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
var DeviceConfigurationService = /** @class */ (function () {
    function DeviceConfigurationService(eventService, eventBinaryService, operationService, alertService, inventoryService, inventoryBinaryService, translateService) {
        this.eventService = eventService;
        this.eventBinaryService = eventBinaryService;
        this.operationService = operationService;
        this.alertService = alertService;
        this.inventoryService = inventoryService;
        this.inventoryBinaryService = inventoryBinaryService;
        this.translateService = translateService;
        this.dateFrom = new Date(0);
        this.dateTo = new Date(Date.now() + 86400000); // 1 day in the future
        this.repositoryConfigListUpdated = new EventEmitter();
    }
    DeviceConfigurationService.prototype.getLatestConfigurationEvent = function (deviceId, type) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var filter, data;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        filter = {
                            source: deviceId,
                            type: type,
                            dateFrom: this.dateFrom.toISOString(),
                            dateTo: this.dateTo.toISOString(),
                            pageSize: 1
                        };
                        return [4 /*yield*/, this.eventService.list(filter)];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data[0]];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.getEventConfigurationBinary = function (event) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var binary, res, ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.eventBinaryService.download(event)];
                    case 1:
                        res = _a.sent();
                        binary = res.text();
                        return [3 /*break*/, 3];
                    case 2:
                        ex_1 = _a.sent();
                        this.alertService.addServerFailure(ex_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/, binary];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.createUploadConfigFileOperation = function (device, configurationType) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var operation, operationText, operationCfg, data, ex_2;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        operationText = gettext('Retrieve {{ configurationType }} configuration snapshot from device {{ deviceName }}');
                        operationCfg = {
                            deviceId: device.id,
                            c8y_UploadConfigFile: {
                                type: configurationType
                            },
                            description: this.translateService.instant(operationText, {
                                configurationType: configurationType,
                                deviceName: device.name
                            })
                        };
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.operationService.create(operationCfg)];
                    case 2:
                        data = (_a.sent()).data;
                        operation = data;
                        return [3 /*break*/, 4];
                    case 3:
                        ex_2 = _a.sent();
                        this.alertService.addServerFailure(ex_2);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, operation];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.getLatestConfigFileOperation = function (deviceId, configType, operationType) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var filter, data;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        filter = {
                            deviceId: deviceId,
                            fragmentType: operationType,
                            dateFrom: this.dateFrom.toISOString(),
                            dateTo: this.dateTo.toISOString(),
                            revert: true,
                            pageSize: 2000
                        };
                        return [4 /*yield*/, this.operationService.list(filter)];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data.find(function (op) { return op[operationType].type === configType; })];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.getSnapshotsFromRepository = function (deviceType, configurationType) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query, filter, deviceTypeFilter;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        filter = {
                            pageSize: 2000,
                            withTotalPages: true
                        };
                        query = {
                            __filter: {
                                __and: [
                                    {
                                        type: 'c8y_ConfigurationDump'
                                    },
                                    {
                                        __or: [
                                            { configurationType: { __eq: configurationType } },
                                            {
                                                __not: {
                                                    __has: 'configurationType'
                                                }
                                            }
                                        ]
                                    }
                                ]
                            },
                            __orderby: [{ configurationType: -1, name: 1 }]
                        };
                        if (deviceType) {
                            deviceTypeFilter = {
                                __or: [
                                    { deviceType: { __eq: deviceType } },
                                    {
                                        __not: {
                                            __has: 'deviceType'
                                        }
                                    }
                                ]
                            };
                            query.__filter.__and.push(deviceTypeFilter);
                        }
                        return [4 /*yield*/, this.inventoryService.listQuery(query, filter)];
                    case 1: return [2 /*return*/, (_a.sent()).data];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.createDownloadConfigFileOperation = function (device, configurationType, binaryUrl) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var operation, operationText, operationCfg, data, ex_3;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        operationText = gettext('Send configuration {{ configurationType }} to device {{ deviceName }}');
                        operationCfg = {
                            deviceId: device.id,
                            c8y_DownloadConfigFile: {
                                url: binaryUrl,
                                type: configurationType
                            },
                            description: this.translateService.instant(operationText, {
                                configurationType: configurationType,
                                deviceName: device.name
                            })
                        };
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.operationService.create(operationCfg)];
                    case 2:
                        data = (_a.sent()).data;
                        operation = data;
                        return [3 /*break*/, 4];
                    case 3:
                        ex_3 = _a.sent();
                        this.alertService.addServerFailure(ex_3);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, operation];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.getConfigurationBinaryFile = function (url) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var binary, moId, binaryMO, res;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        moId = this.inventoryBinaryService.getIdFromUrl(url);
                        if (!moId) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.inventoryService.detail(moId)];
                    case 1:
                        binaryMO = (_a.sent()).data;
                        return [4 /*yield*/, this.inventoryBinaryService.download(binaryMO)];
                    case 2:
                        res = _a.sent();
                        return [4 /*yield*/, res.text()];
                    case 3:
                        binary = _a.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        binary = this.fetchExternalBinary(url);
                        _a.label = 5;
                    case 5: return [2 /*return*/, binary];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.saveToConfigurationRepository = function (config) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var mo, response;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        mo = {};
                        Object.assign(mo, {
                            name: config.name,
                            type: 'c8y_ConfigurationDump',
                            deviceType: config.deviceType,
                            configurationType: config.configurationType,
                            c8y_Global: {}
                        });
                        return [4 /*yield*/, this.inventoryBinaryService.create(new Blob([config.binary]), {
                                c8y_Global: {}
                            })];
                    case 1:
                        response = _a.sent();
                        mo.url = response.data.self;
                        return [2 /*return*/, this.inventoryService.create(mo)];
                }
            });
        });
    };
    DeviceConfigurationService.prototype.updateRepositoryConfigList = function () {
        this.repositoryConfigListUpdated.emit();
    };
    DeviceConfigurationService.prototype.fetchExternalBinary = function (url) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var configBinary, res, ex_4;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        return [4 /*yield*/, fetch(url)];
                    case 1:
                        res = _a.sent();
                        if (!(res.status === 200)) return [3 /*break*/, 3];
                        return [4 /*yield*/, res.text()];
                    case 2:
                        configBinary = _a.sent();
                        _a.label = 3;
                    case 3: return [3 /*break*/, 5];
                    case 4:
                        ex_4 = _a.sent();
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/, configBinary];
                }
            });
        });
    };
    DeviceConfigurationService.ctorParameters = function () { return [
        { type: EventService },
        { type: EventBinaryService },
        { type: OperationService },
        { type: AlertService },
        { type: InventoryService },
        { type: InventoryBinaryService },
        { type: TranslateService }
    ]; };
    DeviceConfigurationService = tslib_1.__decorate([
        Injectable()
    ], DeviceConfigurationService);
    return DeviceConfigurationService;
}());
export { DeviceConfigurationService };
