// @c8y/ngx-components/esm5/device-configuration/device-configuration.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CoreModule, gettext, HOOK_ONCE_ROUTE, ViewContext } from '@c8y/ngx-components';
import { DeviceConfigurationComponent } from './device-configuration.component';
import { ConfigurationListComponent } from './configuration-list.component';
import { ConfigurationOperationComponent } from './configuration-operation.component';
import { DeviceConfigurationService } from './device-configuration.service';
import { SourceCodePreviewComponent } from './source-code-preview.component';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { DeviceConfigurationGuard } from './device-configuration.guard';
import { ConfigurationPreviewComponent } from './configuration-preview.component';
import { ConfigurationFilterPipe } from './configuration-filter.pipe';
import { SaveToRepositoryComponent } from './save-to-repository.component';
var routes = [
    {
        context: ViewContext.Device,
        path: 'device-configuration',
        component: DeviceConfigurationComponent,
        label: gettext('Configuration'),
        icon: 'gears',
        canActivate: [DeviceConfigurationGuard]
    }
];
var ɵ0 = routes;
var DeviceConfigurationModule = /** @class */ (function () {
    function DeviceConfigurationModule() {
    }
    DeviceConfigurationModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                DeviceConfigurationComponent,
                ConfigurationListComponent,
                ConfigurationOperationComponent,
                SourceCodePreviewComponent,
                ConfigurationPreviewComponent,
                ConfigurationFilterPipe,
                SaveToRepositoryComponent
            ],
            exports: [],
            imports: [CoreModule, TooltipModule],
            entryComponents: [DeviceConfigurationComponent, SaveToRepositoryComponent],
            providers: [
                DeviceConfigurationService,
                DeviceConfigurationGuard,
                {
                    provide: HOOK_ONCE_ROUTE,
                    useValue: ɵ0,
                    multi: true
                }
            ]
        })
    ], DeviceConfigurationModule);
    return DeviceConfigurationModule;
}());
export { DeviceConfigurationModule };
export { ɵ0 };
