// @c8y/ngx-components/esm5/device-configuration/save-to-repository.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { AlertService, gettext } from '@c8y/ngx-components';
import { DeviceConfigurationService } from './device-configuration.service';
var SaveToRepositoryComponent = /** @class */ (function () {
    function SaveToRepositoryComponent(modal, alertService, deviceConfigurationService) {
        var _this = this;
        this.modal = modal;
        this.alertService = alertService;
        this.deviceConfigurationService = deviceConfigurationService;
        this.result = new Promise(function (resolve, reject) {
            _this._save = resolve;
            _this._cancel = reject;
        });
    }
    SaveToRepositoryComponent.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.deviceConfigurationService.saveToConfigurationRepository(this.configSnapshot)];
                    case 1:
                        _a.sent();
                        this.alertService.success(gettext('Configuration saved.'));
                        this._save();
                        return [3 /*break*/, 3];
                    case 2:
                        ex_1 = _a.sent();
                        if (ex_1.data) {
                            this.alertService.danger(ex_1.data.message);
                        }
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    SaveToRepositoryComponent.prototype.close = function () {
        this._cancel();
        this.modal.hide();
    };
    SaveToRepositoryComponent.ctorParameters = function () { return [
        { type: BsModalRef },
        { type: AlertService },
        { type: DeviceConfigurationService }
    ]; };
    SaveToRepositoryComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-save-config-to-configuration-repository',
            template: "<div class=\"modal-header bg-primary text-center text-white\">\n  <div style=\"font-size: 62px;\">\n    <span c8yIcon=\"gears\"></span>\n  </div>\n  <h4 class=\"text-uppercase\" translate>\n    Save configuration\n  </h4>\n</div>\n<div class=\"modal-body\">\n  <form #saveConfigurationSnapshot=\"ngForm\" class=\"p-t-24\">\n    <c8y-form-group>\n      <label translate for=\"name\">Name</label>\n      <input\n        id=\"name\"\n        type=\"text\"\n        class=\"form-control\"\n        autocomplete=\"off\"\n        name=\"name\"\n        [(ngModel)]=\"configSnapshot.name\"\n        required\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"deviceType\">Device type</label>\n      <input\n        id=\"deviceType\"\n        class=\"form-control\"\n        rows=\"6\"\n        name=\"deviceType\"\n        [(ngModel)]=\"configSnapshot.deviceType\"\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"configurationType\">Configuration type</label>\n      <input\n        id=\"configurationType\"\n        class=\"form-control\"\n        rows=\"6\"\n        name=\"configurationType\"\n        [(ngModel)]=\"configSnapshot.configurationType\"\n      />\n    </c8y-form-group>\n  </form>\n</div>\n<div class=\"modal-footer\">\n  <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\" translate>\n    Cancel\n  </button>\n\n  <button\n    title=\"{{ 'Add certificate' | translate }}\"\n    class=\"btn btn-primary\"\n    (click)=\"save()\"\n    [disabled]=\"saveConfigurationSnapshot.form.invalid\"\n    translate\n  >\n    Save\n  </button>\n</div>\n"
        })
    ], SaveToRepositoryComponent);
    return SaveToRepositoryComponent;
}());
export { SaveToRepositoryComponent };
