// @c8y/ngx-components/esm2015/operations/bulk/list/type-list/bulk-operation-type-list.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { Router } from '@angular/router';
import { BulkOperationsService } from '../../bulk-operations.service';
let BulkOperationTypeListComponent = class BulkOperationTypeListComponent {
    constructor(router, bulkOpsService) {
        this.router = router;
        this.bulkOpsService = bulkOpsService;
        this.onCustomType = new EventEmitter();
        this.bulkList = [];
    }
    ngOnInit() {
        this.bulkList = this.bulkOpsService.getBulkTypes();
    }
    handleClick(item) {
        if (item.path) {
            this.bulkOpsService.hideNewBulkOperationModal();
            this.router.navigateByUrl(item.path);
        }
        else if (item.component) {
            this.onCustomType.emit(item);
        }
    }
};
BulkOperationTypeListComponent.ctorParameters = () => [
    { type: Router },
    { type: BulkOperationsService }
];
tslib_1.__decorate([
    Output()
], BulkOperationTypeListComponent.prototype, "onCustomType", void 0);
BulkOperationTypeListComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-bulk-operation-type-list',
        template: "<div class=\"list-group list-group-links\">\n  <button\n    *ngFor=\"let item of bulkList\"\n    type=\"button\"\n    class=\"list-group-item flex-row\"\n    (click)=\"handleClick(item)\"\n  >\n    <div *ngIf=\"item.c8yIcon; else icon\" class=\"list-item-icon\">\n      <i [c8yIcon]=\"item.c8yIcon\"></i>\n    </div>\n    <ng-template #icon>\n      <div class=\"list-item-icon\">\n        <i c8yIcon=\"cogs\"></i>\n      </div>\n    </ng-template>\n    <div title=\"{{ item.name | translate }}\" class=\"list-item-body text-truncate\">\n      {{ item.name | translate }}\n    </div>\n  </button>\n</div>\n"
    })
], BulkOperationTypeListComponent);
export { BulkOperationTypeListComponent };
