// @c8y/ngx-components/esm5/core/authentication/password-strength-checker.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { PasswordStrength } from '@c8y/client';
import { filter } from 'lodash-es';
import { gettext } from '../i18n/gettext';
var PasswordStrengthCheckerService = /** @class */ (function () {
    function PasswordStrengthCheckerService() {
        this.GREEN = {
            colorName: 'green',
            color: 'rgb(0, 128, 0)',
            description: gettext('strong'),
            passwordStrength: PasswordStrength.GREEN
        };
        this.YELLOW = {
            colorName: 'yellow',
            color: 'rgb(255, 204, 51)',
            description: gettext('medium'),
            passwordStrength: PasswordStrength.YELLOW
        };
        this.RED = {
            colorName: 'red',
            color: 'rgb(170, 0, 51)',
            description: gettext('weak'),
            passwordStrength: PasswordStrength.RED
        };
    }
    PasswordStrengthCheckerService.prototype.hasLowerCase = function (password) {
        return (password.search(/[a-z]/) !== -1);
    };
    PasswordStrengthCheckerService.prototype.hasUpperCase = function (password) {
        return (password.search(/[A-Z]/) !== -1);
    };
    PasswordStrengthCheckerService.prototype.hasNumbers = function (password) {
        return (password.search(/[0-9]/) !== -1);
    };
    PasswordStrengthCheckerService.prototype.hasSpecialChars = function (password) {
        return (password.search(/[^0-9a-zA-Z]+/) !== -1);
    };
    PasswordStrengthCheckerService.prototype.getStrengthColor = function (password) {
        var passwordStrength = filter([
            this.hasLowerCase(password),
            this.hasUpperCase(password),
            this.hasNumbers(password),
            this.hasSpecialChars(password)
        ]).length;
        if (passwordStrength > 2) {
            return this.GREEN;
        }
        else if (passwordStrength >= 2) {
            return this.YELLOW;
        }
        else {
            return this.RED;
        }
    };
    PasswordStrengthCheckerService = tslib_1.__decorate([
        Injectable()
    ], PasswordStrengthCheckerService);
    return PasswordStrengthCheckerService;
}());
export { PasswordStrengthCheckerService };
