// @c8y/ngx-components/esm5/core/authentication/totp-challenge.component.js
import * as tslib_1 from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { ControlContainer, NgForm } from '@angular/forms';
import { LoginService } from '../login/login.service';
import { UserService, ICredentials } from '@c8y/client';
import { AlertService } from '../alert/alert.service';
var TotpChallengeComponent = /** @class */ (function () {
    function TotpChallengeComponent(loginService, users, alert) {
        this.loginService = loginService;
        this.users = users;
        this.alert = alert;
        /**
         * Calls the verify endpoint if set to true (default true)
         */
        this.verify = true;
        /**
         * Emits the token on success.
         */
        this.onSuccess = new EventEmitter();
        this.model = {
            token: ''
        };
        this.loading = false;
        this.hasError = false;
    }
    TotpChallengeComponent.prototype.verifyCode = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var e_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        this.loading = true;
                        this.hasError = false;
                        if (!this.verify) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.users.verifyTotpCode(this.model.token)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        this.onSuccess.emit(this.model.token);
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        this.hasError = true;
                        this.alert.removeLastDanger();
                        this.loading = false;
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    TotpChallengeComponent.ctorParameters = function () { return [
        { type: LoginService },
        { type: UserService },
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], TotpChallengeComponent.prototype, "verify", void 0);
    tslib_1.__decorate([
        Output()
    ], TotpChallengeComponent.prototype, "onSuccess", void 0);
    tslib_1.__decorate([
        Input()
    ], TotpChallengeComponent.prototype, "loading", void 0);
    tslib_1.__decorate([
        Input()
    ], TotpChallengeComponent.prototype, "hasError", void 0);
    TotpChallengeComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-totp-challenge',
            template: "<form\n  #totpForm=\"ngForm\"\n  role=\"form\"\n  class=\"loginForm\"\n  (ngSubmit)=\"verifyCode()\"\n  novalidate\n>\n\n  <c8y-form-group\n    [hasError]=\"hasError\"\n    [novalidation]=\"true\"\n  >\n    <label\n      translate\n      for=\"totpToken\"\n    >Verification code</label>\n    <input\n      id=\"totpToken\"\n      [(ngModel)]=\"model.token\"\n      name=\"totpToken\"\n      type=\"text\"\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      autocomplete=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{'Verification code' | translate}}\"\n      required\n    >\n    <c8y-messages>\n      <c8y-message *ngIf=\"hasError\" translate>\n        Invalid verification code. In case of key loss, please contact your platform administrator.\n      </c8y-message>\n    </c8y-messages>\n    <p id=\"helpinput\" *ngIf=\"!hasError\" class=\"help-block\" translate>\n      In case of key loss, please contact your platform administrator.\n    </p>\n  </c8y-form-group>\n\n  <button\n    title=\"{{ 'Verify' | translate }}\" \n    [disabled]=\"!totpForm.form.valid || loading\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n    translate\n  >Verify</button>\n\n</form>\n",
            viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
        })
    ], TotpChallengeComponent);
    return TotpChallengeComponent;
}());
export { TotpChallengeComponent };
