// @c8y/ngx-components/esm2015/device-grid/device-grid.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule, DataGridModule, DeviceStatusModule, FormsModule } from '@c8y/ngx-components';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
import { DeviceGridComponent } from './device-grid.component';
import { DeviceGridService } from './device-grid.service';
import { BaseFilteringFormRendererComponent } from './columns/base.filtering-form-renderer.component';
import { GroupFilteringFormRendererComponent } from './columns/group.filtering-form-renderer.component';
import { StatusCellRendererComponent } from './columns/status.cell-renderer.component';
import { NameCellRendererComponent } from './columns/name.cell-renderer.component';
import { ModelCellRendererComponent } from './columns/model.cell-renderer.component';
import { SerialNumberCellRendererComponent } from './columns/serial-number.cell-renderer.component';
import { GroupCellRendererComponent } from './columns/group.cell-renderer.component';
import { RegistrationDateCellRendererComponent } from './columns/registration-date.cell-renderer.component';
import { AlarmsCellRendererComponent } from './columns/alarms.cell-renderer.component';
import { SchemaFormComponent } from './schema-form.component';
let DeviceGridModule = class DeviceGridModule {
};
DeviceGridModule = tslib_1.__decorate([
    NgModule({
        imports: [CommonModule, FormsModule, DeviceStatusModule, DataGridModule, AssetSelectorModule],
        declarations: [
            DeviceGridComponent,
            BaseFilteringFormRendererComponent,
            GroupFilteringFormRendererComponent,
            StatusCellRendererComponent,
            NameCellRendererComponent,
            ModelCellRendererComponent,
            SerialNumberCellRendererComponent,
            GroupCellRendererComponent,
            RegistrationDateCellRendererComponent,
            AlarmsCellRendererComponent,
            SchemaFormComponent
        ],
        exports: [
            DeviceGridComponent,
            BaseFilteringFormRendererComponent,
            GroupFilteringFormRendererComponent,
            StatusCellRendererComponent,
            NameCellRendererComponent,
            ModelCellRendererComponent,
            SerialNumberCellRendererComponent,
            GroupCellRendererComponent,
            RegistrationDateCellRendererComponent,
            AlarmsCellRendererComponent,
            SchemaFormComponent
        ],
        entryComponents: [
            DeviceGridComponent,
            BaseFilteringFormRendererComponent,
            GroupFilteringFormRendererComponent,
            StatusCellRendererComponent,
            NameCellRendererComponent,
            ModelCellRendererComponent,
            SerialNumberCellRendererComponent,
            GroupCellRendererComponent,
            RegistrationDateCellRendererComponent,
            AlarmsCellRendererComponent
        ],
        providers: [DeviceGridService]
    })
], DeviceGridModule);
export { DeviceGridModule };
