// @c8y/ngx-components/esm2015/core/device-grid/columns/group.device-grid-column.js
import { BaseDeviceGridColumn } from './base.device-grid-column';
import { GroupCellRendererComponent } from './group.cell-renderer.component';
import { gettext } from '../../i18n/gettext';
export class GroupDeviceGridColumn extends BaseDeviceGridColumn {
    constructor() {
        super();
        this.name = 'group';
        this.header = gettext('Group');
        this.cellRendererComponent = GroupCellRendererComponent;
        this.filterable = true;
        this.filteringConfig = {
            schema: {
                type: 'object',
                properties: {
                    deviceGroupId: {
                        type: 'integer',
                        format: 'deviceGroupId'
                    }
                }
            },
            form: [
                {
                    key: 'deviceGroupId',
                    title: gettext('Show devices in group')
                }
            ],
            getFilter(model) {
                const filter = {};
                if (model.deviceGroupId) {
                    filter.__bygroupid = model.deviceGroupId;
                }
                return filter;
            }
        };
        this.sortable = false;
    }
}
