// @c8y/ngx-components/esm2015/connectivity/connectivity.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule, CoreModule, FormsModule } from '@c8y/ngx-components';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { ImpactConnectivityService } from './impact/impact-connectivity.service';
import { ImpactProviderSettingsComponent } from './impact/impact-provider-settings.component';
import { StatusDisplayComponent } from './impact/status-display.component';
let ConnectivityModule = class ConnectivityModule {
};
ConnectivityModule = tslib_1.__decorate([
    NgModule({
        declarations: [ImpactProviderSettingsComponent, StatusDisplayComponent],
        exports: [ImpactProviderSettingsComponent],
        imports: [CoreModule, CommonModule, FormsModule, PopoverModule, ReactiveFormsModule],
        entryComponents: [ImpactProviderSettingsComponent],
        providers: [ImpactConnectivityService]
    })
], ConnectivityModule);
export { ConnectivityModule };
