// @c8y/ngx-components/esm5/upgrade/dashboard/component.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetComponent } from './widget.component';
import { DeviceSelectorComponent } from './device-selector.component';
var Ng1ComponentFactoryService = /** @class */ (function () {
    function Ng1ComponentFactoryService(bridge) {
        this.bridge = bridge;
    }
    Ng1ComponentFactoryService.prototype.get = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var components;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.bridge.ng1Components.list()];
                    case 1:
                        components = (_a.sent()).map(function (cmp) {
                            var data = {
                                settings: tslib_1.__assign({}, cmp.options, { configTemplateUrl: cmp.configTemplateUrl, widgetComponent: cmp.widgetComponent, configComponent: cmp.configComponent, transformConfigWithContext: cmp.transformConfigWithContext || cmp.componentTransformConfigWithContext, upgrade: true })
                            };
                            data.settings.templateUrl = cmp.templateUrl; // workaround as in object it creates build error
                            return {
                                id: cmp.name,
                                label: cmp.nameDisplay || cmp.name,
                                description: cmp.description,
                                component: WidgetComponent,
                                configComponent: WidgetComponent,
                                previewImage: cmp.previewImage,
                                data: data
                            };
                        });
                        // add the device selector legacy component
                        components.push({
                            id: 'device.selector.legacy',
                            label: 'Device selector',
                            description: '',
                            data: { settings: { noNewWidgets: true, upgrade: true } },
                            component: DeviceSelectorComponent
                        });
                        return [2 /*return*/, components];
                }
            });
        });
    };
    Ng1ComponentFactoryService.ctorParameters = function () { return [
        { type: DashboardBridgeService }
    ]; };
    Ng1ComponentFactoryService = tslib_1.__decorate([
        Injectable()
    ], Ng1ComponentFactoryService);
    return Ng1ComponentFactoryService;
}());
export { Ng1ComponentFactoryService };
