// @c8y/ngx-components/esm5/upgrade/dashboard/copy-action.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
var CopyActionComponent = /** @class */ (function () {
    function CopyActionComponent(vcRef, dashboardBridgeService) {
        this.vcRef = vcRef;
        this.dashboardBridgeService = dashboardBridgeService;
    }
    CopyActionComponent.prototype.ngOnInit = function () {
        this.vcRef.createEmbeddedView(this.templateCopy);
    };
    CopyActionComponent.prototype.copyDashboard = function () {
        this.dashboardBridgeService.copyDashboard();
    };
    CopyActionComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: DashboardBridgeService }
    ]; };
    tslib_1.__decorate([
        ViewChild('templateCopy', { read: TemplateRef, static: true })
    ], CopyActionComponent.prototype, "templateCopy", void 0);
    CopyActionComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-copy-action-legacy',
            template: "\n    <ng-template #templateCopy>\n      <li>\n        <button title=\"{{'Copy dashboard' | translate}}\" (click)=\"copyDashboard()\">\n          <i class=\"fa fw fa-clone\"></i>&nbsp;<span translate>Copy dashboard</span>\n        </button>\n      </li>\n    </ng-template>\n  "
        })
    ], CopyActionComponent);
    return CopyActionComponent;
}());
export { CopyActionComponent };
