// @c8y/ngx-components/esm5/upgrade/dashboard/device-selector.component.js
import * as tslib_1 from "tslib";
import { Input, ChangeDetectionStrategy, Component, ElementRef, OnInit, OnDestroy } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
var DeviceSelectorComponent = /** @class */ (function () {
    function DeviceSelectorComponent(dashboardBridgeService, el) {
        this.dashboardBridgeService = dashboardBridgeService;
        this.el = el;
    }
    DeviceSelectorComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.ng1Scope = this.dashboardBridgeService.instantiateDeviceSelector(this.el.nativeElement, this.config);
                return [2 /*return*/];
            });
        });
    };
    DeviceSelectorComponent.prototype.ngOnDestroy = function () {
        if (this.ng1Scope) {
            this.ng1Scope.$destroy();
        }
    };
    DeviceSelectorComponent.ctorParameters = function () { return [
        { type: DashboardBridgeService },
        { type: ElementRef }
    ]; };
    DeviceSelectorComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-selector-legacy',
            template: '',
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], DeviceSelectorComponent);
    return DeviceSelectorComponent;
}());
export { DeviceSelectorComponent };
