// @c8y/ngx-components/esm5/connectivity/impact/impact-connectivity.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FetchClient, IFetchOptions, IFetchResponse } from '@c8y/client';
var ImpactConnectivityService = /** @class */ (function () {
    function ImpactConnectivityService(client) {
        this.client = client;
        this.microserviceUrl = '/service/impact/connectivity';
        this.header = { 'Content-Type': 'application/json' };
    }
    ImpactConnectivityService.prototype.getOptions = function () {
        var options = {
            method: 'GET',
            headers: this.header
        };
        return this.client.fetch("" + this.microserviceUrl, options);
    };
    ImpactConnectivityService.prototype.updateOptions = function (impactOptions) {
        var options = {
            method: 'POST',
            headers: this.header,
            body: JSON.stringify(impactOptions)
        };
        return this.client.fetch("" + this.microserviceUrl, options);
    };
    ImpactConnectivityService.prototype.deleteOptions = function () {
        var options = {
            method: 'DELETE',
            headers: this.header
        };
        return this.client.fetch("" + this.microserviceUrl, options);
    };
    ImpactConnectivityService.ctorParameters = function () { return [
        { type: FetchClient }
    ]; };
    ImpactConnectivityService = tslib_1.__decorate([
        Injectable()
    ], ImpactConnectivityService);
    return ImpactConnectivityService;
}());
export { ImpactConnectivityService };
