// @c8y/ngx-components/esm5/core/device-grid/device-grid.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { DeviceGridComponent } from './device-grid.component';
import { DataGridModule } from '../data-grid/data-grid.module';
import { DeviceGridService } from './device-grid.service';
import { DeviceStatusModule } from '../device-status/device-status.module';
import { FilteringFormComponent } from './filtering-form.component';
import { FormsModule } from '../forms/forms.module';
import { StatusCellRendererComponent } from './columns/status.cell-renderer.component';
import { NameCellRendererComponent } from './columns/name.cell-renderer.component';
import { ModelCellRendererComponent } from './columns/model.cell-renderer.component';
import { SerialNumberCellRendererComponent } from './columns/serial-number.cell-renderer.component';
import { GroupCellRendererComponent } from './columns/group.cell-renderer.component';
import { RegistrationDateCellRendererComponent } from './columns/registration-date.cell-renderer.component';
import { AlarmsCellRendererComponent } from './columns/alarms.cell-renderer.component';
var DeviceGridModule = /** @class */ (function () {
    function DeviceGridModule() {
    }
    DeviceGridModule = tslib_1.__decorate([
        NgModule({
            imports: [CommonModule, FormsModule, DeviceStatusModule, DataGridModule],
            declarations: [
                DeviceGridComponent,
                FilteringFormComponent,
                StatusCellRendererComponent,
                NameCellRendererComponent,
                ModelCellRendererComponent,
                SerialNumberCellRendererComponent,
                GroupCellRendererComponent,
                RegistrationDateCellRendererComponent,
                AlarmsCellRendererComponent
            ],
            exports: [
                DeviceGridComponent,
                FilteringFormComponent,
                StatusCellRendererComponent,
                NameCellRendererComponent,
                ModelCellRendererComponent,
                SerialNumberCellRendererComponent,
                GroupCellRendererComponent,
                RegistrationDateCellRendererComponent,
                AlarmsCellRendererComponent
            ],
            entryComponents: [
                DeviceGridComponent,
                FilteringFormComponent,
                StatusCellRendererComponent,
                NameCellRendererComponent,
                ModelCellRendererComponent,
                SerialNumberCellRendererComponent,
                GroupCellRendererComponent,
                RegistrationDateCellRendererComponent,
                AlarmsCellRendererComponent
            ],
            providers: [DeviceGridService]
        })
    ], DeviceGridModule);
    return DeviceGridModule;
}());
export { DeviceGridModule };
