// @c8y/ngx-components/esm5/core/device-grid/filtering-form.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input, ViewChild } from '@angular/core';
import { get, cloneDeep } from 'lodash-es';
import { SchemaFormComponent } from '../forms/schema-form.component';
var FilteringFormComponent = /** @class */ (function () {
    function FilteringFormComponent() {
    }
    FilteringFormComponent.prototype.ngOnInit = function () {
        var column = this.context.property;
        this.schema = cloneDeep(column.filteringConfig.schema);
        this.form = cloneDeep(column.filteringConfig.form);
        this.model = cloneDeep(column.externalFilterQuery || {});
        this.options = { formDefaults: { ngModelOptions: { debounce: 100 } } };
    };
    FilteringFormComponent.prototype.canApply = function () {
        return get(this.schemaFormComponent, '$componentScope.vm.ngForm.$invalid');
    };
    FilteringFormComponent.prototype.applyFilter = function () {
        this.context.applyFilter({
            externalFilterQuery: this.model
        });
    };
    FilteringFormComponent.prototype.resetFilter = function () {
        this.context.resetFilter();
    };
    tslib_1.__decorate([
        Input()
    ], FilteringFormComponent.prototype, "context", void 0);
    tslib_1.__decorate([
        ViewChild(forwardRef(function () { return SchemaFormComponent; }), { static: false })
    ], FilteringFormComponent.prototype, "schemaFormComponent", void 0);
    FilteringFormComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-filtering-form',
            template: "<c8y-schema-form\n  [schema]=\"schema\"\n  [form]=\"form\"\n  [model]=\"model\"\n  [options]=\"options\"\n></c8y-schema-form>\n<div class=\"flex-row p-t-16\">\n  <button\n    class=\"btn btn-default btn-sm m-r-8 flex-grow\"\n    (click)=\"resetFilter()\"\n    title=\"{{ 'Reset' | translate }}\"\n    translate\n  >\n    Reset\n  </button>\n\n  <button\n    class=\"btn btn-primary btn-sm flex-grow\"\n    (click)=\"applyFilter()\"\n    [disabled]=\"canApply()\"\n    title=\"{{ 'Apply' | translate }}\"\n    translate\n  >\n    Apply\n  </button>\n</div>\n"
        })
    ], FilteringFormComponent);
    return FilteringFormComponent;
}());
export { FilteringFormComponent };
