// @c8y/ngx-components/esm5/upgrade/dashboard/paste-action.component.js
import * as tslib_1 from "tslib";
import { Component, OnInit, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
var PasteActionComponent = /** @class */ (function () {
    function PasteActionComponent(vcRef, dashboardBridgeService) {
        this.vcRef = vcRef;
        this.dashboardBridgeService = dashboardBridgeService;
        this.dashboardName = '';
    }
    PasteActionComponent.prototype.ngOnInit = function () {
        this.vcRef.createEmbeddedView(this.templatePaste);
        this.dashboardName = this.dashboardBridgeService.dashboardClipboard.c8y_Dashboard.name;
    };
    PasteActionComponent.prototype.pasteDashboard = function () {
        this.dashboardBridgeService.pasteDashboard();
    };
    PasteActionComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: DashboardBridgeService }
    ]; };
    tslib_1.__decorate([
        ViewChild('templatePaste', { read: TemplateRef, static: true })
    ], PasteActionComponent.prototype, "templatePaste", void 0);
    PasteActionComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-paste-action-legacy',
            template: "\n    <ng-template #templatePaste>\n      <li>\n        <button title=\"{{'Paste dashboard' | translate}}\" (click)=\"pasteDashboard()\">\n          <i class=\"fa fw fa-clipboard\"></i>&nbsp;<span translate>Paste dashboard</span>&nbsp;<span\n            >\"{{ dashboardName }}\"</span\n          >\n        </button>\n      </li>\n    </ng-template>\n  "
        })
    ], PasteActionComponent);
    return PasteActionComponent;
}());
export { PasteActionComponent };
