// @c8y/ngx-components/esm5/operations/bulk/creation/stepper-bulk-type-device-profile.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
import { get } from 'lodash-es';
import { BaseStepperComponent } from '../base-stepper.component';
var StepperBulkTypeDeviceProfile = /** @class */ (function (_super) {
    tslib_1.__extends(StepperBulkTypeDeviceProfile, _super);
    function StepperBulkTypeDeviceProfile() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StepperBulkTypeDeviceProfile.prototype.goToConfirmSelection = function ($event) {
        this.deviceTypes = get(this.selectedDeviceProfile, 'c8y_Filter.type');
        $event.stepper.next();
    };
    StepperBulkTypeDeviceProfile.prototype.onDeviceProfileSelected = function (selectedItem) {
        this.selectedDeviceProfile = selectedItem;
    };
    StepperBulkTypeDeviceProfile.prototype.retrieveOperationPrototype = function () {
        return {
            name: gettext('Apply device profile'),
            description: get(this.selectedDeviceProfile, 'name'),
            prototype: {
                description: "Apply device profile: " + this.selectedDeviceProfile.name,
                profileId: this.selectedDeviceProfile.id,
                profileName: this.selectedDeviceProfile.name,
                c8y_DeviceProfile: this.selectedDeviceProfile.c8y_DeviceProfile
            }
        };
    };
    StepperBulkTypeDeviceProfile = tslib_1.__decorate([
        Component({
            selector: 'c8y-stepper-bulk-type-device-profile',
            template: "<c8y-bulk-operation-stepper>\n  <ng-container\n    *customStep=\"\n      'Select device profile' | translate; \n      completed: !!selectedDeviceProfile;\n      buttonsDisabled: !selectedDeviceProfile; \n      onNext: goToConfirmSelection.bind(this)\"\n  >\n    <c8y-select-device-profile-step\n      (deviceProfile)=\"onDeviceProfileSelected($event)\"\n      class=\"d-contents\"\n    ></c8y-select-device-profile-step>\n  </ng-container>\n  <ng-container *customStep=\"'Confirm selected device profile' | translate\">\n    <c8y-confirm-device-profile-selection-step\n      class=\"d-contents\"\n      [selectedDeviceProfile]=\"selectedDeviceProfile\"\n    ></c8y-confirm-device-profile-selection-step>\n  </ng-container>\n</c8y-bulk-operation-stepper>\n"
        })
    ], StepperBulkTypeDeviceProfile);
    return StepperBulkTypeDeviceProfile;
}(BaseStepperComponent));
export { StepperBulkTypeDeviceProfile };
