// @c8y/ngx-components/esm2015/core/router/router.service.js
import * as tslib_1 from "tslib";
import { Inject, Injectable, InjectionToken, Optional } from '@angular/core';
import { Router } from '@angular/router';
import { fromFactories } from '../common/extension-hooks';
import { TabsService } from '../tabs/tabs.service';
import { ContextRouteService } from './context-route.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../tabs/tabs.service";
import * as i3 from "./context-route.service";
/**
 * A hook to use for Multi Provider extension.
 */
export const HOOK_ONCE_ROUTE = new InjectionToken('RouteFactory');
/**
 * A service which defines routes
 */
let RouterService = class RouterService {
    constructor(factories, router, tabsService, contextRouteService) {
        this.router = router;
        this.tabsService = tabsService;
        this.contextRouteService = contextRouteService;
        fromFactories(factories).subscribe((routes) => {
            this.addRoute(routes);
        });
    }
    /**
     * Refresh the current context
     */
    refresh() {
        this.contextRouteService.refreshContext();
    }
    /**
     * Add a new route to the router configuration or a context.
     * @param route The route to add
     */
    addRoute(route) {
        this.addRoutes(Array.isArray(route) ? route : [route]);
    }
    addRoutes(routes) {
        const emptyRoute = this.router.config.find(r => r.path === '**');
        this.router.resetConfig([
            ...this.router.config.filter(r => r.path !== '**'),
            ...routes.map((r) => this.convertRoute(this.router.config, r)),
            emptyRoute
        ].filter(Boolean));
    }
    convertRoute(initialConfig, route) {
        if (route.context) {
            initialConfig.forEach((r) => {
                if (r.data && r.data.context === route.context) {
                    r.children = [route, ...(r.children || [])];
                }
            });
            return;
        }
        return route;
    }
};
RouterService.ctorParameters = () => [
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [HOOK_ONCE_ROUTE,] }] },
    { type: Router },
    { type: TabsService },
    { type: ContextRouteService }
];
RouterService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function RouterService_Factory() { return new RouterService(i0.ɵɵinject(HOOK_ONCE_ROUTE, 8), i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.TabsService), i0.ɵɵinject(i3.ContextRouteService)); }, token: RouterService, providedIn: "root" });
RouterService = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    }),
    tslib_1.__param(0, Optional()), tslib_1.__param(0, Inject(HOOK_ONCE_ROUTE))
], RouterService);
export { RouterService };
