// @c8y/ngx-components/esm5/operations/bulk/modal/bulk-operations-reschedule-modal.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { AlertService, gettext } from '@c8y/ngx-components';
import { BsModalRef } from 'ngx-bootstrap';
import { Subject } from 'rxjs';
import { BulkOperationsService } from '../bulk-operations.service';
var BulkOperationsRescheduleModalComponent = /** @class */ (function () {
    function BulkOperationsRescheduleModalComponent(bsModalRef, bulkOperationsService, alert) {
        this.bsModalRef = bsModalRef;
        this.bulkOperationsService = bulkOperationsService;
        this.alert = alert;
        this.closeSubject = new Subject();
    }
    BulkOperationsRescheduleModalComponent.prototype.ngOnInit = function () {
        this.modalTitle = this.getTitle();
    };
    BulkOperationsRescheduleModalComponent.prototype.onDateAndDelaySelect = function ($event) {
        this.dateAndDelay = $event;
    };
    BulkOperationsRescheduleModalComponent.prototype.getTitle = function () {
        return this.isRetryOperation ? gettext('Retry operation') : gettext('Edit schedule');
    };
    BulkOperationsRescheduleModalComponent.prototype.onCancel = function () {
        this.bsModalRef.hide();
    };
    BulkOperationsRescheduleModalComponent.prototype.onSave = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var newBulkOperation, er_1, er_2;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.bulkOperation.startDate = this.dateAndDelay.scheduledDate.toISOString();
                        this.bulkOperation.creationRamp = this.dateAndDelay.delayInSeconds;
                        if (!this.isRetryOperation) return [3 /*break*/, 5];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.bulkOperationsService.createBulkOperation(this.bulkOperation)];
                    case 2:
                        newBulkOperation = _a.sent();
                        this.alert.success(gettext('Successfully rescheduled for a retry.'));
                        return [3 /*break*/, 4];
                    case 3:
                        er_1 = _a.sent();
                        this.alert.addServerFailure(er_1);
                        return [3 /*break*/, 4];
                    case 4: return [3 /*break*/, 8];
                    case 5:
                        _a.trys.push([5, 7, , 8]);
                        return [4 /*yield*/, this.bulkOperationsService.updateBulkOperation(this.bulkOperation)];
                    case 6:
                        newBulkOperation = _a.sent();
                        this.alert.success(gettext('Successfully rescheduled.'));
                        return [3 /*break*/, 8];
                    case 7:
                        er_2 = _a.sent();
                        this.alert.addServerFailure(er_2);
                        return [3 /*break*/, 8];
                    case 8:
                        // send back to component from which modal was called
                        this.closeSubject.next(newBulkOperation.data);
                        this.bsModalRef.hide();
                        return [2 /*return*/];
                }
            });
        });
    };
    BulkOperationsRescheduleModalComponent.ctorParameters = function () { return [
        { type: BsModalRef },
        { type: BulkOperationsService },
        { type: AlertService }
    ]; };
    BulkOperationsRescheduleModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-bulk-operations-reschedule-modal',
            template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <h1 class=\"c8y-icon c8y-icon-energy\"></h1>\n    <h4>{{ modalTitle | translate }}</h4>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body\">\n      <div class=\"text-center\">\n        <c8y-operation-summary\n          [description]=\"bulkOperation.operationPrototype.description\"\n        ></c8y-operation-summary>\n      </div>\n      <div class=\"p-t-16 center-block\" style=\"max-width: 300px;\">\n        <c8y-operation-scheduler\n          [(ngModel)]=\"dateAndDelay\"\n          #scheduler=\"ngModel\"\n        ></c8y-operation-scheduler>\n      </div>\n    </div>\n  </div>\n  <div class=\"modal-footer\">\n    <button\n      (click)=\"onCancel()\"\n      type=\"button\"\n      class=\"btn btn-default\"\n      title=\"{{ 'Cancel' | translate }}\"\n    >\n      <span translate>Cancel</span>\n    </button>\n    <button\n      id=\"save-btn\"\n      (click)=\"onSave()\"\n      type=\"button\"\n      class=\"btn btn-primary\"\n      title=\"{{ 'Save' | translate }}\"\n      [disabled]=\"scheduler?.control?.errors\"\n    >\n      <span translate>Save</span>\n    </button>\n  </div>\n</div>\n"
        })
    ], BulkOperationsRescheduleModalComponent);
    return BulkOperationsRescheduleModalComponent;
}());
export { BulkOperationsRescheduleModalComponent };
