// @c8y/ngx-components/esm2015/sensor-phone/connect-smartphone.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { DocLinkFactory, OptionsService, gettext, DocLink } from '@c8y/ngx-components';
import { of } from 'rxjs';
import { SensorPhoneService } from './sensor-phone.service';
let ConnectSmartphoneFactory = class ConnectSmartphoneFactory {
    constructor(optionsService, sensorPhoneService) {
        this.optionsService = optionsService;
        this.sensorPhoneService = sensorPhoneService;
    }
    get() {
        if (this.optionsService.get('sensorPhone', false)) {
            return of([
                {
                    icon: 'c8y-mobile-add',
                    type: 'quicklink',
                    label: gettext('Connect smartphone'),
                    url: undefined,
                    click: () => this.sensorPhoneService.createModal()
                }
            ]);
        }
    }
};
ConnectSmartphoneFactory.ctorParameters = () => [
    { type: OptionsService },
    { type: SensorPhoneService }
];
ConnectSmartphoneFactory = tslib_1.__decorate([
    Injectable()
], ConnectSmartphoneFactory);
export { ConnectSmartphoneFactory };
