// @c8y/ngx-components/esm5/protocol-opcua/address-space.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FetchClient, IFetchOptions } from '@c8y/client';
import { BehaviorSubject } from 'rxjs';
var AddressSpaceService = /** @class */ (function () {
    function AddressSpaceService(fetchClient) {
        this.client = fetchClient;
        this.microserviceUrl = '/service/opcua-mgmt-service/address-space';
        this.header = { 'Content-Type': 'application/json' };
        this.nodeNavigationData$ = new BehaviorSubject({
            node: undefined,
            selectedAncestorIds: []
        });
    }
    AddressSpaceService.prototype.resetTreeToRootNode = function () {
        this.triggerNodeToOpen({ node: undefined, selectedAncestorIds: [] });
    };
    AddressSpaceService.prototype.triggerNodeToOpen = function (nodeNavigationData) {
        this.nodeNavigationData$.next(nodeNavigationData);
    };
    AddressSpaceService.prototype.getNodeNavData$ = function () {
        return this.nodeNavigationData$.asObservable();
    };
    AddressSpaceService.prototype.getNode = function (serverId, nodeId) {
        if (serverId && serverId.length > 0) {
            if (nodeId && nodeId.length > 0) {
                return this.getNodeById(serverId, nodeId);
            }
            return this.getRootNode(serverId);
        }
    };
    AddressSpaceService.prototype.getRootNode = function (serverId) {
        if (serverId && serverId.length > 0) {
            var options = {
                method: 'GET',
                headers: this.header
            };
            return this.client.fetch(this.microserviceUrl + "/" + serverId, options);
        }
    };
    AddressSpaceService.prototype.getNodeById = function (serverId, nodeId) {
        if (serverId && nodeId && serverId.length > 0 && nodeId.length > 0) {
            var options = {
                method: 'GET',
                headers: this.header
            };
            var param = encodeURIComponent(nodeId);
            return this.client.fetch(this.microserviceUrl + "/" + serverId + "?nodeId=" + param, options);
        }
    };
    AddressSpaceService.prototype.getChildrenOf = function (node, serverId) {
        if (serverId && node.nodeId && serverId.length > 0 && node.nodeId.length > 0) {
            var options = {
                method: 'GET',
                headers: this.header
            };
            var param = encodeURIComponent(node.nodeId);
            return this.client.fetch(this.microserviceUrl + "/" + serverId + "/children?nodeId=" + param, options);
        }
    };
    AddressSpaceService.prototype.childrenAvailable = function (nodeReferences) {
        if (!nodeReferences || nodeReferences.length === 0) {
            return false;
        }
        return nodeReferences.some(function (ref) { return !ref.inverse && ref.hierarchical; });
    };
    AddressSpaceService.prototype.getSearchedNodes = function (searchKey, serverId) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var url, options, res;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = "service/opcua-mgmt-service/search/" + serverId + "/";
                        options = {
                            headers: this.header,
                            params: {
                                searchString: '*' + searchKey + '*'
                            }
                        };
                        return [4 /*yield*/, this.client.fetch(url, options)];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, res.json()];
                }
            });
        });
    };
    AddressSpaceService.prototype.getIcon = function (nodeClassName) {
        var iconList = {
            Object: 'cube',
            Variable: 'th-list',
            Method: 'random',
            View: 'window-maximize',
            ObjectType: 'c8y-group',
            VariableType: 'c8y-group',
            ReferenceType: 'c8y-group',
            DataType: 'c8y-group'
        };
        return iconList[nodeClassName] || 'circle';
    };
    AddressSpaceService.ctorParameters = function () { return [
        { type: FetchClient }
    ]; };
    AddressSpaceService = tslib_1.__decorate([
        Injectable()
    ], AddressSpaceService);
    return AddressSpaceService;
}());
export { AddressSpaceService };
