// @c8y/ngx-components/esm2015/api/data.module.js
var DataModule_1;
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { BasicAuth, FetchClient, Realtime, CookieAuth } from '@c8y/client';
import { ApiService } from './api.service';
import * as services from './services';
function toProvider(provide) {
    let deps = [FetchClient, Realtime];
    if (provide === FetchClient) {
        deps = [CookieAuth];
    }
    if (provide === BasicAuth || provide === CookieAuth) {
        deps = [];
    }
    if (provide === Realtime) {
        deps = [FetchClient];
    }
    return { provide, useClass: provide, deps };
}
const ɵ0 = (k) => toProvider(services[k]);
const providers = Object.keys(services).map(ɵ0)
    .concat([
    { provide: ApiService, useClass: ApiService, deps: [FetchClient] }
]);
// @dynamic
let DataModule = DataModule_1 = class DataModule {
    static providers() {
        return providers;
    }
    static forRoot() {
        return {
            ngModule: DataModule_1,
            providers
        };
    }
};
DataModule = DataModule_1 = tslib_1.__decorate([
    NgModule()
], DataModule);
export { DataModule };
export { ɵ0 };
