// @c8y/ngx-components/esm2015/core/forms/json-validation-prettifier.directive.js
import * as tslib_1 from "tslib";
import { Directive, EventEmitter, Output } from '@angular/core';
import { NgControl, AbstractControl } from '@angular/forms';
import { debounceTime, distinctUntilChanged, tap } from 'rxjs/operators';
import { gettext } from '../i18n/public-api';
let JsonValidationPrettifierDirective = class JsonValidationPrettifierDirective {
    constructor(ngCtrl) {
        this.invalidJSON = new EventEmitter();
        this.message = gettext('Must be a valid JSON object.');
        this.debounceTimeInMs = 1000;
        this.abstractCtrl = ngCtrl.control;
    }
    ngOnInit() {
        this.subscription = this.abstractCtrl.valueChanges
            .pipe(tap(() => this.clearValidationMessage()), debounceTime(this.debounceTimeInMs), distinctUntilChanged())
            .subscribe(value => {
            this.validateInputAndPrettify(value);
        });
    }
    ngOnDestroy() {
        if (this.subscription && !this.subscription.closed) {
            this.subscription.unsubscribe();
        }
    }
    validateInputAndPrettify(value) {
        if (value) {
            try {
                const validJsonObject = JSON.parse(value);
                const pretty = JSON.stringify(validJsonObject, undefined, 2);
                this.abstractCtrl.setValue(pretty);
            }
            catch (ex) {
                this.setErrorAndValidationMessage();
            }
        }
    }
    setErrorAndValidationMessage() {
        this.abstractCtrl.setErrors({ invalidBodyTemplate: true });
        this.invalidJSON.emit(this.message);
    }
    clearValidationMessage() {
        this.invalidJSON.emit(undefined);
    }
};
JsonValidationPrettifierDirective.ctorParameters = () => [
    { type: NgControl }
];
tslib_1.__decorate([
    Output()
], JsonValidationPrettifierDirective.prototype, "invalidJSON", void 0);
JsonValidationPrettifierDirective = tslib_1.__decorate([
    Directive({
        selector: 'textarea[prettyValidJson]'
    })
], JsonValidationPrettifierDirective);
export { JsonValidationPrettifierDirective };
