// @c8y/ngx-components/esm2015/core/forms/messages.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren, Input } from '@angular/core';
import { MessageDirective } from './message.directive';
import { ValidationPattern } from './validation-pattern';
import { TranslateService } from '@ngx-translate/core';
let MessagesComponent = class MessagesComponent {
    constructor(translateService) {
        this.translateService = translateService;
        this.show = {};
        this.defaults = ValidationPattern.messages;
        this.messages = [];
        this.messageMap = {};
    }
    ngAfterContentInit() {
        this.messageMap = Object.assign({}, this.defaults, this.customMessages.reduce((messages, directive) => {
            messages[directive.name] = directive.text;
            return messages;
        }, {}));
    }
    changVisibility(show) {
        this.show = show;
        this.ngOnChanges();
    }
    ngOnChanges() {
        if (this.show) {
            this.messages = Object.keys(this.show)
                .map((key) => this.toMessage(key, this.show[key]));
        }
        else {
            this.messages = [];
        }
    }
    toMessage(key, value) {
        return this.translateService.instant(this.messageMap[key], value);
    }
};
MessagesComponent.ctorParameters = () => [
    { type: TranslateService }
];
tslib_1.__decorate([
    Input()
], MessagesComponent.prototype, "show", void 0);
tslib_1.__decorate([
    Input()
], MessagesComponent.prototype, "defaults", void 0);
tslib_1.__decorate([
    ContentChildren(MessageDirective)
], MessagesComponent.prototype, "customMessages", void 0);
MessagesComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-messages',
        template: "<small class=\"form-control-feedback-message\">\n  <div style=\"display:block\" *ngFor=\"let msg of messages\">\n    {{msg | translate}}\n  </div>\n  <ng-content></ng-content>\n</small>\n"
    })
], MessagesComponent);
export { MessagesComponent };
