// @c8y/ngx-components/esm5/trusted-certificates/trusted-certificates.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, HOOK_NAVIGATOR_NODES } from '@c8y/ngx-components';
import { TrustedCertificatesComponent } from './trusted-certificates.component';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { AddTrustedCertificateComponent } from './add-trusted-certificate.component';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { ReactiveFormsModule } from '@angular/forms';
import { TrustedCertificatesNavigationFactory } from './trusted-certificates-navigation.factory';
var trustedCertificatesRoutes = [
    {
        path: 'trusted-certificates',
        component: TrustedCertificatesComponent
    }
];
var TrustedCertificatesModule = /** @class */ (function () {
    function TrustedCertificatesModule() {
    }
    TrustedCertificatesModule = tslib_1.__decorate([
        NgModule({
            declarations: [TrustedCertificatesComponent, AddTrustedCertificateComponent],
            exports: [],
            imports: [
                CoreModule,
                CommonModule,
                RouterModule.forRoot(trustedCertificatesRoutes, { useHash: true }),
                BsDropdownModule.forRoot(),
                TooltipModule,
                ReactiveFormsModule,
                ButtonsModule,
                PopoverModule
            ],
            entryComponents: [TrustedCertificatesComponent, AddTrustedCertificateComponent],
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: TrustedCertificatesNavigationFactory,
                    multi: true
                }
            ]
        })
    ], TrustedCertificatesModule);
    return TrustedCertificatesModule;
}());
export { TrustedCertificatesModule };
