// @c8y/ngx-components/esm5/core/dynamic-component/dynamic-component.service.js
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject, of, Subject } from 'rxjs';
import { filter, mergeMap, take } from 'rxjs/operators';
import { fromFactories } from '../common/extension-hooks';
import { StateService } from '../common/state-service.abstract';
import { HOOK_COMPONENT, HOOK_COMPONENTS } from './dynamic-component.model';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
var DynamicComponentService = /** @class */ (function (_super) {
    tslib_1.__extends(DynamicComponentService, _super);
    function DynamicComponentService(injector, router) {
        var _this = _super.call(this) || this;
        _this.injector = injector;
        _this.refreshTrigger = new Subject();
        _this.state$ = new BehaviorSubject(new Set());
        _this.items$ = fromFactories([function () { return _this.injector.get(HOOK_COMPONENTS, []); }, function () { return _this.getComponentsViaDeprecatedHook(); }], router, false);
        _this.items$.subscribe(function (cmps) {
            cmps.forEach(function (cmp) {
                _this.state.add(cmp);
            });
            _this.emitNewState();
        });
        return _this;
    }
    DynamicComponentService.prototype.getComponentsViaDeprecatedHook = function () {
        var componentsViaDeprecatedHook = this.injector.get(HOOK_COMPONENT, []);
        if (componentsViaDeprecatedHook &&
            componentsViaDeprecatedHook.length &&
            __MODE__ !== 'production') {
            console.warn('HOOK_COMPONENT is now deprecated. Use HOOK_COMPONENTS instead.');
        }
        return componentsViaDeprecatedHook;
    };
    DynamicComponentService.prototype.refresh = function () {
        this.refreshTrigger.next(1);
    };
    Object.defineProperty(DynamicComponentService.prototype, "state", {
        get: function () {
            return this.state$.value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Returns a component by it's id as an observable.
     * @param id The id of the component.
     */
    DynamicComponentService.prototype.getById$ = function (id) {
        return this.state$.pipe(filter(function (cmps) { return cmps.size > 0; }), mergeMap(function (cmps) {
            return of(Array.from(cmps).find(function (cmp) { return cmp.id === id; }));
        }), take(1));
    };
    DynamicComponentService.prototype.add = function (component) {
        this.state.add(component);
        this.emitNewState();
    };
    DynamicComponentService.prototype.remove = function (component) {
        this.state.delete(component);
        this.emitNewState();
    };
    DynamicComponentService.ctorParameters = function () { return [
        { type: Injector },
        { type: Router }
    ]; };
    DynamicComponentService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function DynamicComponentService_Factory() { return new DynamicComponentService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Router)); }, token: DynamicComponentService, providedIn: "root" });
    DynamicComponentService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        })
    ], DynamicComponentService);
    return DynamicComponentService;
}(StateService));
export { DynamicComponentService };
