// @c8y/ngx-components/esm5/sensor-phone/sensor-phone.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_DOCS } from '@c8y/ngx-components';
import { ConnectSmartphoneFactory } from './connect-smartphone.factory';
import { SensorPhoneModalComponent } from './sensor-phone-modal.component';
import { QRCodeModule } from 'angularx-qrcode';
import { SensorPhoneService } from './sensor-phone.service';
var SensorPhoneModule = /** @class */ (function () {
    function SensorPhoneModule() {
    }
    SensorPhoneModule = tslib_1.__decorate([
        NgModule({
            imports: [CoreModule, QRCodeModule],
            declarations: [SensorPhoneModalComponent],
            entryComponents: [SensorPhoneModalComponent],
            providers: [
                {
                    provide: HOOK_DOCS,
                    useClass: ConnectSmartphoneFactory,
                    multi: true
                },
                SensorPhoneService
            ]
        })
    ], SensorPhoneModule);
    return SensorPhoneModule;
}());
export { SensorPhoneModule };
