// @c8y/ngx-components/esm2015/context-dashboard/widget-config.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { clone, cloneDeep, escapeRegExp, get, omit } from 'lodash-es';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { iif, Subject, timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { WIDGET_CONTENT_CLASSES, WIDGET_HEADER_CLASSES } from './context-dashboard.model';
import { WidgetService } from './widget.service';
import { ContextDashboardService } from './context-dashboard.service';
let WidgetConfigComponent = class WidgetConfigComponent {
    constructor(widgetService, modal, contextDashboardService) {
        this.widgetService = widgetService;
        this.modal = modal;
        this.contextDashboardService = contextDashboardService;
        this.mode = 'select';
        this.searchChange$ = new Subject();
        this.searchTerm = '';
        this.styling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.defaultStyling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.possibleStyling = { WIDGET_HEADER_CLASSES, WIDGET_CONTENT_CLASSES };
        this.isUpgrade = false;
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
    }
    get isEdit() {
        return !!this.current;
    }
    ngAfterContentInit() {
        this.components = this.widgetService.getWidgetDefinitions();
        if (this.selected) {
            this.current = clone(this.selected);
            this.select(this.selected, this.isEdit ? 'config' : 'select');
        }
        this.searchSub = this.searchChange$
            .pipe(switchMap((event) => iif(() => event.which !== 13, timer(200), timer(10))))
            .subscribe(() => {
            this.search();
        });
        if (this.mo.c8y_Dashboard.classes) {
            this.styling = this.setDefaultStyle(Object.assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses, (this.isEdit ? this.current.data.classes : {})));
            this.defaultStyling = this.setDefaultStyle(Object.assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses));
        }
    }
    save() {
        const { _x, _y, _width, _height } = this.selected.data;
        if (this.widgetConfig && this.widgetConfig.device) {
            const { id, name } = this.widgetConfig.device;
            this.widgetConfig.device = { id, name };
        }
        const widget = Object.assign({ _x,
            _y,
            _width,
            _height, config: omit(this.widgetConfig, 'settings'), title: this.selected.data.title, componentId: this.selected.id, id: this.isEdit ? this.current.data.id : String(Math.random()).substr(2), classes: this.getStyle() }, (!this.isEdit ? this.widgetConfig.settings.widgetDefaults : {}));
        this._save(widget);
    }
    select(cmp, mode = 'config') {
        cmp.data = cmp.data || {};
        this.selected = cmp;
        this.isUpgrade = !!get(cmp, 'data.settings.upgrade');
        this.contextDashboardService.formDisabled = this.isUpgrade;
        if (this.isEdit) {
            const { _x, _y, _width, _height, classes, title } = this.current.data;
            this.selected.data = Object.assign({}, this.selected.data, { _x, _y, _width, _height, classes, title });
        }
        this.widgetConfig = cloneDeep(this.composeWidgetConfig(this.selected, this.context));
        this.selected.data.title = this.selected.data.title || cmp.label;
        this.componentLabel = cmp.label;
        this.mode = mode;
    }
    search() {
        if (this.searchTerm.length > 0) {
            this.searchResult = this.components.filter(cmp => new RegExp(escapeRegExp(this.searchTerm.trim()), 'i').test(cmp.label));
        }
        else {
            this.resetSearch();
        }
    }
    resetSearch() {
        this.searchTerm = '';
        this.searchResult = undefined;
    }
    changeMode(mode) {
        this.mode = mode;
    }
    close() {
        this._cancel();
        this.modal.hide();
    }
    getStyle(isPreview = false) {
        const cssClasses = {};
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.headerClass)) {
            cssClasses[this.styling.headerClass] = true;
        }
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.contentClass)) {
            cssClasses[this.styling.contentClass] = true;
        }
        if (isPreview) {
            cssClasses[`dashboard-theme-${this.defaultStyling.contentClass.split('-').pop()}`] = true;
        }
        return cssClasses;
    }
    ngOnDestroy() {
        this.contextDashboardService.formDisabled = true;
        if (this.searchSub) {
            this.searchSub.unsubscribe();
        }
    }
    isDashboardDefaultStyle(className) {
        const allClasses = Object.assign({}, this.mo.c8y_Dashboard.classes, this.mo.c8y_Dashboard.widgetClasses);
        const styles = Object.keys(allClasses).map(cssClass => ({ class: cssClass }));
        const style = this.contextDashboardService.getStyling(styles, className.split('-').pop(), undefined);
        return !!style;
    }
    setDefaultStyle(setClasses) {
        let contentClass = this.styling.contentClass;
        let headerClass = this.styling.headerClass;
        const styles = Object.keys(setClasses).map(c => c.split('-').pop());
        styles.forEach(styleName => {
            contentClass = this.contextDashboardService.getStyling(WIDGET_CONTENT_CLASSES, styleName, contentClass);
            headerClass = this.contextDashboardService.getStyling(WIDGET_HEADER_CLASSES, styleName, headerClass);
        });
        return { headerClass, contentClass };
    }
    composeWidgetConfig(selectedComponent, context = {}) {
        const setting = Object.assign({ settings: Object.assign({}, selectedComponent.data.settings, get(selectedComponent.data.settings, 'ng1.options'), get(selectedComponent.data, 'ng1.options'), { context }) }, selectedComponent.data.config, { dashboardMo: this.mo.c8y_Dashboard });
        return this.applyTargetIfDeviceDashboard(setting);
    }
    applyTargetIfDeviceDashboard(widgetConfig) {
        const isDeviceType = this.contextDashboardService.isDeviceType(this.mo);
        if (isDeviceType) {
            widgetConfig.settings.hideTarget = isDeviceType;
            const noDeviceTarget = widgetConfig.settings.ng1
                ? widgetConfig.settings.ng1.options.noDeviceTarget
                : widgetConfig.settings.noDeviceTarget;
            if (!noDeviceTarget) {
                widgetConfig.device = {
                    id: this.context.id,
                    name: this.context.name
                };
            }
        }
        return widgetConfig;
    }
};
WidgetConfigComponent.ctorParameters = () => [
    { type: WidgetService },
    { type: BsModalRef },
    { type: ContextDashboardService }
];
WidgetConfigComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-widget-config',
        template: "<div class=\"modal-header separator\">\n  <h3 *ngIf=\"!current\" translate>Add widget</h3>\n  <h3 *ngIf=\"current\" translate>Edit widget</h3>\n</div>\n<form #configForm=\"ngForm\" name=\"form\">\n  <div class=\"c8y-modal-tabs\">\n    <div class=\"tabContainer\">\n      <ul class=\"nav nav-tabs nav-tabsc8y p-l-24\">\n        <li [class.active]=\"mode === 'select'\">\n          <button type=\"button\" class=\"btn\" (click)=\"changeMode('select'); (false)\">\n            <i c8yIcon=\"th-large\"></i> <span class=\"txt\" translate>Select widget</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'config'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('config'); (false)\"\n          >\n            <i c8yIcon=\"cog\"></i> <span class=\"txt\" translate>Configuration</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'style'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('style'); (false)\"\n          >\n            <i c8yIcon=\"paint-brush\"></i> <span class=\"txt\" translate>Appearance</span>\n          </button>\n        </li>\n      </ul>\n    </div>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <div\n      *ngIf=\"mode === 'select'\"\n      class=\"bg-white p-l-24 p-r-24 p-t-8 p-b-8 sticky-header-top-0\"\n      style=\"z-index: 2;\"\n    >\n      <div class=\"row\">\n        <div class=\"col-sm-6\">\n          <div class=\"input-group input-group-search\">\n            <input\n              class=\"form-control\"\n              placeholder=\"{{ 'Search\u2026' | translate }}\"\n              type=\"text\"\n              [(ngModel)]=\"searchTerm\"\n              [ngModelOptions]=\"{ standalone: true }\"\n              (keydown)=\"searchChange$.next($event)\"\n            />\n            <span class=\"input-group-btn\">\n              <button class=\"btn btn-clean\" (click)=\"resetSearch()\">\n                <i [c8yIcon]=\"searchTerm.length === 0 ? 'search' : 'close'\"></i>\n              </button>\n            </span>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div class=\"modal-body bg-gray-lighter\" *ngIf=\"mode === 'select'\">\n      <div class=\"card-group card-select\" style=\"margin-bottom:-24px;\">\n        <div\n          class=\"col-md-3 col-sm-4 col-xs-6\"\n          *ngFor=\"let cmp of searchResult || components\"\n          title=\"{{ cmp.description | translate }}\"\n        >\n          <div class=\"card p-8\" [class.active]=\"selected === cmp\" (click)=\"select(cmp)\">\n            <div\n              class=\"text-center p-8 m-b-8 flex-col flex-center\"\n              style=\"min-height: 170px; background-color: var(--body-background-color, #f2f3f4);\"\n            >\n              <ng-container *ngIf=\"!cmp.previewImage; else previewImage\">\n                <h1><i c8yIcon=\"file-image-o\"></i></h1>\n                <small translate>Preview not available</small>\n              </ng-container>\n              <ng-template #previewImage>\n                <img class=\"img-responsive\" [src]=\"cmp.previewImage\" />\n              </ng-template>\n            </div>\n            <p class=\"card-title text-truncate\">\n              <c8y-highlight\n                text=\"{{ cmp.label | translate }}\"\n                [pattern]=\"searchTerm\"\n              ></c8y-highlight>\n            </p>\n          </div>\n        </div>\n\n        <div class=\"c8y-empty-state text-center\" *ngIf=\"searchResult && searchResult.length === 0\">\n          <h1 class=\"c8y-icon c8y-icon-device c8y-icon-duocolor\"></h1>\n          <h3 translate>No widgets found.</h3>\n          <div>\n            <p translate>Rephrase your search term.</p>\n            <div>\n              <button class=\"btn btn-primary\" (click)=\"resetSearch()\" translate>\n                Reset search\n              </button>\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n\n    <div class=\"modal-body\" *ngIf=\"mode === 'config'\" style=\"min-height: calc(100vh - 290px);\">\n      <h4 style=\"text-align: left;\">\n        <strong>\n          {{ selected.label | translate }}\n        </strong>\n      </h4>\n      <p class=\"m-b-24\">\n        {{ selected.description | translate }}\n      </p>\n      <div class=\"row\">\n        <!-- change to col-sm-6 when preview is available -->\n        <div class=\"col-sm-12\">\n          <div class=\"legend form-block\" translate>Configuration</div>\n          <c8y-form-group>\n            <label for=\"widgetTitle\" translate>Title</label>\n            <input\n              id=\"widgetTitle\"\n              [(ngModel)]=\"selected.data.title\"\n              type=\"text\"\n              name=\"title\"\n              class=\"form-control\"\n              placeholder=\"{{ 'e.g.' | translate }} {{ componentLabel | translate }}\"\n              required\n            />\n          </c8y-form-group>\n\n          <!-- This is an upgraded component for the device selector and still needs to be migrated -->\n          <c8y-dynamic-component\n            componentId=\"device.selector.legacy\"\n            [config]=\"widgetConfig\"\n            [notFoundError]=\"false\"\n            *ngIf=\"!isUpgrade\"\n          ></c8y-dynamic-component>\n\n          <c8y-dynamic-component\n            [componentId]=\"selected.id\"\n            mode=\"config\"\n            [config]=\"widgetConfig\"\n            [notFoundError]=\"false\"\n          ></c8y-dynamic-component>\n        </div>\n        <!-- markup for the preview \n        \n        <div class=\"col-sm-6 sticky-header-top-0\" >\n          <div class=\"legend form-block\">\n            Preview\n          </div>\n          <div class=\"bg-gray-lighter p-16\">\n            <div class=\"card card-dashboard\" style=\"margin-bottom:0;\">\n              <div class=\"card-header-actions\">\n                <div class=\"card-title text-uppercase\">{{ selected.data.title }}</div>\n                <div class=\"header-actions\">\n                  <div class=\"optionsBtn dropdow\" class=\"btnIcon dropdown-toggle c8y-dropdown\">\n                    <i c8yIcon=\"cog\"></i>\n                  </div>\n                </div>\n              </div>\n              <div class=\"card-inner-scroll\">\n                <div class=\"card-block\" style=\"min-height: 240px;\">\n                  include here the widget\n                </div>\n              </div>\n            </div>\n          </div>\n        </div>\n      --></div>\n    </div>\n\n    <div *ngIf=\"mode === 'style'\" class=\"modal-body p-t-0\" style=\"min-height: calc(100vh - 290px);\">\n      <div class=\"row\">\n        <div class=\"col-xs-6\">\n          <c8y-appearance-settings\n            [(themeClass)]=\"styling.contentClass\"\n            [(headerClass)]=\"styling.headerClass\"\n            [possibleStylingTheme]=\"possibleStyling.WIDGET_CONTENT_CLASSES\"\n            [possibleStylingHeader]=\"possibleStyling.WIDGET_HEADER_CLASSES\"\n            [defaultThemeClass]=\"defaultStyling.contentClass\"\n            [defaultHeaderClass]=\"defaultStyling.headerClass\"\n          >\n          </c8y-appearance-settings>\n        </div>\n        <div class=\"col-xs-6 sticky-header-top-0\">\n          <c8y-widget-preview [previewClasses]=\"getStyle(true)\"></c8y-widget-preview>\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\" translate>\n      Cancel\n    </button>\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      translate\n      [disabled]=\"contextDashboardService.formDisabled || configForm.invalid\"\n    >\n      Save\n    </button>\n  </div>\n</form>\n"
    })
], WidgetConfigComponent);
export { WidgetConfigComponent };
