// @c8y/ngx-components/esm2015/core/forms/phone-validation.directive.js
var PhoneValidationDirective_1;
import * as tslib_1 from "tslib";
import { NG_VALIDATORS } from '@angular/forms';
import { Directive } from '@angular/core';
import { parsePhoneNumberFromString } from 'libphonenumber-js/max';
let PhoneValidationDirective = PhoneValidationDirective_1 = class PhoneValidationDirective {
    validate(control) {
        const valid = this.validateInternationalPhoneNumber(control);
        return valid ? null : {
            internationalPhoneNumber: true
        };
    }
    validateInternationalPhoneNumber(control) {
        let numberValid;
        try {
            const phoneNumber = parsePhoneNumberFromString(control.value);
            numberValid = phoneNumber.isValid();
            if (numberValid && control.value !== phoneNumber.format('E.164')) {
                control.setValue(phoneNumber.format('E.164'), { emitEvent: false, emitModelToViewChange: false });
            }
        }
        catch (e) {
            numberValid = !control.value;
        }
        return numberValid;
    }
};
PhoneValidationDirective = PhoneValidationDirective_1 = tslib_1.__decorate([
    Directive({
        selector: '[c8yPhoneValidation]',
        providers: [{ provide: NG_VALIDATORS, useExisting: PhoneValidationDirective_1, multi: true }]
    })
], PhoneValidationDirective);
export { PhoneValidationDirective };
