// @c8y/ngx-components/esm5/core/breadcrumb/breadcrumb.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren } from '@angular/core';
import { BreadcrumbService } from './breadcrumb.service';
import { BreadcrumbItemComponent } from './breadcrumb-item.component';
var BreadcrumbComponent = /** @class */ (function () {
    function BreadcrumbComponent(breadcrumbService) {
        this.breadcrumbService = breadcrumbService;
    }
    BreadcrumbComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.items = this.contentChildItems.toArray();
        this.viewInitTimeout = setTimeout(function () {
            return _this.breadcrumbService.add(_this);
        });
    };
    BreadcrumbComponent.prototype.ngOnDestroy = function () {
        clearTimeout(this.viewInitTimeout);
        this.breadcrumbService.remove(this);
    };
    BreadcrumbComponent.ctorParameters = function () { return [
        { type: BreadcrumbService }
    ]; };
    tslib_1.__decorate([
        ContentChildren(BreadcrumbItemComponent)
    ], BreadcrumbComponent.prototype, "contentChildItems", void 0);
    BreadcrumbComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-breadcrumb',
            template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>"
        })
    ], BreadcrumbComponent);
    return BreadcrumbComponent;
}());
export { BreadcrumbComponent };
