// @c8y/ngx-components/esm5/core/breadcrumb/breadcrumb.service.js
import * as tslib_1 from "tslib";
import { Injectable, InjectionToken, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject, Subject } from 'rxjs';
import { map, shareReplay } from 'rxjs/operators';
import { fromTrigger, stateToFactory } from '../common/extension-hooks';
import { StateService } from '../common/state-service.abstract';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export var HOOK_BREADCRUMB = new InjectionToken('BreadcrumbFactory');
var BreadcrumbService = /** @class */ (function (_super) {
    tslib_1.__extends(BreadcrumbService, _super);
    function BreadcrumbService(injector, router) {
        var _this = _super.call(this) || this;
        _this.injector = injector;
        _this.refreshTrigger = new Subject();
        _this.state$ = new BehaviorSubject(new Set());
        _this.items$ = fromTrigger(router, _this.refreshTrigger, [
            function () { return _this.injector.get(HOOK_BREADCRUMB, []); },
            stateToFactory(_this.state$)
        ]).pipe(map(function (breadcrumbs) { return _this.sortByPreferredPath(breadcrumbs); }), shareReplay(1));
        return _this;
    }
    Object.defineProperty(BreadcrumbService.prototype, "state", {
        get: function () {
            return this.state$.value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Adds a new item to the action bar in the header.
     * @param item The item to add.
     */
    BreadcrumbService.prototype.add = function (item) {
        this.state.add(item);
        this.emitNewState();
    };
    /**
     * Removes an action bar item from the header.
     * @param item The item to remove.
     */
    BreadcrumbService.prototype.remove = function (item) {
        this.state.delete(item);
        this.emitNewState();
    };
    /**
     * Refresh the extension factories
     */
    BreadcrumbService.prototype.refresh = function () {
        this.refreshTrigger.next(1);
    };
    /**
     * Sets the crumb that contains that path to the top.
     * @param path The path to prefer.
     */
    BreadcrumbService.prototype.selectPreferredByPath = function (path) {
        this.preferredPath = path;
        this.refresh();
    };
    BreadcrumbService.prototype.sortByPreferredPath = function (breadcrumbs) {
        var _this = this;
        if (this.preferredPath) {
            return breadcrumbs.sort(function (bc) {
                return bc.items.find(function (item) { return !!item.path.match(_this.preferredPath); }) ? -1 : 1;
            });
        }
        return breadcrumbs;
    };
    BreadcrumbService.ctorParameters = function () { return [
        { type: Injector },
        { type: Router }
    ]; };
    BreadcrumbService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function BreadcrumbService_Factory() { return new BreadcrumbService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Router)); }, token: BreadcrumbService, providedIn: "root" });
    BreadcrumbService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        })
    ], BreadcrumbService);
    return BreadcrumbService;
}(StateService));
export { BreadcrumbService };
