// @c8y/ngx-components/esm5/core/select-modal/select-modal.component.js
import * as tslib_1 from "tslib";
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { gettext } from '../i18n/gettext';
import { BsModalRef } from 'ngx-bootstrap';
import { ModalSelectionMode } from './select-modal.model';
import { assign } from 'lodash-es';
import { Subject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
var SelectModalComponent = /** @class */ (function () {
    function SelectModalComponent(bsModalRef) {
        var _this = this;
        this.bsModalRef = bsModalRef;
        this.subTitle = gettext('Select from the list of items matching the device type');
        this.mode = ModalSelectionMode.MULTI;
        this.disableSelected = true;
        this.showFilter = true;
        this.areMoreEntries = false;
        this.result = new EventEmitter();
        this.search = new EventEmitter();
        this.selected = false;
        this.filterTerm = '';
        this.listItems = [];
        this.debouncer = new Subject();
        this._labels = { ok: gettext('Confirm'), cancel: gettext('Cancel') };
        this.debouncer.pipe(debounceTime(500)).subscribe(function (value) {
            _this.search.emit(value);
        });
    }
    Object.defineProperty(SelectModalComponent.prototype, "labels", {
        get: function () {
            return this._labels;
        },
        set: function (labels) {
            var _a = labels || {}, _b = _a.ok, ok = _b === void 0 ? this.labels.ok : _b, _c = _a.cancel, cancel = _c === void 0 ? this.labels.cancel : _c;
            this._labels = { ok: ok, cancel: cancel };
        },
        enumerable: true,
        configurable: true
    });
    SelectModalComponent.prototype.ngOnChanges = function (changes) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var itemsPromise, _a;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(changes.items && changes.items.currentValue)) return [3 /*break*/, 2];
                        itemsPromise = changes.items.currentValue.map(function (item) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            var _a, selected;
                            return tslib_1.__generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        _a = item;
                                        return [4 /*yield*/, item.options];
                                    case 1:
                                        _a.options = _b.sent();
                                        selected = item.options.find(function (option) { return option.selected; });
                                        if (selected) {
                                            item.selectedId = selected.obj.id;
                                            if (this.disableSelected) {
                                                item.options.map(function (option) { return assign(option, { disabled: true }); });
                                            }
                                        }
                                        return [2 /*return*/, item];
                                }
                            });
                        }); });
                        _a = this;
                        return [4 /*yield*/, Promise.all(itemsPromise)];
                    case 1:
                        _a.listItems = _b.sent();
                        _b.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    SelectModalComponent.prototype.updatePipe = function (filterTerm) {
        this.debouncer.next(filterTerm);
        this.filterTerm = filterTerm;
    };
    SelectModalComponent.prototype.updateChoice = function (_a) {
        var item = _a.item, id = _a.id;
        if (this.mode === 'single') {
            this.listItems.map(function (value) { return (value.selectedId = undefined); });
        }
        item.selectedId = id;
        this.selected = true;
    };
    SelectModalComponent.prototype.dismiss = function () {
        this.bsModalRef.hide();
    };
    SelectModalComponent.prototype.select = function () {
        var outputArray = this.getOutput();
        var output = this.mode === 'single' ? outputArray[0] : outputArray;
        this.result.emit(output);
        this.bsModalRef.hide();
    };
    SelectModalComponent.prototype.ngOnDestroy = function () {
        this.debouncer.complete();
        this.result.complete();
        this.search.complete();
    };
    SelectModalComponent.prototype.getOutput = function () {
        return this.listItems
            .filter(function (item) { return item.selectedId; })
            .map(function (item) { return item.options.find(function (option) { return item.selectedId === option.obj.id; }); })
            .map(function (selectedOption) { return selectedOption.obj; });
    };
    SelectModalComponent.ctorParameters = function () { return [
        { type: BsModalRef }
    ]; };
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "icon", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "title", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "subTitle", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "mode", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "disableSelected", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "showFilter", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "areMoreEntries", void 0);
    tslib_1.__decorate([
        Input()
    ], SelectModalComponent.prototype, "labels", null);
    tslib_1.__decorate([
        Output()
    ], SelectModalComponent.prototype, "result", void 0);
    tslib_1.__decorate([
        Output()
    ], SelectModalComponent.prototype, "search", void 0);
    SelectModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-select-modal',
            template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <span c8yIcon=\"{{ icon }}\"></span>\n    <h4 class=\"text-uppercase\">\n      {{ title | translate }}\n    </h4>\n  </div>\n  <div class=\"p-16 text-center separator-bottom\">\n    <p class=\"m-b-8\">{{ subTitle | translate }}</p>\n    <c8y-filter *ngIf=\"showFilter\" [icon]=\"'search'\" (onSearch)=\"updatePipe($event)\"></c8y-filter>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"p-l-16 p-r-16\">\n      <div class=\"panel m-t-8 m-b-8\" *ngIf=\"!items || items.length === 0\">\n        <div class=\"c8y-empty-state text-left\">\n          <h1 c8yIcon=\"{{ icon }} \" class=\"c8y-icon-duocolor\"></h1>\n          <p translate>No items to display.</p>\n        </div>\n      </div>\n    </div>\n    <c8y-list-group>\n      <c8y-li *ngFor=\"let item of listItems | selectModalFilterPipe: filterTerm\">\n        <c8y-li-icon>\n          <i c8yIcon=\"{{ icon }}\"></i>\n        </c8y-li-icon>\n\n        <c8y-li-body class=\"content-flex-30\">\n          <div class=\"col-9\">\n            <div *ngFor=\"let bodyPart of item.body\" [ngClass]=\"bodyPart.class\">\n              <c8y-highlight\n                [title]=\"bodyPart.value\"\n                [pattern]=\"filterTerm\"\n                [text]=\"bodyPart.value\"\n              ></c8y-highlight>\n            </div>\n          </div>\n\n          <div class=\"col-3 text-right\" *ngIf=\"item.additionalInformation\">\n            <div [ngClass]=\"item.additionalInformation.class\">\n              {{ item.additionalInformation.value }}\n            </div>\n          </div>\n        </c8y-li-body>\n\n        <c8y-li-collapse>\n          <c8y-list-group>\n            <c8y-li *ngFor=\"let option of item.options\">\n              <c8y-li-radio\n                [name]=\"mode === 'single' ? 'single' : item.groupId\"\n                [selected]=\"option.selected\"\n                (onSelect)=\"updateChoice({ item: item, id: option.obj.id })\"\n                [disabled]=\"option.disabled\"\n              >\n              </c8y-li-radio>\n              <c8y-li-body class=\"content-flex-50\">\n                <div\n                  *ngFor=\"let optionPart of option.body; let i = index\"\n                  [ngClass]=\"optionPart.class\"\n                >\n                  <c8y-highlight [pattern]=\"filterTerm\" [text]=\"optionPart.value\"></c8y-highlight>\n                </div>\n              </c8y-li-body>\n            </c8y-li>\n          </c8y-list-group>\n        </c8y-li-collapse>\n      </c8y-li>\n      <div *ngIf=\"areMoreEntries\">\n        <div class=\"alert alert-info m-t-16 m-r-8 m-l-8\" translate>\n          Some entries might not be shown. Try narrowing search criteria.\n        </div>\n      </div>\n    </c8y-list-group>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      title=\"{{ labels.cancel | translate }}\"\n      *ngIf=\"labels.cancel\"\n      class=\"btn btn-default\"\n      (click)=\"dismiss()\"\n    >\n      {{ labels.cancel | translate }}\n    </button>\n    <button\n      title=\"{{ labels.ok | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"select()\"\n      [disabled]=\"!selected\"\n    >\n      {{ labels.ok | translate }}\n    </button>\n  </div>\n</div>\n"
        })
    ], SelectModalComponent);
    return SelectModalComponent;
}());
export { SelectModalComponent };
