// @c8y/ngx-components/esm5/repository/software-device-tab/device-software-changes.component.js
import * as tslib_1 from "tslib";
import { Component, Output, Input, EventEmitter } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { filter } from 'lodash-es';
var DeviceSoftwareChangesComponent = /** @class */ (function () {
    function DeviceSoftwareChangesComponent() {
        this.clear = new EventEmitter();
        this.apply = new EventEmitter();
        this.changes$ = new BehaviorSubject([]);
        this.installations$ = this.changes$.pipe(map(function (changes) {
            return filter(changes, function (change) {
                var name = change.name, action = change.action;
                var relatedChanges = filter(changes, { name: name });
                return action === 'install' && relatedChanges.length === 1;
            });
        }));
        this.updates$ = this.changes$.pipe(map(function (changes) {
            return filter(changes, function (change) {
                var name = change.name, action = change.action;
                var relatedChanges = filter(changes, { name: name });
                return action === 'install' && relatedChanges.length === 2;
            });
        }));
        this.removals$ = this.changes$.pipe(map(function (changes) {
            return filter(changes, function (change) {
                var name = change.name, action = change.action;
                var relatedChanges = filter(changes, { name: name });
                return action === 'delete' && relatedChanges.length === 1;
            });
        }));
    }
    DeviceSoftwareChangesComponent.prototype.ngOnChanges = function (changes) {
        if (changes.changes) {
            this.changes$.next(changes.changes.currentValue);
        }
    };
    tslib_1.__decorate([
        Input()
    ], DeviceSoftwareChangesComponent.prototype, "changes", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceSoftwareChangesComponent.prototype, "changesInProgress", void 0);
    tslib_1.__decorate([
        Output()
    ], DeviceSoftwareChangesComponent.prototype, "clear", void 0);
    tslib_1.__decorate([
        Output()
    ], DeviceSoftwareChangesComponent.prototype, "apply", void 0);
    DeviceSoftwareChangesComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-software-changes',
            template: "<div class=\"card\">\n  <div class=\"card-header separator\">\n    <h4 class=\"card-title\" translate>Software changes</h4>\n  </div>\n\n  <div class=\"card-block\" *ngIf=\"(changes$ | async).length === 0; else changesList\">\n    <div class=\"c8y-empty-state text-center\">\n      <h1 class=\"c8y-icon c8y-icon-device-control c8y-icon-duocolor\"></h1>\n      <p>\n        <strong translate>No software changes.</strong> <br />\n        <small translate>Selected actions will be displayed here.</small>\n      </p>\n    </div>\n  </div>\n\n  <ng-template #changesList>\n    <fieldset *ngIf=\"(installations$ | async).length > 0\">\n      <div class=\"p-l-16 p-r-16\">\n        <div class=\"form-block legend\" translate>Installations`software`</div>\n      </div>\n      <c8y-device-software-list [softwareList]=\"installations$ | async\"></c8y-device-software-list>\n    </fieldset>\n\n    <fieldset *ngIf=\"(updates$ | async).length > 0\">\n      <div class=\"p-l-16 p-r-16\">\n        <div class=\"form-block legend\" translate>Updates`software,noun`</div>\n      </div>\n      <c8y-device-software-list [softwareList]=\"updates$ | async\"></c8y-device-software-list>\n    </fieldset>\n\n    <fieldset *ngIf=\"(removals$ | async).length > 0\">\n      <div class=\"p-l-16 p-r-16\">\n        <div class=\"form-block legend\" translate>Removals`software`</div>\n      </div>\n      <c8y-device-software-list [softwareList]=\"removals$ | async\"></c8y-device-software-list>\n    </fieldset>\n  </ng-template>\n\n  <div class=\"card-footer p-b-8\" *ngIf=\"(changes$ | async).length > 0\">\n    <div class=\"content-flex-27\">\n      <button\n        class=\"btn btn-default col-6 m-b-8\"\n        (click)=\"clear.emit()\"\n        title=\"{{ 'Clear changes' | translate }}\"\n        [disabled]=\"changesInProgress\"\n        translate\n      >\n        Clear changes\n      </button>\n      <button\n        class=\"btn btn-primary col-6 btn-block m-b-8\"\n        [ngClass]=\"{ 'btn-pending': changesInProgress }\"\n        [disabled]=\"changesInProgress\"\n        (click)=\"apply.emit()\"\n        title=\"{{ 'Apply changes' | translate }}\"\n        translate\n      >\n        Apply changes\n      </button>\n    </div>\n  </div>\n</div>\n"
        })
    ], DeviceSoftwareChangesComponent);
    return DeviceSoftwareChangesComponent;
}());
export { DeviceSoftwareChangesComponent };
