// @c8y/ngx-components/esm5/core/forms/ip-range-input-list.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
var IpRangeInputListComponent = /** @class */ (function () {
    function IpRangeInputListComponent() {
    }
    Object.defineProperty(IpRangeInputListComponent.prototype, "data", {
        get: function () {
            return this.ipRangeList;
        },
        set: function (list) {
            this.ipRangeList = list;
            if (this.ipRangeList.length === 0) {
                this.add();
            }
        },
        enumerable: true,
        configurable: true
    });
    IpRangeInputListComponent.prototype.add = function () {
        this.ipRangeList.push({ ip: '', cidr: undefined });
    };
    IpRangeInputListComponent.prototype.remove = function (index) {
        this.ipRangeList.splice(index, 1);
    };
    tslib_1.__decorate([
        Input()
    ], IpRangeInputListComponent.prototype, "data", null);
    IpRangeInputListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-ip-range-input-list',
            template: "<c8y-form-group class=\"m-t-8 m-b-8\">\n  <ul c8yInputGroupListContainer class=\"list-unstyled\">\n    <li class=\"m-b-8\" *ngFor=\"let item of data; let i = index\">\n      <c8y-input-group-list [index]=\"i\" (onAdd)=\"add()\" (onRemove)=\"remove($event)\">\n        <input\n          type=\"text\"\n          name=\"ipAddress\"\n          class=\"form-control m-r-4\"\n          placeholder=\"{{ 'IP range start, e.g.' | translate }} 192.168.0.1\"\n          style=\"flex-grow: 2;\"\n          [(ngModel)]=\"item.ip\"\n          [required]=\"true\"\n          #ipAddress\n        />\n        <input\n          type=\"number\"\n          name=\"cidr\"\n          [min]=\"0\"\n          class=\"form-control\"\n          placeholder=\"{{ 'e.g.' | translate }} 20\"\n          [(ngModel)]=\"item.cidr\"\n          [required]=\"true\"\n          #cidr\n        />\n      </c8y-input-group-list>\n    </li>\n  </ul>\n</c8y-form-group>\n"
        })
    ], IpRangeInputListComponent);
    return IpRangeInputListComponent;
}());
export { IpRangeInputListComponent };
