// @c8y/ngx-components/esm5/trusted-certificates/trusted-certificates-navigation.factory.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { TenantOptionsService } from '@c8y/client';
import { TrustedCertificatesGuard } from './trusted-certificates.guard';
var TrustedCertificatesNavigationFactory = /** @class */ (function () {
    function TrustedCertificatesNavigationFactory(tenantOptionsService, trustedCertificatesGuard) {
        this.tenantOptionsService = tenantOptionsService;
        this.trustedCertificatesGuard = trustedCertificatesGuard;
        this.nav = [];
    }
    TrustedCertificatesNavigationFactory.prototype.get = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var isFeatureEnabled;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.trustedCertificatesGuard.isFeatureEnabled()];
                    case 1:
                        isFeatureEnabled = _a.sent();
                        if (isFeatureEnabled && this.nav.length === 0) {
                            this.nav.push(new NavigatorNode({
                                label: gettext('Trusted certificates'),
                                icon: 'certificate',
                                path: '/trusted-certificates',
                                parent: gettext('Management'),
                                priority: 100
                            }));
                        }
                        return [2 /*return*/, this.nav];
                }
            });
        });
    };
    TrustedCertificatesNavigationFactory.ctorParameters = function () { return [
        { type: TenantOptionsService },
        { type: TrustedCertificatesGuard }
    ]; };
    TrustedCertificatesNavigationFactory = tslib_1.__decorate([
        Injectable()
    ], TrustedCertificatesNavigationFactory);
    return TrustedCertificatesNavigationFactory;
}());
export { TrustedCertificatesNavigationFactory };
