// @c8y/ngx-components/esm5/core/date-picker/date-picker.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
var DatePickerComponent = /** @class */ (function () {
    function DatePickerComponent() {
        this.onDateSelected = new EventEmitter();
    }
    DatePickerComponent.prototype.ngOnInit = function () {
        this.fgDatePicker = new FormGroup({
            dateFrom: new FormControl(),
            dateTo: new FormControl()
        });
    };
    DatePickerComponent.prototype.filter = function () {
        this.onDateSelected.emit(this.fgDatePicker.value);
    };
    DatePickerComponent.prototype.clearFilter = function () {
        this.fgDatePicker.setValue({
            dateFrom: null,
            dateTo: null
        });
        this.onDateSelected.emit(null);
    };
    tslib_1.__decorate([
        Output()
    ], DatePickerComponent.prototype, "onDateSelected", void 0);
    DatePickerComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-date-picker',
            template: "<form [formGroup]=\"fgDatePicker\">\n  <c8y-form-group class=\"datepicker\">\n    <input\n      formControlName=\"dateFrom\"\n      class=\"form-control\"\n      placeholder=\"{{ 'Date from' | translate }}\"\n      bsDatepicker\n      [bsConfig]=\"{ customTodayClass: 'today' }\"\n    />\n  </c8y-form-group>\n  <c8y-form-group class=\"datepicker\">\n    <input\n      formControlName=\"dateTo\"\n      class=\"form-control\"\n      placeholder=\"{{ 'Date to' | translate }}\"\n      bsDatepicker\n      [bsConfig]=\"{ customTodayClass: 'today' }\"\n    />\n  </c8y-form-group>\n  <button\n    *ngIf=\"fgDatePicker.get('dateFrom').value || fgDatePicker.get('dateTo').value\"\n    title=\"{{ 'Apply' | translate }}\"\n    type=\"submit\"\n    class=\"btn btn-link\"\n    (click)=\"filter()\"\n  >\n    <i c8y-icon=\"filter\"></i>\n    {{ 'Apply`filters`' | translate }}\n  </button>\n  <button\n    *ngIf=\"fgDatePicker.get('dateFrom').value || fgDatePicker.get('dateTo').value\"\n    title=\"{{ 'Clear' | translate }}\"\n    type=\"button\"\n    class=\"btn btn-link\"\n    (click)=\"clearFilter()\"\n  >\n    <i c8y-icon=\"times\"></i>\n    {{ 'Clear`filters`' | translate }}\n  </button>\n</form>\n"
        })
    ], DatePickerComponent);
    return DatePickerComponent;
}());
export { DatePickerComponent };
