// @c8y/ngx-components/esm5/repository/configuration-device-tab/device-configuration-list.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var DeviceConfigurationListComponent = /** @class */ (function () {
    function DeviceConfigurationListComponent() {
        this.configSelected = new EventEmitter();
        this.filterTerm = '';
    }
    DeviceConfigurationListComponent.prototype.showConfigurationTypePreview = function (config) {
        this.selectedConfig = config;
        this.configSelected.emit(config);
    };
    DeviceConfigurationListComponent.prototype.updatePipe = function (filterTerm) {
        this.filterTerm = filterTerm;
    };
    tslib_1.__decorate([
        Input()
    ], DeviceConfigurationListComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceConfigurationListComponent.prototype, "itemIcon", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceConfigurationListComponent.prototype, "emptyState", void 0);
    tslib_1.__decorate([
        Input()
    ], DeviceConfigurationListComponent.prototype, "isFilterEnabled", void 0);
    tslib_1.__decorate([
        Output()
    ], DeviceConfigurationListComponent.prototype, "configSelected", void 0);
    DeviceConfigurationListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-device-configuration-list',
            template: "<div class=\"p-l-16 m-b-8\" style=\"margin-top: -16px;\" *ngIf=\"isFilterEnabled\">\n  <c8y-filter [icon]=\"'search'\" (onSearch)=\"updatePipe($event)\"></c8y-filter>\n</div>\n\n<!-- EMPTY STATE -->\n<div class=\"c8y-empty-state text-left\" *ngIf=\"items?.length === 0\">\n  <h1 [c8yIcon]=\"emptyState.icon\"></h1>\n  <p>\n    <strong>{{ emptyState.title | translate }}</strong\n    ><br />\n    <small>{{ emptyState.text | translate }}</small>\n  </p>\n</div>\n\n<!-- CONFIGURATIONS AVAILABLE -->\n<div class=\"c8y-nav-stacked\">\n  <div\n    class=\"c8y-stacked-item flex-row\"\n    [class.active]=\"config === selectedConfig\"\n    *ngFor=\"let config of items | configurationFilterPipe: filterTerm\"\n    (click)=\"showConfigurationTypePreview(config)\"\n  >\n    <div class=\"list-item-icon\">\n      <i [c8yIcon]=\"itemIcon\"></i>\n    </div>\n    <div class=\"list-item-body text-truncate\">\n      <div class=\"d-flex\">\n        <span class=\"text-truncate\" title=\"{{ config.name }}\">{{ config.name }}</span>\n        <span class=\"text-label-small flex-item-right flex-item-middle\">{{ config.deviceType }}</span>\n      </div>\n    </div>\n  </div>\n</div>\n\n<!-- for Carlos: config.configurationType to differentiate whether a config matches configuration type. -->\n"
        })
    ], DeviceConfigurationListComponent);
    return DeviceConfigurationListComponent;
}());
export { DeviceConfigurationListComponent };
