// @c8y/ngx-components/esm5/repository/configuration-device-tab/save-to-repository.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { RepositoryType } from '../repository.model';
import { AlertService, gettext } from '@c8y/ngx-components';
import { RepositoryService } from '../repository.service';
var SaveToRepositoryComponent = /** @class */ (function () {
    function SaveToRepositoryComponent(modal, alertService, repositoryService) {
        var _this = this;
        this.modal = modal;
        this.alertService = alertService;
        this.repositoryService = repositoryService;
        this.result = new Promise(function (resolve, reject) {
            _this._save = resolve;
            _this._cancel = reject;
        });
    }
    SaveToRepositoryComponent.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var configSnapshotData, ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        configSnapshotData = {
                            selected: {
                                configurationType: this.configSnapshot.configurationType
                            },
                            version: this.configSnapshot.name,
                            deviceType: this.configSnapshot.deviceType,
                            description: this.configSnapshot.description,
                            binary: {
                                file: new File([this.configSnapshot.binary], this.configSnapshot.name)
                            }
                        };
                        return [4 /*yield*/, this.repositoryService.save(configSnapshotData, RepositoryType.CONFIGURATION)];
                    case 1:
                        _a.sent();
                        this.alertService.success(gettext('Configuration saved.'));
                        this._save();
                        return [3 /*break*/, 3];
                    case 2:
                        ex_1 = _a.sent();
                        this.alertService.addServerFailure(ex_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    SaveToRepositoryComponent.prototype.close = function () {
        this._cancel();
        this.modal.hide();
    };
    SaveToRepositoryComponent.ctorParameters = function () { return [
        { type: BsModalRef },
        { type: AlertService },
        { type: RepositoryService }
    ]; };
    SaveToRepositoryComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-save-config-to-configuration-repository',
            template: "<div class=\"modal-header bg-primary text-center text-white\">\n  <div style=\"font-size: 62px;\">\n    <span c8yIcon=\"gears\"></span>\n  </div>\n  <h4 class=\"text-uppercase\" translate>\n    Save configuration\n  </h4>\n</div>\n<div class=\"modal-body\">\n  <form #saveConfigurationSnapshot=\"ngForm\" class=\"p-t-24\">\n    <c8y-form-group>\n      <label translate for=\"name\">Name</label>\n      <input\n        id=\"name\"\n        type=\"text\"\n        class=\"form-control\"\n        autocomplete=\"off\"\n        name=\"name\"\n        [(ngModel)]=\"configSnapshot.name\"\n        required\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"deviceType\">Device type</label>\n      <input\n        id=\"deviceType\"\n        class=\"form-control\"\n        rows=\"6\"\n        name=\"deviceType\"\n        [(ngModel)]=\"configSnapshot.deviceType\"\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"description\">Description</label>\n      <input\n        type=\"text\"\n        id=\"description\"\n        class=\"form-control\"\n        maxlength=\"254\"\n        autocomplete=\"off\"\n        name=\"description\"\n        [(ngModel)]=\"configSnapshot.description\"\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"configurationType\">Configuration type</label>\n      <input\n        id=\"configurationType\"\n        class=\"form-control\"\n        rows=\"6\"\n        name=\"configurationType\"\n        [(ngModel)]=\"configSnapshot.configurationType\"\n      />\n    </c8y-form-group>\n  </form>\n</div>\n<div class=\"modal-footer\">\n  <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\" translate>\n    Cancel\n  </button>\n\n  <button\n    title=\"{{ 'Add certificate' | translate }}\"\n    class=\"btn btn-primary\"\n    (click)=\"save()\"\n    [disabled]=\"saveConfigurationSnapshot.form.invalid\"\n    translate\n  >\n    Save\n  </button>\n</div>\n"
        })
    ], SaveToRepositoryComponent);
    return SaveToRepositoryComponent;
}());
export { SaveToRepositoryComponent };
