// @c8y/ngx-components/esm2015/core/audit-log/audit-log.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { AuditService, IAuditRecord, IResultList } from '@c8y/client';
let AuditLogComponent = class AuditLogComponent {
    constructor(audit) {
        this.audit = audit;
    }
    ngOnInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.source) {
                this.records = yield this.audit.list({
                    withTotalPages: true,
                    source: this.source
                });
            }
        });
    }
};
AuditLogComponent.ctorParameters = () => [
    { type: AuditService }
];
tslib_1.__decorate([
    Input()
], AuditLogComponent.prototype, "source", void 0);
AuditLogComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-audit-log',
        template: "<div class=\"inner-scroll\">\n  <ul class=\"timeline-list list-condensed m-t-16\">\n    <li class=\"timeline-list-item flex-row\" *c8yFor=\"let record of records\">\n      <small class=\"timeline-item-date text-right\">\n        <span>{{ record.time | c8yDate }}</span>\n      </small>\n      <div class=\"timeline-item-content flex-row\">\n        <div class=\"list-item-body\">\n          <div class=\"row text-break-word\">\n            <div class=\"col-md-5\">\n              <span class=\"text-truncate\">{{ record.activity | translate }}</span>\n              <small class=\"text-muted\">{{ record.user }}</small>\n            </div>\n            <div class=\"col-md-7\">\n              <div class=\"small text-truncate-wrap\">{{ record.text | translate }}</div>\n            </div>\n          </div>\n        </div>\n      </div>\n    </li>\n  </ul>\n</div>\n"
    })
], AuditLogComponent);
export { AuditLogComponent };
