// @c8y/ngx-components/esm2015/core/authentication/password-strength.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { PasswordService } from './password.service';
let PasswordStrengthComponent = class PasswordStrengthComponent {
    constructor(passwordStrengthCheckerService) {
        this.passwordStrengthCheckerService = passwordStrengthCheckerService;
        this.strength = new EventEmitter();
        this.password = '';
    }
    set password(password) {
        if (typeof password === 'string') {
            this.model = this.passwordStrengthCheckerService.getStrengthColor(password);
            this.strength.emit(this.model);
        }
    }
};
PasswordStrengthComponent.ctorParameters = () => [
    { type: PasswordService }
];
tslib_1.__decorate([
    Input('password')
], PasswordStrengthComponent.prototype, "password", null);
tslib_1.__decorate([
    Output()
], PasswordStrengthComponent.prototype, "strength", void 0);
PasswordStrengthComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-password-strength',
        template: "<div class=\"password-strength\">\r\n    <span class=\"password-strength-label\" translate>Password strength</span>\r\n    <span class=\"pull-right ng-binding\" [ngStyle]=\"{'color': model.color}\">{{model.description | translate}}</span>\r\n    <div>\r\n        <div class=\"password-{{model.colorName}}\" [ngStyle]=\"{'background-color': model.color}\"></div>\r\n    </div>\r\n</div>"
    })
], PasswordStrengthComponent);
export { PasswordStrengthComponent };
