// @c8y/ngx-components/esm2015/core/authentication/password-confirm.directive.js
var PasswordConfirm_1;
import * as tslib_1 from "tslib";
import { Directive, Attribute } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
let PasswordConfirm = PasswordConfirm_1 = class PasswordConfirm {
    constructor(passwordConfirm) {
        this.passwordConfirm = passwordConfirm;
    }
    validate(abControl) {
        const value = abControl.value;
        const controlToCompareWith = abControl.root.get(this.passwordConfirm);
        if (controlToCompareWith && value !== controlToCompareWith.value) {
            return { passwordConfirm: true };
        }
        return null;
    }
};
PasswordConfirm.ctorParameters = () => [
    { type: String, decorators: [{ type: Attribute, args: ['passwordConfirm',] }] }
];
PasswordConfirm = PasswordConfirm_1 = tslib_1.__decorate([
    Directive({
        selector: '[passwordConfirm]',
        providers: [
            {
                provide: NG_VALIDATORS,
                useExisting: PasswordConfirm_1,
                multi: true
            }
        ]
    }),
    tslib_1.__param(0, Attribute('passwordConfirm'))
], PasswordConfirm);
export { PasswordConfirm };
