// @c8y/ngx-components/esm5/device-grid/columns/base.filtering-form-renderer.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef, ViewChild } from '@angular/core';
import { get, cloneDeep } from 'lodash-es';
import { FilteringFormRendererContext } from '@c8y/ngx-components';
import { SchemaFormComponent } from '../schema-form.component';
var BaseFilteringFormRendererComponent = /** @class */ (function () {
    function BaseFilteringFormRendererComponent(context) {
        this.context = context;
    }
    BaseFilteringFormRendererComponent.prototype.ngOnInit = function () {
        var column = this.context.property;
        this.schema = cloneDeep(column.filteringConfig.schema);
        this.form = cloneDeep(column.filteringConfig.form);
        this.model = cloneDeep(column.externalFilterQuery || {});
        this.options = { formDefaults: { ngModelOptions: { debounce: 100 } } };
    };
    BaseFilteringFormRendererComponent.prototype.canApply = function () {
        return get(this.schemaFormComponent, '$componentScope.vm.ngForm.$invalid');
    };
    BaseFilteringFormRendererComponent.prototype.applyFilter = function () {
        this.context.applyFilter({
            externalFilterQuery: this.model
        });
    };
    BaseFilteringFormRendererComponent.prototype.resetFilter = function () {
        this.context.resetFilter();
    };
    BaseFilteringFormRendererComponent.ctorParameters = function () { return [
        { type: FilteringFormRendererContext }
    ]; };
    tslib_1.__decorate([
        ViewChild(forwardRef(function () { return SchemaFormComponent; }), { static: false })
    ], BaseFilteringFormRendererComponent.prototype, "schemaFormComponent", void 0);
    BaseFilteringFormRendererComponent = tslib_1.__decorate([
        Component({
            template: "<c8y-schema-form\n  [schema]=\"schema\"\n  [form]=\"form\"\n  [model]=\"model\"\n  [options]=\"options\"\n></c8y-schema-form>\n<div class=\"data-grid__dropdown__footer d-flex separator-top\">\n  <button\n    class=\"btn btn-default btn-sm m-r-8 flex-grow\"\n    (click)=\"resetFilter()\"\n    title=\"{{ 'Reset' | translate }}\"\n    translate\n  >\n    Reset\n  </button>\n\n  <button\n    class=\"btn btn-primary btn-sm flex-grow\"\n    (click)=\"applyFilter()\"\n    [disabled]=\"canApply()\"\n    title=\"{{ 'Apply' | translate }}\"\n    translate\n  >\n    Apply\n  </button>\n</div>\n"
        })
    ], BaseFilteringFormRendererComponent);
    return BaseFilteringFormRendererComponent;
}());
export { BaseFilteringFormRendererComponent };
