// @c8y/ngx-components/esm5/upgrade/ng1/navigator-node-root-legacy.js
import * as tslib_1 from "tslib";
import { NavigatorNodeRoot } from '@c8y/ngx-components';
import { assign } from 'lodash-es';
var NavigatorNodeRootLegacy = /** @class */ (function (_super) {
    tslib_1.__extends(NavigatorNodeRootLegacy, _super);
    function NavigatorNodeRootLegacy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NavigatorNodeRootLegacy.prototype.addRoot = function (nodeData) {
        var duplicate;
        if (nodeData.path === '') {
            nodeData.path = '/';
        }
        nodeData.label = nodeData.name;
        if (typeof nodeData.parent === 'object') {
            nodeData.parent.label = nodeData.parent.name;
        }
        if (nodeData.preventDuplicates) {
            duplicate = this.find(function (_a) {
                var path = _a.path, parents = _a.parents, label = _a.label;
                return path === nodeData.path &&
                    (label === nodeData.label) &&
                    parents.some(function (p) { return p.label === nodeData.parent; });
            });
            if (duplicate) {
                duplicate.routerLinkExact = false;
            }
        }
        return duplicate || _super.prototype.addRoot.call(this, nodeData);
    };
    NavigatorNodeRootLegacy.prototype.createNode = function (node) {
        var newNode = _super.prototype.createNode.call(this, node);
        var update = newNode.update.bind(newNode);
        var root = this;
        return Object.defineProperties(assign(newNode, {
            realName: newNode.name || newNode.label,
            _parent: true,
            update: function (data) {
                if (this._parent === undefined) {
                    // _parent was deleted somene instead to put this in root
                    root.addRoot(this);
                    // put it back so it can be deleted again
                    this._parent = true;
                }
                update(data);
            },
            addChild: function (nodeChild) {
                this.add(root.createNode(nodeChild));
            }
        }), {
            label: {
                get: function () {
                    return this.realName || '';
                },
                set: function (name) {
                    this.realName = name;
                }
            },
            name: {
                get: function () {
                    return this.realName || '';
                },
                set: function (name) {
                    this.realName = name;
                }
            },
            show: {
                get: function () {
                    return !this.hidden;
                },
                set: function (show) {
                    this.hidden = !show;
                }
            }
        });
    };
    return NavigatorNodeRootLegacy;
}(NavigatorNodeRoot));
export { NavigatorNodeRootLegacy };
