// @c8y/ngx-components/esm5/operations/bulk/list/single-operations-list.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { StatusFilterComponent } from '../../status-filter/status-filter.component';
import { BulkOperationsService } from '../bulk-operations.service';
import { SINGLE_OP_STATUS_NAMES, SINGLE_OP_STATUS_PROPERTIES } from './single-operation-item.model';
var SingleOperationsListComponent = /** @class */ (function () {
    function SingleOperationsListComponent(bulkOperationsService) {
        this.bulkOperationsService = bulkOperationsService;
        this.readOnly = false;
        this.onRetryFailedOperations = new EventEmitter();
        this.singleOperationStatusProperties = SINGLE_OP_STATUS_PROPERTIES;
        this.singleOperationStatusNames = SINGLE_OP_STATUS_NAMES;
    }
    SingleOperationsListComponent.prototype.ngAfterViewInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                if (this.statusFilter) {
                    this.filterOperationsByType();
                }
                return [2 /*return*/];
            });
        });
    };
    SingleOperationsListComponent.prototype.ngOnChanges = function (changes) {
        if (changes.bulkOperation && !changes.bulkOperation.firstChange) {
            this.filterOperationsByType();
        }
    };
    SingleOperationsListComponent.prototype.filterOperationsByType = function () {
        if (this.statusFilter) {
            var typeFilter = this.getFilterTypeOfSingleOperations();
            this.statusFilter.preset(typeFilter === this.singleOperationStatusNames.ALL
                ? []
                : [{ label: typeFilter }]);
        }
        else {
            setTimeout(this.filterOperationsByType.bind(this));
        }
    };
    SingleOperationsListComponent.prototype.getFilterTypeOfSingleOperations = function () {
        var progress = this.bulkOperation.progress;
        if (progress.failed > 0) {
            return this.singleOperationStatusNames.FAILED;
        }
        else if (progress.failed === 0 && progress.successful === 0 && progress.pending === 0) {
            return this.singleOperationStatusNames.EXECUTING;
        }
        else if (progress.failed === 0 && progress.successful === 0 && progress.executing === 0) {
            return this.singleOperationStatusNames.PENDING;
        }
        else if (progress.failed === 0 && progress.pending === 0 && progress.executing === 0) {
            return this.singleOperationStatusNames.SUCCESSFUL;
        }
        return this.singleOperationStatusNames.ALL;
    };
    SingleOperationsListComponent.prototype.getOperationsByStatus = function (filter) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, this.bulkOperationsService.getSingleOperationsByStatus(filter[0] && filter[0].label, this.bulkOperation.id)];
                    case 1:
                        _a.singleOperations = _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    SingleOperationsListComponent.prototype.retryBulkOperation = function () {
        this.onRetryFailedOperations.emit(this.bulkOperation);
    };
    SingleOperationsListComponent.ctorParameters = function () { return [
        { type: BulkOperationsService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], SingleOperationsListComponent.prototype, "bulkOperation", void 0);
    tslib_1.__decorate([
        Input()
    ], SingleOperationsListComponent.prototype, "readOnly", void 0);
    tslib_1.__decorate([
        Output()
    ], SingleOperationsListComponent.prototype, "onRetryFailedOperations", void 0);
    tslib_1.__decorate([
        ViewChild(StatusFilterComponent, { static: false })
    ], SingleOperationsListComponent.prototype, "statusFilter", void 0);
    SingleOperationsListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-single-operations-list',
            template: "<ng-container\n  *ngIf=\"\n    bulkOperation.progress.failed ||\n    bulkOperation.progress.executing ||\n    bulkOperation.progress.pending ||\n    bulkOperation.progress.successful\n  \"\n>\n  <div class=\"legend form-block\" translate>\n    Operations\n  </div>\n  <div class=\"flex-row p-b-8\">\n    <div class=\"flex-grow p-b-8\">\n      <span class=\"m-r-4\" translate>Filter by status</span>\n      <c8y-status-filter\n        small\n        [options]=\"singleOperationStatusProperties\"\n        (onFilterChanged)=\"getOperationsByStatus($event)\"\n      ></c8y-status-filter>\n    </div>\n    <div class=\"p-b-8\" *ngIf=\"!readOnly && bulkOperation.progress.failed > 0\">\n      <button\n        class=\"btn btn-navbar\"\n        (click)=\"retryBulkOperation()\"\n        title=\"{{ 'Retry failed operations' | translate }}\"\n      >\n        <i class=\"fa fa-repeat m-r-4\"></i> <span translate>Retry failed operations</span>\n      </button>\n    </div>\n  </div>\n  <div class=\"inner-scroll\">\n    <c8y-list-group>\n      <div class=\"d-contents\" *c8yFor=\"let operation of singleOperations; loadMore: 'auto'\">\n        <c8y-single-operation-item\n          [operation]=\"operation\"\n          [readOnly]=\"readOnly\"\n          style=\"display: contents;\"\n        >\n        </c8y-single-operation-item>\n      </div>\n\n      <c8y-li *ngIf=\"singleOperations && singleOperations.data.length === 0\">\n        <c8y-li-body>\n          <div class=\"c8y-empty-state text-center\">\n            <h2 class=\"c8y-icon c8y-icon-energy c8y-icon-duocolor m-b-16\"></h2>\n            <div>\n              <h3 translate>No single operations of the selected status to display.</h3>\n              <p class=\"m-b-16\" translate>Single operations will be displayed here</p>\n            </div>\n          </div>\n        </c8y-li-body>\n      </c8y-li>\n    </c8y-list-group>\n  </div>\n</ng-container>\n"
        })
    ], SingleOperationsListComponent);
    return SingleOperationsListComponent;
}());
export { SingleOperationsListComponent };
