// @c8y/ngx-components/esm5/repository/firmware/add-firmware-patch-modal.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild, Output, EventEmitter } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { RepositoryService } from '../repository.service';
import { RepositoryType } from '../repository.model';
import { AlertService, gettext } from '@c8y/ngx-components';
import { isUndefined } from 'lodash-es';
import { BehaviorSubject, of, from, pipe, merge } from 'rxjs';
import { debounceTime, tap, switchMap, distinctUntilChanged, shareReplay, map } from 'rxjs/operators';
var AddFirmwarePatchModalComponent = /** @class */ (function () {
    function AddFirmwarePatchModalComponent(modal, repository, alert) {
        var _this = this;
        this.modal = modal;
        this.repository = repository;
        this.alert = alert;
        this.saved = new EventEmitter();
        this.model = {
            selected: undefined,
            dependency: null,
            patchVersion: undefined,
            binary: {
                file: undefined,
                url: undefined
            }
        };
        this.firmwareInput$ = new BehaviorSubject('');
        this.firmwares$ = this.firmwareInput$.pipe(debounceTime(300), distinctUntilChanged(), switchMap(function (searchStr) {
            return from(_this.repository.listRepositoryEntries(RepositoryType.FIRMWARE, {
                partialName: searchStr,
                skipLegacy: true
            }));
        }), shareReplay(1));
        this.firmwareSelected$ = new BehaviorSubject(null);
        this.patchDependencyInput$ = new BehaviorSubject('');
        this.saving = false;
        this.firmwarePreselected = false;
        this.baseVersions$ = merge(this.firmwareInput$.pipe(tap(function () {
            _this.model.dependency = null;
            if (_this.form) {
                _this.form.form.get('patchDependency').reset();
            }
        }), switchMap(function () { return of(null); })), this.firmwareSelected$).pipe(switchMap(function (selectedFirmware) {
            return selectedFirmware ? _this.repository.listBaseVersions(selectedFirmware) : of(null);
        }), shareReplay(1));
        this.baseVersionsFilterPipe = pipe(switchMap(function (data) {
            return _this.patchDependencyInput$.pipe(map(function (partialVersion) {
                return data.filter(function (mo) {
                    var version = mo.c8y_Firmware.version.toLowerCase();
                    return (partialVersion.length === 0 || version.indexOf(partialVersion.toLowerCase()) > -1);
                });
            }));
        }));
    }
    AddFirmwarePatchModalComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.setInitialState();
                return [2 /*return*/];
            });
        });
    };
    AddFirmwarePatchModalComponent.prototype.setInitialState = function () {
        if (this.model.selected) {
            this.firmwarePreselected = true;
            this.firmwareSelected$.next(this.model.selected);
        }
    };
    AddFirmwarePatchModalComponent.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                this.saving = true;
                this.repository
                    .create(this.model, RepositoryType.FIRMWARE)
                    .then(function (savedFirmware) {
                    _this.successMsg();
                    _this.saving = false;
                    _this.saved.next(savedFirmware);
                    _this.cancel();
                })
                    .catch(function (e) {
                    _this.saving = false;
                    _this.saved.error(e);
                    _this.cancel();
                });
                return [2 /*return*/];
            });
        });
    };
    AddFirmwarePatchModalComponent.prototype.successMsg = function () {
        var msg = gettext('Firmware patch added.');
        this.alert.success(msg);
    };
    AddFirmwarePatchModalComponent.prototype.cancel = function () {
        this.modal.hide();
        this.saved.complete();
    };
    AddFirmwarePatchModalComponent.prototype.onFile = function (dropped) {
        if (!isUndefined(dropped.url)) {
            this.model.binary = {
                url: dropped.url
            };
            return;
        }
        else if (!isUndefined(dropped.droppedFiles)) {
            this.model.binary = {
                file: dropped.droppedFiles[0].file
            };
            return;
        }
        else {
            this.model.binary = {
                file: undefined,
                url: undefined
            };
        }
    };
    AddFirmwarePatchModalComponent.ctorParameters = function () { return [
        { type: BsModalRef },
        { type: RepositoryService },
        { type: AlertService }
    ]; };
    tslib_1.__decorate([
        Output()
    ], AddFirmwarePatchModalComponent.prototype, "saved", void 0);
    tslib_1.__decorate([
        ViewChild('dropdown', { static: false })
    ], AddFirmwarePatchModalComponent.prototype, "dropdown", void 0);
    tslib_1.__decorate([
        ViewChild('firmwarePatchForm', { static: false })
    ], AddFirmwarePatchModalComponent.prototype, "form", void 0);
    AddFirmwarePatchModalComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-add-firmware-patch-modal.component',
            template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i [c8yIcon]=\"'c8y-firmware'\"></i>\n    <h4 translate>\n      Add firmware patch\n    </h4>\n  </div>\n  <div class=\"p-16 text-center separator-bottom\">\n    <p class=\"lead m-0\" translate>\n      Select a firmware version\n    </p>\n  </div>\n\n  <form class=\"d-contents\" autocomplete=\"off\" #firmwarePatchForm=\"ngForm\" (ngSubmit)=\"save()\">\n    <div class=\"modal-inner-scroll\">\n      <div class=\"modal-body\">\n        <div [hidden]=\"firmwarePreselected\">\n          <c8y-form-group>\n            <label for=\"firmwareName\" translate>Firmware</label>\n            <c8y-typeahead\n              [ngModel]=\"model.selected\"\n              name=\"firmwareName\"\n              placeholder=\"{{ 'Select or enter' | translate }}\"\n              (onSearch)=\"firmwareInput$.next($event)\"\n              [allowFreeEntries]=\"false\"\n              [required]=\"true\"\n            >\n              <c8y-li\n                *c8yFor=\"let firmware of firmwares$ | async; loadMore: 'auto'\"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"model.selected = firmware; firmwareSelected$.next(firmware)\"\n                [active]=\"model.selected === firmware\"\n              >\n                <c8y-highlight\n                  [text]=\"firmware.name || '--'\"\n                  [pattern]=\"firmwareInput$ | async\"\n                ></c8y-highlight>\n              </c8y-li>\n            </c8y-typeahead>\n            <c8y-messages\n              ><c8y-message\n                name=\"notExisting\"\n                [text]=\"'Select one of the existing firmwares.' | translate\"\n              ></c8y-message>\n            </c8y-messages>\n          </c8y-form-group>\n        </div>\n\n        <c8y-form-group>\n          <label for=\"patchDependency\" class=\"m-r-8\" translate>Version</label>\n          <c8y-typeahead\n            [ngModel]=\"model.dependency\"\n            name=\"patchDependency\"\n            placeholder=\"{{ 'Select or enter' | translate }}\"\n            (onSearch)=\"patchDependencyInput$.next($event)\"\n            [displayProperty]=\"'c8y_Firmware.version'\"\n            [allowFreeEntries]=\"false\"\n            [disabled]=\"\n              (baseVersions$ | async) === null || (baseVersions$ | async)?.data.length === 0\n            \"\n            [required]=\"true\"\n          >\n            <c8y-li\n              *c8yFor=\"\n                let baseVersion of baseVersions$;\n                loadMore: 'auto';\n                pipe: baseVersionsFilterPipe\n              \"\n              class=\"p-l-8 p-r-8 c8y-list__item--link\"\n              (click)=\"model.dependency = baseVersion\"\n              [active]=\"model.dependency === baseVersion\"\n            >\n              <c8y-highlight\n                [text]=\"baseVersion.c8y_Firmware.version || '--'\"\n                [pattern]=\"patchDependencyInput$ | async\"\n              ></c8y-highlight>\n            </c8y-li>\n          </c8y-typeahead>\n          <c8y-messages\n            ><c8y-message\n              name=\"notExisting\"\n              [text]=\"'Select one of the existing versions.' | translate\"\n            ></c8y-message>\n          </c8y-messages>\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label for=\"patchVersion\" translate>Patch</label>\n          <input\n            id=\"patchVersion\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            name=\"patchVersion\"\n            [(ngModel)]=\"model.patchVersion\"\n            placeholder=\"{{ 'e.g.' | translate }} 1.0.0\"\n            required\n          />\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <div class=\"legend form-block top-m-xxl\" translate>Patch file</div>\n          <c8y-file-picker [maxAllowedFiles]=\"1\" (onFilesPicked)=\"onFile($event)\">\n          </c8y-file-picker>\n        </c8y-form-group>\n      </div>\n    </div>\n    <div class=\"modal-footer\">\n      <button\n        title=\"{{ 'Cancel' | translate }}\"\n        class=\"btn btn-default\"\n        type=\"button\"\n        (click)=\"cancel()\"\n        [disabled]=\"saving\"\n        translate\n      >\n        Cancel\n      </button>\n      <button\n        title=\"{{ 'Add firmware patch' | translate }}\"\n        class=\"btn btn-primary\"\n        [ngClass]=\"{ 'btn-pending': saving }\"\n        type=\"submit\"\n        [disabled]=\"\n          !firmwarePatchForm.form.valid ||\n          firmwarePatchForm.form.pristine ||\n          (!model.binary?.url && !model.binary?.file) ||\n          saving\n        \"\n        translate\n      >\n        Add firmware patch\n      </button>\n    </div>\n  </form>\n</div>\n"
        })
    ], AddFirmwarePatchModalComponent);
    return AddFirmwarePatchModalComponent;
}());
export { AddFirmwarePatchModalComponent };
