// @c8y/ngx-components/esm5/repository/firmware/firmware-list.component.js
import * as tslib_1 from "tslib";
import { Router, ActivatedRoute } from '@angular/router';
import { Component } from '@angular/core';
import { BsModalService, ModalOptions } from 'ngx-bootstrap/modal';
import { TranslateService } from '@ngx-translate/core';
import { distinctUntilChanged, tap, debounceTime, switchMap, shareReplay, map } from 'rxjs/operators';
import { BehaviorSubject, combineLatest, from } from 'rxjs';
import { ModalService, gettext, Status, AlertService, memoize } from '@c8y/ngx-components';
import { RepositoryService } from '../repository.service';
import { RepositoryType } from '../repository.model';
import { AddFirmwareModalComponent } from './add-firmware-modal.component';
import { AddFirmwarePatchModalComponent } from './add-firmware-patch-modal.component';
import { property, get } from 'lodash-es';
var FirmwareListComponent = /** @class */ (function () {
    function FirmwareListComponent(repositoryService, modalService, bsModalService, translateService, alertService, router, activatedRoute) {
        var _this = this;
        this.repositoryService = repositoryService;
        this.modalService = modalService;
        this.bsModalService = bsModalService;
        this.translateService = translateService;
        this.alertService = alertService;
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.textFilter$ = new BehaviorSubject('');
        this.reload$ = new BehaviorSubject(null);
        this.reloading = false;
        this.firmwares$ = combineLatest(this.textFilter$.pipe(debounceTime(400), distinctUntilChanged()), this.reload$).pipe(tap(function () {
            _this.reloading = true;
        }), switchMap(function (_a) {
            var _b = tslib_1.__read(_a, 1), text = _b[0];
            return _this.getFirmwares(text);
        }), tap(function () {
            _this.reloading = false;
        }), shareReplay(1));
        this.canAddFirmwarePatch$ = from(this.repositoryService.listRepositoryEntries(RepositoryType.FIRMWARE, { skipLegacy: true })).pipe(map(function (_a) {
            var data = _a.data;
            return data.length > 0;
        }));
        this.isLegacy = this.repositoryService.isLegacyEntry.bind(this.repositoryService);
        this.DEVICE_TYPE_NOT_AVAILABLE = gettext('No device type available');
    }
    FirmwareListComponent.prototype.getFirmwares = function (partialName) {
        return this.repositoryService.listRepositoryEntries(RepositoryType.FIRMWARE, { partialName: partialName });
    };
    FirmwareListComponent.prototype.addFirmware = function () {
        var _this = this;
        var config = { class: 'modal-sm', ignoreBackdropClick: true };
        var modalRef = this.bsModalService.show(AddFirmwareModalComponent, config);
        modalRef.content.saved.subscribe(function (savedFirmware) { return _this.editFirmware(savedFirmware); });
    };
    FirmwareListComponent.prototype.addFirmwarePatch = function () {
        var _this = this;
        var config = { class: 'modal-sm', ignoreBackdropClick: true };
        var modalRef = this.bsModalService.show(AddFirmwarePatchModalComponent, config);
        modalRef.content.saved.subscribe(function (savedFirmware) { return _this.editFirmware(savedFirmware); });
    };
    FirmwareListComponent.prototype.editFirmware = function (firmware) {
        this.router.navigate([firmware.id], { relativeTo: this.activatedRoute });
    };
    FirmwareListComponent.prototype.deleteFirmware = function (firmware) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var title, body, labels, ex_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        title = gettext('Delete firmware');
                        body = "\n        " + this.translateService.instant(gettext('You are about to delete firmware "{{ name }}" with all its versions and patches.'), { name: firmware.name }) + "\n        " + this.translateService.instant(gettext('This operation is irreversible.')) + "\n        " + this.translateService.instant(gettext('Do you want to proceed?')) + "\n      ";
                        labels = {
                            ok: gettext('Delete')
                        };
                        return [4 /*yield*/, this.modalService.confirm(title, body, Status.DANGER, labels)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.repositoryService.delete(firmware)];
                    case 2:
                        _a.sent();
                        this.alertService.success(gettext('Firmware deleted.'));
                        this.reload$.next();
                        return [3 /*break*/, 4];
                    case 3:
                        ex_1 = _a.sent();
                        // only if not cancel from modal
                        if (ex_1) {
                            this.alertService.addServerFailure(ex_1);
                        }
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    FirmwareListComponent.prototype.getBaseVersionsCount$ = function (firmware) {
        var _this = this;
        return this.firmwares$.pipe(switchMap(function () { return _this.repositoryService.getBaseVersionsCount$(firmware); }), shareReplay(1));
    };
    FirmwareListComponent.prototype.getDeviceTypeTitle = function (firmware) {
        return get(firmware, 'c8y_Filter.type', this.translateService.instant(this.DEVICE_TYPE_NOT_AVAILABLE));
    };
    FirmwareListComponent.ctorParameters = function () { return [
        { type: RepositoryService },
        { type: ModalService },
        { type: BsModalService },
        { type: TranslateService },
        { type: AlertService },
        { type: Router },
        { type: ActivatedRoute }
    ]; };
    tslib_1.__decorate([
        memoize(property('id'))
    ], FirmwareListComponent.prototype, "getBaseVersionsCount$", null);
    FirmwareListComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-firmware-list',
            template: "<c8y-title>\n  {{ 'Firmware repository' | translate }}\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form\">\n  <div class=\"input-group input-group-search\">\n    <input\n      type=\"search\"\n      class=\"form-control\"\n      title=\"{{ 'Filter firmware\u2026' | translate }}\"\n      placeholder=\"{{ 'Filter firmware\u2026' | translate }}\"\n      [ngModel]=\"textFilter$ | async\"\n      (ngModelChange)=\"textFilter$.next($event)\"\n    />\n    <span class=\"input-group-addon\">\n      <i c8yIcon=\"filter\" *ngIf=\"(textFilter$ | async).length === 0\"></i>\n      <i\n        c8yIcon=\"times\"\n        class=\"text-muted\"\n        *ngIf=\"(textFilter$ | async).length > 0\"\n        (click)=\"textFilter$.next('')\"\n      ></i>\n    </span>\n  </div>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Add firmware' | translate }}\" (click)=\"addFirmware()\">\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add firmware' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    *ngIf=\"canAddFirmwarePatch$ | async\"\n    class=\"btn btn-link\"\n    title=\"{{ 'Add firmware patch' | translate }}\"\n    (click)=\"addFirmwarePatch()\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add firmware patch' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\" (click)=\"reload$.next()\">\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'fa-spin': reloading }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"c8y-empty-state text-center\" *ngIf=\"(firmwares$ | async)?.data.length === 0\">\n  <h1 c8yIcon=\"c8y-firmware\" class=\"c8y-icon-duocolor\"></h1>\n  <h3 translate>No firmware to display.</h3>\n  <p translate>Add a new firmware by clicking below.</p>\n  <p>\n    <button\n      class=\"btn btn-primary\"\n      title=\"{{ 'Add firmware' | translate }}\"\n      (click)=\"addFirmware()\"\n      translate\n    >\n      Add firmware\n    </button>\n  </p>\n</div>\n\n<c8y-list-group class=\"elevation-1\">\n  <c8y-li *c8yFor=\"let firmware of firmwares$ | async; let i = index; loadMore: 'auto'\">\n    <c8y-li-icon>\n      <i c8yIcon=\"c8y-firmware\"></i>\n    </c8y-li-icon>\n\n    <c8y-li-body class=\"content-flex-50\">\n      <div class=\"col-5 pointer\" (click)=\"editFirmware(firmware)\">\n        <div class=\"text-truncate\" title=\"{{ firmware.name }}\">\n          {{ firmware.name }}\n        </div>\n      </div>\n      <div class=\"col-5\">\n        <div class=\"text-truncate\" title=\"{{ 'Device type' | translate }}: {{ getDeviceTypeTitle(firmware) }}\">\n          <span class=\"text-label-small m-r-8\" translate>\n            Device type\n          </span>\n          <span *ngIf=\"firmware.c8y_Filter?.type; else emptyText\">\n            {{ firmware.c8y_Filter?.type }}\n          </span>\n          <ng-template #emptyText> --- </ng-template>\n        </div>\n      </div>\n      <div class=\"col-2\">\n        <span *ngIf=\"isLegacy(firmware)\" class=\"label label-warning flex-item-right-sm\">\n          <span translate>Legacy</span>\n        </span>\n\n        <span *ngIf=\"!isLegacy(firmware)\">\n          <span *ngIf=\"(getBaseVersionsCount$(firmware) | async) === null\">\n            <span class=\"label label-info\">\n              <i c8yIcon=\"circle-o-notch\" class=\"fa-spin\"></i>\n            </span>\n          </span>\n          <span *ngIf=\"(getBaseVersionsCount$(firmware) | async) !== null\">\n            <span [ngPlural]=\"getBaseVersionsCount$(firmware) | async\">\n              <ng-template ngPluralCase=\"=0\">\n                <span class=\"label label-default flex-item-right-sm\">\n                  <span translate>No versions</span>\n                </span>\n              </ng-template>\n              <ng-template ngPluralCase=\"=1\">\n                <span class=\"label label-info\"><span translate>1 version</span></span>\n              </ng-template>\n              <ng-template ngPluralCase=\"other\">\n                <span class=\"label label-info\">\n                  <span\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{ count: getBaseVersionsCount$(firmware) | async }\"\n                    >{{ count }} versions</span\n                  >\n                </span>\n              </ng-template>\n            </span>\n          </span>\n        </span>\n      </div>\n    </c8y-li-body>\n\n    <c8y-li-action (click)=\"editFirmware(firmware)\" icon=\"pencil\">\n      {{ 'Edit' | translate }}\n    </c8y-li-action>\n\n    <c8y-li-action (click)=\"deleteFirmware(firmware)\" icon=\"trash\">\n      {{ 'Delete' | translate }}\n    </c8y-li-action>\n  </c8y-li>\n</c8y-list-group>\n"
        })
    ], FirmwareListComponent);
    return FirmwareListComponent;
}());
export { FirmwareListComponent };
