// @c8y/ngx-components/esm5/sms-gateway/sms-gateway.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ModalService, AlertService, gettext, Status } from '@c8y/ngx-components';
import { SmsGatewayService } from './sms-gateway.service';
import { pick } from 'lodash-es';
var SmsGatewayComponent = /** @class */ (function () {
    function SmsGatewayComponent(modalService, alertService, smsGatewayService) {
        this.modalService = modalService;
        this.alertService = alertService;
        this.smsGatewayService = smsGatewayService;
        this.smsProviders = [
            { id: 'openit', name: 'OpenIT' },
            { id: 'sms77', name: 'sms77' }
        ];
        this.supportedConfigOptions = [
            'provider',
            'openit.username',
            'credentials.openit.password',
            'credentials.sms77.api.key'
        ];
        this.pageTitle = gettext('SMS provider');
    }
    SmsGatewayComponent.prototype.ngOnInit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var res;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.smsConfig = this.getEmptyConfig();
                        this.oldConfig = this.getEmptyConfig();
                        return [4 /*yield*/, this.smsGatewayService.getProviderConfig()];
                    case 1:
                        res = _a.sent();
                        if (this.isSupportedProvider(res.data)) {
                            this.smsConfig = this.getConfigurableProperties(res.data);
                            this.oldConfig = this.smsConfig;
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    SmsGatewayComponent.prototype.getConfigurableProperties = function (config) {
        return pick(config, this.supportedConfigOptions);
    };
    SmsGatewayComponent.prototype.isSupportedProvider = function (config) {
        return this.smsProviders.some(function (prov) { return prov.id === config.provider; });
    };
    SmsGatewayComponent.prototype.saveSMSGatewayConfiguration = function (form) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var res, data, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.smsGatewayService.saveProviderConfig(this.smsConfig)];
                    case 1:
                        res = _b.sent();
                        if (!(res && res.status !== 200)) return [3 /*break*/, 5];
                        if (!res.json) return [3 /*break*/, 3];
                        return [4 /*yield*/, res.json()];
                    case 2:
                        _a = _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        _a = undefined;
                        _b.label = 4;
                    case 4:
                        data = _a;
                        this.alertService.addServerFailure({ data: data, res: res });
                        return [3 /*break*/, 6];
                    case 5:
                        this.alertService.success(gettext('Credentials updated.'));
                        Object.assign(this.oldConfig, this.smsConfig);
                        form.pristine = true;
                        _b.label = 6;
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    SmsGatewayComponent.prototype.deleteSMSGatewayConfiguration = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.modalService.confirm(gettext('Delete credentials'), gettext("You are about to delete SMS provider credentials. Deleting credentials will lock out any users with\n      SMS-based two-factor authentication and deactivate the SMS-based smart rules and device communication.\n      Do you want to proceed?"), Status.DANGER, { ok: gettext('Delete') })];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.smsGatewayService.deleteProviderConfig(this.oldConfig)];
                    case 2:
                        _a.sent();
                        this.alertService.success(gettext('Credentials deleted.'));
                        this.smsConfig = this.getEmptyConfig();
                        this.oldConfig = this.getEmptyConfig();
                        return [2 /*return*/];
                }
            });
        });
    };
    SmsGatewayComponent.prototype.getEmptyConfig = function () {
        return {
            provider: undefined
        };
    };
    SmsGatewayComponent.ctorParameters = function () { return [
        { type: ModalService },
        { type: AlertService },
        { type: SmsGatewayService }
    ]; };
    SmsGatewayComponent = tslib_1.__decorate([
        Component({
            selector: 'c8y-sms-gateway',
            template: "<c8y-title>{{ pageTitle | translate }}</c8y-title>\n<div class=\"row\">\n  <div class=\"col-xs-12 col-sm-10 col-md-8 col-lg-6\">\n    <form\n      class=\"card\"\n      #smsGatewayForm=\"ngForm\"\n      (ngSubmit)=\"smsGatewayForm.form.valid && saveSMSGatewayConfiguration(smsGatewayForm.form)\"\n    >\n      <div class=\"card-header\">\n        <h4 class=\"card-title\" translate>Credentials</h4>\n      </div>\n      <div class=\"card-block\">\n        <p class=\"bottom-m\" translate>\n          Enter your credentials from your SMS provider. This enables platform features that utilize\n          SMS services (e.g. two-factor authentication and user notifications).\n        </p>\n        <label translate>SMS provider</label>\n        <div class=\"form-group\">\n          <ul class=\"list-unstyled\">\n            <li *ngFor=\"let provider of smsProviders\">\n              <label class=\"c8y-radio\">\n                <input\n                  class=\"form-control\"\n                  type=\"radio\"\n                  name=\"providerName\"\n                  [(ngModel)]=\"smsConfig.provider\"\n                  [value]=\"provider.id\"\n                />\n                <span></span>\n                {{ provider.name | translate }}\n              </label>\n            </li>\n          </ul>\n        </div>\n        <div class=\"form-group\" *ngIf=\"smsConfig.provider === 'openit'\">\n          <label translate for=\"userName\">Username</label>\n          <input\n            id=\"userName\"\n            class=\"form-control\"\n            type=\"text\"\n            autocomplete=\"off\"\n            name=\"userName\"\n            [(ngModel)]=\"smsConfig['openit.username']\"\n            placeholder=\"{{ 'e.g. joe`LOCALIZE`' | translate }}\"\n            required\n          />\n        </div>\n        <div class=\"form-group\" *ngIf=\"smsConfig.provider === 'openit'\">\n          <label translate for=\"password\">Password</label>\n          <input\n            id=\"password\"\n            class=\"form-control\"\n            type=\"password\"\n            autocomplete=\"off\"\n            name=\"password\"\n            [(ngModel)]=\"smsConfig['credentials.openit.password']\"\n            placeholder=\"{{ 'e.g. my_password' | translate }}\"\n            required\n          />\n        </div>\n        <div class=\"form-group\" *ngIf=\"smsConfig.provider === 'sms77'\">\n          <label translate for=\"apiKey\">API key</label>\n          <input\n            id=\"apiKey\"\n            class=\"form-control\"\n            type=\"text\"\n            autocomplete=\"off\"\n            name=\"apiKey\"\n            [(ngModel)]=\"smsConfig['credentials.sms77.api.key']\"\n            required\n          />\n        </div>\n      </div>\n      <div class=\"card-footer\">\n        <button\n          type=\"button\"\n          class=\"btn btn-default\"\n          (click)=\"deleteSMSGatewayConfiguration()\"\n          [disabled]=\"!(oldConfig && oldConfig.provider)\"\n          title=\"{{ 'Delete' | translate }}\"\n          translate\n        >\n          Delete\n        </button>\n        <button\n          type=\"submit\"\n          class=\"btn btn-primary\"\n          [disabled]=\"!smsGatewayForm.form.valid || smsGatewayForm.form.pristine\"\n          title=\"{{ 'Save' | translate }}\"\n          translate\n        >\n          Save\n        </button>\n      </div>\n    </form>\n  </div>\n</div>\n"
        })
    ], SmsGatewayComponent);
    return SmsGatewayComponent;
}());
export { SmsGatewayComponent };
