// @c8y/ngx-components/esm5/sms-gateway/sms-gateway.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { FetchClient, IFetchOptions, TenantOptionsService } from '@c8y/client';
var SmsGatewayService = /** @class */ (function () {
    function SmsGatewayService(client, tenantOptionsService) {
        this.client = client;
        this.tenantOptionsService = tenantOptionsService;
        this.category = 'messaging';
        this.basePath = 'service/register/messaging';
        this.providerTemplates = {
            cepConfig: {
                'sms.senderAddress': 'cumulocity',
                'sms.senderName': 'cumulocity'
            },
            openit: {
                provider: 'openit',
                'openit.baseUrl': 'https://sms.plusserver.com/put.php'
            },
            sms77: {
                provider: 'sms77',
                'sms77.url': 'https://gateway.sms77.io/api/sms'
            }
        };
    }
    SmsGatewayService.prototype.getProviderConfig = function () {
        return this.tenantOptionsService.detail({ category: this.category, key: '' });
    };
    SmsGatewayService.prototype.saveProviderConfig = function (newConfig) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var providerTemplate, cepConfig, config;
            return tslib_1.__generator(this, function (_a) {
                providerTemplate = this.providerTemplates[newConfig.provider];
                cepConfig = this.providerTemplates.cepConfig;
                Object.assign(newConfig, providerTemplate, cepConfig);
                config = {
                    method: 'PUT',
                    headers: { 'content-type': 'application/json' },
                    body: JSON.stringify(newConfig)
                };
                return [2 /*return*/, this.client.fetch(this.basePath, config)];
            });
        });
    };
    SmsGatewayService.prototype.deleteProviderConfig = function (config) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var providerTemplate;
            return tslib_1.__generator(this, function (_a) {
                providerTemplate = this.providerTemplates[config.provider];
                Object.assign(config, providerTemplate);
                return [2 /*return*/, this.deleteProviderOptions(config)];
            });
        });
    };
    SmsGatewayService.prototype.deleteProviderOptions = function (options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, _b, optionKey, e_1_1;
            var e_1, _c;
            return tslib_1.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 5, 6, 7]);
                        _a = tslib_1.__values(Object.keys(options)), _b = _a.next();
                        _d.label = 1;
                    case 1:
                        if (!!_b.done) return [3 /*break*/, 4];
                        optionKey = _b.value;
                        return [4 /*yield*/, this.tenantOptionsService.delete({ category: this.category, key: optionKey })];
                    case 2:
                        _d.sent();
                        _d.label = 3;
                    case 3:
                        _b = _a.next();
                        return [3 /*break*/, 1];
                    case 4: return [3 /*break*/, 7];
                    case 5:
                        e_1_1 = _d.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 7];
                    case 6:
                        try {
                            if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    SmsGatewayService.ctorParameters = function () { return [
        { type: FetchClient },
        { type: TenantOptionsService }
    ]; };
    SmsGatewayService = tslib_1.__decorate([
        Injectable()
    ], SmsGatewayService);
    return SmsGatewayService;
}());
export { SmsGatewayService };
