// @c8y/ngx-components/esm2015/upgrade/ng1/downgraded.services.js
import { downgradeInjectable } from '@angular/upgrade/static';
import { AlertService, AppStateService, DocsService, HeaderService, PasswordService, UserMenuService } from '@c8y/ngx-components';
import { ApiService } from '@c8y/ngx-components/api';
import { BulkOperationDetailsService } from '@c8y/ngx-components/operations';
import { TranslateService } from '@ngx-translate/core';
import { BridgeService } from '../bridge.service';
import { ServerMessagesService } from './serverMessages.service';
export const bridgeService = downgradeInjectable(BridgeService);
export const appStateService = downgradeInjectable(AppStateService);
export const headerService = downgradeInjectable(HeaderService);
export const alertsService = downgradeInjectable(AlertService);
export const userMenuService = downgradeInjectable(UserMenuService);
export const apiService = downgradeInjectable(ApiService);
export const docsService = downgradeInjectable(DocsService);
export const passwordService = downgradeInjectable(PasswordService);
export const translateService = downgradeInjectable(TranslateService);
export const serverMessagesService = downgradeInjectable(ServerMessagesService);
export const bulkOperationDetailsService = downgradeInjectable(BulkOperationDetailsService);
