// @c8y/ngx-components/esm2015/core/search/search-result-empty.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { gettext } from '../i18n/gettext';
let SearchResultEmptyComponent = class SearchResultEmptyComponent {
    constructor() {
        this.NO_RESULT_FOUND = gettext('No results found.');
        this.msg = this.NO_RESULT_FOUND;
    }
    set _message(msg) {
        if (msg && msg.length > 0) {
            this.msg = msg;
        }
        else {
            this.msg = this.NO_RESULT_FOUND;
        }
    }
    get message() {
        return this.msg;
    }
};
tslib_1.__decorate([
    Input('message')
], SearchResultEmptyComponent.prototype, "_message", null);
SearchResultEmptyComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-search-result-empty',
        template: "<div class=\"no-results\">\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"search\"></h1>\n    <div class=\"p-t-8\">\n      <p><strong>{{ message | translate }}</strong></p>\n      <small translate>Refine your search terms.</small>\n    </div>\n  </div>\n</div>\n"
    })
], SearchResultEmptyComponent);
export { SearchResultEmptyComponent };
