// @c8y/ngx-components/esm2015/core/search/search.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, ViewChild, EventEmitter } from '@angular/core';
import { SearchService } from './search.service';
let SearchComponent = class SearchComponent {
    constructor(searchService) {
        this.searchService = searchService;
        this.icon = 'search';
        this.priority = 0;
        this.search = new EventEmitter();
        this.term = '';
    }
    onSearch() {
        this.search.emit(this);
    }
    ngAfterViewInit() {
        this.template = undefined;
        this.viewInitTimeout = setTimeout(() => this.searchService.add(this));
    }
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.searchService.remove(this);
    }
};
SearchComponent.ctorParameters = () => [
    { type: SearchService }
];
tslib_1.__decorate([
    Input()
], SearchComponent.prototype, "name", void 0);
tslib_1.__decorate([
    Input()
], SearchComponent.prototype, "icon", void 0);
tslib_1.__decorate([
    Input()
], SearchComponent.prototype, "priority", void 0);
tslib_1.__decorate([
    ViewChild('template', { static: false })
], SearchComponent.prototype, "template", void 0);
tslib_1.__decorate([
    Output()
], SearchComponent.prototype, "search", void 0);
tslib_1.__decorate([
    Input()
], SearchComponent.prototype, "term", void 0);
SearchComponent = tslib_1.__decorate([
    Component({
        selector: 'c8y-search',
        template: "<ng-template #template>\n  <ng-content></ng-content>\n</ng-template>"
    })
], SearchComponent);
export { SearchComponent };
